// @zapier/zapier-sdk/dist/plugins/tables/listTableFields/index.js
import { createFunction } from "../../../utils/function-utils";
import { ListTableFieldsOptionsSchema, ListTableFieldsOptionsInputSchema, ListTableFieldsApiResponseSchema, FieldItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver } from "../../../resolvers";
import { resolveFieldKeys, transformFieldItem } from "../utils";
export const listTableFieldsPlugin = ({ context }) => {
    async function listTableFields(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const fieldKeys = "fields" in options ? options.fields : options.fieldKeys;
        const searchParams = {};
        if (fieldKeys && fieldKeys.length > 0) {
            const numericIds = await resolveFieldKeys({
                api,
                tableId,
                fieldKeys,
            });
            searchParams.field_ids = numericIds.join(",");
        }
        const rawResponse = await api.get(`/tables/api/v1/tables/${tableId}/fields`, {
            searchParams: Object.keys(searchParams).length > 0 ? searchParams : undefined,
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to access this table or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to access this table. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 404) {
                    return new Error(`Table not found: ${tableId}`);
                }
                return undefined;
            },
            authRequired: true,
        });
        const response = ListTableFieldsApiResponseSchema.parse(rawResponse);
        return {
            data: response.data.map(transformFieldItem),
        };
    }
    const listTableFieldsDefinition = createFunction(listTableFields, ListTableFieldsOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, listTableFields.name));
    return {
        listTableFields: listTableFieldsDefinition,
        context: {
            meta: {
                listTableFields: {
                    categories: ["table"],
                    type: "list",
                    itemType: "Field",
                    inputSchema: ListTableFieldsOptionsSchema,
                    outputSchema: FieldItemSchema,
                    resolvers: {
                        table: tableIdResolver,
                    },
                },
            },
        },
    };
};
