// @zapier/zapier-sdk/dist/plugins/tables/listTableFields/schemas.js
import { z } from "zod";
import { withFormatter } from "../../../utils/schema-utils";
import { TablePropertySchema, FieldsPropertySchema, } from "../../../types/properties";
export const FieldTypeSchema = z.enum([
    "string",
    "multiple_string",
    "labeled_string",
    "multiple_labeled_string",
    "text",
    "multiple_text",
    "boolean",
    "multiple_boolean",
    "number",
    "multiple_number",
    "decimal",
    "multiple_decimal",
    "datetime",
    "multiple_datetime",
    "uuid",
    "multiple_uuid",
    "json",
    "multiple_json",
    "formula",
    "button_trigger_zap",
    "button_continue_zap",
    "email",
    "multiple_email",
    "link",
    "multiple_link",
    "currency",
    "phone_number",
    "ai_formula",
    "linked_record",
    "multiple_linked_record",
]);
export const FieldApiItemSchema = z.object({
    id: z.number(),
    type: FieldTypeSchema,
    name: z.string(),
    created_at: z.string().optional(),
    edited_at: z.string().optional(),
    options: z.record(z.string(), z.unknown()).optional(),
    config: z.record(z.string(), z.unknown()).optional(),
    is_order_field: z.boolean().optional(),
    is_filter_field: z.boolean().optional(),
    is_selected_field: z.boolean().optional(),
});
export const ListTableFieldsApiResponseSchema = z.object({
    data: z.array(FieldApiItemSchema),
});
const FieldItemSchemaBase = z.object({
    id: z.string(),
    type: FieldTypeSchema,
    name: z.string(),
    created_at: z.string().optional(),
    edited_at: z.string().optional(),
    options: z.record(z.string(), z.unknown()).optional(),
    config: z.record(z.string(), z.unknown()).optional(),
});
export const FieldItemSchema = withFormatter(FieldItemSchemaBase, {
    format: (item) => ({
        title: item.name,
        id: item.id,
        details: [{ text: `Type: ${item.type}`, style: "dim" }],
    }),
});
const ListTableFieldsDescription = "List fields for a table";
// Base schema with optional fields (new + deprecated)
const ListTableFieldsOptionsBaseSchema = z.object({
    fields: FieldsPropertySchema.optional(),
    fieldKeys: z
        .array(z.union([z.string(), z.number()]))
        .optional()
        .describe('Filter by specific fields. Accepts field names (e.g., "Email") or IDs (e.g., "f6", "6", or 6).')
        .meta({ deprecated: true }),
});
// Schema for docs/registry (new param names only)
export const ListTableFieldsOptionsSchema = z
    .object({
    table: TablePropertySchema,
})
    .merge(ListTableFieldsOptionsBaseSchema)
    .describe(ListTableFieldsDescription)
    .meta({ aliases: { tableId: "table", fieldKeys: "fields" } });
// Deprecated schema (backward compat)
const ListTableFieldsOptionsSchemaDeprecated = z
    .object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
})
    .merge(ListTableFieldsOptionsBaseSchema);
// Union for runtime validation
export const ListTableFieldsOptionsInputSchema = z
    .union([ListTableFieldsOptionsSchema, ListTableFieldsOptionsSchemaDeprecated])
    .describe(ListTableFieldsDescription);
