// @zapier/zapier-sdk/dist/plugins/getProfile/index.js
import { createFunction } from "../../utils/function-utils";
import { GetProfileSchema } from "./schemas";
import { UserProfileItemSchema } from "../../schemas/UserProfile";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
// Direct plugin function - takes options + sdk + context in one object
export const getProfilePlugin = ({ context }) => {
    async function getProfile() {
        const profile = await context.api.get("/zapier/api/v4/profile/", {
            authRequired: true,
        });
        return {
            data: {
                id: String(profile.public_id ?? profile.id),
                first_name: profile.first_name,
                last_name: profile.last_name,
                full_name: `${profile.first_name} ${profile.last_name}`,
                email: profile.email,
                email_confirmed: profile.email_confirmed,
                timezone: profile.timezone,
            },
        };
    }
    const getProfileDefinition = createFunction(getProfile, GetProfileSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, getProfile.name));
    // Return flat structure - getProfile goes directly to SDK
    return {
        getProfile: getProfileDefinition,
        context: {
            meta: {
                getProfile: {
                    categories: ["account"],
                    type: "item",
                    itemType: "Profile",
                    inputSchema: GetProfileSchema,
                    outputSchema: UserProfileItemSchema,
                },
            },
        },
    };
};
