// @zapier/zapier-sdk-cli-login/dist/src/index.js
/**
 * Zapier SDK CLI Login Package
 *
 * Handles login, logout, token storage and refresh for Zapier SDK CLI.
 * Provides getToken function that can be optionally imported by zapier-sdk.
 */
import Conf from "conf";
import { existsSync } from "fs";
import * as jwt from "jsonwebtoken";
import { getTokensFromKeychain, setTokensInKeychain, clearTokensFromKeychain, } from "./keychain";
/**
 * Authentication error for token refresh failures.
 * This is a standalone clone of ZapierAuthenticationError from the SDK.
 * We can't import from SDK because cli-login must remain a standalone package
 * with no SDK dependencies. The SDK recognizes this error by name.
 */
export class ZapierAuthenticationError extends Error {
    constructor(message) {
        super(message);
        this.name = "ZapierAuthenticationError";
    }
}
let config = null;
// Default OAuth client ID
const DEFAULT_AUTH_CLIENT_ID = "grwWZD5hUWGvb4V8ODBuOtXer3h0DBEZ2HR8aay6";
// Buffer time before token expiration to trigger refresh (5 minutes)
const TOKEN_REFRESH_BUFFER_MS = 5 * 60 * 1000;
/**
 * Creates a debug logger function.
 * Returns a no-op function when debug is disabled.
 */
function createDebugLog(enabled) {
    if (!enabled) {
        return () => { };
    }
    return (message, data) => {
        if (data === undefined) {
            console.log(`[Zapier SDK CLI Login] ${message}`);
        }
        else {
            console.log(`[Zapier SDK CLI Login] ${message}`, data);
        }
    };
}
/**
 * Censors sensitive values in headers for debug logging.
 */
function censorHeaderValue(value) {
    if (value.length > 12) {
        return `${value.substring(0, 4)}...${value.substring(value.length - 4)}`;
    }
    return `${value.charAt(0)}...`;
}
// Get auth client ID - uses passed value or falls back to default.
// Env var resolution is handled by the SDK before calling cli-login.
function getAuthClientId(clientId) {
    return clientId || DEFAULT_AUTH_CLIENT_ID;
}
export const AUTH_MODE_HEADER = "X-Auth";
// Default auth base URL
const DEFAULT_AUTH_BASE_URL = "https://zapier.com";
/**
 * Gets the OAuth token endpoint URL.
 * baseUrl should be the auth base URL (e.g., https://zapier.com), already resolved by SDK.
 */
export function getAuthTokenUrl(options) {
    const authBaseUrl = options?.baseUrl || DEFAULT_AUTH_BASE_URL;
    return `${authBaseUrl}/oauth/token/`;
}
/**
 * Gets the OAuth authorization endpoint URL.
 * baseUrl should be the auth base URL (e.g., https://zapier.com), already resolved by SDK.
 */
export function getAuthAuthorizeUrl(options) {
    const authBaseUrl = options?.baseUrl || DEFAULT_AUTH_BASE_URL;
    return `${authBaseUrl}/oauth/authorize/`;
}
/**
 * Gets all configuration needed for the PKCE login flow.
 * Credentials should have baseUrl already resolved by SDK.
 */
export function getPkceLoginConfig(options) {
    return {
        clientId: getAuthClientId(options?.credentials?.clientId),
        tokenUrl: getAuthTokenUrl({ baseUrl: options?.credentials?.baseUrl }),
        authorizeUrl: getAuthAuthorizeUrl({
            baseUrl: options?.credentials?.baseUrl,
        }),
    };
}
let cachedLogin;
export function getConfig() {
    if (!config) {
        // Conf does not create the file until the first set() call.
        config = new Conf({ projectName: "zapier-sdk-cli" });
        // If the config file already exists but has no storage mode marker, an older
        // SDK version created it. Set the marker to "config" so the login flow will
        // prompt the user to upgrade to keychain storage. Otherwise, stamp "keychain"
        // so that any future writes to the config don't leave it unmarked.
        if (!config.has("login_storage_mode")) {
            config.set("login_storage_mode", existsSync(config.path) ? "config" : "keychain");
        }
    }
    return config;
}
/**
 * Drops the cached Conf instance so the next getConfig() call re-initializes
 * from disk (including re-running the pre-existing file detection).
 */
export function unloadConfig() {
    config = null;
    cachedLogin = undefined;
}
export async function updateLogin(loginData, options = {}) {
    const debugLog = createDebugLog(options.debug ?? false);
    const storage = options.storage ?? cachedLogin?.storage ?? "keychain";
    const expiresAt = Date.now() + loginData.expires_in * 1000;
    const cfg = getConfig();
    cfg.set("login_storage_mode", storage);
    if (storage === "keychain") {
        await setTokensInKeychain({
            data: {
                login_jwt: loginData.access_token,
                login_refresh_token: loginData.refresh_token,
            },
            debugLog,
        });
        cfg.set("login_expires_at", expiresAt);
        cfg.delete("login_jwt");
        cfg.delete("login_refresh_token");
    }
    else {
        cfg.set("login_jwt", loginData.access_token);
        cfg.set("login_refresh_token", loginData.refresh_token);
        cfg.set("login_expires_at", expiresAt);
        await clearTokensFromKeychain({ debugLog });
    }
    cachedLogin = {
        jwt: loginData.access_token,
        refreshToken: loginData.refresh_token,
        expiresAt,
        storage,
    };
}
// JWT utility functions
function decodeJwtOrThrow(token) {
    if (typeof token !== "string") {
        throw new Error("Expected JWT to be a string");
    }
    const decodedJwt = jwt.decode(token, { complete: true });
    if (!decodedJwt) {
        throw new Error("Could not decode JWT");
    }
    if (typeof decodedJwt.payload === "string") {
        throw new Error("Did not expect JWT payload to be a string");
    }
    return decodedJwt;
}
/**
 * Refreshes an expired JWT token using the refresh token.
 * Returns the new access token or throws an error.
 */
async function refreshJwt(refreshToken, options = {}) {
    const { onEvent, fetch = globalThis.fetch, credentials, debug = false, } = options;
    const debugLog = createDebugLog(debug);
    const tokenUrl = getAuthTokenUrl({ baseUrl: credentials?.baseUrl });
    const clientId = getAuthClientId(credentials?.clientId);
    const startTime = Date.now();
    try {
        onEvent?.({
            type: "auth_refreshing",
            payload: {
                message: "Refreshing your token...",
                operation: "token_refresh",
            },
            timestamp: Date.now(),
        });
        debugLog(`→ POST ${tokenUrl}`, {
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                [AUTH_MODE_HEADER]: "no",
            },
            body: {
                client_id: clientId,
                refresh_token: censorHeaderValue(refreshToken),
                grant_type: "refresh_token",
            },
        });
        const response = await fetch(tokenUrl, {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                [AUTH_MODE_HEADER]: "no",
            },
            body: new URLSearchParams({
                client_id: clientId,
                refresh_token: refreshToken,
                grant_type: "refresh_token",
            }),
        });
        const duration = Date.now() - startTime;
        if (!response.ok) {
            debugLog(`← ${response.status} ${response.statusText} (${duration}ms)`);
            throw new Error(`Token refresh failed: ${response.status} ${response.statusText}`);
        }
        debugLog(`← ${response.status} ${response.statusText} (${duration}ms)`);
        const data = await response.json();
        // Update stored login data
        await updateLogin(data, { debug });
        debugLog(`Token refreshed and saved to ${cachedLogin?.storage ?? "keychain"}`);
        onEvent?.({
            type: "auth_success",
            payload: {
                message: "Token refreshed successfully",
                operation: "token_refresh",
            },
            timestamp: Date.now(),
        });
        return data.access_token;
    }
    catch (error) {
        const duration = Date.now() - startTime;
        debugLog(`✖ Token refresh failed (${duration}ms)`, {
            error: error instanceof Error ? error.message : error,
        });
        // Clear in-memory cache so the next call re-reads from storage.
        // Another process may have refreshed successfully in the meantime.
        cachedLogin = undefined;
        const errorMessage = `Token refresh failed: ${error instanceof Error ? error.message : "Unknown error"}`;
        onEvent?.({
            type: "auth_error",
            payload: {
                message: errorMessage,
                error: errorMessage,
                operation: "token_refresh",
            },
            timestamp: Date.now(),
        });
        throw error;
    }
}
/**
 * In-flight refresh promise to prevent duplicate concurrent refreshes.
 * Only one refresh can be in progress at a time within a single process.
 */
let pendingRefresh = null;
let pendingResolve = null;
/**
 * Resolves stored login credentials from keychain or config.
 * Caches the result in memory so the keychain is only read once per process.
 * Deduplicates concurrent calls so only one keychain read is in flight at a time.
 */
async function resolveStoredLogin(debugLog) {
    if (cachedLogin) {
        debugLog("Using in-memory cached credentials");
        return cachedLogin;
    }
    if (pendingResolve) {
        debugLog("Waiting for existing keychain read to complete");
        return pendingResolve;
    }
    pendingResolve = resolveStoredLoginFromStorage(debugLog).finally(() => {
        pendingResolve = null;
    });
    return pendingResolve;
}
async function resolveStoredLoginFromStorage(debugLog) {
    let cfg;
    try {
        cfg = getConfig();
    }
    catch (error) {
        debugLog("Failed to load config", {
            error: error instanceof Error ? error.message : error,
        });
        return undefined;
    }
    const expiresAt = cfg.get("login_expires_at");
    // Check for legacy config-only login (all three keys present)
    const configJwt = cfg.get("login_jwt");
    const configRefresh = cfg.get("login_refresh_token");
    if (configJwt && configRefresh && typeof expiresAt === "number") {
        debugLog("Loaded credentials from config (legacy format)");
        cachedLogin = {
            jwt: configJwt,
            refreshToken: configRefresh,
            expiresAt,
            storage: "config",
        };
        return cachedLogin;
    }
    if (typeof expiresAt !== "number") {
        debugLog("No stored login credentials found");
        return undefined;
    }
    // expires_at exists without config tokens — read from keychain
    const keychainData = await getTokensFromKeychain({ debugLog });
    if (!keychainData) {
        debugLog("No tokens found in keychain");
        return undefined;
    }
    debugLog("Loaded credentials from keychain");
    cachedLogin = {
        jwt: keychainData.login_jwt,
        refreshToken: keychainData.login_refresh_token,
        expiresAt,
        storage: "keychain",
    };
    return cachedLogin;
}
/**
 * Resolves a valid token, refreshing if expired. Returns undefined if no
 * stored login exists. Throws on refresh failure.
 */
async function resolveOrRefreshToken(options = {}) {
    const { debug = false } = options;
    const debugLog = createDebugLog(debug);
    const stored = await resolveStoredLogin(debugLog);
    if (!stored) {
        return undefined;
    }
    const { jwt: storedJwt, refreshToken, expiresAt } = stored;
    if (expiresAt > Date.now() + TOKEN_REFRESH_BUFFER_MS) {
        debugLog("Using cached token (still valid)");
        return storedJwt;
    }
    debugLog("Token expired, refreshing...");
    if (pendingRefresh) {
        debugLog("Waiting for existing refresh to complete");
        return pendingRefresh;
    }
    pendingRefresh = refreshJwt(refreshToken, options).finally(() => {
        pendingRefresh = null;
    });
    return await pendingRefresh;
}
/**
 * Main function exported for zapier-sdk to optionally use.
 * Returns the stored token from CLI configuration (with auto-refresh).
 *
 * Note: Environment variable handling (ZAPIER_CREDENTIALS, ZAPIER_TOKEN)
 * is done by the SDK before calling this function.
 *
 * Returns undefined if no valid token is found.
 */
export async function getToken(options = {}) {
    try {
        return await resolveOrRefreshToken(options);
    }
    catch (error) {
        const message = error instanceof Error ? error.message : "Token refresh failed";
        throw new ZapierAuthenticationError(`${message}\nPlease run 'login' to authenticate again.`);
    }
}
/**
 * Gets the logged-in user information from JWT token.
 * Automatically refreshes token if expired.
 */
export async function getLoggedInUser(options = {}) {
    const jwt = await getToken(options).catch(() => undefined);
    if (!jwt) {
        throw new Error("No valid authentication token available. Please login first.");
    }
    let decodedJwt = decodeJwtOrThrow(jwt);
    if (decodedJwt.payload["sub_type"] == "service") {
        decodedJwt = decodeJwtOrThrow(decodedJwt.payload["njwt"]);
    }
    if (typeof decodedJwt.payload["zap:acc"] !== "string") {
        throw new Error("JWT payload does not contain accountId");
    }
    const accountId = parseInt(decodedJwt.payload["zap:acc"], 10);
    if (isNaN(accountId)) {
        throw new Error("JWT accountId is not a number");
    }
    if (decodedJwt.payload["sub_type"] !== "customuser" ||
        typeof decodedJwt.payload["sub"] !== "string") {
        throw new Error("JWT payload does not contain customUserId");
    }
    const customUserId = parseInt(decodedJwt.payload["sub"], 10);
    if (isNaN(customUserId)) {
        throw new Error("JWT customUserId is not a number");
    }
    const email = decodedJwt.payload["zap:uname"];
    if (typeof email !== "string") {
        throw new Error("JWT payload does not contain email");
    }
    return {
        accountId,
        customUserId,
        email,
    };
}
/**
 * Checks config to determine current storage mode without reading keychain.
 * Credential keys in config are ground truth — if login_jwt exists, we're in
 * config mode regardless of the marker. The marker only matters when no
 * credential keys are present (e.g. after logout or when using keychain).
 */
export function getLoginStorageMode() {
    const cfg = getConfig();
    // Config tokens are ground truth — overrides any marker
    if (typeof cfg.get("login_jwt") === "string") {
        return "config";
    }
    // No config tokens — use the explicit marker (set by getConfig on init or
    // by updateLogin on write)
    const explicitMode = cfg.get("login_storage_mode");
    if (explicitMode === "keychain" || explicitMode === "config") {
        return explicitMode;
    }
    return "keychain";
}
/**
 * Clears stored login information.
 */
export async function logout(options = {}) {
    const { onEvent } = options;
    // Snapshot the storage mode before clearing tokens, so the marker survives
    // logout. Without this, an old-SDK config login would lose its "config"
    // signal once login_jwt is deleted.
    const mode = getLoginStorageMode();
    cachedLogin = undefined;
    await clearTokensFromKeychain();
    const cfg = getConfig();
    cfg.set("login_storage_mode", mode);
    cfg.delete("login_expires_at");
    cfg.delete("login_jwt");
    cfg.delete("login_refresh_token");
    onEvent?.({
        type: "auth_logout",
        payload: { message: "Logged out successfully", operation: "logout" },
        timestamp: Date.now(),
    });
}
/**
 * Gets the path to the configuration file.
 */
export function getConfigPath() {
    const cfg = getConfig();
    return cfg.path;
}
