// @zapier/zapier-sdk-cli-login/dist/src/keychain.js
import { getPassword, setPassword, deletePassword, getKeyring, } from "cross-keychain";
const SERVICE = "zapier-sdk-cli";
const ACCOUNT = "login";
let cachedBackendInfo;
async function getBackendInfo() {
    if (!cachedBackendInfo) {
        const keyring = await getKeyring();
        cachedBackendInfo = `${keyring.name} (${keyring.id})`;
    }
    return cachedBackendInfo;
}
// Serialize all keychain operations so concurrent reads and writes never
// overlap. Mostly, this is to ensure we don't try to read user info while
// in the middle of a logout. We don't really need to make reads serial, but
// we cache those elsewhere anyway, so we use a simple serial queue.
let keychainQueue = Promise.resolve();
function enqueue(fn) {
    // fn is passed as both fulfillment and rejection handler so a failed
    // operation never blocks subsequent ones from running.
    const result = keychainQueue.then(fn, fn);
    keychainQueue = result.then(() => { }, () => { });
    return result;
}
export async function getTokensFromKeychain({ debugLog, } = {}) {
    return enqueue(async () => {
        const backendInfo = await getBackendInfo();
        debugLog?.(`Keychain read via ${backendInfo}`);
        const startTime = Date.now();
        const raw = await getPassword(SERVICE, ACCOUNT);
        debugLog?.(`Keychain read completed in ${Date.now() - startTime}ms`);
        if (!raw) {
            debugLog?.("Keychain returned no data");
            return undefined;
        }
        let parsed;
        try {
            parsed = JSON.parse(raw);
        }
        catch {
            debugLog?.("Keychain data is not valid JSON");
            return undefined;
        }
        if (typeof parsed.login_jwt === "string" &&
            typeof parsed.login_refresh_token === "string") {
            return {
                login_jwt: parsed.login_jwt,
                login_refresh_token: parsed.login_refresh_token,
            };
        }
        debugLog?.("Keychain data has invalid shape", parsed);
        return undefined;
    });
}
export async function setTokensInKeychain({ data, debugLog, }) {
    return enqueue(async () => {
        const backendInfo = await getBackendInfo();
        debugLog?.(`Keychain write via ${backendInfo}`);
        const startTime = Date.now();
        await setPassword(SERVICE, ACCOUNT, JSON.stringify(data));
        debugLog?.(`Keychain write completed in ${Date.now() - startTime}ms`);
    });
}
export async function clearTokensFromKeychain({ debugLog, } = {}) {
    return enqueue(async () => {
        try {
            const backendInfo = await getBackendInfo();
            debugLog?.(`Keychain clear via ${backendInfo}`);
            await deletePassword(SERVICE, ACCOUNT);
        }
        catch {
            // Best-effort: keychain may not have an entry or may be unavailable
        }
    });
}
