// @zapier/zapier-sdk/dist/plugins/tables/createTableFields/index.js
import { createFunction } from "../../../utils/function-utils";
import { CreateTableFieldsOptionsSchema, CreateTableFieldsOptionsInputSchema, CreateTableFieldsApiResponseSchema, FieldItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver, tableFieldsResolver } from "../../../resolvers";
import { transformFieldItem } from "../utils";
export const createTableFieldsPlugin = ({ context }) => {
    async function createTableFields(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const rawResponse = await api.post(`/tables/api/v1/tables/${tableId}/fields`, {
            new_fields: options.fields,
        }, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to create fields or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to create fields. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
        });
        const response = CreateTableFieldsApiResponseSchema.parse(rawResponse);
        // Extract the 'new' field from each changeset item
        return {
            data: response.data.map((changeset) => transformFieldItem(changeset.new)),
        };
    }
    const createTableFieldsDefinition = createFunction(createTableFields, CreateTableFieldsOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, createTableFields.name));
    return {
        createTableFields: createTableFieldsDefinition,
        context: {
            meta: {
                createTableFields: {
                    categories: ["table"],
                    type: "create",
                    itemType: "Field",
                    inputSchema: CreateTableFieldsOptionsSchema,
                    outputSchema: FieldItemSchema,
                    resolvers: {
                        table: tableIdResolver,
                        fields: tableFieldsResolver,
                    },
                },
            },
        },
    };
};
