// @zapier/zapier-sdk/dist/plugins/tables/createTableFields/schemas.js
import { z } from "zod";
import { TablePropertySchema } from "../../../types/properties";
import { FieldTypeSchema, FieldApiItemSchema, FieldItemSchema, } from "../listTableFields/schemas";
// Re-export for consumers
export { FieldTypeSchema, FieldItemSchema };
// The API returns a changeset format with new/old/kind for each field
const FieldChangesetItemSchema = z.object({
    new: FieldApiItemSchema,
    old: FieldApiItemSchema.nullable(),
    kind: z.enum(["created", "updated", "deleted"]),
});
export const CreateTableFieldsApiResponseSchema = z.object({
    data: z.array(FieldChangesetItemSchema),
    meta: z
        .object({
        did_change: z.boolean(),
    })
        .optional(),
});
export const NewFieldSchema = z.object({
    type: FieldTypeSchema.describe("The data type of the field"),
    name: z.string().min(1).describe("The display name of the field"),
    options: z
        .record(z.string(), z.unknown())
        .optional()
        .describe("Data configuration options for the field"),
    config: z
        .record(z.string(), z.unknown())
        .optional()
        .describe("Display configuration for the field"),
});
const CreateTableFieldsDescription = "Create one or more fields in a table";
const CreateTableFieldsBase = z.object({
    fields: z
        .array(NewFieldSchema)
        .min(1)
        .describe("Array of field definitions to create"),
});
// Schema for docs/registry (new param names only)
export const CreateTableFieldsOptionsSchema = z
    .object({
    table: TablePropertySchema,
})
    .merge(CreateTableFieldsBase)
    .describe(CreateTableFieldsDescription)
    .meta({ aliases: { tableId: "table" } });
// Deprecated schema (backward compat)
const CreateTableFieldsOptionsSchemaDeprecated = z
    .object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
})
    .merge(CreateTableFieldsBase);
// Union for runtime validation
export const CreateTableFieldsOptionsInputSchema = z
    .union([
    CreateTableFieldsOptionsSchema,
    CreateTableFieldsOptionsSchemaDeprecated,
])
    .describe(CreateTableFieldsDescription);
