// @zapier/zapier-sdk-core/dist/v0/schemas/actions.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/v0/schemas/actions.ts
var actions_exports = {};
__export(actions_exports, {
  ActionItemSchema: () => ActionItemSchema,
  ActionSchema: () => ActionSchema,
  ActionTypeSchema: () => ActionTypeSchema,
  ImplementationWithActionsSchema: () => ImplementationWithActionsSchema,
  ImplementationsWithActionsResponseSchema: () => ImplementationsWithActionsResponseSchema,
  ListActionsQuerySchema: () => ListActionsQuerySchema,
  ListActionsResponseSchema: () => ListActionsResponseSchema
});
module.exports = __toCommonJS(actions_exports);
var import_zod = require("zod");
var ActionTypeSchema = import_zod.z.enum([
  "filter",
  "read",
  "read_bulk",
  "run",
  "search",
  "search_and_write",
  "search_or_write",
  "write"
]);
var ActionSchema = import_zod.z.object({
  id: import_zod.z.string().optional().describe("Unique identifier for the action"),
  type: ActionTypeSchema.describe("The type of action"),
  key: import_zod.z.string().describe("Unique key identifier for the action"),
  name: import_zod.z.string().describe("Display name of the action"),
  description: import_zod.z.string().describe("Description of what the action does"),
  is_important: import_zod.z.boolean().optional().describe("Whether this action is marked as important"),
  is_hidden: import_zod.z.boolean().optional().describe("Whether this action is hidden from listings"),
  selected_api: import_zod.z.string().optional().describe("The implementation ID this action belongs to")
});
var ActionItemSchema = import_zod.z.object({
  id: import_zod.z.string().optional().describe("Unique identifier for the action"),
  key: import_zod.z.string().describe("Unique key identifier for the action"),
  description: import_zod.z.string().describe("Description of what the action does"),
  is_important: import_zod.z.boolean().optional().describe("Whether this action is marked as important"),
  is_hidden: import_zod.z.boolean().optional().describe("Whether this action is hidden from listings"),
  app_key: import_zod.z.string().describe("App key extracted from implementation ID (without version)"),
  app_version: import_zod.z.string().optional().describe("App version extracted from implementation ID"),
  action_type: ActionTypeSchema.describe("The type of action"),
  title: import_zod.z.string().describe("Display name of the action (mapped from name)"),
  type: import_zod.z.literal("action").describe("Type identifier for this item")
}).describe("Normalized action item returned by list actions endpoint");
var ImplementationWithActionsSchema = import_zod.z.object({
  selected_api: import_zod.z.string().describe("The implementation ID"),
  app_id: import_zod.z.number().optional().describe("Numeric app ID"),
  auth_type: import_zod.z.string().optional().describe("Authentication type"),
  actions: import_zod.z.array(ActionSchema).optional().describe("Array of actions for this implementation"),
  is_deprecated: import_zod.z.boolean().optional().describe("Whether the implementation is deprecated"),
  is_private_only: import_zod.z.boolean().optional().describe("Whether the implementation is private only"),
  is_invite_only: import_zod.z.boolean().optional().describe("Whether the implementation is invite only"),
  is_beta: import_zod.z.boolean().optional().default(false).describe("Whether the implementation is in beta"),
  is_premium: import_zod.z.boolean().optional().default(false).describe("Whether the implementation requires premium"),
  name: import_zod.z.string().optional().describe("Name of the implementation"),
  slug: import_zod.z.string().optional().describe("URL-friendly slug")
});
var ImplementationsWithActionsResponseSchema = import_zod.z.object({
  count: import_zod.z.number().describe("Total number of results"),
  next: import_zod.z.string().nullable().optional().describe("URL for the next page of results"),
  previous: import_zod.z.string().nullable().optional().describe("URL for the previous page of results"),
  results: import_zod.z.array(ImplementationWithActionsSchema).describe("Array of implementations with their actions")
});
var ListActionsQuerySchema = import_zod.z.object({
  appKey: import_zod.z.string().describe(
    "App key to list actions for (e.g., 'SlackCLIAPI@1.21.1' or 'SlackCLIAPI')"
  ),
  actionType: ActionTypeSchema.optional().describe(
    "Filter actions by type (e.g., 'read', 'write', 'search')"
  )
}).describe("Query parameters for listing actions");
var ListActionsResponseSchema = import_zod.z.object({
  data: import_zod.z.array(ActionItemSchema).describe("Array of action items")
}).describe("Response schema for listActions");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ActionItemSchema,
  ActionSchema,
  ActionTypeSchema,
  ImplementationWithActionsSchema,
  ImplementationsWithActionsResponseSchema,
  ListActionsQuerySchema,
  ListActionsResponseSchema
});
