// @zapier/zapier-sdk-core/dist/v0/schemas/actions.js
// src/v0/schemas/actions.ts
import { z } from "zod";
var ActionTypeSchema = z.enum([
  "filter",
  "read",
  "read_bulk",
  "run",
  "search",
  "search_and_write",
  "search_or_write",
  "write"
]);
var ActionSchema = z.object({
  id: z.string().optional().describe("Unique identifier for the action"),
  type: ActionTypeSchema.describe("The type of action"),
  key: z.string().describe("Unique key identifier for the action"),
  name: z.string().describe("Display name of the action"),
  description: z.string().describe("Description of what the action does"),
  is_important: z.boolean().optional().describe("Whether this action is marked as important"),
  is_hidden: z.boolean().optional().describe("Whether this action is hidden from listings"),
  selected_api: z.string().optional().describe("The implementation ID this action belongs to")
});
var ActionItemSchema = z.object({
  id: z.string().optional().describe("Unique identifier for the action"),
  key: z.string().describe("Unique key identifier for the action"),
  description: z.string().describe("Description of what the action does"),
  is_important: z.boolean().optional().describe("Whether this action is marked as important"),
  is_hidden: z.boolean().optional().describe("Whether this action is hidden from listings"),
  app_key: z.string().describe("App key extracted from implementation ID (without version)"),
  app_version: z.string().optional().describe("App version extracted from implementation ID"),
  action_type: ActionTypeSchema.describe("The type of action"),
  title: z.string().describe("Display name of the action (mapped from name)"),
  type: z.literal("action").describe("Type identifier for this item")
}).describe("Normalized action item returned by list actions endpoint");
var ImplementationWithActionsSchema = z.object({
  selected_api: z.string().describe("The implementation ID"),
  app_id: z.number().optional().describe("Numeric app ID"),
  auth_type: z.string().optional().describe("Authentication type"),
  actions: z.array(ActionSchema).optional().describe("Array of actions for this implementation"),
  is_deprecated: z.boolean().optional().describe("Whether the implementation is deprecated"),
  is_private_only: z.boolean().optional().describe("Whether the implementation is private only"),
  is_invite_only: z.boolean().optional().describe("Whether the implementation is invite only"),
  is_beta: z.boolean().optional().default(false).describe("Whether the implementation is in beta"),
  is_premium: z.boolean().optional().default(false).describe("Whether the implementation requires premium"),
  name: z.string().optional().describe("Name of the implementation"),
  slug: z.string().optional().describe("URL-friendly slug")
});
var ImplementationsWithActionsResponseSchema = z.object({
  count: z.number().describe("Total number of results"),
  next: z.string().nullable().optional().describe("URL for the next page of results"),
  previous: z.string().nullable().optional().describe("URL for the previous page of results"),
  results: z.array(ImplementationWithActionsSchema).describe("Array of implementations with their actions")
});
var ListActionsQuerySchema = z.object({
  appKey: z.string().describe(
    "App key to list actions for (e.g., 'SlackCLIAPI@1.21.1' or 'SlackCLIAPI')"
  ),
  actionType: ActionTypeSchema.optional().describe(
    "Filter actions by type (e.g., 'read', 'write', 'search')"
  )
}).describe("Query parameters for listing actions");
var ListActionsResponseSchema = z.object({
  data: z.array(ActionItemSchema).describe("Array of action items")
}).describe("Response schema for listActions");
export {
  ActionItemSchema,
  ActionSchema,
  ActionTypeSchema,
  ImplementationWithActionsSchema,
  ImplementationsWithActionsResponseSchema,
  ListActionsQuerySchema,
  ListActionsResponseSchema
};
