// @zapier/zapier-sdk-core/dist/v0/schemas/client-credentials.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/v0/schemas/client-credentials.ts
var client_credentials_exports = {};
__export(client_credentials_exports, {
  ClientCredentialsCreatedItemSchema: () => ClientCredentialsCreatedItemSchema,
  ClientCredentialsItemSchema: () => ClientCredentialsItemSchema,
  ClientCredentialsScopeSchema: () => ClientCredentialsScopeSchema,
  CreateClientCredentialsRequestSchema: () => CreateClientCredentialsRequestSchema,
  CreateClientCredentialsResponseSchema: () => CreateClientCredentialsResponseSchema,
  DeleteClientCredentialsParamSchema: () => DeleteClientCredentialsParamSchema,
  IdentityClientCredentialsCreatedItemSchema: () => IdentityClientCredentialsCreatedItemSchema,
  IdentityClientCredentialsItemSchema: () => IdentityClientCredentialsItemSchema,
  IdentityClientCredentialsListResponseSchema: () => IdentityClientCredentialsListResponseSchema,
  ListClientCredentialsQuerySchema: () => ListClientCredentialsQuerySchema,
  ListClientCredentialsResponseSchema: () => ListClientCredentialsResponseSchema
});
module.exports = __toCommonJS(client_credentials_exports);
var import_zod = require("zod");
var ClientCredentialsScopeSchema = import_zod.z.enum(["credentials", "external"]);
var IdentityClientCredentialsBaseSchema = import_zod.z.object({
  client_id: import_zod.z.string(),
  name: import_zod.z.string()
});
var IdentityClientCredentialsItemSchema = IdentityClientCredentialsBaseSchema.extend({
  allowed_scopes: import_zod.z.array(import_zod.z.string()).nullable().optional(),
  created: import_zod.z.string().datetime().nullable().optional(),
  updated: import_zod.z.string().datetime().optional()
});
var IdentityClientCredentialsCreatedItemSchema = IdentityClientCredentialsBaseSchema.extend({
  client_secret: import_zod.z.string()
});
var IdentityClientCredentialsListResponseSchema = import_zod.z.object({
  count: import_zod.z.number(),
  next: import_zod.z.string().nullable().optional(),
  previous: import_zod.z.string().nullable().optional(),
  results: import_zod.z.array(IdentityClientCredentialsItemSchema)
});
var ClientCredentialsBaseSchema = import_zod.z.object({
  client_id: import_zod.z.string().describe("The public identifier (Client ID) of the OAuth application"),
  name: import_zod.z.string().describe("Human-readable name of the OAuth application")
});
var ClientCredentialsItemSchema = ClientCredentialsBaseSchema.extend({
  allowed_scopes: import_zod.z.array(ClientCredentialsScopeSchema).describe(
    "List of OAuth scopes that this application is allowed to request"
  ),
  created_at: import_zod.z.string().datetime().nullable().optional().describe("When the application was created (ISO 8601)"),
  updated_at: import_zod.z.string().datetime().optional().describe("When the application was last updated (ISO 8601)")
});
var ClientCredentialsCreatedItemSchema = ClientCredentialsBaseSchema.extend({
  client_secret: import_zod.z.string().describe("The client secret (only shown once on creation)")
});
var ListClientCredentialsQuerySchema = import_zod.z.object({
  pageSize: import_zod.z.coerce.number().int().min(1).max(100).optional().describe("Number of items per page (default: 100, max: 100)"),
  offset: import_zod.z.coerce.number().int().min(0).optional().describe("Pagination offset")
}).describe("Query parameters for listing client credentials");
var ListClientCredentialsResponseSchema = import_zod.z.object({
  data: import_zod.z.array(ClientCredentialsItemSchema).describe("Array of client credentials"),
  links: import_zod.z.object({
    next: import_zod.z.string().nullable().optional().describe("URL for the next page of results (if available)")
  }).describe("Pagination links"),
  meta: import_zod.z.object({
    count: import_zod.z.number().describe("Total number of items"),
    limit: import_zod.z.number().describe("Number of items per page"),
    offset: import_zod.z.number().describe("Offset of the current page")
  }).describe("Pagination metadata")
}).describe("Response for listing client credentials");
var CreateClientCredentialsRequestSchema = import_zod.z.object({
  name: import_zod.z.string().min(1).max(255).describe("Human-readable name for the client credentials"),
  allowed_scopes: import_zod.z.array(ClientCredentialsScopeSchema).min(1).describe(
    "List of OAuth scopes that this application will be allowed to request"
  )
}).describe("Request body for creating client credentials");
var CreateClientCredentialsResponseSchema = import_zod.z.object({
  data: ClientCredentialsCreatedItemSchema.describe(
    "The created client credentials (includes client_secret)"
  )
}).describe("Response for creating client credentials");
var DeleteClientCredentialsParamSchema = import_zod.z.object({
  clientId: import_zod.z.string().describe("The client ID of the credentials to delete")
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClientCredentialsCreatedItemSchema,
  ClientCredentialsItemSchema,
  ClientCredentialsScopeSchema,
  CreateClientCredentialsRequestSchema,
  CreateClientCredentialsResponseSchema,
  DeleteClientCredentialsParamSchema,
  IdentityClientCredentialsCreatedItemSchema,
  IdentityClientCredentialsItemSchema,
  IdentityClientCredentialsListResponseSchema,
  ListClientCredentialsQuerySchema,
  ListClientCredentialsResponseSchema
});
