// @zapier/zapier-sdk-core/dist/v0/schemas/client-credentials.js
// src/v0/schemas/client-credentials.ts
import { z } from "zod";
var ClientCredentialsScopeSchema = z.enum(["credentials", "external"]);
var IdentityClientCredentialsBaseSchema = z.object({
  client_id: z.string(),
  name: z.string()
});
var IdentityClientCredentialsItemSchema = IdentityClientCredentialsBaseSchema.extend({
  allowed_scopes: z.array(z.string()).nullable().optional(),
  created: z.string().datetime().nullable().optional(),
  updated: z.string().datetime().optional()
});
var IdentityClientCredentialsCreatedItemSchema = IdentityClientCredentialsBaseSchema.extend({
  client_secret: z.string()
});
var IdentityClientCredentialsListResponseSchema = z.object({
  count: z.number(),
  next: z.string().nullable().optional(),
  previous: z.string().nullable().optional(),
  results: z.array(IdentityClientCredentialsItemSchema)
});
var ClientCredentialsBaseSchema = z.object({
  client_id: z.string().describe("The public identifier (Client ID) of the OAuth application"),
  name: z.string().describe("Human-readable name of the OAuth application")
});
var ClientCredentialsItemSchema = ClientCredentialsBaseSchema.extend({
  allowed_scopes: z.array(ClientCredentialsScopeSchema).describe(
    "List of OAuth scopes that this application is allowed to request"
  ),
  created_at: z.string().datetime().nullable().optional().describe("When the application was created (ISO 8601)"),
  updated_at: z.string().datetime().optional().describe("When the application was last updated (ISO 8601)")
});
var ClientCredentialsCreatedItemSchema = ClientCredentialsBaseSchema.extend({
  client_secret: z.string().describe("The client secret (only shown once on creation)")
});
var ListClientCredentialsQuerySchema = z.object({
  pageSize: z.coerce.number().int().min(1).max(100).optional().describe("Number of items per page (default: 100, max: 100)"),
  offset: z.coerce.number().int().min(0).optional().describe("Pagination offset")
}).describe("Query parameters for listing client credentials");
var ListClientCredentialsResponseSchema = z.object({
  data: z.array(ClientCredentialsItemSchema).describe("Array of client credentials"),
  links: z.object({
    next: z.string().nullable().optional().describe("URL for the next page of results (if available)")
  }).describe("Pagination links"),
  meta: z.object({
    count: z.number().describe("Total number of items"),
    limit: z.number().describe("Number of items per page"),
    offset: z.number().describe("Offset of the current page")
  }).describe("Pagination metadata")
}).describe("Response for listing client credentials");
var CreateClientCredentialsRequestSchema = z.object({
  name: z.string().min(1).max(255).describe("Human-readable name for the client credentials"),
  allowed_scopes: z.array(ClientCredentialsScopeSchema).min(1).describe(
    "List of OAuth scopes that this application will be allowed to request"
  )
}).describe("Request body for creating client credentials");
var CreateClientCredentialsResponseSchema = z.object({
  data: ClientCredentialsCreatedItemSchema.describe(
    "The created client credentials (includes client_secret)"
  )
}).describe("Response for creating client credentials");
var DeleteClientCredentialsParamSchema = z.object({
  clientId: z.string().describe("The client ID of the credentials to delete")
});
export {
  ClientCredentialsCreatedItemSchema,
  ClientCredentialsItemSchema,
  ClientCredentialsScopeSchema,
  CreateClientCredentialsRequestSchema,
  CreateClientCredentialsResponseSchema,
  DeleteClientCredentialsParamSchema,
  IdentityClientCredentialsCreatedItemSchema,
  IdentityClientCredentialsItemSchema,
  IdentityClientCredentialsListResponseSchema,
  ListClientCredentialsQuerySchema,
  ListClientCredentialsResponseSchema
};
