// @zapier/zapier-sdk-core/dist/v0/schemas/connections.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/v0/schemas/connections.ts
var connections_exports = {};
__export(connections_exports, {
  ConnectionItemSchema: () => ConnectionItemSchema,
  ConnectionSchema: () => ConnectionSchema,
  ConnectionsResponseSchema: () => ConnectionsResponseSchema,
  GetConnectionParamSchema: () => GetConnectionParamSchema,
  ListConnectionsQuerySchema: () => ListConnectionsQuerySchema,
  ListConnectionsResponseSchema: () => ListConnectionsResponseSchema
});
module.exports = __toCommonJS(connections_exports);
var import_zod = require("zod");
var ConnectionSchema = import_zod.z.object({
  id: import_zod.z.number().describe("Unique identifier for the connection"),
  date: import_zod.z.string().describe("Date created"),
  lastchanged: import_zod.z.string().optional().describe("Date last changed"),
  account_id: import_zod.z.number().describe("Account ID associated with this connection"),
  customuser_id: import_zod.z.number().optional().describe("Custom user ID (if applicable)"),
  selected_api: import_zod.z.string().describe("Selected API key (internal identifier)"),
  destination_selected_api: import_zod.z.string().nullable().optional().describe("Destination API key (if applicable)"),
  is_invite_only: import_zod.z.boolean().describe("Whether the connection is invite-only"),
  is_private: import_zod.z.boolean().describe("Whether the connection is private"),
  shared_with_all: import_zod.z.boolean().describe("Whether the connection is shared with all users"),
  is_stale: import_zod.z.string().optional().describe("Stale status string"),
  is_shared: import_zod.z.string().optional().describe("Shared status string"),
  marked_stale_at: import_zod.z.string().nullable().optional().describe("Date when marked stale"),
  label: import_zod.z.string().nullable().optional().describe("User label for the connection"),
  identifier: import_zod.z.string().nullable().optional().describe("Identifier"),
  title: import_zod.z.string().nullable().optional().describe("Title of the connection"),
  url: import_zod.z.string().optional().describe("URL to the connection resource"),
  groups: import_zod.z.array(
    import_zod.z.record(import_zod.z.string(), import_zod.z.unknown()).describe("Groups associated with the connection")
  ).optional().describe("Array of groups associated with the connection"),
  members: import_zod.z.string().optional().describe("Members associated with the connection"),
  permissions: import_zod.z.record(import_zod.z.string(), import_zod.z.boolean()).optional().describe("Permissions for the connection")
});
var ConnectionItemSchema = ConnectionSchema.omit({
  selected_api: true,
  customuser_id: true
}).extend({
  // Override numeric IDs with string versions (converted by normalizeConnectionItem)
  id: import_zod.z.string().describe("Unique identifier for the connection"),
  account_id: import_zod.z.string().describe("Account ID associated with this connection"),
  // Renamed fields
  implementation_id: import_zod.z.string().optional().describe("Implementation ID (was selected_api)"),
  profile_id: import_zod.z.string().optional().describe("Profile ID (was customuser_id)"),
  // Mapped fields (originals preserved in ...restOfAuth)
  is_expired: import_zod.z.string().optional().describe("Whether the connection is expired (mapped from is_stale)"),
  expired_at: import_zod.z.string().nullable().optional().describe("Date when connection expired (mapped from marked_stale_at)"),
  // Computed fields
  app_key: import_zod.z.string().optional().describe("App Key extracted from implementation_id"),
  app_version: import_zod.z.string().optional().describe("App Version extracted from implementation_id")
});
var ConnectionsResponseSchema = import_zod.z.object({
  count: import_zod.z.number().describe("Total number of items"),
  next: import_zod.z.string().nullable().optional().describe("Cursor for the next page of results (if available)"),
  previous: import_zod.z.string().nullable().optional().describe("Cursor for the previous page of results (if available)"),
  results: import_zod.z.array(ConnectionSchema).describe("Array of connection items")
});
var GetConnectionParamSchema = import_zod.z.object({
  connectionId: import_zod.z.string().describe("Connection ID to retrieve")
}).describe("Get a specific connection by ID");
var ListConnectionsQuerySchema = import_zod.z.object({
  app_key: import_zod.z.string().optional().describe(
    "Filter connections by app key (e.g., 'SlackCLIAPI' or slug like 'github')"
  ),
  connection_ids: import_zod.z.string().optional().describe("Comma-separated list of connection IDs to filter by"),
  search: import_zod.z.string().optional().describe("Search term to filter connections by title"),
  title: import_zod.z.string().optional().describe(
    "Filter connections by exact title match (searches first, then filters locally)"
  ),
  account_id: import_zod.z.string().optional().describe("Filter connections by account ID"),
  owner: import_zod.z.string().optional().describe(
    "Filter by owner, 'me' for your own connections or a specific user ID"
  ),
  is_expired: import_zod.z.boolean().optional().describe(
    "Filter by expired status (true = expired only, false = non-expired only)"
  ),
  page_size: import_zod.z.number().min(1).optional().describe("Number of connections per page"),
  offset: import_zod.z.string().optional().describe("Pagination offset from previous response")
}).describe("Query parameters for listing connections");
var ListConnectionsResponseSchema = import_zod.z.object({
  data: import_zod.z.array(ConnectionItemSchema).describe("Array of connection items"),
  links: import_zod.z.object({
    next: import_zod.z.string().nullable().optional().describe(
      "Fully qualified URL for the next page of results (if available), e.g. https://sdkapi.zapier.com/api/v0/connections?offset=100&page_size=50"
    )
  }).describe("Pagination links for navigating through results"),
  meta: import_zod.z.object({
    count: import_zod.z.number().describe("Total number of items"),
    limit: import_zod.z.number().describe("Number of items per page"),
    offset: import_zod.z.number().describe("Offset of the current page")
  }).describe("Metadata for the paginated result set")
}).describe("Response schema for listing connections");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ConnectionItemSchema,
  ConnectionSchema,
  ConnectionsResponseSchema,
  GetConnectionParamSchema,
  ListConnectionsQuerySchema,
  ListConnectionsResponseSchema
});
