// @zapier/zapier-sdk-core/dist/v0/schemas/errors.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/v0/schemas/errors.ts
var errors_exports = {};
__export(errors_exports, {
  ErrorCodeSchema: () => ErrorCodeSchema,
  ErrorObjectSchema: () => ErrorObjectSchema,
  ErrorSourceSchema: () => ErrorSourceSchema,
  ErrorsResponseSchema: () => ErrorsResponseSchema
});
module.exports = __toCommonJS(errors_exports);
var import_zod = require("zod");
var ErrorCodeSchema = import_zod.z.string().describe("An application-specific error code, expressed as a string value.");
var ErrorSourceSchema = import_zod.z.object({
  pointer: import_zod.z.string().optional().describe(
    "A JSON Pointer [RFC6901](https://tools.ietf.org/html/rfc6901) to the associated entity in the request document [e.g. `/data` for a primary data object, or `/data/attributes/title` for a specific attribute]."
  ),
  parameter: import_zod.z.string().optional().describe(
    "A string indicating which URI query parameter caused the error."
  ),
  header: import_zod.z.string().optional().describe("A string indicating the header that caused the error.")
}).strict().nullable().describe(
  "Identifies the source of the error within the request payload, if relevant."
);
var ErrorObjectSchema = import_zod.z.object({
  id: import_zod.z.string().nullable().optional().describe(
    "A unique identifier for this particular occurrence of the problem."
  ),
  links: import_zod.z.object({
    about: import_zod.z.string().url().describe(
      "A link that leads to further details about this particular occurrence of the problem."
    )
  }).strict().nullable().optional().describe("Relevant links about the error"),
  status: import_zod.z.string().optional().describe(
    "The HTTP status code applicable to this problem, expressed as a string value."
  ),
  code: ErrorCodeSchema.optional(),
  title: import_zod.z.string().optional().describe(
    "A short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of the problem, except for purposes of localization."
  ),
  detail: import_zod.z.string().optional().describe(
    "A human-readable explanation specific to this occurrence of the problem. Like `title`, this field's value can be localized."
  ),
  source: ErrorSourceSchema.optional()
}).strict().describe(
  "An error object provides additional information about problems encountered while performing an operation. Error objects MUST be returned as an array keyed by `errors` in the top level of a JSON:API document."
);
var ErrorsResponseSchema = import_zod.z.object({
  errors: import_zod.z.array(ErrorObjectSchema).describe("A collection of the errors returned.")
}).describe(
  "A JSON:API error response document containing an array of error objects."
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ErrorCodeSchema,
  ErrorObjectSchema,
  ErrorSourceSchema,
  ErrorsResponseSchema
});
