// @zapier/zapier-sdk-core/dist/v0/schemas/implementations.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/v0/schemas/implementations.ts
var implementations_exports = {};
__export(implementations_exports, {
  ImplementationMetaSchema: () => ImplementationMetaSchema,
  ImplementationsMetaResponseSchema: () => ImplementationsMetaResponseSchema
});
module.exports = __toCommonJS(implementations_exports);
var import_zod = require("zod");
var ImplementationMetaSchema = import_zod.z.object({
  id: import_zod.z.string().describe("Unique identifier for the implementation"),
  name: import_zod.z.string().describe("Display name of the implementation"),
  slug: import_zod.z.string().describe("URL-friendly slug identifier"),
  age_in_days: import_zod.z.number().optional().describe("Number of days since the implementation was created"),
  auth_type: import_zod.z.string().optional().describe("Authentication type (e.g., oauth2, api_key)"),
  banner: import_zod.z.string().optional().describe("Banner message or status indicator"),
  categories: import_zod.z.array(
    import_zod.z.object({
      id: import_zod.z.number().describe("Unique identifier for the category"),
      name: import_zod.z.string().describe("Display name of the category"),
      slug: import_zod.z.string().describe("URL-friendly slug for the category")
    })
  ).optional().describe("Categories the implementation belongs to"),
  images: import_zod.z.object({
    url_16x16: import_zod.z.string().optional().describe("16x16 pixel icon URL"),
    url_32x32: import_zod.z.string().optional().describe("32x32 pixel icon URL"),
    url_64x64: import_zod.z.string().optional().describe("64x64 pixel icon URL"),
    url_128x128: import_zod.z.string().optional().describe("128x128 pixel icon URL")
  }).optional().describe("Icon images at various sizes"),
  popularity: import_zod.z.number().optional().describe("Popularity score for ranking apps"),
  has_filters: import_zod.z.boolean().optional().describe("Whether the app has filter actions"),
  has_reads: import_zod.z.boolean().optional().describe("Whether the app has read actions"),
  has_searches: import_zod.z.boolean().optional().describe("Whether the app has search actions"),
  has_searches_or_writes: import_zod.z.boolean().optional().describe("Whether the app has search or write actions"),
  has_upfront_fields: import_zod.z.boolean().optional().describe("Whether the app has upfront input fields"),
  has_writes: import_zod.z.boolean().optional().describe("Whether the app has write actions"),
  is_beta: import_zod.z.boolean().optional().describe("Whether the app is in beta"),
  is_built_in: import_zod.z.boolean().optional().describe("Whether the app is a built-in Zapier app"),
  is_deprecated: import_zod.z.boolean().optional().describe("Whether the app is deprecated"),
  is_featured: import_zod.z.boolean().optional().describe("Whether the app is featured"),
  is_hidden: import_zod.z.boolean().optional().describe("Whether the app is hidden from listings"),
  is_invite: import_zod.z.boolean().optional().describe("Whether the app is invite-only"),
  is_premium: import_zod.z.boolean().optional().describe("Whether the app requires a premium plan"),
  is_public: import_zod.z.boolean().optional().describe("Whether the app is publicly available"),
  is_upcoming: import_zod.z.boolean().optional().describe("Whether the app is upcoming/not yet released"),
  version: import_zod.z.string().optional().describe("Version string of the app"),
  visibility: import_zod.z.string().optional().describe("Visibility status (e.g., public, private)"),
  actions: import_zod.z.object({
    read: import_zod.z.number().optional().describe("Number of read actions"),
    read_bulk: import_zod.z.number().optional().describe("Number of bulk read actions"),
    write: import_zod.z.number().optional().describe("Number of write actions"),
    search: import_zod.z.number().optional().describe("Number of search actions"),
    search_or_write: import_zod.z.number().optional().describe("Number of search-or-write actions"),
    search_and_write: import_zod.z.number().optional().describe("Number of search-and-write actions"),
    filter: import_zod.z.number().optional().describe("Number of filter actions")
  }).optional().describe("Count of available actions by type"),
  description: import_zod.z.string().optional().describe("Description of the app"),
  primary_color: import_zod.z.string().optional().describe("Primary brand color (hex)"),
  secondary_color: import_zod.z.string().optional().describe("Secondary brand color (hex)"),
  classification: import_zod.z.string().optional().describe("App classification category"),
  api_docs_url: import_zod.z.string().optional().describe("URL to API documentation"),
  image: import_zod.z.string().optional().describe("Default image URL for the app")
});
var ImplementationsMetaResponseSchema = import_zod.z.object({
  count: import_zod.z.number().describe("Total number of results"),
  next: import_zod.z.string().nullable().optional().describe("URL for the next page of results"),
  previous: import_zod.z.string().nullable().optional().describe("URL for the previous page of results"),
  results: import_zod.z.array(ImplementationMetaSchema).describe("Array of implementation metadata")
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ImplementationMetaSchema,
  ImplementationsMetaResponseSchema
});
