// @zapier/zapier-sdk/dist/plugins/listClientCredentials/index.js
import { ListClientCredentialsQuerySchema, } from "./schemas";
import { createPaginatedFunction } from "../../utils/function-utils";
import { DEFAULT_PAGE_SIZE } from "../../constants";
import { ClientCredentialsItemSchema } from "../../schemas/ClientCredentials";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { stripPageSuffix } from "../../utils/string-utils";
import { ZapierAuthenticationError } from "../../types/errors";
export const listClientCredentialsPlugin = ({ context }) => {
    async function listClientCredentialsPage(options) {
        const { api } = context;
        const searchParams = {};
        if (options.pageSize !== undefined) {
            searchParams.pageSize = options.pageSize.toString();
        }
        if (options.cursor) {
            searchParams.offset = options.cursor;
        }
        const response = await api.get("/api/v0/client-credentials", {
            searchParams,
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to access client credentials or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to list client credentials. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
            requiredScopes: ["credentials"],
        });
        return response;
    }
    const methodName = stripPageSuffix(listClientCredentialsPage.name);
    const listClientCredentialsDefinition = createPaginatedFunction(listClientCredentialsPage, ListClientCredentialsQuerySchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName), methodName, DEFAULT_PAGE_SIZE);
    return {
        listClientCredentials: listClientCredentialsDefinition,
        context: {
            meta: {
                listClientCredentials: {
                    categories: ["client-credentials"],
                    type: "list",
                    itemType: "ClientCredentials",
                    inputSchema: ListClientCredentialsQuerySchema,
                    outputSchema: ClientCredentialsItemSchema,
                },
            },
        },
    };
};
