// @zapier/zapier-sdk/dist/plugins/tables/updateTableRecords/index.js
import { createFunction } from "../../../utils/function-utils";
import { UpdateTableRecordsOptionsSchema, UpdateTableRecordsOptionsInputSchema, UpdateTableRecordsApiResponseSchema, RecordItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver, tableUpdateRecordsResolver, } from "../../../resolvers";
import { tableRecordFormatter } from "../../../formatters";
import { createFieldKeyTranslator, transformRecordItem, throwOnRecordErrors, throwOnResponseErrors, } from "../utils";
export const updateTableRecordsPlugin = ({ context }) => {
    async function updateTableRecords(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const translator = await createFieldKeyTranslator({
            api,
            tableId,
            keyMode: options.keyMode,
        });
        const rawResponse = await api.patch(`/tables/api/v1/tables/${tableId}/records`, {
            updated_records: options.records.map((record) => ({
                id: record.id,
                data: translator.translateInput(record.data),
            })),
        }, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to update records or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to update records. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 404) {
                    return new Error(`Table not found: ${tableId}`);
                }
                return undefined;
            },
            authRequired: true,
        });
        throwOnResponseErrors(rawResponse);
        const response = UpdateTableRecordsApiResponseSchema.parse(rawResponse);
        for (const changeset of response.data) {
            throwOnRecordErrors(changeset.new);
        }
        return {
            data: response.data.map((changeset) => ({
                ...transformRecordItem(changeset.new),
                data: translator.translateOutput(changeset.new.data),
            })),
        };
    }
    const updateTableRecordsDefinition = createFunction(updateTableRecords, UpdateTableRecordsOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, updateTableRecords.name));
    return {
        updateTableRecords: updateTableRecordsDefinition,
        context: {
            meta: {
                updateTableRecords: {
                    categories: ["table"],
                    type: "update",
                    itemType: "Record",
                    inputSchema: UpdateTableRecordsOptionsSchema,
                    outputSchema: RecordItemSchema,
                    resolvers: {
                        table: tableIdResolver,
                        records: tableUpdateRecordsResolver,
                    },
                    formatter: tableRecordFormatter,
                },
            },
        },
    };
};
