// @zapier/zapier-sdk/dist/plugins/tables/updateTableRecords/schemas.js
import { z } from "zod";
import { TablePropertySchema } from "../../../types/properties";
import { RecordApiItemSchema, RecordItemSchema, } from "../getTableRecord/schemas";
import { KeyModeSchema } from "../utils";
// Re-export for consumers
export { RecordItemSchema };
const RecordChangesetSchema = z.object({
    change_id: z.string(),
    old: RecordApiItemSchema.nullable(),
    new: RecordApiItemSchema,
    changed_field_ids: z.array(z.number()).nullable(),
    record_id: z.string(),
});
export const UpdateTableRecordsApiResponseSchema = z.object({
    data: z.array(RecordChangesetSchema),
});
export const UpdateRecordSchema = z.object({
    id: z.string().describe("The record ID to update"),
    data: z
        .record(z.string(), z.unknown())
        .describe("The field values to update, keyed by field key"),
});
const UpdateTableRecordsDescription = "Update one or more records in a table";
const UpdateTableRecordsBase = z.object({
    records: z
        .array(UpdateRecordSchema)
        .min(1)
        .max(100)
        .describe("Array of records to update (max 100)"),
    keyMode: KeyModeSchema,
});
// Schema for docs/registry (new param names only)
export const UpdateTableRecordsOptionsSchema = z
    .object({
    table: TablePropertySchema,
})
    .merge(UpdateTableRecordsBase)
    .describe(UpdateTableRecordsDescription)
    .meta({ aliases: { tableId: "table" } });
// Deprecated schema (backward compat)
const UpdateTableRecordsOptionsSchemaDeprecated = z
    .object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
})
    .merge(UpdateTableRecordsBase);
// Union for runtime validation
export const UpdateTableRecordsOptionsInputSchema = z
    .union([
    UpdateTableRecordsOptionsSchema,
    UpdateTableRecordsOptionsSchemaDeprecated,
])
    .describe(UpdateTableRecordsDescription);
