// @zapier/zapier-sdk-core/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AuthenticationItemSchema: () => AuthenticationItemSchema,
  AuthenticationSchema: () => AuthenticationSchema,
  ErrorCodeSchema: () => ErrorCodeSchema,
  ErrorObjectSchema: () => ErrorObjectSchema,
  ErrorSourceSchema: () => ErrorSourceSchema,
  ErrorsResponseSchema: () => ErrorsResponseSchema,
  GetAuthenticationParamSchema: () => GetAuthenticationParamSchema,
  openapiSpecPath: () => openapiSpecPath
});
module.exports = __toCommonJS(index_exports);

// src/v0/schemas/authentications.ts
var import_zod = require("zod");
var AuthenticationSchema = import_zod.z.object({
  id: import_zod.z.number().describe("Unique identifier for the authentication"),
  date: import_zod.z.string().describe("Date created"),
  lastchanged: import_zod.z.string().optional().describe("Date last changed"),
  account_id: import_zod.z.number().describe("Account ID associated with this authentication"),
  customuser_id: import_zod.z.number().optional().describe("Custom user ID (if applicable)"),
  selected_api: import_zod.z.string().describe("Selected API key (internal identifier)"),
  destination_selected_api: import_zod.z.string().nullable().optional().describe("Destination API key (if applicable)"),
  is_invite_only: import_zod.z.boolean().describe("Whether the authentication is invite-only"),
  is_private: import_zod.z.boolean().describe("Whether the authentication is private"),
  shared_with_all: import_zod.z.boolean().describe("Whether the authentication is shared with all users"),
  is_stale: import_zod.z.string().optional().describe("Stale status string"),
  is_shared: import_zod.z.string().optional().describe("Shared status string"),
  marked_stale_at: import_zod.z.string().nullable().optional().describe("Date when marked stale"),
  label: import_zod.z.string().nullable().optional().describe("User label for the authentication"),
  identifier: import_zod.z.string().nullable().optional().describe("Identifier"),
  title: import_zod.z.string().nullable().optional().describe("Title of the authentication"),
  url: import_zod.z.string().optional().describe("URL to the authentication resource"),
  groups: import_zod.z.array(
    import_zod.z.record(import_zod.z.string(), import_zod.z.unknown()).describe("Groups associated with the authentication")
  ).optional().describe("Array of groups associated with the authentication"),
  members: import_zod.z.string().optional().describe("Members associated with the authentication"),
  permissions: import_zod.z.record(import_zod.z.string(), import_zod.z.boolean()).optional().describe("Permissions for the authentication")
});
var AuthenticationItemSchema = AuthenticationSchema.omit({
  selected_api: true,
  customuser_id: true
}).extend({
  // Override numeric IDs with string versions (converted by normalizeAuthenticationItem)
  id: import_zod.z.string().describe("Unique identifier for the authentication"),
  account_id: import_zod.z.string().describe("Account ID associated with this authentication"),
  // Renamed fields
  implementation_id: import_zod.z.string().optional().describe("Implementation ID (was selected_api)"),
  profile_id: import_zod.z.string().optional().describe("Profile ID (was customuser_id)"),
  // Mapped fields (originals preserved in ...restOfAuth)
  is_expired: import_zod.z.string().optional().describe("Whether the authentication is expired (mapped from is_stale)"),
  expired_at: import_zod.z.string().nullable().optional().describe("Date when authentication expired (mapped from marked_stale_at)"),
  // Computed fields
  app_key: import_zod.z.string().optional().describe("App Key extracted from implementation_id"),
  app_version: import_zod.z.string().optional().describe("App Version extracted from implementation_id")
});
var AuthenticationsResponseSchema = import_zod.z.object({
  count: import_zod.z.number().describe("Total number of items"),
  next: import_zod.z.string().nullable().optional().describe("Cursor for the next page of results (if available)"),
  previous: import_zod.z.string().nullable().optional().describe("Cursor for the previous page of results (if available)"),
  results: import_zod.z.array(AuthenticationSchema).describe("Array of authentication items")
});
var GetAuthenticationParamSchema = import_zod.z.object({
  authenticationId: import_zod.z.string().describe("Authentication ID to retrieve")
}).describe("Get a specific authentication by ID");
var ListAuthenticationsQuerySchema = import_zod.z.object({
  appKey: import_zod.z.string().optional().describe(
    "Filter authentications by app key (e.g., 'SlackCLIAPI' or slug like 'github')"
  ),
  authenticationIds: import_zod.z.string().optional().describe("Comma-separated list of authentication IDs to filter by"),
  search: import_zod.z.string().optional().describe("Search term to filter authentications by title"),
  title: import_zod.z.string().optional().describe(
    "Filter authentications by exact title match (searches first, then filters locally)"
  ),
  accountId: import_zod.z.string().optional().describe("Filter authentications by account ID"),
  owner: import_zod.z.string().optional().describe(
    "Filter by owner, 'me' for your own authentications or a specific user ID"
  ),
  isExpired: import_zod.z.boolean().optional().describe(
    "Filter by expired status (true = expired only, false = non-expired only)"
  ),
  pageSize: import_zod.z.number().min(1).optional().describe("Number of authentications per page"),
  offset: import_zod.z.string().optional().describe("Pagination offset from previous response")
}).describe("Query parameters for listing authentications");
var ListAuthenticationsResponseSchema = import_zod.z.object({
  data: import_zod.z.array(AuthenticationItemSchema).describe("Array of authentication items"),
  links: import_zod.z.object({
    next: import_zod.z.string().nullable().optional().describe(
      "Fully qualified URL for the next page of results (if available), e.g. https://sdkapi.zapier.com/api/v0/authentications?offset=100&pageSize=50"
    )
  }).describe("Pagination links for navigating through results"),
  meta: import_zod.z.object({
    count: import_zod.z.number().describe("Total number of items"),
    limit: import_zod.z.number().describe("Number of items per page"),
    offset: import_zod.z.number().describe("Offset of the current page")
  }).describe("Metadata for the paginated result set")
}).describe("Response schema for listing authentications");

// src/v0/schemas/errors.ts
var import_zod2 = require("zod");
var ErrorCodeSchema = import_zod2.z.string().describe("An application-specific error code, expressed as a string value.");
var ErrorSourceSchema = import_zod2.z.object({
  pointer: import_zod2.z.string().optional().describe(
    "A JSON Pointer [RFC6901](https://tools.ietf.org/html/rfc6901) to the associated entity in the request document [e.g. `/data` for a primary data object, or `/data/attributes/title` for a specific attribute]."
  ),
  parameter: import_zod2.z.string().optional().describe(
    "A string indicating which URI query parameter caused the error."
  ),
  header: import_zod2.z.string().optional().describe("A string indicating the header that caused the error.")
}).strict().nullable().describe(
  "Identifies the source of the error within the request payload, if relevant."
);
var ErrorObjectSchema = import_zod2.z.object({
  id: import_zod2.z.string().nullable().optional().describe(
    "A unique identifier for this particular occurrence of the problem."
  ),
  links: import_zod2.z.object({
    about: import_zod2.z.string().url().describe(
      "A link that leads to further details about this particular occurrence of the problem."
    )
  }).strict().nullable().optional().describe("Relevant links about the error"),
  status: import_zod2.z.string().optional().describe(
    "The HTTP status code applicable to this problem, expressed as a string value."
  ),
  code: ErrorCodeSchema.optional(),
  title: import_zod2.z.string().optional().describe(
    "A short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of the problem, except for purposes of localization."
  ),
  detail: import_zod2.z.string().optional().describe(
    "A human-readable explanation specific to this occurrence of the problem. Like `title`, this field's value can be localized."
  ),
  source: ErrorSourceSchema.optional()
}).strict().describe(
  "An error object provides additional information about problems encountered while performing an operation. Error objects MUST be returned as an array keyed by `errors` in the top level of a JSON:API document."
);
var ErrorsResponseSchema = import_zod2.z.object({
  errors: import_zod2.z.array(ErrorObjectSchema).describe("A collection of the errors returned.")
}).describe(
  "A JSON:API error response document containing an array of error objects."
);

// src/index.ts
var import_node_url = require("url");
var import_node_path = require("path");
var import_meta = {};
var openapiSpecPath = (() => {
  if (typeof import_meta?.url === "string") {
    return (0, import_node_url.fileURLToPath)(new URL("../openapi.yaml", import_meta.url));
  }
  const currentDir = typeof __dirname !== "undefined" ? __dirname : (0, import_node_path.dirname)(require.resolve("@zapier/zapier-sdk-core/package.json"));
  return (0, import_node_path.resolve)(currentDir, "../openapi.yaml");
})();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AuthenticationItemSchema,
  AuthenticationSchema,
  ErrorCodeSchema,
  ErrorObjectSchema,
  ErrorSourceSchema,
  ErrorsResponseSchema,
  GetAuthenticationParamSchema,
  openapiSpecPath
});
