// @zapier/zapier-sdk/dist/plugins/createClientCredentials/index.js
import { CreateClientCredentialsSchema, } from "./schemas";
import { createFunction } from "../../utils/function-utils";
import { ClientCredentialsCreatedItemSchema } from "../../schemas/ClientCredentials";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../types/errors";
import { clientCredentialsNameResolver } from "../../resolvers";
export const createClientCredentialsPlugin = ({ context }) => {
    async function createClientCredentials(options) {
        const { api } = context;
        const response = await api.post("/api/v0/client-credentials", {
            name: options.name,
            allowed_scopes: options.allowedScopes,
        }, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to create client credentials or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to create client credentials. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
            requiredScopes: ["credentials"],
        });
        return {
            data: response.data,
        };
    }
    const createClientCredentialsDefinition = createFunction(createClientCredentials, CreateClientCredentialsSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, createClientCredentials.name));
    return {
        createClientCredentials: createClientCredentialsDefinition,
        context: {
            meta: {
                createClientCredentials: {
                    categories: ["client-credentials"],
                    type: "create",
                    itemType: "ClientCredentials",
                    inputSchema: CreateClientCredentialsSchema,
                    outputSchema: ClientCredentialsCreatedItemSchema,
                    confirm: "create-secret",
                    resolvers: {
                        name: clientCredentialsNameResolver,
                    },
                },
            },
        },
    };
};
