// @zapier/zapier-sdk-core/dist/v0/config/metadata.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/v0/config/metadata.ts
var metadata_exports = {};
__export(metadata_exports, {
  apiInfo: () => apiInfo,
  openApiMetadata: () => openApiMetadata,
  securitySchemes: () => securitySchemes,
  servers: () => servers,
  tags: () => tags
});
module.exports = __toCommonJS(metadata_exports);
var apiInfo = {
  title: "Zapier SDK API",
  version: "1.0.0",
  contact: {
    email: "engineering@zapier.com"
  },
  description: "The Zapier SDK API serves as a backend for the Zapier SDK."
};
var servers = [];
var tags = [
  {
    name: "Actions",
    description: "Action-related routes"
  },
  {
    name: "Apps",
    description: "App-related routes"
  },
  {
    name: "Authentications",
    description: "Authentication-related routes (deprecated, use Connections)"
  },
  {
    name: "Client Credentials",
    description: "Client credentials management routes"
  },
  {
    name: "Connections",
    description: "Connection-related routes"
  },
  {
    name: "Deduplication",
    description: "Deduplication-related routes"
  },
  {
    name: "Documentation",
    description: "Documentation-related routes"
  }
];
var securitySchemes = {
  userJwt: {
    type: "apiKey",
    in: "header",
    name: "Authorization",
    description: "Format should be `JWT <your-jwt>` (you must ensure that the JWT prefix is included in your requests).\\n\\nExample:  `Authorization: JWT your.jwt.value.here`"
  }
};
var openApiMetadata = {
  openapi: "3.1.0",
  info: apiInfo,
  servers,
  tags
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  apiInfo,
  openApiMetadata,
  securitySchemes,
  servers,
  tags
});
