// @zapier/zapier-sdk-core/dist/v0/config/metadata.js
// src/v0/config/metadata.ts
var apiInfo = {
  title: "Zapier SDK API",
  version: "1.0.0",
  contact: {
    email: "engineering@zapier.com"
  },
  description: "The Zapier SDK API serves as a backend for the Zapier SDK."
};
var servers = [];
var tags = [
  {
    name: "Actions",
    description: "Action-related routes"
  },
  {
    name: "Apps",
    description: "App-related routes"
  },
  {
    name: "Authentications",
    description: "Authentication-related routes (deprecated, use Connections)"
  },
  {
    name: "Client Credentials",
    description: "Client credentials management routes"
  },
  {
    name: "Connections",
    description: "Connection-related routes"
  },
  {
    name: "Deduplication",
    description: "Deduplication-related routes"
  },
  {
    name: "Documentation",
    description: "Documentation-related routes"
  }
];
var securitySchemes = {
  userJwt: {
    type: "apiKey",
    in: "header",
    name: "Authorization",
    description: "Format should be `JWT <your-jwt>` (you must ensure that the JWT prefix is included in your requests).\\n\\nExample:  `Authorization: JWT your.jwt.value.here`"
  }
};
var openApiMetadata = {
  openapi: "3.1.0",
  info: apiInfo,
  servers,
  tags
};
export {
  apiInfo,
  openApiMetadata,
  securitySchemes,
  servers,
  tags
};
