// @zapier/zapier-sdk-cli-login/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AUTH_MODE_HEADER: () => AUTH_MODE_HEADER,
  ZapierAuthenticationError: () => ZapierAuthenticationError,
  getAuthAuthorizeUrl: () => getAuthAuthorizeUrl,
  getAuthTokenUrl: () => getAuthTokenUrl,
  getConfig: () => getConfig,
  getConfigPath: () => getConfigPath,
  getLoggedInUser: () => getLoggedInUser,
  getLoginStorageMode: () => getLoginStorageMode,
  getPkceLoginConfig: () => getPkceLoginConfig,
  getToken: () => getToken,
  logout: () => logout,
  unloadConfig: () => unloadConfig,
  updateLogin: () => updateLogin
});
module.exports = __toCommonJS(index_exports);
var import_conf = __toESM(require("conf"));
var import_fs = require("fs");
var jwt = __toESM(require("jsonwebtoken"));

// src/keychain.ts
var import_cross_keychain = require("cross-keychain");
var SERVICE = "zapier-sdk-cli";
var ACCOUNT = "login";
var cachedBackendInfo;
async function getBackendInfo() {
  if (!cachedBackendInfo) {
    const keyring = await (0, import_cross_keychain.getKeyring)();
    cachedBackendInfo = `${keyring.name} (${keyring.id})`;
  }
  return cachedBackendInfo;
}
var keychainQueue = Promise.resolve();
function enqueue(fn) {
  const result = keychainQueue.then(fn, fn);
  keychainQueue = result.then(
    () => {
    },
    () => {
    }
  );
  return result;
}
async function getTokensFromKeychain({
  debugLog
} = {}) {
  return enqueue(async () => {
    const backendInfo = await getBackendInfo();
    debugLog?.(`Keychain read via ${backendInfo}`);
    const startTime = Date.now();
    const raw = await (0, import_cross_keychain.getPassword)(SERVICE, ACCOUNT);
    debugLog?.(`Keychain read completed in ${Date.now() - startTime}ms`);
    if (!raw) {
      debugLog?.("Keychain returned no data");
      return void 0;
    }
    let parsed;
    try {
      parsed = JSON.parse(raw);
    } catch {
      debugLog?.("Keychain data is not valid JSON");
      return void 0;
    }
    if (typeof parsed.login_jwt === "string" && typeof parsed.login_refresh_token === "string") {
      return {
        login_jwt: parsed.login_jwt,
        login_refresh_token: parsed.login_refresh_token
      };
    }
    debugLog?.("Keychain data has invalid shape", parsed);
    return void 0;
  });
}
async function setTokensInKeychain({
  data,
  debugLog
}) {
  return enqueue(async () => {
    const backendInfo = await getBackendInfo();
    debugLog?.(`Keychain write via ${backendInfo}`);
    const startTime = Date.now();
    await (0, import_cross_keychain.setPassword)(SERVICE, ACCOUNT, JSON.stringify(data));
    debugLog?.(`Keychain write completed in ${Date.now() - startTime}ms`);
  });
}
async function clearTokensFromKeychain({
  debugLog
} = {}) {
  return enqueue(async () => {
    try {
      const backendInfo = await getBackendInfo();
      debugLog?.(`Keychain clear via ${backendInfo}`);
      await (0, import_cross_keychain.deletePassword)(SERVICE, ACCOUNT);
    } catch {
    }
  });
}

// src/index.ts
var ZapierAuthenticationError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "ZapierAuthenticationError";
  }
};
var config = null;
var DEFAULT_AUTH_CLIENT_ID = "grwWZD5hUWGvb4V8ODBuOtXer3h0DBEZ2HR8aay6";
var TOKEN_REFRESH_BUFFER_MS = 5 * 60 * 1e3;
function createDebugLog(enabled) {
  if (!enabled) {
    return () => {
    };
  }
  return (message, data) => {
    if (data === void 0) {
      console.log(`[Zapier SDK CLI Login] ${message}`);
    } else {
      console.log(`[Zapier SDK CLI Login] ${message}`, data);
    }
  };
}
function censorHeaderValue(value) {
  if (value.length > 12) {
    return `${value.substring(0, 4)}...${value.substring(value.length - 4)}`;
  }
  return `${value.charAt(0)}...`;
}
function getAuthClientId(clientId) {
  return clientId || DEFAULT_AUTH_CLIENT_ID;
}
var AUTH_MODE_HEADER = "X-Auth";
var DEFAULT_AUTH_BASE_URL = "https://zapier.com";
function getAuthTokenUrl(options) {
  const authBaseUrl = options?.baseUrl || DEFAULT_AUTH_BASE_URL;
  return `${authBaseUrl}/oauth/token/`;
}
function getAuthAuthorizeUrl(options) {
  const authBaseUrl = options?.baseUrl || DEFAULT_AUTH_BASE_URL;
  return `${authBaseUrl}/oauth/authorize/`;
}
function getPkceLoginConfig(options) {
  return {
    clientId: getAuthClientId(options?.credentials?.clientId),
    tokenUrl: getAuthTokenUrl({ baseUrl: options?.credentials?.baseUrl }),
    authorizeUrl: getAuthAuthorizeUrl({
      baseUrl: options?.credentials?.baseUrl
    })
  };
}
var cachedLogin;
function getConfig() {
  if (!config) {
    config = new import_conf.default({ projectName: "zapier-sdk-cli" });
    if (!config.has("login_storage_mode")) {
      config.set(
        "login_storage_mode",
        (0, import_fs.existsSync)(config.path) ? "config" : "keychain"
      );
    }
  }
  return config;
}
function unloadConfig() {
  config = null;
  cachedLogin = void 0;
}
async function updateLogin(loginData, options = {}) {
  const debugLog = createDebugLog(options.debug ?? false);
  const storage = options.storage ?? cachedLogin?.storage ?? "keychain";
  const expiresAt = Date.now() + loginData.expires_in * 1e3;
  const cfg = getConfig();
  cfg.set("login_storage_mode", storage);
  if (storage === "keychain") {
    await setTokensInKeychain({
      data: {
        login_jwt: loginData.access_token,
        login_refresh_token: loginData.refresh_token
      },
      debugLog
    });
    cfg.set("login_expires_at", expiresAt);
    cfg.delete("login_jwt");
    cfg.delete("login_refresh_token");
  } else {
    cfg.set("login_jwt", loginData.access_token);
    cfg.set("login_refresh_token", loginData.refresh_token);
    cfg.set("login_expires_at", expiresAt);
    await clearTokensFromKeychain({ debugLog });
  }
  cachedLogin = {
    jwt: loginData.access_token,
    refreshToken: loginData.refresh_token,
    expiresAt,
    storage
  };
}
function decodeJwtOrThrow(token) {
  if (typeof token !== "string") {
    throw new Error("Expected JWT to be a string");
  }
  const decodedJwt = jwt.decode(token, { complete: true });
  if (!decodedJwt) {
    throw new Error("Could not decode JWT");
  }
  if (typeof decodedJwt.payload === "string") {
    throw new Error("Did not expect JWT payload to be a string");
  }
  return decodedJwt;
}
async function refreshJwt(refreshToken, options = {}) {
  const {
    onEvent,
    fetch = globalThis.fetch,
    credentials,
    debug = false
  } = options;
  const debugLog = createDebugLog(debug);
  const tokenUrl = getAuthTokenUrl({ baseUrl: credentials?.baseUrl });
  const clientId = getAuthClientId(credentials?.clientId);
  const startTime = Date.now();
  try {
    onEvent?.({
      type: "auth_refreshing",
      payload: {
        message: "Refreshing your token...",
        operation: "token_refresh"
      },
      timestamp: Date.now()
    });
    debugLog(`\u2192 POST ${tokenUrl}`, {
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        [AUTH_MODE_HEADER]: "no"
      },
      body: {
        client_id: clientId,
        refresh_token: censorHeaderValue(refreshToken),
        grant_type: "refresh_token"
      }
    });
    const response = await fetch(tokenUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        [AUTH_MODE_HEADER]: "no"
      },
      body: new URLSearchParams({
        client_id: clientId,
        refresh_token: refreshToken,
        grant_type: "refresh_token"
      })
    });
    const duration = Date.now() - startTime;
    if (!response.ok) {
      debugLog(`\u2190 ${response.status} ${response.statusText} (${duration}ms)`);
      throw new Error(
        `Token refresh failed: ${response.status} ${response.statusText}`
      );
    }
    debugLog(`\u2190 ${response.status} ${response.statusText} (${duration}ms)`);
    const data = await response.json();
    await updateLogin(data, { debug });
    debugLog(
      `Token refreshed and saved to ${cachedLogin?.storage ?? "keychain"}`
    );
    onEvent?.({
      type: "auth_success",
      payload: {
        message: "Token refreshed successfully",
        operation: "token_refresh"
      },
      timestamp: Date.now()
    });
    return data.access_token;
  } catch (error) {
    const duration = Date.now() - startTime;
    debugLog(`\u2716 Token refresh failed (${duration}ms)`, {
      error: error instanceof Error ? error.message : error
    });
    cachedLogin = void 0;
    const errorMessage = `Token refresh failed: ${error instanceof Error ? error.message : "Unknown error"}`;
    onEvent?.({
      type: "auth_error",
      payload: {
        message: errorMessage,
        error: errorMessage,
        operation: "token_refresh"
      },
      timestamp: Date.now()
    });
    throw error;
  }
}
var pendingRefresh = null;
var pendingResolve = null;
async function resolveStoredLogin(debugLog) {
  if (cachedLogin) {
    debugLog("Using in-memory cached credentials");
    return cachedLogin;
  }
  if (pendingResolve) {
    debugLog("Waiting for existing keychain read to complete");
    return pendingResolve;
  }
  pendingResolve = resolveStoredLoginFromStorage(debugLog).finally(() => {
    pendingResolve = null;
  });
  return pendingResolve;
}
async function resolveStoredLoginFromStorage(debugLog) {
  let cfg;
  try {
    cfg = getConfig();
  } catch (error) {
    debugLog("Failed to load config", {
      error: error instanceof Error ? error.message : error
    });
    return void 0;
  }
  const expiresAt = cfg.get("login_expires_at");
  const configJwt = cfg.get("login_jwt");
  const configRefresh = cfg.get("login_refresh_token");
  if (configJwt && configRefresh && typeof expiresAt === "number") {
    debugLog("Loaded credentials from config (legacy format)");
    cachedLogin = {
      jwt: configJwt,
      refreshToken: configRefresh,
      expiresAt,
      storage: "config"
    };
    return cachedLogin;
  }
  if (typeof expiresAt !== "number") {
    debugLog("No stored login credentials found");
    return void 0;
  }
  const keychainData = await getTokensFromKeychain({ debugLog });
  if (!keychainData) {
    debugLog("No tokens found in keychain");
    return void 0;
  }
  debugLog("Loaded credentials from keychain");
  cachedLogin = {
    jwt: keychainData.login_jwt,
    refreshToken: keychainData.login_refresh_token,
    expiresAt,
    storage: "keychain"
  };
  return cachedLogin;
}
async function resolveOrRefreshToken(options = {}) {
  const { debug = false } = options;
  const debugLog = createDebugLog(debug);
  const stored = await resolveStoredLogin(debugLog);
  if (!stored) {
    return void 0;
  }
  const { jwt: storedJwt, refreshToken, expiresAt } = stored;
  if (expiresAt > Date.now() + TOKEN_REFRESH_BUFFER_MS) {
    debugLog("Using cached token (still valid)");
    return storedJwt;
  }
  debugLog("Token expired, refreshing...");
  if (pendingRefresh) {
    debugLog("Waiting for existing refresh to complete");
    return pendingRefresh;
  }
  pendingRefresh = refreshJwt(refreshToken, options).finally(() => {
    pendingRefresh = null;
  });
  return await pendingRefresh;
}
async function getToken(options = {}) {
  try {
    return await resolveOrRefreshToken(options);
  } catch (error) {
    const message = error instanceof Error ? error.message : "Token refresh failed";
    throw new ZapierAuthenticationError(
      `${message}
Please run 'login' to authenticate again.`
    );
  }
}
async function getLoggedInUser(options = {}) {
  const jwt2 = await getToken(options).catch(() => void 0);
  if (!jwt2) {
    throw new Error(
      "No valid authentication token available. Please login first."
    );
  }
  let decodedJwt = decodeJwtOrThrow(jwt2);
  if (decodedJwt.payload["sub_type"] == "service") {
    decodedJwt = decodeJwtOrThrow(decodedJwt.payload["njwt"]);
  }
  if (typeof decodedJwt.payload["zap:acc"] !== "string") {
    throw new Error("JWT payload does not contain accountId");
  }
  const accountId = parseInt(decodedJwt.payload["zap:acc"], 10);
  if (isNaN(accountId)) {
    throw new Error("JWT accountId is not a number");
  }
  if (decodedJwt.payload["sub_type"] !== "customuser" || typeof decodedJwt.payload["sub"] !== "string") {
    throw new Error("JWT payload does not contain customUserId");
  }
  const customUserId = parseInt(decodedJwt.payload["sub"], 10);
  if (isNaN(customUserId)) {
    throw new Error("JWT customUserId is not a number");
  }
  const email = decodedJwt.payload["zap:uname"];
  if (typeof email !== "string") {
    throw new Error("JWT payload does not contain email");
  }
  return {
    accountId,
    customUserId,
    email
  };
}
function getLoginStorageMode() {
  const cfg = getConfig();
  if (typeof cfg.get("login_jwt") === "string") {
    return "config";
  }
  const explicitMode = cfg.get("login_storage_mode");
  if (explicitMode === "keychain" || explicitMode === "config") {
    return explicitMode;
  }
  return "keychain";
}
async function logout(options = {}) {
  const { onEvent } = options;
  const mode = getLoginStorageMode();
  cachedLogin = void 0;
  await clearTokensFromKeychain();
  const cfg = getConfig();
  cfg.set("login_storage_mode", mode);
  cfg.delete("login_expires_at");
  cfg.delete("login_jwt");
  cfg.delete("login_refresh_token");
  onEvent?.({
    type: "auth_logout",
    payload: { message: "Logged out successfully", operation: "logout" },
    timestamp: Date.now()
  });
}
function getConfigPath() {
  const cfg = getConfig();
  return cfg.path;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AUTH_MODE_HEADER,
  ZapierAuthenticationError,
  getAuthAuthorizeUrl,
  getAuthTokenUrl,
  getConfig,
  getConfigPath,
  getLoggedInUser,
  getLoginStorageMode,
  getPkceLoginConfig,
  getToken,
  logout,
  unloadConfig,
  updateLogin
});
