// @zapier/zapier-sdk-cli/dist/src/telemetry/builders.js
/**
 * CLI-specific event builders
 *
 * Provides builder functions for CLI command telemetry that auto-populate
 * common CLI fields and system information.
 */
import { generateEventId, getCurrentTimestamp, getReleaseId, getOsInfo, getPlatformVersions, isCi, getCiPlatform, } from "@zapier/zapier-sdk";
import cliPackageJson from "../../package.json";
// Create base event for CLI events
function createCliBaseEvent(context = {}) {
    return {
        event_id: generateEventId(),
        timestamp_ms: getCurrentTimestamp(),
        release_id: getReleaseId(),
        customuser_id: context.customuser_id ?? null,
        account_id: context.account_id ?? null,
        identity_id: context.identity_id ?? null,
        visitor_id: context.visitor_id ?? null,
        correlation_id: context.correlation_id ?? null,
    };
}
export function buildCliCommandExecutedEvent({ data, context = {}, cliVersion = cliPackageJson.version, }) {
    const osInfo = getOsInfo();
    const platformVersions = getPlatformVersions();
    return {
        ...createCliBaseEvent(context),
        system_name: "zapier-sdk-cli",
        session_id: context.session_id ?? null,
        cli_version: data.cli_version ?? cliVersion,
        cli_arguments: data.cli_arguments ?? null,
        cli_primary_command: data.cli_primary_command,
        os_platform: osInfo.platform,
        os_release: osInfo.release,
        os_architecture: osInfo.architecture,
        platform_versions: platformVersions,
        selected_api: context.selected_api ?? null,
        app_id: context.app_id ?? null,
        app_version_id: context.app_version_id ?? null,
        execution_duration_ms: data.execution_duration_ms ?? null,
        success_flag: data.success_flag,
        exit_code: data.exit_code ?? (data.success_flag ? 0 : 1),
        error_message: data.error_message ?? null,
        command_category: data.command_category ?? null,
        requires_auth: data.requires_auth ?? null,
        is_ci_environment: isCi(),
        ci_platform: getCiPlatform(),
        package_manager: data.package_manager ?? "pnpm", // Default based on project setup
        made_network_requests: data.made_network_requests ?? null,
        files_modified_count: data.files_modified_count ?? null,
        files_created_count: data.files_created_count ?? null,
        files_processed_size_bytes: data.files_processed_size_bytes ?? null,
        cpu_time_ms: data.cpu_time_ms ?? null,
        subprocess_count: data.subprocess_count ?? null,
    };
}
