// @zapier/zapier-sdk/dist/plugins/request/index.js
import { RelayRequestSchema } from "./schemas";
import { createFunction } from "../../utils/function-utils";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { logDeprecation } from "../../utils/logging";
/**
 * @deprecated Use `sdk.fetch(url, init?)` instead. This plugin now delegates to the fetch plugin
 * and is kept only for backward compatibility. It is hidden from the CLI/MCP registries.
 */
export const requestPlugin = ({ sdk, context }) => {
    async function request(options) {
        logDeprecation("request() is deprecated. Use fetch() instead.");
        const { url, method, body, headers, connection, connectionId, authenticationId, callbackUrl, } = options;
        return sdk.fetch(url, {
            method,
            body: body,
            headers: headers,
            connection: connection ?? connectionId ?? authenticationId,
            callbackUrl,
        });
    }
    const requestDefinition = createFunction(request, RelayRequestSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, request.name));
    return {
        request: requestDefinition,
        context: {
            meta: {
                request: {
                    packages: ["cli", "mcp"],
                    categories: ["http", "deprecated"],
                    returnType: "Response",
                    inputSchema: RelayRequestSchema,
                },
            },
        },
    };
};
