// @zapier/zapier-sdk/dist/plugins/request/schemas.js
import { z } from "zod";
import { ConnectionIdPropertySchema, ConnectionPropertySchema, AuthenticationIdPropertySchema, } from "../../types/properties";
// Pure Zod schema for the request function
export const RelayRequestSchema = z
    .object({
    url: z
        .string()
        .url()
        .describe("The URL to request (will be proxied through Relay)"),
    method: z
        .enum(["GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"])
        .optional()
        .describe("HTTP method"),
    body: z.any().optional().describe("Request body as a string"),
    connection: ConnectionPropertySchema.optional(),
    connectionId: ConnectionIdPropertySchema.optional().meta({
        deprecated: true,
    }),
    /** @deprecated Use connection instead */
    authenticationId: AuthenticationIdPropertySchema.optional().meta({
        deprecated: true,
    }),
    callbackUrl: z
        .string()
        .url()
        .optional()
        .describe("URL to send async response to (makes request async)"),
    headers: z
        .union([
        z.record(z.string(), z.string()),
        z.instanceof(Headers),
        z.array(z.tuple([z.string(), z.string()])),
    ])
        .optional()
        .describe("Request headers"),
})
    .describe("Make authenticated HTTP requests through Zapier's Relay service")
    .meta({
    aliases: { connectionId: "connection", authenticationId: "connection" },
});
// Legacy exports for compatibility
export const RelayFetchSchema = RelayRequestSchema;
