// @zapier/zapier-sdk/dist/plugins/manifest/schemas.js
import { z } from "zod";
import { ConnectionIdPropertySchema, AuthenticationIdPropertySchema, } from "../../types/properties";
import { ConnectionsMapSchema, } from "../../types/connections";
export const DEFAULT_CONFIG_PATH = ".zapierrc";
/**
 * Action entry for storing saved action configurations
 * The key in the actions record IS the user-provided name
 */
export const ActionEntrySchema = z.object({
    appKey: z.string().describe("App key (slug or implementation name)"),
    actionKey: z.string().describe("Action key identifier"),
    actionType: z
        .string()
        .describe("Action type (e.g., 'read', 'write', 'search')"),
    connectionId: ConnectionIdPropertySchema.nullable().optional(),
    /** @deprecated Use connection instead */
    authenticationId: AuthenticationIdPropertySchema.nullable()
        .optional()
        .meta({ deprecated: true }),
    inputs: z
        .record(z.string(), z.unknown())
        .optional()
        .describe("Resolved input values"),
    schema: z
        .record(z.string(), z.unknown())
        .describe("Complete JSON Schema from getInputFieldsSchema (includes $schema, type, properties, required, etc.)"),
    createdAt: z.string().describe("ISO 8601 timestamp when created"),
});
/**
 * Manifest schema for version locking and saved action configurations
 */
export const ManifestSchema = z
    .object({
    apps: z
        .record(z.string(), z.object({
        implementationName: z
            .string()
            .describe("Base implementation name without version (e.g., 'SlackCLIAPI')"),
        version: z.string().describe("Version string (e.g., '1.21.1')"),
    }))
        .optional(),
    actions: z
        .record(z.string(), ActionEntrySchema)
        .optional()
        .describe("Saved action configurations with their schemas"),
    canIncludeSharedConnections: z
        .boolean()
        .optional()
        .describe("Allow listing shared connections"),
    canIncludeSharedTables: z
        .boolean()
        .optional()
        .describe("Allow listing shared tables"),
    canDeleteTables: z.boolean().optional().describe("Allow deleting tables"),
    connections: ConnectionsMapSchema.optional().describe("Named connections map. Keys are connection names, values contain a connectionId."),
})
    .describe("Manifest for app version locking, action configurations, and named connections");
export const ManifestPluginOptionsSchema = z.object({
    manifestPath: z.string().optional().describe("Path to manifest file"),
    manifest: ManifestSchema.optional().describe("Direct manifest object"),
});
