// @zapier/zapier-sdk/dist/plugins/eventEmission/transport.js
/**
 * Transport abstraction layer for event emission
 *
 * Provides configurable transport mechanisms for emitting telemetry events.
 * All transports implement silent failure to prevent SDK disruption.
 */
// Constants for transport configuration
const DEFAULT_RETRY_ATTEMPTS = 2;
const DEFAULT_RETRY_DELAY_MS = 300;
// HTTP Transport - sends events to remote endpoint
export function createHttpTransport(config) {
    const delay = async (ms) => {
        return new Promise((resolve) => {
            const timer = setTimeout(resolve, ms);
            if (typeof timer.unref === "function") {
                timer.unref();
            }
        });
    };
    const emitWithRetry = async (subject, event, attemptsLeft) => {
        try {
            const payload = {
                subject,
                properties: event,
            };
            const response = await fetch(config.endpoint, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    ...config.headers,
                },
                body: JSON.stringify(payload),
            });
            if (!response.ok && attemptsLeft > 1) {
                await delay(config.retryDelayMs || DEFAULT_RETRY_DELAY_MS);
                return emitWithRetry(subject, event, attemptsLeft - 1);
            }
        }
        catch (error) {
            if (attemptsLeft > 1) {
                await delay(config.retryDelayMs || DEFAULT_RETRY_DELAY_MS);
                return emitWithRetry(subject, event, attemptsLeft - 1);
            }
            throw error;
        }
    };
    return {
        async emit(subject, event) {
            try {
                await emitWithRetry(subject, event, config.retryAttempts || DEFAULT_RETRY_ATTEMPTS);
            }
            catch {
                // Silent failure - never throw
            }
        },
    };
}
// Console Transport - logs events to console (for development)
export function createConsoleTransport() {
    return {
        async emit(subject, event) {
            try {
                console.log("[SDK Telemetry]", JSON.stringify({ subject, properties: event }, null, 2));
            }
            catch {
                // Silent failure - never throw
            }
        },
    };
}
// No-op Transport - discards all events (for testing/disabled state)
export function createNoopTransport() {
    return {
        async emit(_subject, _event) {
            // Intentionally do nothing
        },
    };
}
// Transport factory
export function createTransport(config) {
    try {
        switch (config.type) {
            case "http":
                if (!config.endpoint) {
                    throw new Error("HTTP transport requires endpoint");
                }
                return createHttpTransport({
                    endpoint: config.endpoint,
                    headers: config.headers,
                    retryAttempts: config.retryAttempts,
                    retryDelayMs: config.retryDelayMs,
                });
            case "console":
                return createConsoleTransport();
            case "noop":
            default:
                return createNoopTransport();
        }
    }
    catch {
        // If transport creation fails, return noop to maintain silent operation
        return createNoopTransport();
    }
}
