// @zapier/zapier-sdk/dist/plugins/eventEmission/utils.js
/**
 * Simple utility functions for event emission
 * These are pure functions that can be used to populate common event fields
 */
// Intentional dynamic require for browser compatibility. The os module is only
// available in Node; in browser environments this gracefully falls back to nulls.
let osModule = null;
try {
    osModule = require("os");
}
catch {
    osModule = null;
}
/**
 * Generate a unique event ID
 */
export function generateEventId() {
    return crypto.randomUUID();
}
/**
 * Get current timestamp in milliseconds since epoch
 */
export function getCurrentTimestamp() {
    return Date.now();
}
/**
 * Get release ID (git SHA) - in production this would come from build process
 */
export function getReleaseId() {
    return globalThis.process?.env?.SDK_RELEASE_ID || "development";
}
/**
 * Get operating system information
 */
export function getOsInfo() {
    if (!osModule) {
        return { platform: null, release: null, architecture: null };
    }
    try {
        return {
            platform: osModule.platform() || null,
            release: osModule.release() || null,
            architecture: osModule.arch() || null,
        };
    }
    catch {
        return { platform: null, release: null, architecture: null };
    }
}
/**
 * Get platform versions (Node.js, npm, etc.)
 */
export function getPlatformVersions() {
    const versions = {};
    if (typeof globalThis.process?.versions === "object") {
        for (const [key, value] of Object.entries(globalThis.process.versions)) {
            versions[key] = value || null;
        }
    }
    return versions;
}
/**
 * Check if running in CI environment
 */
export function isCi() {
    return !!(globalThis.process?.env?.CI ||
        globalThis.process?.env?.CONTINUOUS_INTEGRATION ||
        globalThis.process?.env?.GITHUB_ACTIONS ||
        globalThis.process?.env?.JENKINS_URL ||
        globalThis.process?.env?.GITLAB_CI ||
        globalThis.process?.env?.CIRCLECI ||
        globalThis.process?.env?.TRAVIS ||
        globalThis.process?.env?.BUILDKITE ||
        globalThis.process?.env?.DRONE ||
        globalThis.process?.env?.BITBUCKET_PIPELINES_UUID);
}
/**
 * Get CI platform name if running in CI
 */
export function getCiPlatform() {
    if (globalThis.process?.env?.GITHUB_ACTIONS)
        return "github-actions";
    if (globalThis.process?.env?.JENKINS_URL)
        return "jenkins";
    if (globalThis.process?.env?.GITLAB_CI)
        return "gitlab-ci";
    if (globalThis.process?.env?.CIRCLECI)
        return "circleci";
    if (globalThis.process?.env?.TRAVIS)
        return "travis";
    if (globalThis.process?.env?.BUILDKITE)
        return "buildkite";
    if (globalThis.process?.env?.DRONE)
        return "drone";
    if (globalThis.process?.env?.BITBUCKET_PIPELINES_UUID)
        return "bitbucket-pipelines";
    if (globalThis.process?.env?.CI ||
        globalThis.process?.env?.CONTINUOUS_INTEGRATION)
        return "unknown-ci";
    return null;
}
/**
 * Get memory usage in bytes
 */
export function getMemoryUsage() {
    if (globalThis.process?.memoryUsage) {
        const usage = globalThis.process.memoryUsage();
        return usage.rss || null; // Resident Set Size
    }
    return null;
}
/**
 * Get CPU time in milliseconds
 */
export function getCpuTime() {
    if (globalThis.process?.cpuUsage) {
        const usage = globalThis.process.cpuUsage();
        return Math.round((usage.user + usage.system) / 1000); // Convert to milliseconds
    }
    return null;
}
