// @zapier/zapier-sdk/dist/plugins/listActions/index.js
import { normalizeActionItem } from "../../utils/domain-utils";
import { ListActionsSchema, ListActionsInputSchema, } from "./schemas";
import { ActionItemSchema } from "../../schemas/Action";
import { ZapierConfigurationError, ZapierAuthenticationError, } from "../../types/errors";
import { createPaginatedFunction } from "../../utils/function-utils";
import { DEFAULT_PAGE_SIZE } from "../../constants";
import { appKeyResolver, actionTypeResolver } from "../../resolvers";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { stripPageSuffix } from "../../utils/string-utils";
import { setMethodMetadata } from "../../utils/telemetry-context";
export const listActionsPlugin = ({ context }) => {
    const methodMeta = {
        categories: ["action"],
        type: "list",
        itemType: "Action",
        inputSchema: ListActionsSchema,
        outputSchema: ActionItemSchema,
        resolvers: {
            app: appKeyResolver,
            actionType: actionTypeResolver,
        },
    };
    async function listActionsPage(options) {
        const { api, getVersionedImplementationId } = context;
        const appKey = "app" in options ? options.app : options.appKey;
        // Use getVersionedImplementationId (optimized to check manifest first)
        const selectedApi = await getVersionedImplementationId(appKey);
        if (!selectedApi) {
            throw new ZapierConfigurationError("No current_implementation_id found for app", { configType: "current_implementation_id" });
        }
        // operationKey is intentionally omitted — listActions is a list-level operation with no specific action key.
        setMethodMetadata({
            selectedApi,
            operationType: options.actionType ?? null,
        });
        const searchParams = {
            global: "true",
            public_only: "true",
            selected_apis: selectedApi,
        };
        const data = await api.get("/zapier/api/v4/implementations/", {
            searchParams,
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to access implementations or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to list implementations. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
        });
        let allActions = [];
        for (const implementation of data.results || []) {
            if (implementation.actions) {
                for (const action of implementation.actions) {
                    const actionWithContext = {
                        ...action,
                        selected_api: action.selected_api || implementation.selected_api,
                    };
                    allActions.push(normalizeActionItem(actionWithContext));
                }
            }
        }
        if (options.actionType) {
            allActions = allActions.filter((action) => action.action_type === options.actionType);
        }
        return {
            data: allActions,
            nextCursor: undefined,
        };
    }
    const methodName = stripPageSuffix(listActionsPage.name);
    const listActionsDefinition = createPaginatedFunction(listActionsPage, ListActionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName), methodName, DEFAULT_PAGE_SIZE);
    return {
        listActions: listActionsDefinition,
        context: {
            meta: {
                listActions: methodMeta,
            },
        },
    };
};
