// @zapier/zapier-sdk/dist/plugins/listActions/schemas.js
import { z } from "zod";
import { AppKeyPropertySchema, AppPropertySchema, ActionTypePropertySchema, } from "../../types/properties";
const ListActionsDescription = "List all actions for a specific app";
// Base schema for shared optional properties
const ListActionsBaseSchema = z.object({
    pageSize: z.number().min(1).optional().describe("Number of actions per page"),
    maxItems: z
        .number()
        .min(1)
        .optional()
        .describe("Maximum total items to return across all pages"),
    cursor: z.string().optional().describe("Cursor to start from"),
});
// Schema for docs/registry (new param names only)
export const ListActionsSchema = z
    .object({
    app: AppPropertySchema.describe("App key of actions to list (e.g., 'SlackCLIAPI' or slug like 'github')"),
    actionType: ActionTypePropertySchema.optional().describe("Filter actions by type"),
})
    .merge(ListActionsBaseSchema)
    .describe(ListActionsDescription)
    .meta({ aliases: { appKey: "app" } });
// Deprecated schema (backward compat)
const ListActionsSchemaDeprecated = z
    .object({
    appKey: AppKeyPropertySchema.describe("App key of actions to list (e.g., 'SlackCLIAPI' or slug like 'github')"),
    actionType: ActionTypePropertySchema.optional().describe("Filter actions by type"),
})
    .merge(ListActionsBaseSchema);
// Union for runtime validation (accepts either new or old names)
export const ListActionsInputSchema = z
    .union([ListActionsSchema, ListActionsSchemaDeprecated])
    .describe(ListActionsDescription);
