// @zapier/zapier-sdk/dist/define/define.js
function toJsonSchema(schema) {
    return schema.toJSONSchema();
}
function buildDescribe(meta) {
    return () => ({
        kind: meta.kind,
        name: meta.name,
        description: meta.description,
        inputSchema: meta.inputSchema ? toJsonSchema(meta.inputSchema) : undefined,
        connection: meta.connection,
        connections: meta.connections,
    });
}
export function defineTool(config) {
    const execute = () => {
        throw new Error(`defineTool("${config.name}") is not callable yet. ` +
            "Runtime resolution is coming in Phase 2.");
    };
    Object.defineProperty(execute, "name", {
        value: config.name,
        writable: false,
    });
    return Object.assign(execute, {
        kind: "tool",
        description: config.description,
        connection: config.connection,
        connections: config.connections,
        inputSchema: config.inputSchema,
        inputDependencies: config.inputDependencies,
        run: config.run,
        options: config.options,
        schema: config.schema,
        digest: config.digest,
        describe: buildDescribe({
            kind: "tool",
            name: config.name,
            description: config.description,
            inputSchema: config.inputSchema,
            connection: config.connection,
            connections: config.connections,
        }),
    });
}
export function defineDurable(config) {
    const execute = () => {
        throw new Error(`defineDurable("${config.name}") is not callable yet. ` +
            "Runtime resolution is coming in Phase 2.");
    };
    Object.defineProperty(execute, "name", {
        value: config.name,
        writable: false,
    });
    return Object.assign(execute, {
        kind: "durable",
        description: config.description,
        connection: config.connection,
        connections: config.connections,
        inputSchema: config.inputSchema,
        inputDependencies: config.inputDependencies,
        run: config.run,
        describe: buildDescribe({
            kind: "durable",
            name: config.name,
            description: config.description,
            inputSchema: config.inputSchema,
            connection: config.connection,
            connections: config.connections,
        }),
    });
}
export function defineComponent(config) {
    const component = () => {
        throw new Error(`defineComponent("${config.name}") is not callable yet. ` +
            "Runtime resolution is coming in Phase 2.");
    };
    Object.defineProperty(component, "name", {
        value: config.name,
        writable: false,
    });
    return Object.assign(component, {
        kind: "component",
        description: config.description,
        connection: "connection" in config ? config.connection : undefined,
        connections: "connections" in config ? config.connections : undefined,
        describe: buildDescribe({
            kind: "component",
            name: config.name,
            description: config.description,
            connection: "connection" in config ? config.connection : undefined,
            connections: "connections" in config ? config.connections : undefined,
        }),
    });
}
