// cross-keychain/dist/index.cjs
'use strict';

var fs = require('fs');
var os = require('os');
var path = require('path');
var child_process = require('child_process');
var url = require('url');
var crypto = require('crypto');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
/**
 *
 */
class KeyringError extends Error {
}
/**
 *
 */
class PasswordSetError extends KeyringError {
}
/**
 *
 */
class PasswordDeleteError extends KeyringError {
}
/**
 *
 */
class InitError extends KeyringError {
}
/**
 *
 */
class KeyringLockedError extends KeyringError {
}
/**
 *
 */
class NoKeyringError extends KeyringError {
}

var errors = /*#__PURE__*/Object.freeze({
    __proto__: null,
    InitError: InitError,
    KeyringError: KeyringError,
    KeyringLockedError: KeyringLockedError,
    NoKeyringError: NoKeyringError,
    PasswordDeleteError: PasswordDeleteError,
    PasswordSetError: PasswordSetError
});

/**
 *
 */
function getConfigFile() {
    return path.join(getConfigRoot(), "keyring.config.json");
}
/**
 *
 */
function getConfigRoot() {
    if (process.env.XDG_CONFIG_HOME) {
        return path.join(process.env.XDG_CONFIG_HOME, "keyring");
    }
    if (process.platform === "win32") {
        const root = process.env.LOCALAPPDATA || process.env.APPDATA || os.homedir();
        return path.join(root, "Keyring");
    }
    const fallback = path.join(os.homedir(), ".config");
    return path.join(fallback, "keyring");
}
/**
 *
 */
function getDataRoot() {
    if (process.env.XDG_DATA_HOME) {
        return path.join(process.env.XDG_DATA_HOME, "keyring");
    }
    if (process.platform === "win32") {
        const root = process.env.LOCALAPPDATA || process.env.ProgramData || os.homedir();
        return path.join(root, "Keyring");
    }
    const fallback = path.join(os.homedir(), ".local", "share");
    return path.join(fallback, "keyring");
}
/**
 *
 */
async function ensureParent(file) {
    await fs.promises.mkdir(path.dirname(file), { recursive: true, mode: 0o700 });
}
/**
 *
 */
async function readConfig() {
    const file = getConfigFile();
    const raw = await fs.promises.readFile(file, "utf8");
    return JSON.parse(raw);
}

const __filename$1 = url.fileURLToPath((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href)));
const __dirname$1 = path.dirname(__filename$1);
async function runCommandImpl(command, args, options = {}) {
    return await new Promise((resolve, reject) => {
        const child = child_process.spawn(command, args, {
            stdio: "pipe",
            env: options.env ?? process.env,
            cwd: options.cwd,
        });
        let stdout = "";
        let stderr = "";
        let timer;
        if (options.timeoutMs && options.timeoutMs > 0) {
            timer = setTimeout(() => {
                child.kill("SIGKILL");
                reject(new KeyringError(`Command timed out after ${options.timeoutMs}ms: ${command}`));
            }, options.timeoutMs);
        }
        child.stdout?.on("data", (chunk) => {
            stdout += chunk.toString();
        });
        child.stderr?.on("data", (chunk) => {
            stderr += chunk.toString();
        });
        child.on("error", (error) => {
            if (timer)
                clearTimeout(timer);
            reject(error);
        });
        child.on("close", (code) => {
            if (timer)
                clearTimeout(timer);
            resolve({
                code: code ?? 0,
                stdout,
                stderr,
            });
        });
        if (options.input !== undefined) {
            child.stdin?.end(options.input);
        }
        else {
            child.stdin?.end();
        }
    });
}
async function executableExistsImpl(command) {
    const delimiter = process.platform === "win32" ? ";" : ":";
    const paths = (process.env.PATH ?? "").split(delimiter).filter(Boolean);
    if (paths.length === 0) {
        return false;
    }
    const extensions = process.platform === "win32"
        ? (process.env.PATHEXT ?? ".EXE;.CMD;.BAT;.COM").split(";")
        : [""];
    for (const base of paths) {
        for (const ext of extensions) {
            const candidate = path.join(base, command + ext);
            try {
                await fs.promises.access(candidate, fs.constants.X_OK);
                return true;
            }
            catch {
                if (process.platform === "win32") {
                    try {
                        await fs.promises.access(candidate, fs.constants.F_OK);
                        return true;
                    }
                    catch {
                        // ignore and continue
                    }
                }
            }
        }
    }
    return false;
}
let credmanBootstrap = null;
/**
 * Loads the Windows Credential Manager PowerShell bootstrap script.
 * The script is cached after first load for performance.
 *
 * @returns The PowerShell script content
 */
async function getCredmanBootstrap() {
    if (!credmanBootstrap) {
        const scriptPath = path.join(__dirname$1, "scripts", "credman.ps1");
        credmanBootstrap = await fs.promises.readFile(scriptPath, "utf-8");
    }
    return credmanBootstrap;
}
const runtime = {
    runCommand: runCommandImpl,
    executableExists: executableExistsImpl,
    async runPowerShell(script, options) {
        const command = (await runtime.executableExists("powershell"))
            ? "powershell"
            : (await runtime.executableExists("pwsh"))
                ? "pwsh"
                : null;
        if (!command) {
            throw new InitError("PowerShell is required for Windows credential operations");
        }
        return runtime.runCommand(command, ["-NoProfile", "-NonInteractive", "-Command", script], options);
    },
};
const defaultRunPowerShell = runtime.runPowerShell;
/**
 *
 */
function __resetRuntimeForTests() {
    runtime.runCommand = runCommandImpl;
    runtime.executableExists = executableExistsImpl;
    runtime.runPowerShell = defaultRunPowerShell;
}
process.env.NODE_ENV === "test"
    ? {
        }
    : undefined;

const ENV_PROPERTY_PREFIX = "KEYRING_PROPERTY_";
/**
 * Abstract base class for all keyring backends.
 * Provides common functionality for property management, validation, and credential handling.
 * Backends extend this class and implement the core storage operations.
 */
class ConfigurableBackend {
    properties;
    /**
     * Creates a new backend instance with optional configuration properties.
     * Properties can be overridden by environment variables prefixed with KEYRING_PROPERTY_.
     *
     * @param properties - Optional configuration properties for this backend
     */
    constructor(properties) {
        this.properties = { ...(properties ?? {}) };
        this.applyEnvOverrides();
    }
    /**
     * Retrieves a credential (username and password) for a service.
     * If account is specified, retrieves that specific credential.
     * If account is not specified, returns the first credential found for the service.
     *
     * @param service - The service identifier
     * @param account - Optional account/username identifier
     * @returns The credential object, or null if not found
     */
    async getCredential(service, account) {
        if (account) {
            const password = await this.getPassword(service, account);
            return password === null ? null : { username: account, password };
        }
        const usernames = await this.lookupUsernames(service);
        const username = usernames[0];
        if (!username) {
            return null;
        }
        const password = await this.getPassword(service, username);
        return password === null ? null : { username, password };
    }
    /**
     * Creates a new instance of this backend with merged properties.
     * Useful for runtime configuration changes without recreating the backend registry.
     *
     * @param properties - Properties to merge with the current backend's properties
     * @returns A new backend instance with the merged properties
     */
    withProperties(properties) {
        const ctor = this.constructor;
        return new ctor({ ...this.properties, ...properties });
    }
    /**
     * Returns diagnostic information about this backend.
     * Includes backend ID, name, and priority. Subclasses can override to add more details.
     *
     * @returns Diagnostic information as key-value pairs
     */
    async diagnose() {
        return {
            name: this.name,
            id: this.id,
            priority: this.priority,
        };
    }
    async ensureDir(dir) {
        await fs.promises.mkdir(dir, { recursive: true, mode: 0o700 });
    }
    applyEnvOverrides() {
        const entries = Object.entries(process.env)
            .filter(([key]) => key.startsWith(ENV_PROPERTY_PREFIX))
            .map(([key, value]) => ({
            property: key.replace(ENV_PROPERTY_PREFIX, "").toLowerCase(),
            value,
        }))
            .filter((entry) => entry.value !== undefined);
        for (const { property, value } of entries) {
            this.properties[property] = value;
        }
    }
    async lookupUsernames(_service) {
        return [];
    }
    normalizeString(value) {
        return value.normalize("NFC");
    }
    validateIdentifier(value, name) {
        const normalized = this.normalizeString(value);
        if (!normalized || normalized.length === 0) {
            throw new KeyringError(`${name} cannot be empty`);
        }
        if (normalized.length > 255) {
            throw new KeyringError(`${name} exceeds maximum length of 255 characters`);
        }
        const identifierPattern = /^[a-zA-Z0-9._@-]+$/;
        if (!identifierPattern.test(normalized)) {
            throw new KeyringError(`${name} contains invalid characters. Only alphanumeric characters, dots, underscores, @ symbols, and hyphens are allowed`);
        }
    }
    validatePassword(password) {
        const normalized = this.normalizeString(password);
        if (!normalized || normalized.length === 0) {
            throw new KeyringError("Password cannot be empty");
        }
        if (normalized.length > 4096) {
            throw new KeyringError("Password exceeds maximum length of 4096 characters");
        }
    }
}

let NativeKeyringEntry$2 = null;
let nativeModuleLoadAttempted$2 = false;
async function loadNativeModule$2() {
    if (nativeModuleLoadAttempted$2)
        return;
    nativeModuleLoadAttempted$2 = true;
    try {
        const nativeModule = await import('@napi-rs/keyring');
        NativeKeyringEntry$2 = nativeModule.Entry;
    }
    catch {
        NativeKeyringEntry$2 = null;
    }
}
/**
 * Backend for macOS Keychain using native bindings via @napi-rs/keyring.
 * This backend provides the best performance and integration with macOS Keychain.
 * Falls back gracefully to CLI-based backend if native module is not available.
 * Has the highest priority among macOS backends.
 */
class NativeKeychainBackend extends ConfigurableBackend {
    id = "native-macos";
    name = "Native macOS Keychain";
    priority = 10;
    /**
     * Checks if this backend is supported on the current platform.
     * Requires macOS and the @napi-rs/keyring native module.
     *
     * @returns True if macOS and the native module are available
     */
    static async isSupported() {
        if (process.platform !== "darwin")
            return false;
        await loadNativeModule$2();
        return NativeKeyringEntry$2 !== null;
    }
    /**
     * Retrieves a password from macOS Keychain using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @returns The stored password, or null if not found
     */
    async getPassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        if (!NativeKeyringEntry$2) {
            throw new KeyringError("Native keyring module not available");
        }
        try {
            const entry = new NativeKeyringEntry$2(service, account);
            return entry.getPassword();
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            if (message.includes("not found") ||
                message.includes("No such") ||
                message.includes("not exist")) {
                return null;
            }
            throw new KeyringError(`Native keychain error: ${message}`);
        }
    }
    /**
     * Stores a password in macOS Keychain using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @param password - The password to store
     */
    async setPassword(service, account, password) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        this.validatePassword(password);
        if (!NativeKeyringEntry$2) {
            throw new KeyringError("Native keyring module not available");
        }
        try {
            const entry = new NativeKeyringEntry$2(service, account);
            entry.setPassword(password);
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            throw new PasswordSetError(`Native keychain error: ${message}`);
        }
    }
    /**
     * Deletes a password from macOS Keychain using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @throws {PasswordDeleteError} If the password does not exist
     */
    async deletePassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        if (!NativeKeyringEntry$2) {
            throw new KeyringError("Native keyring module not available");
        }
        const exists = await this.getPassword(service, account);
        if (exists === null) {
            throw new PasswordDeleteError("Password not found");
        }
        try {
            const entry = new NativeKeyringEntry$2(service, account);
            entry.deletePassword();
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            throw new PasswordDeleteError(`Native keychain error: ${message}`);
        }
    }
    /**
     * Lists all usernames stored for a given service.
     * Uses the @napi-rs/keyring findCredentials API.
     *
     * @param service - The service identifier
     * @returns Array of usernames that have credentials stored for this service
     */
    async lookupUsernames(service) {
        if (!NativeKeyringEntry$2) {
            return [];
        }
        try {
            // Import findCredentials dynamically
            const { findCredentials } = await import('@napi-rs/keyring');
            const credentials = findCredentials(service);
            return credentials.map((cred) => cred.account);
        }
        catch {
            // If findCredentials fails or is not available, return empty array
            return [];
        }
    }
    /**
     * Returns diagnostic information about this backend.
     * Includes information about the native implementation and fallback availability.
     *
     * @returns Diagnostic information including implementation details
     */
    async diagnose() {
        const details = await super.diagnose();
        const { MacOSKeychainBackend } = await Promise.resolve().then(function () { return macos; });
        return {
            ...details,
            implementation: "Native Security.framework bindings",
            fallbackAvailable: await MacOSKeychainBackend.isSupported(),
        };
    }
}

let NativeKeyringEntry$1 = null;
let nativeModuleLoadAttempted$1 = false;
async function loadNativeModule$1() {
    if (nativeModuleLoadAttempted$1)
        return;
    nativeModuleLoadAttempted$1 = true;
    try {
        const nativeModule = await import('@napi-rs/keyring');
        NativeKeyringEntry$1 = nativeModule.Entry;
    }
    catch {
        NativeKeyringEntry$1 = null;
    }
}
/**
 * Backend for Windows Credential Manager using native bindings via @napi-rs/keyring.
 * This backend provides the best performance and security by directly accessing the
 * Windows DPAPI through native bindings. Falls back gracefully to PowerShell-based
 * backend if native module is not available. Has the highest priority among Windows backends.
 */
class NativeWindowsBackend extends ConfigurableBackend {
    id = "native-windows";
    name = "Native Windows Credential Manager";
    priority = 10;
    /**
     * Checks if this backend is supported on the current platform.
     * Requires Windows and the @napi-rs/keyring native module.
     *
     * @returns True if Windows and the native module are available
     */
    static async isSupported() {
        if (process.platform !== "win32")
            return false;
        await loadNativeModule$1();
        return NativeKeyringEntry$1 !== null;
    }
    /**
     * Retrieves a password from Windows Credential Manager using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @returns The stored password, or null if not found
     */
    async getPassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        if (!NativeKeyringEntry$1) {
            throw new KeyringError("Native keyring module not available");
        }
        try {
            const entry = new NativeKeyringEntry$1(service, account);
            return entry.getPassword();
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            if (message.includes("not found") ||
                message.includes("No such") ||
                message.includes("not exist")) {
                return null;
            }
            throw new KeyringError(`Native credential manager error: ${message}`);
        }
    }
    /**
     * Stores a password in Windows Credential Manager using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @param password - The password to store
     */
    async setPassword(service, account, password) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        this.validatePassword(password);
        if (!NativeKeyringEntry$1) {
            throw new KeyringError("Native keyring module not available");
        }
        try {
            const entry = new NativeKeyringEntry$1(service, account);
            entry.setPassword(password);
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            throw new PasswordSetError(`Native credential manager error: ${message}`);
        }
    }
    /**
     * Deletes a password from Windows Credential Manager using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @throws {PasswordDeleteError} If the password does not exist
     */
    async deletePassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        if (!NativeKeyringEntry$1) {
            throw new KeyringError("Native keyring module not available");
        }
        const exists = await this.getPassword(service, account);
        if (exists === null) {
            throw new PasswordDeleteError("Password not found");
        }
        try {
            const entry = new NativeKeyringEntry$1(service, account);
            entry.deletePassword();
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            throw new PasswordDeleteError(`Native credential manager error: ${message}`);
        }
    }
    /**
     * Lists all usernames stored for a given service.
     * Note: The @napi-rs/keyring library does not support credential enumeration.
     *
     * @param _service - The service identifier
     * @returns Empty array (enumeration not supported by native backend)
     */
    async lookupUsernames(_service) {
        return [];
    }
    /**
     * Returns diagnostic information about this backend.
     * Includes information about the native implementation and fallback availability.
     *
     * @returns Diagnostic information including implementation details
     */
    async diagnose() {
        const details = await super.diagnose();
        const { WindowsCredentialBackend } = await Promise.resolve().then(function () { return windows; });
        return {
            ...details,
            implementation: "Native DPAPI bindings",
            fallbackAvailable: await WindowsCredentialBackend.isSupported(),
        };
    }
}

let NativeKeyringEntry = null;
let nativeModuleLoadAttempted = false;
async function loadNativeModule() {
    if (nativeModuleLoadAttempted)
        return;
    nativeModuleLoadAttempted = true;
    try {
        const nativeModule = await import('@napi-rs/keyring');
        NativeKeyringEntry = nativeModule.Entry;
    }
    catch {
        NativeKeyringEntry = null;
    }
}
/**
 * Backend for Freedesktop Secret Service using native bindings via @napi-rs/keyring.
 * This backend provides the best performance and security by directly accessing the
 * Secret Service D-Bus API through native bindings. Falls back gracefully to
 * secret-tool based backend if native module is not available.
 * Has the highest priority among Linux backends.
 */
class NativeLinuxBackend extends ConfigurableBackend {
    id = "native-linux";
    name = "Native Freedesktop Secret Service";
    priority = 10;
    /**
     * Checks if this backend is supported on the current platform.
     * Requires Linux and the @napi-rs/keyring native module.
     *
     * @returns True if Linux and the native module are available
     */
    static async isSupported() {
        if (process.platform !== "linux")
            return false;
        await loadNativeModule();
        return NativeKeyringEntry !== null;
    }
    /**
     * Retrieves a password from the Secret Service using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @returns The stored password, or null if not found
     */
    async getPassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        if (!NativeKeyringEntry) {
            throw new KeyringError("Native keyring module not available");
        }
        try {
            const entry = new NativeKeyringEntry(service, account);
            return entry.getPassword();
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            if (message.includes("not found") ||
                message.includes("No such") ||
                message.includes("not exist")) {
                return null;
            }
            throw new KeyringError(`Native secret service error: ${message}`);
        }
    }
    /**
     * Stores a password in the Secret Service using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @param password - The password to store
     */
    async setPassword(service, account, password) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        this.validatePassword(password);
        if (!NativeKeyringEntry) {
            throw new KeyringError("Native keyring module not available");
        }
        try {
            const entry = new NativeKeyringEntry(service, account);
            entry.setPassword(password);
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            throw new PasswordSetError(`Native secret service error: ${message}`);
        }
    }
    /**
     * Deletes a password from the Secret Service using native bindings.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @throws {PasswordDeleteError} If the password does not exist
     */
    async deletePassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        if (!NativeKeyringEntry) {
            throw new KeyringError("Native keyring module not available");
        }
        const exists = await this.getPassword(service, account);
        if (exists === null) {
            throw new PasswordDeleteError("Password not found");
        }
        try {
            const entry = new NativeKeyringEntry(service, account);
            entry.deletePassword();
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            throw new PasswordDeleteError(`Native secret service error: ${message}`);
        }
    }
    /**
     * Lists all usernames stored for a given service.
     * Note: The @napi-rs/keyring library does not support credential enumeration.
     *
     * @param _service - The service identifier
     * @returns Empty array (enumeration not supported by native backend)
     */
    async lookupUsernames(_service) {
        return [];
    }
    /**
     * Returns diagnostic information about this backend.
     * Includes information about the native implementation and fallback availability.
     *
     * @returns Diagnostic information including implementation details
     */
    async diagnose() {
        const details = await super.diagnose();
        const { SecretServiceBackend } = await Promise.resolve().then(function () { return linux; });
        return {
            ...details,
            implementation: "Native DBus bindings",
            fallbackAvailable: await SecretServiceBackend.isSupported(),
        };
    }
}

/**
 * Backend for macOS Keychain using the `security` command-line tool.
 * Stores credentials in the system keychain using generic password items.
 * Supports custom keychain paths via the `keychain` property.
 */
class MacOSKeychainBackend extends ConfigurableBackend {
    id = "macos";
    name = "macOS Keychain (CLI)";
    priority = 5;
    /**
     * Checks if this backend is supported on the current platform.
     * Requires macOS and the `security` command-line tool.
     *
     * @returns True if macOS and security tool are available
     */
    static async isSupported() {
        return (process.platform === "darwin" &&
            (await runtime.executableExists("security")));
    }
    get keychain() {
        const kc = this.properties["keychain"];
        return typeof kc === "string" && kc.length ? kc : undefined;
    }
    /**
     * Retrieves a password from the macOS Keychain.
     * Uses the `security find-generic-password` command.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @returns The stored password, or null if not found
     */
    async getPassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        const args = ["find-generic-password", "-s", service, "-a", account, "-w"];
        this.appendKeychain(args);
        const result = await runtime.runCommand("security", args, {
            timeoutMs: 10000,
        });
        if (result.code === 44) {
            return null;
        }
        if (result.code !== 0) {
            throw new KeyringError(`Keychain operation failed with code ${result.code}`);
        }
        return result.stdout.trim();
    }
    /**
     * Stores a password in the macOS Keychain.
     * Uses the `security add-generic-password` command with the -U flag to update existing entries.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @param password - The password to store
     */
    async setPassword(service, account, password) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        this.validatePassword(password);
        const args = [
            "add-generic-password",
            "-s",
            service,
            "-a",
            account,
            "-U",
            "-w",
            password,
        ];
        this.appendKeychain(args);
        const result = await runtime.runCommand("security", args, {
            timeoutMs: 10000,
        });
        if (result.code !== 0) {
            throw new PasswordSetError(`Keychain operation failed with code ${result.code}`);
        }
    }
    /**
     * Deletes a password from the macOS Keychain.
     * Uses the `security delete-generic-password` command.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @throws {PasswordDeleteError} If the password does not exist
     */
    async deletePassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        const args = ["delete-generic-password", "-s", service, "-a", account];
        this.appendKeychain(args);
        const result = await runtime.runCommand("security", args, {
            timeoutMs: 10000,
        });
        if (result.code === 44) {
            throw new PasswordDeleteError("Password not found");
        }
        if (result.code !== 0) {
            throw new PasswordDeleteError(`Keychain operation failed with code ${result.code}`);
        }
    }
    async lookupUsernames(service) {
        const args = ["find-generic-password", "-s", service];
        this.appendKeychain(args);
        const result = await runtime.runCommand("security", args, {
            timeoutMs: 10000,
        });
        if (result.code !== 0) {
            return [];
        }
        const matches = [...result.stdout.matchAll(/"acct"<blob>="([^"]+)"/g)];
        return matches.map((match) => match[1]);
    }
    /**
     * Returns diagnostic information about this backend.
     * Includes the keychain path being used (or "default" for the system keychain).
     *
     * @returns Diagnostic information including backend details and keychain path
     */
    async diagnose() {
        const details = await super.diagnose();
        return {
            ...details,
            keychain: this.keychain ?? "default",
        };
    }
    appendKeychain(args) {
        if (this.keychain) {
            args.push(this.keychain);
        }
    }
}

var macos = /*#__PURE__*/Object.freeze({
    __proto__: null,
    MacOSKeychainBackend: MacOSKeychainBackend
});

/**
 * Backend for Linux Secret Service using the `secret-tool` command.
 * Implements the Freedesktop Secret Service specification.
 * Compatible with GNOME Keyring, KWallet, and other Secret Service implementations.
 * Supports custom application IDs and collection names via properties.
 */
class SecretServiceBackend extends ConfigurableBackend {
    id = "secret-service";
    name = "Freedesktop Secret Service";
    priority = 4.8;
    /**
     * Checks if this backend is supported on the current platform.
     * Requires Linux and the `secret-tool` command.
     *
     * @returns True if Linux and secret-tool are available
     */
    static async isSupported() {
        return (process.platform === "linux" &&
            (await runtime.executableExists("secret-tool")));
    }
    get appId() {
        const value = this.properties["application"] ?? this.properties["appid"];
        return typeof value === "string" && value.length ? value : "ts-keyring";
    }
    get collection() {
        const value = this.properties["collection"] ?? this.properties["preferred_collection"];
        return typeof value === "string" && value.length ? value : undefined;
    }
    /**
     * Retrieves a password from the Secret Service.
     * Uses the `secret-tool lookup` command with service, username, and application attributes.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @returns The stored password, or null if not found
     */
    async getPassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        const args = [
            "lookup",
            "service",
            service,
            "username",
            account,
            "application",
            this.appId,
        ];
        if (this.collection) {
            args.splice(1, 0, `--collection=${this.collection}`);
        }
        const result = await runtime.runCommand("secret-tool", args, {
            timeoutMs: 10000,
        });
        if (result.code === 1) {
            return null;
        }
        if (result.code !== 0) {
            throw new KeyringError(`Secret service operation failed with code ${result.code}`);
        }
        return result.stdout.replace(/\r?\n$/, "");
    }
    /**
     * Stores a password in the Secret Service.
     * Uses the `secret-tool store` command with a descriptive label.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @param password - The password to store
     */
    async setPassword(service, account, password) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        this.validatePassword(password);
        const args = [
            "store",
            "--label",
            `Password for '${account.replace(/'/g, "\\'")}' on '${service.replace(/'/g, "\\'")}' (${this.appId})`,
        ];
        if (this.collection) {
            args.push(`--collection=${this.collection}`);
        }
        args.push("service", service, "username", account, "application", this.appId);
        const result = await runtime.runCommand("secret-tool", args, {
            input: `${password}\n`,
            timeoutMs: 10000,
        });
        if (result.code !== 0) {
            throw new PasswordSetError(`Secret service operation failed with code ${result.code}`);
        }
    }
    /**
     * Deletes a password from the Secret Service.
     * Uses the `secret-tool clear` command.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @throws {PasswordDeleteError} If the password does not exist
     */
    async deletePassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        const args = [
            "clear",
            "service",
            service,
            "username",
            account,
            "application",
            this.appId,
        ];
        if (this.collection) {
            args.splice(1, 0, `--collection=${this.collection}`);
        }
        const result = await runtime.runCommand("secret-tool", args, {
            timeoutMs: 10000,
        });
        if (result.code === 1) {
            throw new PasswordDeleteError("Password not found");
        }
        if (result.code !== 0) {
            throw new PasswordDeleteError(`Secret service operation failed with code ${result.code}`);
        }
    }
    async lookupUsernames(service) {
        this.validateIdentifier(service, "service");
        const args = ["search", "service", service, "application", this.appId];
        if (this.collection) {
            args.splice(1, 0, `--collection=${this.collection}`);
        }
        const result = await runtime.runCommand("secret-tool", args, {
            timeoutMs: 10000,
        });
        if (result.code !== 0) {
            return [];
        }
        const usernames = new Set();
        for (const line of result.stdout.split(/\r?\n/)) {
            const trimmed = line.trim();
            if (!trimmed) {
                continue;
            }
            const parts = trimmed.split("=");
            if (parts.length < 2) {
                continue;
            }
            const key = parts[0].trim().toLowerCase();
            const value = parts.slice(1).join("=").trim();
            if (key.endsWith("username") && value) {
                usernames.add(value);
            }
        }
        return Array.from(usernames);
    }
}

var linux = /*#__PURE__*/Object.freeze({
    __proto__: null,
    SecretServiceBackend: SecretServiceBackend
});

/**
 * Backend for Windows Credential Manager using PowerShell and Windows API.
 * Stores credentials in the Windows Credential Vault as generic credentials.
 * Uses PowerShell with P/Invoke to access the native Windows Credential Manager API.
 */
class WindowsCredentialBackend extends ConfigurableBackend {
    id = "windows";
    name = "Windows Credential Manager";
    priority = 5;
    /**
     * Checks if this backend is supported on the current platform.
     * Requires Windows and functional PowerShell.
     *
     * @returns True if Windows and PowerShell are available
     */
    static async isSupported() {
        if (process.platform !== "win32") {
            return false;
        }
        try {
            await runtime.runPowerShell("Write-Output 'ok'");
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Retrieves a password from Windows Credential Manager.
     * Uses PowerShell to call CredRead from the Windows API.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @returns The stored password, or null if not found
     */
    async getPassword(service, account) {
        this.validateIdentifier(account, "account");
        const target = this.buildTarget(service, account);
        const targetB64 = Buffer.from(target, "utf8").toString("base64");
        const bootstrap = await getCredmanBootstrap();
        const script = `${bootstrap}
$target = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String('${targetB64}'))
$ptr = [IntPtr]::Zero
if (-not [CredMan.CredentialManager]::CredRead($target, 1, 0, [ref]$ptr)) {
  exit 2
}
try {
  $cred = [Runtime.InteropServices.Marshal]::PtrToStructure($ptr, [Type]'CredMan.CredentialManager+CREDENTIAL')
  if ($cred.CredentialBlob -eq [IntPtr]::Zero -or $cred.CredentialBlobSize -le 0) {
    Write-Output ''
  } else {
    Write-Output ([Runtime.InteropServices.Marshal]::PtrToStringUni($cred.CredentialBlob, [int]($cred.CredentialBlobSize / 2)))
  }
} finally {
  [CredMan.CredentialManager]::CredFree($ptr)
}`;
        const result = await runtime.runPowerShell(script, { timeoutMs: 10000 });
        if (result.code === 2) {
            return null;
        }
        if (result.code !== 0) {
            throw new KeyringError(`Credential Manager operation failed with code ${result.code}`);
        }
        return result.stdout.replace(/\r?\n$/, "");
    }
    /**
     * Stores a password in Windows Credential Manager.
     * Uses PowerShell to call CredWrite from the Windows API.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @param password - The password to store
     */
    async setPassword(service, account, password) {
        this.validateIdentifier(account, "account");
        this.validatePassword(password);
        const target = this.buildTarget(service, account);
        const persistValue = this.persistence;
        this.validatePersistence(persistValue);
        const targetB64 = Buffer.from(target, "utf8").toString("base64");
        const accountB64 = Buffer.from(account, "utf8").toString("base64");
        const passwordB64 = Buffer.from(password, "utf8").toString("base64");
        const bootstrap = await getCredmanBootstrap();
        const script = `${bootstrap}
$target = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String('${targetB64}'))
$username = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String('${accountB64}'))
$password = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String('${passwordB64}'))
$persist = ${persistValue}
$bytes = [System.Text.Encoding]::Unicode.GetBytes($password)
$cred = New-Object CredMan.CredentialManager+CREDENTIAL
$cred.Type = 1
$cred.TargetName = $target
$cred.UserName = $username
$cred.CredentialBlobSize = $bytes.Length
$cred.Persist = $persist
$cred.CredentialBlob = [Runtime.InteropServices.Marshal]::AllocHGlobal($bytes.Length)
try {
  [Runtime.InteropServices.Marshal]::Copy($bytes, 0, $cred.CredentialBlob, $bytes.Length)
  if (-not [CredMan.CredentialManager]::CredWrite([ref]$cred, 0)) {
    $code = [Runtime.InteropServices.Marshal]::GetLastWin32Error()
    throw "CredWrite failed: $code"
  }
} finally {
  if ($cred.CredentialBlob -ne [IntPtr]::Zero) {
    [Runtime.InteropServices.Marshal]::FreeHGlobal($cred.CredentialBlob)
  }
}`;
        const result = await runtime.runPowerShell(script, { timeoutMs: 10000 });
        if (result.code !== 0) {
            throw new PasswordSetError(`Credential Manager operation failed with code ${result.code}`);
        }
    }
    /**
     * Deletes a password from Windows Credential Manager.
     * Uses PowerShell to call CredDelete from the Windows API.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @throws {PasswordDeleteError} If the password does not exist
     */
    async deletePassword(service, account) {
        this.validateIdentifier(account, "account");
        const target = this.buildTarget(service, account);
        const targetB64 = Buffer.from(target, "utf8").toString("base64");
        const bootstrap = await getCredmanBootstrap();
        const script = `${bootstrap}
$target = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String('${targetB64}'))
if (-not [CredMan.CredentialManager]::CredDelete($target, 1, 0)) {
  $code = [Runtime.InteropServices.Marshal]::GetLastWin32Error()
  if ($code -eq 1168) {
    exit 2
  }
  throw "CredDelete failed: $code"
}`;
        const result = await runtime.runPowerShell(script, { timeoutMs: 10000 });
        if (result.code === 2) {
            throw new PasswordDeleteError("Password not found");
        }
        if (result.code !== 0) {
            throw new PasswordDeleteError(`Credential Manager operation failed with code ${result.code}`);
        }
    }
    buildTarget(service, account) {
        this.validateIdentifier(service, "service");
        return `${service}:${account}`;
    }
    /**
     *
     */
    async diagnose() {
        const details = await super.diagnose();
        return {
            ...details,
            persistence: this.persistence,
        };
    }
    get persistence() {
        const value = this.properties["persist"];
        if (typeof value === "number" && Number.isFinite(value)) {
            return Math.floor(value);
        }
        if (typeof value === "string" && value.length) {
            const normalized = value.toLowerCase();
            if (["session", "cred_persist_session"].includes(normalized)) {
                return 1;
            }
            if ([
                "local",
                "local_machine",
                "localmachine",
                "cred_persist_local_machine",
            ].includes(normalized)) {
                return 2;
            }
            if (["enterprise", "cred_persist_enterprise"].includes(normalized)) {
                return 3;
            }
        }
        return 3;
    }
    validatePersistence(value) {
        if (value < 1 || value > 3) {
            throw new KeyringError("Persistence must be 1 (Session), 2 (LocalMachine), or 3 (Enterprise)");
        }
    }
}

var windows = /*#__PURE__*/Object.freeze({
    __proto__: null,
    WindowsCredentialBackend: WindowsCredentialBackend
});

/**
 * A backend that stores secrets in an encrypted local file.
 * Uses AES-256-GCM for authenticated encryption with a randomly generated key.
 * The encryption key is stored in a separate file with 0600 permissions.
 * This backend serves as a fallback when native platform keychains are unavailable.
 */
class FileSystemBackend extends ConfigurableBackend {
    id = "file";
    name = "Encrypted file storage (AES-256-GCM)";
    priority = 0.5;
    get filePath() {
        const custom = this.properties["file_path"];
        if (typeof custom === "string" && custom.length > 0) {
            const normalized = path.normalize(custom);
            const dangerousPaths = [
                "/etc",
                "/sys",
                "/proc",
                "/dev",
                "/root",
                "C:\\Windows",
                "C:\\System",
            ];
            for (const dangerous of dangerousPaths) {
                if (normalized.toLowerCase().startsWith(dangerous.toLowerCase())) {
                    throw new KeyringError(`File path cannot be in protected system directory: ${dangerous}`);
                }
            }
            if (path.isAbsolute(normalized)) {
                const dataRoot = getDataRoot();
                if (!normalized.startsWith(dataRoot)) {
                    console.warn(`Warning: Using file path outside data directory: ${normalized}`);
                }
            }
            return normalized;
        }
        return path.join(getDataRoot(), "secrets.json");
    }
    get keyFilePath() {
        const custom = this.properties["key_file_path"];
        if (typeof custom === "string" && custom.length > 0) {
            return path.normalize(custom);
        }
        return path.join(getConfigRoot(), "file.key");
    }
    async getKeyMaterial() {
        const envKey = process.env.KEYRING_FILE_MASTER_KEY;
        if (envKey) {
            if (envKey.length !== 64) {
                throw new KeyringError("KEYRING_FILE_MASTER_KEY must be 64 hex characters (32 bytes)");
            }
            return Buffer.from(envKey, "hex");
        }
        const keyFile = this.keyFilePath;
        try {
            const key = await fs.promises.readFile(keyFile);
            if (key.length !== 32) {
                throw new KeyringError("Key file must contain exactly 32 bytes");
            }
            return key;
        }
        catch (error) {
            if (error.code === "ENOENT") {
                const key = crypto.randomBytes(32);
                await this.ensureDir(path.dirname(keyFile));
                await fs.promises.writeFile(keyFile, key, { mode: 0o600 });
                return key;
            }
            throw error;
        }
    }
    async encryptStore(store) {
        const plaintext = JSON.stringify(store);
        const key = await this.getKeyMaterial();
        const iv = crypto.randomBytes(12);
        const cipher = crypto.createCipheriv("aes-256-gcm", key, iv);
        const encrypted = Buffer.concat([
            cipher.update(plaintext, "utf8"),
            cipher.final(),
        ]);
        const authTag = cipher.getAuthTag();
        const version = Buffer.from([1]);
        return Buffer.concat([version, iv, authTag, encrypted]);
    }
    async decryptStore(data) {
        const version = data[0];
        if (version === 1) {
            const iv = data.subarray(1, 13);
            const authTag = data.subarray(13, 29);
            const encrypted = data.subarray(29);
            const key = await this.getKeyMaterial();
            const decipher = crypto.createDecipheriv("aes-256-gcm", key, iv);
            decipher.setAuthTag(authTag);
            const decrypted = Buffer.concat([
                decipher.update(encrypted),
                decipher.final(),
            ]);
            return JSON.parse(decrypted.toString("utf8"));
        }
        throw new KeyringError(`Unsupported store format version: ${version}`);
    }
    /**
     * Retrieves a password from the encrypted file store.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @returns The stored password, or null if not found
     */
    async getPassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        const store = await this.readStore();
        return store[service]?.[account] ?? null;
    }
    /**
     * Stores a password in the encrypted file store.
     * Creates the storage file and encryption key if they don't exist.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @param password - The password to store
     */
    async setPassword(service, account, password) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        this.validatePassword(password);
        const file = this.filePath;
        await this.ensureDir(path.dirname(file));
        const store = await this.readStore();
        const serviceEntry = store[service] ?? {};
        serviceEntry[account] = password;
        store[service] = serviceEntry;
        await this.writeStore(store);
    }
    /**
     * Deletes a password from the encrypted file store.
     *
     * @param service - The service identifier
     * @param account - The account identifier
     * @throws {PasswordDeleteError} If the password does not exist
     */
    async deletePassword(service, account) {
        this.validateIdentifier(service, "service");
        this.validateIdentifier(account, "account");
        const store = await this.readStore();
        const serviceEntry = store[service];
        if (!serviceEntry || !(account in serviceEntry)) {
            throw new PasswordDeleteError("Password not found");
        }
        delete serviceEntry[account];
        if (Object.keys(serviceEntry).length === 0) {
            delete store[service];
        }
        else {
            store[service] = serviceEntry;
        }
        await this.writeStore(store);
    }
    async lookupUsernames(service) {
        const store = await this.readStore();
        const serviceEntry = store[service];
        return serviceEntry ? Object.keys(serviceEntry) : [];
    }
    async readStore() {
        const file = this.filePath;
        try {
            const data = await fs.promises.readFile(file);
            return this.decryptStore(data);
        }
        catch (error) {
            if (error.code === "ENOENT") {
                return {};
            }
            throw error;
        }
    }
    async atomicWrite(file, buf) {
        const dir = path.dirname(file);
        await this.ensureDir(dir);
        const tmp = path.join(dir, `.tmp-${process.pid}-${Date.now()}-${Math.random().toString(36).slice(2)}`);
        await fs.promises.writeFile(tmp, buf, { mode: 0o600 });
        await fs.promises.rename(tmp, file);
    }
    async writeStore(store) {
        const file = this.filePath;
        const encrypted = await this.encryptStore(store);
        await this.atomicWrite(file, encrypted);
    }
}

/**
 * A no-op backend that does not store any credentials.
 * Used when credential storage needs to be explicitly disabled.
 * All get operations return null, set operations succeed silently, and delete operations throw errors.
 * This backend has a negative priority so it's only used when explicitly configured.
 */
class NullBackend extends ConfigurableBackend {
    id = "null";
    name = "Null keyring";
    priority = -1;
    /**
     * Always returns null, as no credentials are stored.
     *
     * @param _service - The service identifier (unused)
     * @param _account - The account identifier (unused)
     * @returns Always returns null
     */
    async getPassword(_service, _account) {
        return null;
    }
    /**
     * Silently succeeds without storing anything.
     *
     * @param _service - The service identifier (unused)
     * @param _account - The account identifier (unused)
     * @param _password - The password (unused)
     */
    async setPassword(_service, _account, _password) {
        // Intentionally left blank
    }
    /**
     * Always throws an error, as there are no stored credentials to delete.
     *
     * @param _service - The service identifier (unused)
     * @param _account - The account identifier (unused)
     * @throws {PasswordDeleteError} Always thrown since no passwords are stored
     */
    async deletePassword(_service, _account) {
        throw new PasswordDeleteError("Null backend does not store passwords");
    }
}

const ENV_BACKEND = "TS_KEYRING_BACKEND";
let cachedBackends;
let activeBackend;
let backendLimit;
const builtinFactories = [
    NativeKeychainBackend,
    NativeWindowsBackend,
    NativeLinuxBackend,
    MacOSKeychainBackend,
    SecretServiceBackend,
    WindowsCredentialBackend,
    FileSystemBackend,
    NullBackend,
];
async function instantiate(factory) {
    if (factory.isSupported) {
        const supported = await factory.isSupported();
        if (!supported) {
            return null;
        }
    }
    try {
        return new factory();
    }
    catch (error) {
        if (error instanceof InitError) {
            return null;
        }
        throw error;
    }
}
/**
 * Registers a custom backend factory with the keyring system.
 * The backend will be added to the list of available backends for automatic detection.
 *
 * @param factory - The backend factory class to register
 *
 * @example
 * ```typescript
 * class CustomBackend extends ConfigurableBackend {
 *   // ... implementation
 * }
 * registerBackend(CustomBackend);
 * ```
 */
function registerBackend(factory) {
    builtinFactories.push(factory);
    cachedBackends = undefined;
}
/**
 * Returns all available backends that are supported on the current platform.
 * Backends are instantiated and cached on first call.
 * Unsupported backends are filtered out based on their isSupported() method.
 *
 * @returns Array of all supported backend instances
 */
async function getAllBackends() {
    if (!cachedBackends) {
        const instances = await Promise.all(builtinFactories.map((factory) => instantiate(factory)));
        cachedBackends = instances.filter(Boolean);
    }
    return cachedBackends;
}
/**
 * Sets the active keyring backend.
 * This overrides automatic backend detection and forces all operations to use this backend.
 *
 * @param backend - The backend instance to use
 */
function setKeyring(backend) {
    activeBackend = backend;
}
/**
 * Returns the currently active keyring backend.
 * If no backend is active, initializes one using automatic detection.
 * Detection order: environment variable, config file, platform detection.
 *
 * @returns The active backend instance
 * @throws {NoKeyringError} If no backend can be initialized
 */
async function getKeyring() {
    if (!activeBackend) {
        await initBackend(backendLimit);
    }
    if (!activeBackend) {
        throw new NoKeyringError("No keyring backend could be initialized");
    }
    return activeBackend;
}
/**
 * Initializes the active backend using automatic detection.
 * Detection order: environment variable (TS_KEYRING_BACKEND), config file, platform detection.
 *
 * @param limit - Optional filter function to restrict which backends are allowed
 */
async function initBackend(limit) {
    backendLimit = limit;
    const backend = (await loadBackendFromEnv(limit)) ||
        (await loadBackendFromConfig(limit)) ||
        (await detectBackend(limit));
    activeBackend = backend;
}
async function loadBackendFromEnv(limit) {
    const backendId = process.env[ENV_BACKEND];
    if (!backendId) {
        return null;
    }
    const backend = await loadBackendById(backendId, limit);
    if (!backend) {
        throw new InitError(`Backend ${backendId} from ${ENV_BACKEND} is not available`);
    }
    return backend;
}
async function loadBackendFromConfig(limit) {
    try {
        const config = await readConfig();
        if (!config.defaultBackend) {
            return null;
        }
        return loadBackendById(config.defaultBackend, limit, config.backendProperties?.[config.defaultBackend]);
    }
    catch (error) {
        if (error.code === "ENOENT") {
            return null;
        }
        throw error;
    }
}
/**
 * Loads a backend by its ID and optionally applies configuration overrides.
 * Returns null if the backend is not available or doesn't pass the limit filter.
 *
 * @param backendId - The ID of the backend to load (e.g., "macos", "file", "null")
 * @param limit - Optional filter function to check if backend is allowed
 * @param overrides - Optional configuration properties to apply to the backend
 * @returns The backend instance, or null if not available
 */
async function loadBackendById(backendId, limit, overrides) {
    const all = await getAllBackends();
    const backend = all.find((candidate) => candidate.id === backendId);
    if (!backend) {
        return null;
    }
    if (limit && !limit(backend)) {
        return null;
    }
    return overrides ? backend.withProperties(overrides) : backend;
}
async function detectBackend(limit) {
    const backends = await getAllBackends();
    const filtered = limit ? backends.filter(limit) : backends;
    if (!filtered.length) {
        return new NullBackend();
    }
    return filtered.reduce((selected, candidate) => candidate.priority > selected.priority ? candidate : selected);
}
/**
 * Resets all registry state for testing purposes.
 * Clears the backend cache, active backend, and backend limit.
 * Internal function used by the test suite to ensure test isolation.
 */
function __resetRegistryForTests() {
    cachedBackends = undefined;
    activeBackend = undefined;
    backendLimit = undefined;
}

const testingExports = {
    runCommand: runtime.runCommand,
    executableExists: runtime.executableExists,
    runPowerShell: runtime.runPowerShell,
    NativeKeychainBackend,
    NativeWindowsBackend,
    NativeLinuxBackend,
    MacOSKeychainBackend,
    SecretServiceBackend,
    WindowsCredentialBackend,
    FileSystemBackend,
    NullBackend,
};
const __testing = process.env.NODE_ENV === "test" ? testingExports : undefined;
/**
 * Returns the root directory where keyring data is stored.
 * This is typically platform-specific (e.g., ~/.local/share/cross-keychain on Linux).
 *
 * @returns The absolute path to the data root directory
 */
function dataRoot() {
    return getDataRoot();
}
/**
 * Returns the root directory where keyring configuration is stored.
 * This is typically platform-specific (e.g., ~/.config/cross-keychain on Linux).
 *
 * @returns The absolute path to the config root directory
 */
function configRoot() {
    return getConfigRoot();
}
/**
 * Retrieves a password from the keyring.
 * This is the primary method for reading stored credentials.
 *
 * @param service - The service identifier (e.g., "github", "npm")
 * @param account - The account/username identifier
 * @returns The stored password, or null if not found
 *
 * @example
 * ```typescript
 * const password = await getPassword("github", "myusername");
 * if (password) {
 *   console.log("Password found!");
 * }
 * ```
 */
async function getPassword(service, account) {
    const backend = await getKeyring();
    return backend.getPassword(service, account);
}
/**
 * Stores a password in the keyring.
 * If a password already exists for the service/account combination, it will be updated.
 *
 * @param service - The service identifier (e.g., "github", "npm")
 * @param account - The account/username identifier
 * @param password - The password to store
 *
 * @example
 * ```typescript
 * await setPassword("github", "myusername", "secret123");
 * ```
 */
async function setPassword(service, account, password) {
    const backend = await getKeyring();
    await backend.setPassword(service, account, password);
}
/**
 * Deletes a password from the keyring.
 * Throws an error if the password does not exist.
 *
 * @param service - The service identifier
 * @param account - The account/username identifier
 *
 * @example
 * ```typescript
 * await deletePassword("github", "myusername");
 * ```
 */
async function deletePassword(service, account) {
    const backend = await getKeyring();
    await backend.deletePassword(service, account);
}
/**
 * Retrieves a credential (username and password) from the keyring.
 * If account is not specified, returns the first credential found for the service.
 *
 * @param service - The service identifier
 * @param account - Optional account/username identifier
 * @returns The credential object with username and password, or null if not found
 *
 * @example
 * ```typescript
 * // Get specific credential
 * const cred = await getCredential("github", "myusername");
 *
 * // Get any credential for the service
 * const anyCred = await getCredential("github");
 * ```
 */
async function getCredential(service, account) {
    const backend = await getKeyring();
    return backend.getCredential(service, account);
}
/**
 * Disables credential storage by configuring the null backend.
 * Creates a configuration file that prevents any credentials from being stored.
 * Throws an error if a configuration file already exists.
 */
async function disable() {
    const file = getConfigFile();
    try {
        await fs.promises.access(file);
        throw new KeyringError(`Refusing to overwrite existing configuration at ${file}`);
    }
    catch (error) {
        if (error.code !== "ENOENT") {
            throw error;
        }
    }
    const config = {
        defaultBackend: "null",
    };
    await ensureParent(file);
    await fs.promises.writeFile(file, JSON.stringify(config, null, 2), {
        encoding: "utf8",
        mode: 0o600,
    });
}
/**
 * Returns diagnostic information about the current keyring configuration.
 * Useful for debugging and troubleshooting keyring issues.
 *
 * @returns An object containing configuration paths, active backend info, and platform details
 *
 * @example
 * ```typescript
 * const info = await diagnose();
 * console.log("Active backend:", info.name);
 * console.log("Config path:", info.configPath);
 * ```
 */
async function diagnose() {
    const backend = await getKeyring();
    const diagnosis = await backend.diagnose();
    return {
        configPath: getConfigFile(),
        dataRoot: getDataRoot(),
        ...diagnosis,
    };
}
/**
 * Resets all keyring state for testing purposes.
 * Internal function used by the test suite to ensure test isolation.
 */
function __resetKeyringStateForTests() {
    __resetRuntimeForTests();
    __resetRegistryForTests();
}
/**
 * Lists all available keyring backends with their metadata.
 * Includes backend ID, name, and priority for each detected backend.
 *
 * @returns Array of backend information objects
 *
 * @example
 * ```typescript
 * const backends = await listBackends();
 * backends.forEach(b => {
 *   console.log(`${b.id}: ${b.name} (priority: ${b.priority})`);
 * });
 * ```
 */
async function listBackends() {
    const backends = await getAllBackends();
    return backends.map((backend) => ({
        id: backend.id,
        name: backend.name,
        priority: backend.priority,
    }));
}
/**
 * Forces the use of a specific backend by ID.
 * Useful for testing or when you want to use a specific storage mechanism.
 * Throws an error if the backend is not available.
 *
 * @param backendId - The ID of the backend to use (e.g., "macos", "file", "null")
 * @param overrides - Optional configuration properties for the backend
 *
 * @example
 * ```typescript
 * // Use file backend with custom path
 * await useBackend("file", { file_path: "/custom/path/secrets.json" });
 * ```
 */
async function useBackend(backendId, overrides) {
    const backend = await loadBackendById(backendId, undefined, overrides);
    if (!backend) {
        const { NoKeyringError } = await Promise.resolve().then(function () { return errors; });
        throw new NoKeyringError(`Backend ${backendId} is not available`);
    }
    setKeyring(backend);
}

exports.InitError = InitError;
exports.KeyringError = KeyringError;
exports.KeyringLockedError = KeyringLockedError;
exports.NoKeyringError = NoKeyringError;
exports.PasswordDeleteError = PasswordDeleteError;
exports.PasswordSetError = PasswordSetError;
exports.__resetKeyringStateForTests = __resetKeyringStateForTests;
exports.__testing = __testing;
exports.configRoot = configRoot;
exports.dataRoot = dataRoot;
exports.deletePassword = deletePassword;
exports.diagnose = diagnose;
exports.disable = disable;
exports.getAllBackends = getAllBackends;
exports.getCredential = getCredential;
exports.getKeyring = getKeyring;
exports.getPassword = getPassword;
exports.initBackend = initBackend;
exports.listBackends = listBackends;
exports.registerBackend = registerBackend;
exports.setKeyring = setKeyring;
exports.setPassword = setPassword;
exports.useBackend = useBackend;
