// @zapier/zapier-sdk/dist/plugins/listInputFields/schemas.js
import { z } from "zod";
import { AppKeyPropertySchema, AppPropertySchema, ActionTypePropertySchema, ActionKeyPropertySchema, ActionPropertySchema, ConnectionIdPropertySchema, ConnectionPropertySchema, AuthenticationIdPropertySchema, InputsPropertySchema, } from "../../types/properties";
const ListInputFieldsDescription = "Get the input fields required for a specific action";
// Base schema for shared (optional) properties
const ListInputFieldsBaseSchema = z.object({
    connection: ConnectionPropertySchema.optional().describe("Connection alias (string) or numeric connectionId. Strings are resolved from the connections map; numbers are used directly. Mutually exclusive with connectionId."),
    connectionId: ConnectionIdPropertySchema.nullable()
        .optional()
        .describe("Connection ID to use when listing input fields. Required if the action needs a connection to determine available fields.")
        .meta({ deprecated: true }),
    /** @deprecated Use connection instead */
    authenticationId: AuthenticationIdPropertySchema.nullable()
        .optional()
        .meta({ deprecated: true }),
    inputs: InputsPropertySchema.optional().describe("Current input values that may affect available fields"),
    pageSize: z
        .number()
        .min(1)
        .optional()
        .describe("Number of input fields per page"),
    maxItems: z
        .number()
        .min(1)
        .optional()
        .describe("Maximum total items to return across all pages"),
    cursor: z.string().optional().describe("Cursor to start from"),
});
// Schema for docs/registry (new param names only)
export const ListInputFieldsSchema = z
    .object({
    app: AppPropertySchema,
    actionType: ActionTypePropertySchema,
    action: ActionPropertySchema,
})
    .merge(ListInputFieldsBaseSchema)
    .describe(ListInputFieldsDescription)
    .meta({
    aliases: {
        appKey: "app",
        actionKey: "action",
        connectionId: "connection",
        authenticationId: "connection",
    },
});
// Deprecated schema (backward compat)
const ListInputFieldsSchemaDeprecated = z
    .object({
    appKey: AppKeyPropertySchema,
    actionType: ActionTypePropertySchema,
    actionKey: ActionKeyPropertySchema,
})
    .merge(ListInputFieldsBaseSchema);
// Union for runtime validation (accepts either new or old names)
export const ListInputFieldsInputSchema = z
    .union([ListInputFieldsSchema, ListInputFieldsSchemaDeprecated])
    .describe(ListInputFieldsDescription);
