// @zapier/zapier-sdk/dist/resolvers/index.js
// Re-export individual resolvers for plugin use
export { appKeyResolver } from "./appKey";
export { actionTypeResolver } from "./actionType";
export { actionKeyResolver } from "./actionKey";
export { connectionIdResolver, connectionIdGenericResolver, } from "./connectionId";
// Deprecated aliases for backward compatibility
/** @deprecated Use connectionIdResolver instead */
export { connectionIdResolver as authenticationIdResolver } from "./connectionId";
/** @deprecated Use connectionIdGenericResolver instead */
export { connectionIdGenericResolver as authenticationIdGenericResolver } from "./connectionId";
export { inputsResolver } from "./inputs";
export { inputFieldKeyResolver } from "./inputFieldKey";
export { inputsAllOptionalResolver } from "./inputs";
export { clientCredentialsNameResolver } from "./clientCredentialsName";
export { clientIdResolver } from "./clientId";
export { tableIdResolver } from "./tableId";
export { tableRecordIdResolver, tableRecordIdsResolver } from "./tableRecordId";
export { tableFieldIdsResolver } from "./tableFieldIds";
export { tableNameResolver } from "./tableName";
export { tableFieldsResolver } from "./tableFields";
export { tableRecordsResolver, tableUpdateRecordsResolver, } from "./tableRecords";
export { tableFiltersResolver } from "./tableFilters";
export { tableSortResolver } from "./tableSort";
