// @zapier/zapier-sdk/dist/resolvers/inputs.js
// Helper function to recursively make all input fields optional while preserving structure
function makeFieldsOptional(fields) {
    return fields.map((field) => {
        if (field.type === "input_field") {
            return {
                ...field,
                is_required: false,
            };
        }
        else if (field.type === "fieldset") {
            return {
                ...field,
                fields: makeFieldsOptional(field.fields),
            };
        }
        else {
            // Info fields remain unchanged
            return field;
        }
    });
}
export const inputsResolver = {
    type: "fields",
    depends: ["app", "action", "actionType", "connection"],
    fetch: async (sdk, resolvedParams) => {
        const fieldsResponse = await sdk.listInputFields({
            app: resolvedParams.app,
            action: resolvedParams.action,
            actionType: resolvedParams.actionType,
            connection: resolvedParams.connection,
            inputs: resolvedParams.inputs, // Pass along currently resolved inputs
        });
        return fieldsResponse.data;
    },
};
export const inputsAllOptionalResolver = {
    type: "fields",
    depends: ["app", "action", "actionType", "connection"],
    fetch: async (sdk, resolvedParams) => {
        const fieldsResponse = await sdk.listInputFields({
            app: resolvedParams.app,
            action: resolvedParams.action,
            actionType: resolvedParams.actionType,
            connection: resolvedParams.connection,
            inputs: resolvedParams.inputs, // Pass along currently resolved inputs
        });
        // Return nested structure with all fields marked as optional.
        return makeFieldsOptional(fieldsResponse.data);
    },
};
