// @zapier/zapier-sdk/dist/resolvers/tableSort.js
import { NESTED_COMPONENTS } from "../plugins/tables/utils";
export const tableSortResolver = {
    type: "fields",
    depends: ["table"],
    fetch: async (sdk, params) => {
        const { data: fields } = await sdk.listTableFields({
            table: params.table,
        });
        const fieldChoices = fields.flatMap((field) => {
            const components = NESTED_COMPONENTS[field.type];
            if (components) {
                return [...components].map((component) => ({
                    label: `${field.name}__${component} (${field.id}, ${field.type})`,
                    value: `${field.name}__${component}`,
                }));
            }
            return {
                label: `${field.name} (${field.id})`,
                value: field.name,
            };
        });
        const fieldItems = [
            {
                type: "input_field",
                key: "fieldKey",
                title: "Field",
                is_required: true,
                choices: fieldChoices,
            },
            {
                type: "input_field",
                key: "direction",
                title: "Direction",
                is_required: true,
                choices: [
                    { label: "Ascending", value: "asc" },
                    { label: "Descending", value: "desc" },
                ],
            },
        ];
        return fieldItems;
    },
};
