// @zapier/zapier-sdk/dist/resolvers/tableFilters.js
import { NESTED_COMPONENTS } from "../plugins/tables/utils";
const FILTER_OPERATORS = [
    { label: "equals", value: "exact" },
    { label: "not equals", value: "different" },
    { label: "contains", value: "contains" },
    { label: "contains (case-insensitive)", value: "icontains" },
    { label: "starts with", value: "startswith" },
    { label: "search", value: "search" },
    { label: "greater than", value: "gt" },
    { label: "greater than or equal", value: "gte" },
    { label: "less than", value: "lt" },
    { label: "less than or equal", value: "lte" },
    { label: "in range", value: "range" },
    { label: "in list", value: "in" },
    { label: "is empty", value: "isnull" },
    { label: "is within (date)", value: "is_within" },
];
function createSingleFilterResolver({ sdk, table, }) {
    return {
        type: "fields",
        fetch: async () => {
            const { data: fields } = await sdk.listTableFields({ table });
            const fieldChoices = fields.flatMap((field) => {
                const components = NESTED_COMPONENTS[field.type];
                if (components && components.size > 1) {
                    return [...components].map((component) => ({
                        label: `${field.name}__${component} (${field.id}, ${field.type})`,
                        value: `${field.name}__${component}`,
                    }));
                }
                return {
                    label: `${field.name} (${field.id}${components ? `, ${field.type}` : ""})`,
                    value: components
                        ? `${field.name}__${[...components][0]}`
                        : field.name,
                };
            });
            const fieldItems = [
                {
                    type: "input_field",
                    key: "fieldKey",
                    title: "Field",
                    is_required: true,
                    choices: fieldChoices,
                },
                {
                    type: "input_field",
                    key: "operator",
                    title: "Operator",
                    is_required: true,
                    choices: FILTER_OPERATORS,
                },
                {
                    type: "input_field",
                    key: "value",
                    title: "Value",
                    is_required: false,
                },
            ];
            return fieldItems;
        },
    };
}
export const tableFiltersResolver = {
    type: "array",
    depends: ["table"],
    fetch: async (sdk, params) => {
        return createSingleFilterResolver({
            sdk,
            table: params.table,
        });
    },
    minItems: 0,
};
