// @zapier/zapier-sdk/dist/resolvers/tableRecords.js
import { tableRecordIdResolver } from "./tableRecordId";
function tableFieldValueType(fieldType) {
    if (fieldType.startsWith("multiple_"))
        return "array";
    switch (fieldType) {
        case "number":
        case "decimal":
        case "currency":
            return "number";
        case "boolean":
            return "boolean";
        default:
            return undefined;
    }
}
// Creates an inner resolver that prompts for a single record's field values.
// When includeRecordId is true, also prompts for a record ID (for updates).
function createSingleRecordResolver({ sdk, table, includeRecordId, }) {
    return {
        type: "fields",
        fetch: async () => {
            const { data: fields } = await sdk.listTableFields({ table });
            const inputFields = fields.map((field) => ({
                type: "input_field",
                key: field.id,
                title: field.name,
                is_required: false,
                ...(tableFieldValueType(field.type) && {
                    value_type: tableFieldValueType(field.type),
                }),
            }));
            if (includeRecordId) {
                return [
                    {
                        type: "input_field",
                        key: "_record_id",
                        title: "Record",
                        is_required: true,
                        resolver: tableRecordIdResolver,
                    },
                    ...inputFields,
                ];
            }
            return inputFields;
        },
        transform: (values) => {
            if (includeRecordId) {
                const { _record_id: id, ...data } = values;
                return { id, data };
            }
            return { data: values };
        },
    };
}
export const tableRecordsResolver = {
    type: "array",
    depends: ["table"],
    fetch: async (sdk, params) => {
        return createSingleRecordResolver({
            sdk,
            table: params.table,
        });
    },
    minItems: 1,
};
export const tableUpdateRecordsResolver = {
    type: "array",
    depends: ["table"],
    fetch: async (sdk, params) => {
        return createSingleRecordResolver({
            sdk,
            table: params.table,
            includeRecordId: true,
        });
    },
    minItems: 1,
};
