// @zapier/zapier-sdk/dist/plugins/findUniqueConnection/index.js
import { FindUniqueConnectionSchema, } from "./schemas";
import { ZapierResourceNotFoundError, ZapierValidationError, } from "../../types/errors";
import { createFunction } from "../../utils/function-utils";
import { ConnectionItemSchema } from "../../schemas/Connection";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
export const findUniqueConnectionPlugin = ({ sdk, context }) => {
    async function findUniqueConnection(options = {}) {
        // Use listConnections with maxItems 2 to check for uniqueness
        const connectionsResponse = await sdk.listConnections({
            ...options,
            maxItems: 2, // Get up to 2 to check for uniqueness
        });
        if (connectionsResponse.data.length === 0) {
            throw new ZapierResourceNotFoundError("No connection found matching the specified criteria", { resourceType: "Connection" });
        }
        if (connectionsResponse.data.length > 1) {
            throw new ZapierValidationError("Multiple connections found matching the specified criteria. Expected exactly one.");
        }
        return {
            data: connectionsResponse.data[0],
        };
    }
    const findUniqueConnectionDefinition = createFunction(findUniqueConnection, FindUniqueConnectionSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, findUniqueConnection.name));
    return {
        findUniqueConnection: findUniqueConnectionDefinition,
        context: {
            meta: {
                findUniqueConnection: {
                    categories: ["connection"],
                    type: "item",
                    itemType: "Connection",
                    inputSchema: FindUniqueConnectionSchema,
                    outputSchema: ConnectionItemSchema,
                },
            },
        },
    };
};
