// @zapier/zapier-sdk/dist/plugins/tables/deleteTableFields/index.js
import { createFunction } from "../../../utils/function-utils";
import { DeleteTableFieldsOptionsSchema, DeleteTableFieldsOptionsInputSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver, tableFieldIdsResolver } from "../../../resolvers";
import { resolveFieldKeys } from "../utils";
export const deleteTableFieldsPlugin = ({ context }) => {
    async function deleteTableFields(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const fieldKeys = "fields" in options ? options.fields : options.fieldKeys;
        const numericFieldIds = await resolveFieldKeys({
            api,
            tableId,
            fieldKeys,
        });
        await api.delete(`/tables/api/v1/tables/${tableId}/fields`, { field_ids: numericFieldIds }, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to delete fields or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to delete fields. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
        });
        return {
            success: true,
        };
    }
    const deleteTableFieldsDefinition = createFunction(deleteTableFields, DeleteTableFieldsOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, deleteTableFields.name));
    return {
        deleteTableFields: deleteTableFieldsDefinition,
        context: {
            meta: {
                deleteTableFields: {
                    categories: ["table"],
                    type: "delete",
                    itemType: "Field",
                    inputSchema: DeleteTableFieldsOptionsSchema,
                    resolvers: {
                        table: tableIdResolver,
                        fields: tableFieldIdsResolver,
                    },
                    confirm: "delete",
                },
            },
        },
    };
};
