// @zapier/zapier-sdk/dist/plugins/tables/deleteTableFields/schemas.js
import { z } from "zod";
import { TablePropertySchema, FieldsPropertySchema, } from "../../../types/properties";
const DeleteTableFieldsDescription = "Delete one or more fields from a table";
// Schema for docs/registry (new param names only)
export const DeleteTableFieldsOptionsSchema = z
    .object({
    table: TablePropertySchema,
    fields: FieldsPropertySchema.min(1),
})
    .describe(DeleteTableFieldsDescription)
    .meta({ aliases: { tableId: "table", fieldKeys: "fields" } });
// Deprecated schema (backward compat)
const DeleteTableFieldsOptionsSchemaDeprecated = z.object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
    fieldKeys: z
        .array(z.union([z.string(), z.number()]))
        .min(1)
        .describe('Fields to delete. Accepts field names (e.g., "Email") or IDs (e.g., "f6", "6", or 6).'),
});
// Union for runtime validation
export const DeleteTableFieldsOptionsInputSchema = z
    .union([
    DeleteTableFieldsOptionsSchema,
    DeleteTableFieldsOptionsSchemaDeprecated,
])
    .describe(DeleteTableFieldsDescription);
