// @zapier/zapier-sdk-cli/dist/src/utils/cli-renderer.js
import { ZapierCliMissingParametersError, ZapierCliExitError } from "./errors";
import { formatErrorMessage, ZapierError } from "@zapier/zapier-sdk";
import { formatItemsFromSchema, formatJsonOutput } from "./schema-formatter";
import { toKebabCase } from "./string-utils";
import chalk from "chalk";
import inquirer from "inquirer";
// ============================================================================
// Shared error formatting
// ============================================================================
function formatMissingParamsError(error) {
    return [
        "Missing required parameters:",
        ...error.params.map(({ name, isPositional }) => isPositional
            ? `  • <${toKebabCase(name)}>`
            : `  • --${toKebabCase(name)}`),
    ].join("\n");
}
// ============================================================================
// Shared error serialization (used by JSON renderer)
// ============================================================================
export function buildJsonErrors(error) {
    if (error instanceof ZapierCliMissingParametersError) {
        return error.params.map(({ name, isPositional }) => ({
            code: error.code,
            message: isPositional
                ? `<${toKebabCase(name)}> is required in non-interactive mode`
                : `--${toKebabCase(name)} is required in non-interactive mode`,
        }));
    }
    const code = error instanceof ZapierError ? error.code : "UNKNOWN_ERROR";
    const message = error instanceof Error ? error.message : String(error);
    return [{ code, message }];
}
// ============================================================================
// HTTP response unwrapping
// ============================================================================
async function unwrapHttpResponse(response) {
    const text = await response.text().catch(() => "[unable to read body]");
    try {
        return {
            statusCode: response.status,
            headers: Object.fromEntries(response.headers.entries()),
            body: JSON.parse(text),
        };
    }
    catch {
        throw new Error(`Failed to parse response as JSON (status: ${response.status}). Body: ${text.slice(0, 500)}`);
    }
}
// ============================================================================
// JSON renderer
// ============================================================================
function outputJson(envelope) {
    console.log(JSON.stringify(envelope, null, 2));
}
export function createJsonRenderer() {
    return {
        async renderPaginatedList(source, _functionInfo) {
            let page;
            if (Symbol.asyncIterator in Object(source)) {
                const iter = source[Symbol.asyncIterator]();
                const result = await iter.next();
                page = result.done ? undefined : result.value;
            }
            else {
                page = await source;
            }
            const data = Array.isArray(page?.data) ? page.data : [];
            outputJson({
                data,
                ...(page?.nextCursor && { nextCursor: page.nextCursor }),
                errors: [],
            });
        },
        renderCollectedList(items, _options) {
            outputJson({ data: items, errors: [] });
        },
        renderItem(value, options) {
            if (value === undefined && !options?.outputFile) {
                return;
            }
            outputJson({
                data: options?.outputFile
                    ? { outputFile: options.outputFile }
                    : (value ?? null),
                errors: [],
            });
        },
        async renderResponse(response) {
            // Leave envelope as is for backwards compatibility
            // until we fully deprecate fetch from CLI
            console.log(JSON.stringify(await unwrapHttpResponse(response), null, 2));
        },
        renderError(error) {
            outputJson({ data: null, errors: buildJsonErrors(error) });
            throw new ZapierCliExitError(error instanceof Error ? error.message : String(error), 1);
        },
    };
}
// ============================================================================
// Interactive renderer
// ============================================================================
function getItemName(functionInfo) {
    if (functionInfo?.itemType)
        return `${functionInfo.itemType} items`;
    return "items";
}
function getListTitle(functionInfo) {
    if (functionInfo.itemType)
        return `Available ${functionInfo.itemType} items`;
    return "items";
}
function renderItemsForDisplay(items, functionInfo, startingNumber = 0) {
    if (functionInfo?.inputSchema) {
        formatItemsFromSchema(functionInfo, items, startingNumber);
    }
    else {
        items.forEach((item, index) => {
            const obj = item;
            const name = obj?.name || obj?.key || obj?.id || "Item";
            console.log(`${chalk.gray(`${startingNumber + index + 1}.`)} ${chalk.cyan(String(name))}`);
            if (obj?.description)
                console.log(`   ${chalk.dim(String(obj.description))}`);
            console.log();
        });
    }
}
export function createInteractiveRenderer() {
    return {
        async renderPaginatedList(source, functionInfo) {
            const itemName = getItemName(functionInfo);
            if (!(Symbol.asyncIterator in Object(source))) {
                const items = source?.data;
                if (!Array.isArray(items) || items.length === 0) {
                    console.log(chalk.yellow(`No ${itemName} found.`));
                    return;
                }
                renderItemsForDisplay(items, functionInfo, 0);
                console.log(chalk.green(`\n✅ Showing ${items.length} ${itemName}`));
                return;
            }
            let totalShown = 0;
            let pageCount = 0;
            console.log(chalk.blue(`📋 ${getListTitle(functionInfo)}\n`));
            for await (const page of source) {
                const items = page.data ?? [];
                pageCount++;
                if (items.length === 0 && pageCount === 1) {
                    console.log(chalk.yellow(`No ${itemName} found.`));
                    return;
                }
                if (items.length === 0)
                    break;
                if (pageCount > 1) {
                    console.clear();
                    console.log(chalk.blue(`📋 ${getListTitle(functionInfo)}\n`));
                }
                renderItemsForDisplay(items, functionInfo, totalShown);
                totalShown += items.length;
                console.log(chalk.green(`\n✅ Showing ${totalShown} ${itemName} (page ${pageCount})`));
                if (page.nextCursor) {
                    const { continueReading } = await inquirer.prompt([
                        {
                            type: "confirm",
                            name: "continueReading",
                            message: "Load next page?",
                            default: true,
                        },
                    ]);
                    if (!continueReading)
                        break;
                }
                else {
                    break;
                }
            }
            console.log(chalk.gray(`\n📄 Finished browsing ${itemName}`));
        },
        renderCollectedList(items, { maxItems, userSpecifiedMaxItems, functionInfo } = {}) {
            if (!Array.isArray(items)) {
                formatJsonOutput(items);
                return;
            }
            const itemName = getItemName(functionInfo);
            if (items.length === 0) {
                console.log(chalk.yellow(`No ${itemName} found.`));
                return;
            }
            console.log(chalk.green(`\n✅ Found ${items.length} ${itemName}:\n`));
            renderItemsForDisplay(items, functionInfo);
            if (userSpecifiedMaxItems && maxItems) {
                console.log(chalk.gray(`\n📄 Showing up to ${maxItems} ${itemName} (--max-items ${maxItems})`));
            }
            else {
                console.log(chalk.gray(`\n📄 All available ${itemName} shown`));
            }
        },
        renderItem(value, options) {
            if (options?.outputFile) {
                const label = options.commandName
                    ? `✅ ${options.commandName} completed successfully!`
                    : "✅ Done!";
                console.log(chalk.green(label));
                console.log(chalk.gray(`Output written to: ${options.outputFile}`));
            }
            else {
                formatJsonOutput(value);
            }
        },
        async renderResponse(response) {
            formatJsonOutput(await unwrapHttpResponse(response));
        },
        renderError(error) {
            if (error instanceof ZapierCliMissingParametersError) {
                console.error(chalk.red("❌ " + formatMissingParamsError(error)));
                console.error("\n" + chalk.dim("Use --help to see available options"));
                throw new ZapierCliExitError(error.message, 1);
            }
            if (error instanceof ZapierError) {
                const formattedMessage = formatErrorMessage(error);
                console.error(chalk.red("❌ Error:"), formattedMessage);
                throw new ZapierCliExitError(formattedMessage, 1);
            }
            const msg = error instanceof Error ? error.message : "Unknown error";
            console.error(chalk.red("❌ Error:"), msg);
            throw new ZapierCliExitError(msg, 1);
        },
    };
}
