// @zapier/zapier-sdk-cli/dist/src/utils/cli-generator-utils.js
import { z } from "zod";
import { isPositional } from "@zapier/zapier-sdk";
// ============================================================================
// Schema Analysis
// ============================================================================
export function analyzeZodSchema(schema, functionInfo) {
    const parameters = [];
    if (schema instanceof z.ZodObject) {
        const shape = schema.shape;
        for (const [key, fieldSchema] of Object.entries(shape)) {
            const param = analyzeZodField(key, fieldSchema, functionInfo);
            if (param) {
                parameters.push(param);
            }
        }
    }
    return parameters;
}
function analyzeZodField(name, schema, functionInfo) {
    let baseSchema = schema;
    let required = true;
    let defaultValue = undefined;
    // Unwrap optional and default wrappers
    if (baseSchema instanceof z.ZodOptional) {
        required = false;
        baseSchema = baseSchema._zod.def.innerType;
    }
    if (baseSchema instanceof z.ZodDefault) {
        required = false;
        defaultValue = baseSchema._zod.def.defaultValue();
        baseSchema = baseSchema._zod.def.innerType;
    }
    // Determine parameter type
    let paramType = "string";
    let choices;
    if (baseSchema instanceof z.ZodString) {
        paramType = "string";
    }
    else if (baseSchema instanceof z.ZodNumber) {
        paramType = "number";
    }
    else if (baseSchema instanceof z.ZodBoolean) {
        paramType = "boolean";
    }
    else if (baseSchema instanceof z.ZodArray) {
        paramType = "array";
    }
    else if (baseSchema instanceof z.ZodEnum) {
        paramType = "string";
        choices = baseSchema.options;
    }
    else if (baseSchema instanceof z.ZodRecord) {
        // Handle Record<string, any> as JSON string input
        paramType = "string";
    }
    // Check if this parameter has a resolver
    let paramHasResolver = false;
    // Check function-specific resolvers first
    if (functionInfo?.resolvers?.[name]) {
        paramHasResolver = true;
    }
    // Extract resolver metadata
    return {
        name,
        type: paramType,
        required,
        description: schema.description,
        default: defaultValue,
        choices,
        hasResolver: paramHasResolver,
        isPositional: isPositional(schema),
    };
}
// ============================================================================
// Parameter Conversion
// ============================================================================
export function convertCliArgsToSdkParams(parameters, positionalArgs, options) {
    const sdkParams = {};
    // Handle positional arguments (required parameters or optional positional parameters)
    let argIndex = 0;
    parameters.forEach((param) => {
        if ((param.required || param.isPositional) &&
            argIndex < positionalArgs.length) {
            // Use the original camelCase parameter name for the SDK
            sdkParams[param.name] = convertValue(positionalArgs[argIndex], param.type);
            argIndex++;
        }
    });
    // Handle option flags
    Object.entries(options).forEach(([key, value]) => {
        // Convert kebab-case back to camelCase
        const camelKey = key.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
        const param = parameters.find((p) => p.name === camelKey);
        if (param && value !== undefined) {
            sdkParams[camelKey] = convertValue(value, param.type);
        }
    });
    return sdkParams;
}
function convertValue(value, type) {
    switch (type) {
        case "number":
            return Number(value);
        case "boolean":
            return Boolean(value);
        case "array":
            return Array.isArray(value) ? value : [value];
        case "string":
        default:
            // Handle JSON string for objects
            if (typeof value === "string" &&
                (value.startsWith("{") || value.startsWith("["))) {
                try {
                    return JSON.parse(value);
                }
                catch {
                    return value;
                }
            }
            return value;
    }
}
