// @zapier/zapier-sdk-cli/dist/src/utils/cli-options.js
export var ReservedCliParameter;
(function (ReservedCliParameter) {
    ReservedCliParameter["Version"] = "version";
    ReservedCliParameter["Help"] = "help";
})(ReservedCliParameter || (ReservedCliParameter = {}));
/** CLI options at the framework level, not defined by any SDK command. */
export const RESERVED_CLI_OPTIONS = [
    {
        name: ReservedCliParameter.Version,
        flag: "--version",
        short: "-V",
        description: "Display version number",
    },
    {
        name: ReservedCliParameter.Help,
        flag: "--help",
        short: "-h",
        description: "Display help for command",
    },
];
export function getReservedCliOption(name) {
    const option = RESERVED_CLI_OPTIONS.find((opt) => opt.name === name);
    if (!option)
        throw new Error(`Reserved CLI option not found: ${name}`);
    return option;
}
/** CLI options automatically added to every SDK-generated command. */
export const SHARED_COMMAND_CLI_OPTIONS = [
    {
        name: "json",
        flag: "--json",
        description: "Output raw JSON instead of formatted results",
    },
];
