// @zapier/zapier-sdk-cli/dist/src/utils/errors.js
import { ZapierError } from "@zapier/zapier-sdk";
export class ZapierCliError extends ZapierError {
}
export class ZapierCliUserCancellationError extends ZapierCliError {
    constructor(message = "Operation cancelled by user") {
        super(message);
        this.name = "ZapierCliUserCancellationError";
        this.code = "ZAPIER_CLI_USER_CANCELLATION";
        this.exitCode = 0;
    }
}
export class ZapierCliExitError extends ZapierCliError {
    constructor(message, exitCode = 1) {
        super(message);
        this.name = "ZapierCliExitError";
        this.code = "ZAPIER_CLI_EXIT";
        this.exitCode = exitCode;
    }
}
export class ZapierCliValidationError extends ZapierCliError {
    constructor(message) {
        super(message);
        this.name = "ZapierCliValidationError";
        this.code = "ZAPIER_CLI_VALIDATION_ERROR";
        this.exitCode = 1;
    }
}
export class ZapierCliMissingParametersError extends ZapierCliError {
    constructor(params) {
        super(`Missing required parameters: ${params.map((p) => p.name).join(", ")}`);
        this.name = "ZapierCliMissingParametersError";
        this.code = "ZAPIER_CLI_MISSING_PARAMETERS";
        this.exitCode = 1;
        this.params = params;
    }
}
