// @zapier/zapier-sdk-cli/dist/src/utils/parameter-resolver.js
import inquirer from "inquirer";
import chalk from "chalk";
import ora from "ora";
import { z } from "zod";
import { runWithTelemetryContext, } from "@zapier/zapier-sdk";
import { isPositional, buildCapabilityMessage } from "@zapier/zapier-sdk";
import { ZapierCliUserCancellationError, ZapierCliMissingParametersError, ZapierCliValidationError, } from "./errors";
// Zod validation errors are structured objects; flatten them into a
// readable "field: reason" string for CLI error messages.
function formatZodError(error) {
    return error.issues
        .map((issue) => {
        const field = issue.path.length > 0 ? issue.path.join(".") : "input";
        return `${field}: ${issue.message}`;
    })
        .join(", ");
}
// ============================================================================
// Local Resolution Helper Functions
// ============================================================================
/**
 * Resolve dependency chain for a parameter using local resolvers
 * Returns parameters in the order they need to be resolved
 */
function getLocalResolutionOrder(paramName, resolvers, resolved = new Set()) {
    const resolver = resolvers[paramName];
    if (!resolver || resolver.type === "static") {
        return [paramName];
    }
    const order = [];
    if ("depends" in resolver && resolver.depends) {
        for (const dependency of resolver.depends) {
            if (!resolved.has(dependency)) {
                order.push(...getLocalResolutionOrder(dependency, resolvers, resolved));
                resolved.add(dependency);
            }
        }
    }
    if (!resolved.has(paramName)) {
        order.push(paramName);
        resolved.add(paramName);
    }
    return order;
}
/**
 * Get resolution order for multiple parameters using local resolvers
 */
function getLocalResolutionOrderForParams(paramNames, resolvers) {
    const resolved = new Set();
    const order = [];
    for (const paramName of paramNames) {
        const paramOrder = getLocalResolutionOrder(paramName, resolvers, resolved);
        for (const param of paramOrder) {
            if (!order.includes(param)) {
                order.push(param);
            }
        }
    }
    return order;
}
// ============================================================================
// Schema Parameter Resolver
// ============================================================================
export class SchemaParameterResolver {
    constructor() {
        this.debug = false;
        this.spinner = null;
    }
    debugLog(message) {
        if (this.debug) {
            this.stopSpinner();
            console.log(chalk.gray(`[Zapier CLI] ${message}`));
        }
    }
    startSpinner() {
        if (!this.debug && !this.spinner) {
            this.spinner = ora({ text: "", spinner: "dots" }).start();
        }
    }
    stopSpinner() {
        if (this.spinner) {
            this.spinner.stop();
            this.spinner = null;
        }
    }
    async resolveParameters(schema, providedParams, sdk, functionName, options) {
        return runWithTelemetryContext(async () => {
            this.debug = options?.debug ?? false;
            const interactiveMode = (options?.interactiveMode ?? true) && !!process.stdin.isTTY;
            // 1. Try to parse with current parameters
            const parseResult = schema.safeParse(providedParams);
            // Get all schema parameters to check which ones have resolvers
            const allParams = this.extractParametersFromSchema(schema);
            const resolvableParams = allParams.filter((param) => this.hasResolver(param.name, sdk, functionName));
            // Get all missing parameters that have resolvers
            const missingResolvable = resolvableParams.filter((param) => {
                const hasValue = this.getNestedValue(providedParams, param.path) !== undefined;
                return !hasValue;
            });
            // Split missing resolvable params into required vs optional.
            // "inputs" is treated as required in interactive mode.
            const required = missingResolvable.filter((param) => {
                if (param.isRequired)
                    return true;
                if (param.name === "inputs")
                    return interactiveMode;
                return false;
            });
            const optional = missingResolvable.filter((param) => !required.includes(param));
            if (parseResult.success &&
                required.length === 0 &&
                optional.length === 0) {
                return parseResult.data;
            }
            if (required.length === 0 && optional.length === 0) {
                if (!parseResult.success) {
                    throw new ZapierCliValidationError(formatZodError(parseResult.error));
                }
                return parseResult.data;
            }
            // 2. Resolve required parameters
            const resolvedParams = { ...providedParams };
            const context = {
                sdk,
                currentParams: providedParams,
                resolvedParams,
                functionName,
            };
            const localResolvers = this.getLocalResolvers(sdk, functionName);
            if (required.length > 0) {
                const requiredParamNames = required.map((p) => p.name);
                const requiredResolutionOrder = getLocalResolutionOrderForParams(requiredParamNames, localResolvers);
                // Find all parameters that need to be resolved (including dependencies)
                const orderedRequiredParams = requiredResolutionOrder
                    .map((paramName) => {
                    let param = required.find((p) => p.name === paramName);
                    if (!param) {
                        param = optional.find((p) => p.name === paramName);
                    }
                    return param;
                })
                    .filter((param) => param !== undefined);
                if (!interactiveMode) {
                    await this.resolveRequiredParamsNonInteractive(orderedRequiredParams, context, resolvedParams, functionName);
                }
                else {
                    for (const param of orderedRequiredParams) {
                        try {
                            const value = await this.resolveParameter(param, context, functionName, { isOptional: !param.isRequired });
                            if (param.isRequired || value !== undefined) {
                                this.setNestedValue(resolvedParams, param.path, value);
                                context.resolvedParams = resolvedParams;
                            }
                        }
                        catch (error) {
                            if (this.isUserCancellation(error)) {
                                console.log(chalk.yellow("\n\nOperation cancelled by user"));
                                throw new ZapierCliUserCancellationError();
                            }
                            throw error;
                        }
                    }
                }
                // Remove resolved dependencies from optional to avoid double-prompting
                const resolvedParamNames = new Set(orderedRequiredParams.map((p) => p.name));
                optional.splice(0, optional.length, ...optional.filter((p) => !resolvedParamNames.has(p.name)));
            }
            // 3. Resolve optional parameters individually (each is skippable).
            // Skipped in non-interactive mode.
            if (interactiveMode && optional.length > 0) {
                const optionalParamNames = optional.map((p) => p.name);
                const optionalResolutionOrder = getLocalResolutionOrderForParams(optionalParamNames, localResolvers);
                const orderedOptionalParams = optionalResolutionOrder
                    .map((paramName) => optional.find((p) => p.name === paramName))
                    .filter((param) => param !== undefined);
                for (const param of orderedOptionalParams) {
                    try {
                        const value = await this.resolveParameter(param, context, functionName, { isOptional: true });
                        if (value !== undefined) {
                            this.setNestedValue(resolvedParams, param.path, value);
                            context.resolvedParams = resolvedParams;
                        }
                    }
                    catch (error) {
                        if (this.isUserCancellation(error)) {
                            console.log(chalk.yellow("\n\nOperation cancelled by user"));
                            throw new ZapierCliUserCancellationError();
                        }
                        throw error;
                    }
                }
            }
            // 4. Validate final parameters
            const finalResult = schema.safeParse(resolvedParams);
            if (!finalResult.success) {
                throw new ZapierCliValidationError(`Parameter validation failed: ${formatZodError(finalResult.error)}`);
            }
            return finalResult.data;
        });
    }
    extractParametersFromSchema(schema) {
        const parameters = [];
        // Only handle ZodObject at the top level
        if (schema instanceof z.ZodObject) {
            const shape = schema.shape;
            for (const [key, fieldSchema] of Object.entries(shape)) {
                const param = this.analyzeFieldSchema(key, fieldSchema);
                if (param) {
                    parameters.push(param);
                }
            }
        }
        return parameters;
    }
    analyzeFieldSchema(fieldName, fieldSchema) {
        let baseSchema = fieldSchema;
        let isRequired = true;
        // Check if field is optional or has default
        if (baseSchema instanceof z.ZodOptional) {
            isRequired = false;
            baseSchema = baseSchema._zod.def.innerType;
        }
        if (baseSchema instanceof z.ZodDefault) {
            isRequired = false;
            baseSchema = baseSchema._zod.def.innerType;
        }
        return this.createResolvableParameter([fieldName], baseSchema, isRequired);
    }
    createResolvableParameter(path, schema, isRequired) {
        if (path.length === 0)
            return null;
        const name = path[path.length - 1];
        return {
            name,
            path,
            schema,
            description: schema.description,
            isRequired,
        };
    }
    /**
     * Calls `tryResolveWithoutPrompt` on a dynamic resolver.
     * Returns the resolution result object, or null if unresolvable / throws.
     * Note: { resolvedValue: null } is a valid result (e.g. connectionId when app has no auth);
     * only a null return from the resolver itself means "could not auto-resolve".
     */
    async tryAutoResolve(dynamicResolver, context) {
        if (!dynamicResolver.tryResolveWithoutPrompt)
            return null;
        try {
            return await dynamicResolver.tryResolveWithoutPrompt(context.sdk, context.resolvedParams);
        }
        catch (err) {
            console.warn(`Auto-resolver threw unexpectedly; treating as unresolved. Error: ${err instanceof Error ? err.message : String(err)}`);
            return null;
        }
    }
    /**
     * Non-interactive resolution: auto-resolves what it can via tryAutoResolve,
     * throws ZapierCliMissingParametersError for anything that requires user input.
     */
    async resolveRequiredParamsNonInteractive(params, context, resolvedParams, functionName) {
        const missingParams = [];
        for (const param of params) {
            const resolver = this.getResolver(param.name, context.sdk, functionName);
            const autoResolution = resolver?.type === "dynamic"
                ? await this.tryAutoResolve(resolver, context)
                : null;
            if (autoResolution != null) {
                this.setNestedValue(resolvedParams, param.path, autoResolution.resolvedValue);
                context.resolvedParams = resolvedParams;
            }
            else {
                missingParams.push({
                    name: param.name,
                    // Required params render as positional CLI args (<name>); so do explicitly positional optional params.
                    isPositional: param.isRequired || isPositional(param.schema),
                });
            }
        }
        if (missingParams.length > 0) {
            throw new ZapierCliMissingParametersError(missingParams);
        }
    }
    async resolveParameter(param, context, functionName, options) {
        const resolver = this.getResolver(param.name, context.sdk, functionName);
        if (!resolver) {
            throw new Error(`No resolver found for parameter: ${param.name}`);
        }
        return this.resolveWithResolver(resolver, param, context, {
            isOptional: options?.isOptional,
        });
    }
    async resolveWithResolver(resolver, param, context, options = {}) {
        const { arrayIndex, isOptional } = options;
        const inArrayContext = arrayIndex != null;
        const promptLabel = inArrayContext
            ? `${param.name}[${arrayIndex}]`
            : param.name;
        const promptName = inArrayContext ? "value" : param.name;
        this.debugLog(`Resolving ${promptLabel}${isOptional ? " (optional)" : ""}`);
        if (resolver.type === "static") {
            const staticResolver = resolver;
            const promptConfig = {
                type: staticResolver.inputType === "password" ? "password" : "input",
                name: promptName,
                message: `Enter ${promptLabel}${isOptional ? " (optional)" : ""}:`,
                ...(staticResolver.placeholder && {
                    default: staticResolver.placeholder,
                }),
            };
            this.stopSpinner();
            const answers = await inquirer.prompt([promptConfig]);
            const value = answers[promptName];
            if (isOptional &&
                (value === undefined ||
                    value === "" ||
                    value === staticResolver.placeholder)) {
                return undefined;
            }
            return value;
        }
        else if (resolver.type === "dynamic") {
            const dynamicResolver = resolver;
            this.startSpinner();
            const autoResolution = await this.tryAutoResolve(dynamicResolver, context);
            if (autoResolution != null) {
                this.stopSpinner();
                return autoResolution.resolvedValue;
            }
            this.debugLog(`Fetching options for ${promptLabel}`);
            const fetchResult = await dynamicResolver.fetch(context.sdk, context.resolvedParams);
            // Check if the result is an AsyncIterable of pages (from toIterable())
            let pageIterator = null;
            let items;
            let hasMore = false;
            if (fetchResult != null &&
                typeof fetchResult === "object" &&
                Symbol.asyncIterator in fetchResult) {
                // Paginated: pull the first page from the iterator
                pageIterator = fetchResult[Symbol.asyncIterator]();
                const first = await pageIterator.next();
                if (!first.done && first.value) {
                    items = first.value.data;
                    hasMore = first.value.nextCursor != null;
                }
                else {
                    items = [];
                }
            }
            else if (fetchResult != null &&
                typeof fetchResult === "object" &&
                "data" in fetchResult) {
                const page = fetchResult;
                items = page.data;
                hasMore = page.nextCursor != null;
                if (hasMore) {
                    this.debugLog(`Resolver for ${promptLabel} has more pages but no iterator. ` +
                        `Use toIterable() to enable "Load more..." support.`);
                }
            }
            else {
                items = fetchResult || [];
                pageIterator = null;
            }
            const LOAD_MORE_SENTINEL = Symbol("LOAD_MORE");
            const SKIP_SENTINEL = Symbol("SKIP");
            let newItemsStartIndex = -1;
            this.stopSpinner();
            while (true) {
                const promptConfig = dynamicResolver.prompt(items, context.resolvedParams);
                promptConfig.name = promptName;
                if (isOptional && promptConfig.choices) {
                    promptConfig.choices.unshift({
                        name: chalk.dim("(Skip)"),
                        value: SKIP_SENTINEL,
                    });
                }
                // Add "Load more..." option if paginated
                if (hasMore && pageIterator && promptConfig.choices) {
                    promptConfig.choices.push({
                        name: chalk.dim("(Load more...)"),
                        value: LOAD_MORE_SENTINEL,
                    });
                }
                // Show hints for capabilities that would expand results
                if (!hasMore &&
                    promptConfig.choices &&
                    dynamicResolver.requireCapabilities) {
                    const capContext = context.sdk.getContext();
                    if (capContext.hasCapability) {
                        for (const cap of dynamicResolver.requireCapabilities) {
                            const enabled = await capContext.hasCapability(cap);
                            if (!enabled) {
                                promptConfig.choices.push({
                                    name: chalk.dim(buildCapabilityMessage(cap)),
                                    value: SKIP_SENTINEL,
                                    disabled: true,
                                });
                            }
                        }
                    }
                }
                // After loading more, jump cursor to first new item.
                // Offset by the number of injected choices before the data items (e.g. Skip).
                if (newItemsStartIndex >= 0 && promptConfig.choices) {
                    const injectedBefore = isOptional ? 1 : 0;
                    const adjustedIndex = newItemsStartIndex + injectedBefore;
                    if (promptConfig.choices[adjustedIndex]) {
                        promptConfig.default = promptConfig.choices[adjustedIndex].value;
                    }
                    newItemsStartIndex = -1;
                }
                const answers = await inquirer.prompt([promptConfig]);
                let selected = answers[promptName];
                if (selected === SKIP_SENTINEL) {
                    return undefined;
                }
                const wantsMore = Array.isArray(selected)
                    ? selected.includes(LOAD_MORE_SENTINEL)
                    : selected === LOAD_MORE_SENTINEL;
                if (wantsMore && pageIterator) {
                    if (Array.isArray(selected)) {
                        selected = selected.filter((v) => v !== LOAD_MORE_SENTINEL);
                    }
                    const prevLength = items.length;
                    this.startSpinner();
                    this.debugLog("Fetching more options...");
                    const next = await pageIterator.next();
                    this.stopSpinner();
                    if (!next.done && next.value) {
                        items = [...items, ...next.value.data];
                        hasMore = next.value.nextCursor != null;
                        newItemsStartIndex = prevLength;
                    }
                    else {
                        hasMore = false;
                    }
                    continue;
                }
                return selected;
            }
        }
        else if (resolver.type === "fields") {
            if (isOptional && !inArrayContext) {
                this.stopSpinner();
                const { confirm } = await inquirer.prompt([
                    {
                        type: "confirm",
                        name: "confirm",
                        message: `Add ${promptLabel}?`,
                        default: false,
                    },
                ]);
                if (!confirm) {
                    return undefined;
                }
            }
            return await this.resolveFieldsRecursively(resolver, context, param, { inArrayContext });
        }
        else if (resolver.type === "array") {
            return await this.resolveArrayRecursively(resolver, context, param);
        }
        throw new Error(`Unknown resolver type for ${promptLabel}`);
    }
    async resolveFieldsRecursively(resolver, context, param, options = {}) {
        const inputs = {};
        let processedFieldKeys = new Set();
        let iteration = 0;
        const maxIterations = 10; // Prevent infinite loops
        while (iteration < maxIterations) {
            iteration++;
            // Update context with current inputs so they're passed to listInputFields
            const updatedContext = {
                ...context,
                resolvedParams: {
                    ...context.resolvedParams,
                    inputs,
                },
            };
            this.debugLog(`Fetching input fields for ${param.name}${iteration > 1 ? ` (iteration ${iteration})` : ""}`);
            this.startSpinner();
            const rootFieldItems = await resolver.fetch(updatedContext.sdk, updatedContext.resolvedParams);
            this.stopSpinner();
            if (!rootFieldItems || rootFieldItems.length === 0) {
                if (iteration === 1) {
                    console.log(chalk.yellow(`No input fields required for this action.`));
                }
                break;
            }
            // Process fields recursively, maintaining fieldset structure
            const fieldStats = await this.processFieldItems(rootFieldItems, inputs, processedFieldKeys, [], iteration, updatedContext, { inArrayContext: options.inArrayContext });
            // If no new fields were processed, we're done
            if (fieldStats.newRequired === 0 && fieldStats.newOptional === 0) {
                break;
            }
            // If we only processed optional fields and skipped them, no need to re-fetch
            if (fieldStats.newRequired === 0 && fieldStats.optionalSkipped) {
                break;
            }
        }
        if (iteration >= maxIterations) {
            console.log(chalk.yellow(`\n⚠️  Maximum field resolution iterations reached. Some dynamic fields may not have been discovered.`));
        }
        // Apply transform if provided
        if (resolver.transform) {
            return resolver.transform(inputs);
        }
        return inputs;
    }
    /**
     * Resolves an array parameter by repeatedly prompting for items until user says no
     */
    async resolveArrayRecursively(resolver, context, param) {
        const items = [];
        const minItems = resolver.minItems ?? 0;
        const maxItems = resolver.maxItems ?? Infinity;
        while (items.length < maxItems) {
            const currentIndex = items.length;
            // Skip confirmation if we haven't hit minItems yet - just collect the item
            if (currentIndex >= minItems) {
                this.stopSpinner();
                const confirmAnswer = await inquirer.prompt([
                    {
                        type: "confirm",
                        name: "addItem",
                        message: `Add ${param.name}[${currentIndex}]?`,
                        default: false,
                    },
                ]);
                if (!confirmAnswer.addItem) {
                    break;
                }
            }
            // Fetch the inner resolver for this item
            const innerResolver = await resolver.fetch(context.sdk, context.resolvedParams);
            const itemValue = await this.resolveWithResolver(innerResolver, param, context, { arrayIndex: currentIndex });
            items.push(itemValue);
            // Update context with current array for subsequent iterations
            context.resolvedParams = {
                ...context.resolvedParams,
                [param.name]: items,
            };
        }
        if (items.length >= maxItems) {
            console.log(chalk.gray(`Maximum of ${maxItems} items reached.`));
        }
        return items;
    }
    /**
     * Recursively processes fieldsets and their fields, maintaining natural structure
     * and creating nested inputs as needed (e.g., fieldset "foo" becomes inputs.foo = [{}])
     */
    async processFieldItems(items, targetInputs, processedFieldKeys, fieldsetPath = [], iteration = 1, context, options = {}) {
        let newRequiredCount = 0;
        let newOptionalCount = 0;
        let optionalSkipped = false;
        for (const item of items) {
            const typedItem = item;
            if (typedItem.type === "fieldset" && typedItem.fields && typedItem.key) {
                // Show fieldset context to user
                const fieldsetTitle = typedItem.title || typedItem.key;
                const pathDisplay = fieldsetPath.length > 0 ? ` (in ${fieldsetPath.join(" > ")})` : "";
                console.log(chalk.cyan(`\n📁 Processing fieldset: ${fieldsetTitle}${pathDisplay}`));
                // Create fieldset array in target inputs if it doesn't exist
                if (!targetInputs[typedItem.key]) {
                    targetInputs[typedItem.key] = [{}];
                }
                // Process fields within this fieldset recursively
                const fieldsetTarget = targetInputs[typedItem.key][0];
                const nestedPath = [...fieldsetPath, fieldsetTitle];
                const nestedStats = await this.processFieldItems(typedItem.fields, fieldsetTarget, processedFieldKeys, nestedPath, iteration, context, options);
                newRequiredCount += nestedStats.newRequired;
                newOptionalCount += nestedStats.newOptional;
                if (nestedStats.optionalSkipped) {
                    optionalSkipped = true;
                }
            }
            else if (typedItem.type === "input_field" && typedItem.key) {
                // Skip if already processed
                if (processedFieldKeys.has(typedItem.key)) {
                    continue;
                }
                const isRequired = typedItem.is_required || false;
                if (isRequired) {
                    // Process required field immediately
                    newRequiredCount++;
                    if (typedItem.resolver && context) {
                        const param = {
                            name: typedItem.key,
                            path: [typedItem.key],
                            schema: z.unknown(),
                            isRequired: true,
                        };
                        targetInputs[typedItem.key] = await this.resolveWithResolver(typedItem.resolver, param, context, { isOptional: false });
                    }
                    else {
                        await this.promptForField(typedItem, targetInputs, context);
                    }
                    processedFieldKeys.add(typedItem.key);
                }
                else {
                    // Collect optional fields for batch processing
                    newOptionalCount++;
                }
            }
            // Skip info fields - they're for display only
        }
        // Handle optional fields after processing all required fields
        if (newOptionalCount > 0) {
            const optionalFields = items.filter((item) => {
                const typedItem = item;
                return (typedItem.type === "input_field" &&
                    typedItem.key &&
                    !typedItem.is_required &&
                    !processedFieldKeys.has(typedItem.key));
            });
            if (optionalFields.length > 0) {
                const pathContext = fieldsetPath.length > 0 ? ` in ${fieldsetPath.join(" > ")}` : "";
                // In array context, prompt for all fields directly without confirmation
                if (options.inArrayContext) {
                    for (const field of optionalFields) {
                        await this.promptForField(field, targetInputs, context);
                        const typedField = field;
                        processedFieldKeys.add(typedField.key);
                    }
                }
                else {
                    console.log(chalk.gray(`\nThere are ${optionalFields.length} ${iteration === 1 ? "" : "additional "}optional field(s) available${pathContext}.`));
                    try {
                        const shouldConfigureOptional = await inquirer.prompt([
                            {
                                type: "confirm",
                                name: "configure",
                                message: `Would you like to configure ${iteration === 1 ? "" : "these additional "}optional fields${pathContext}?`,
                                default: false,
                            },
                        ]);
                        if (shouldConfigureOptional.configure) {
                            console.log(chalk.cyan(`\nOptional fields${pathContext}:`));
                            for (const field of optionalFields) {
                                await this.promptForField(field, targetInputs, context);
                                const typedField = field;
                                processedFieldKeys.add(typedField.key);
                            }
                        }
                        else {
                            optionalSkipped = true;
                            // Mark these fields as processed even if skipped to avoid re-asking
                            optionalFields.forEach((field) => {
                                const typedField = field;
                                processedFieldKeys.add(typedField.key);
                            });
                        }
                    }
                    catch (error) {
                        if (this.isUserCancellation(error)) {
                            console.log(chalk.yellow("\n\nOperation cancelled by user"));
                            throw new ZapierCliUserCancellationError();
                        }
                        throw error;
                    }
                }
            }
        }
        return {
            newRequired: newRequiredCount,
            newOptional: newOptionalCount,
            optionalSkipped,
        };
    }
    getNestedValue(obj, path) {
        return path.reduce((current, key) => current?.[key], obj);
    }
    setNestedValue(obj, path, value) {
        const lastKey = path[path.length - 1];
        const parent = path.slice(0, -1).reduce((current, key) => {
            const currentObj = current;
            if (!(key in currentObj)) {
                currentObj[key] = {};
            }
            return currentObj[key];
        }, obj);
        parent[lastKey] = value;
    }
    /**
     * Extract and normalize field metadata from raw field object
     */
    extractFieldMetadata(field) {
        const fieldObj = field;
        const valueType = fieldObj.value_type || "string";
        return {
            key: fieldObj.key,
            title: fieldObj.title || fieldObj.label || fieldObj.key,
            description: fieldObj.description || fieldObj.helpText,
            isRequired: fieldObj.is_required || false,
            defaultValue: fieldObj.default_value ?? fieldObj.default,
            valueType,
            hasDropdown: fieldObj.format === "SELECT" || Boolean(fieldObj.choices),
            isMultiSelect: Boolean(valueType === "array" ||
                (fieldObj.items && fieldObj.items.type !== undefined)),
            inlineChoices: fieldObj.choices,
        };
    }
    /**
     * Fetch a page of choices for a dropdown field
     */
    async fetchChoices(fieldMeta, inputs, context, cursor) {
        try {
            this.debugLog(cursor
                ? `Fetching more choices for ${fieldMeta.title}`
                : `Fetching choices for ${fieldMeta.title}`);
            this.startSpinner();
            const page = await context.sdk.listInputFieldChoices({
                app: context.resolvedParams.app,
                action: context.resolvedParams.action,
                actionType: context.resolvedParams.actionType,
                connection: context.resolvedParams.connection,
                inputField: fieldMeta.key,
                inputs,
                ...(cursor && { cursor }),
            });
            this.stopSpinner();
            const choices = page.data.map((choice) => ({
                label: choice.label || choice.key || String(choice.value),
                value: choice.value ?? choice.key,
            }));
            if (choices.length === 0 && !cursor) {
                console.log(chalk.yellow(`No choices available for ${fieldMeta.title}`));
            }
            return {
                choices,
                nextCursor: page.nextCursor,
            };
        }
        catch (error) {
            this.stopSpinner();
            console.warn(chalk.yellow(`Failed to fetch choices for ${fieldMeta.title}:`), error);
            return { choices: [] };
        }
    }
    /**
     * Prompt user with choices (handles both single and multi-select with pagination)
     */
    async promptWithChoices({ fieldMeta, choices: initialChoices, nextCursor: initialCursor, inputs, context, }) {
        this.stopSpinner();
        const choices = [...initialChoices];
        let nextCursor = initialCursor;
        const LOAD_MORE_SENTINEL = Symbol("LOAD_MORE");
        // Progressive loading loop
        while (true) {
            const promptChoices = choices.map((choice) => ({
                name: choice.label,
                value: choice.value,
            }));
            // Add "(Load more...)" option if there are more pages
            if (nextCursor) {
                promptChoices.push({
                    name: chalk.dim("(Load more...)"),
                    value: LOAD_MORE_SENTINEL,
                });
            }
            // Add skip option for optional fields (single-select only)
            if (!fieldMeta.isRequired && !fieldMeta.isMultiSelect) {
                promptChoices.push({ name: "(Skip)", value: undefined });
            }
            const promptConfig = {
                type: fieldMeta.isMultiSelect ? "checkbox" : "list",
                name: fieldMeta.key,
                message: `${fieldMeta.title}${fieldMeta.isRequired ? " (required)" : " (optional)"}:`,
                choices: promptChoices,
                ...(fieldMeta.isMultiSelect && {
                    validate: (input) => {
                        if (fieldMeta.isRequired && (!input || input.length === 0)) {
                            return "At least one selection is required";
                        }
                        return true;
                    },
                }),
            };
            const answer = await inquirer.prompt([promptConfig]);
            let selectedValue = answer[fieldMeta.key];
            // Check if user selected "Load more..."
            const wantsMore = fieldMeta.isMultiSelect
                ? Array.isArray(selectedValue) &&
                    selectedValue.includes(LOAD_MORE_SENTINEL)
                : selectedValue === LOAD_MORE_SENTINEL;
            if (wantsMore && nextCursor && context) {
                // Remove sentinel from multi-select
                if (fieldMeta.isMultiSelect && Array.isArray(selectedValue)) {
                    selectedValue = selectedValue.filter((v) => v !== LOAD_MORE_SENTINEL);
                }
                // Fetch next page
                const result = await this.fetchChoices(fieldMeta, inputs, context, nextCursor);
                choices.push(...result.choices);
                nextCursor = result.nextCursor;
                // Re-prompt with updated choices
                continue;
            }
            return selectedValue;
        }
    }
    /**
     * Prompt user for free-form input (text or boolean)
     */
    async promptFreeForm(fieldMeta) {
        const promptConfig = {
            name: fieldMeta.key,
            message: `${fieldMeta.title}${fieldMeta.isRequired ? " (required)" : " (optional)"}:`,
        };
        if (fieldMeta.valueType === "boolean") {
            promptConfig.type = "confirm";
            promptConfig.default =
                fieldMeta.defaultValue !== undefined
                    ? Boolean(fieldMeta.defaultValue)
                    : undefined;
        }
        else if (fieldMeta.valueType === "array") {
            promptConfig.type = "input";
            promptConfig.default = fieldMeta.defaultValue;
            promptConfig.message = `${fieldMeta.title}${fieldMeta.isRequired ? " (required)" : " (optional)"} (JSON array or comma-separated):`;
            promptConfig.validate = (input) => {
                if (fieldMeta.isRequired && !input) {
                    return "This field is required";
                }
                return true;
            };
            promptConfig.filter = (input) => {
                if (!input)
                    return input;
                const trimmed = input.trim();
                if (trimmed.startsWith("[")) {
                    try {
                        return JSON.parse(trimmed);
                    }
                    catch {
                        // Fall through to comma-separated
                    }
                }
                return trimmed.split(",").map((s) => s.trim());
            };
        }
        else {
            promptConfig.type = "input";
            promptConfig.default = fieldMeta.defaultValue;
            promptConfig.validate = (input) => {
                if (fieldMeta.isRequired && !input) {
                    return "This field is required";
                }
                return true;
            };
        }
        // Add help text if available
        if (fieldMeta.description) {
            promptConfig.prefix = chalk.gray(`ℹ  ${fieldMeta.description}\n`);
        }
        try {
            const answer = await inquirer.prompt([promptConfig]);
            return answer[fieldMeta.key];
        }
        catch (error) {
            if (this.isUserCancellation(error)) {
                console.log(chalk.yellow("\n\nOperation cancelled by user"));
                throw new ZapierCliUserCancellationError();
            }
            throw error;
        }
    }
    /**
     * Store field value in inputs object with validation
     */
    storeFieldValue(inputs, key, value, isRequired) {
        try {
            if (value !== undefined && value !== "") {
                inputs[key] = value;
            }
            else if (isRequired) {
                throw new Error(`Required field ${key} cannot be empty`);
            }
        }
        catch (error) {
            if (this.isUserCancellation(error)) {
                console.log(chalk.yellow("\n\nOperation cancelled by user"));
                throw new ZapierCliUserCancellationError();
            }
            throw error;
        }
    }
    async promptForField(field, inputs, context) {
        const fieldMeta = this.extractFieldMetadata(field);
        // Get choices - either inline or fetched from API
        let choices = [];
        let nextCursor;
        if (fieldMeta.inlineChoices) {
            // Use inline choices directly
            choices = fieldMeta.inlineChoices;
        }
        else if (fieldMeta.hasDropdown && context) {
            // Fetch choices from API
            const result = await this.fetchChoices(fieldMeta, inputs, context);
            choices = result.choices;
            nextCursor = result.nextCursor;
        }
        // Prompt user based on field type
        let selectedValue;
        if (choices.length > 0) {
            selectedValue = await this.promptWithChoices({
                fieldMeta,
                choices,
                nextCursor,
                inputs,
                context,
            });
        }
        else {
            selectedValue = await this.promptFreeForm(fieldMeta);
        }
        // Store result
        this.storeFieldValue(inputs, fieldMeta.key, selectedValue, fieldMeta.isRequired);
    }
    isUserCancellation(error) {
        const errorObj = error;
        return (errorObj?.name === "ExitPromptError" ||
            errorObj?.message?.includes("User force closed") ||
            errorObj?.isTTYError === true);
    }
    hasResolver(paramName, sdk, functionName) {
        // Check plugin-specific resolvers first
        if (functionName && typeof sdk.getRegistry === "function") {
            const registry = sdk.getRegistry({ package: "cli" });
            const functionInfo = registry.functions.find((f) => f.name === functionName);
            if (functionInfo && functionInfo.resolvers?.[paramName]) {
                return true;
            }
        }
        // No global registry fallback
        return false;
    }
    getResolver(paramName, sdk, functionName) {
        // Check plugin-specific resolvers first
        if (functionName && typeof sdk.getRegistry === "function") {
            const registry = sdk.getRegistry({ package: "cli" });
            const functionInfo = registry.functions.find((f) => f.name === functionName);
            if (functionInfo && functionInfo.resolvers?.[paramName]) {
                return functionInfo.resolvers[paramName];
            }
        }
        // No global registry fallback
        return null;
    }
    getLocalResolvers(sdk, functionName) {
        if (!functionName || typeof sdk.getRegistry !== "function") {
            return {};
        }
        const registry = sdk.getRegistry();
        const functionInfo = registry.functions.find((f) => f.name === functionName);
        return functionInfo?.resolvers || {};
    }
}
