// @zapier/zapier-sdk-cli/dist/src/utils/schema-formatter.js
import chalk from "chalk";
import util from "util";
import wrapAnsi from "wrap-ansi";
// These functions are internal to SDK, implementing basic formatting fallback
// TODO: Consider exposing these utilities or implementing proper CLI formatting
function getFormatMetadata(schema) {
    return schema?._zod
        ?.def?.formatMeta;
}
function getOutputSchema(schema) {
    return schema?._zod?.def
        ?.outputSchema;
}
// ============================================================================
// JSON Formatting
// ============================================================================
export function formatJsonOutput(data) {
    // Don't print anything for undefined results (commands that just perform actions)
    if (data === undefined) {
        return;
    }
    // Use util.inspect for colored output
    console.log(util.inspect(data, { colors: true, depth: null, breakLength: 80 }));
}
// ============================================================================
// Generic Schema-Driven Formatter
// ============================================================================
export async function formatItemsFromSchema(functionInfo, items, startingNumber = 0, options) {
    // If a registry-level OutputFormatter is provided, use it
    if (options?.formatter) {
        let context;
        if (options.formatter.fetch && options.sdk && options.params) {
            for (const item of items) {
                context = await options.formatter.fetch(options.sdk, options.params, item, context);
            }
        }
        items.forEach((item, index) => {
            const formatted = options.formatter.format(item, context);
            formatSingleItem(formatted, startingNumber + index);
        });
        return;
    }
    // Get the output schema from function info or fall back to input schema output schema
    const outputSchema = functionInfo.outputSchema || getOutputSchema(functionInfo.inputSchema);
    if (!outputSchema) {
        // Fallback to generic formatting if no output schema
        formatItemsGeneric(items, startingNumber);
        return;
    }
    const formatMeta = getFormatMetadata(outputSchema);
    if (!formatMeta) {
        // Fallback to generic formatting if no format metadata
        formatItemsGeneric(items, startingNumber);
        return;
    }
    // Format each item using the schema metadata
    items.forEach((item, index) => {
        const formatted = formatMeta.format(item);
        formatSingleItem(formatted, startingNumber + index);
    });
}
function formatSingleItem(formatted, itemNumber) {
    // Build the main title line with optional subtitle
    let titleLine = `${chalk.gray(`${itemNumber + 1}.`)} ${chalk.cyan(formatted.title)}`;
    // Generate subtitle - always show key if available, then ID if available
    const subtitleParts = [];
    if (formatted.keys) {
        subtitleParts.push(...formatted.keys);
    }
    else if (formatted.key) {
        subtitleParts.push(formatted.key);
    }
    if (formatted.id) {
        subtitleParts.push(formatted.id);
    }
    // Remove duplicates while preserving order
    const uniqueParts = [...new Set(subtitleParts)];
    if (uniqueParts.length > 0) {
        titleLine += ` ${chalk.gray(`(${uniqueParts.join(", ")})`)}`;
    }
    console.log(titleLine);
    // Show description if available
    if (formatted.description) {
        console.log(`   ${chalk.dim(formatted.description)}`);
    }
    // If data is provided, use JSON formatting instead of details
    if (formatted.data !== undefined) {
        formatJsonOutput(formatted.data);
        console.log(); // Empty line between items
        return;
    }
    // Format detail lines
    for (const detail of formatted.details) {
        if (detail.label) {
            const isMultiline = detail.text.includes("\n");
            if (isMultiline) {
                console.log(`   ${chalk.gray(detail.label + ":")}`);
                const displayText = formatDetailText(detail.text, DETAIL_INDENT + "   ");
                const styledText = applyStyle(displayText, detail.style);
                console.log(`${DETAIL_INDENT}   ${styledText}`);
            }
            else {
                const styledValue = applyStyle(detail.text, detail.style);
                console.log(`   ${chalk.gray(detail.label + ":")} ${styledValue}`);
            }
        }
        else {
            const displayText = formatDetailText(detail.text, DETAIL_INDENT);
            const styledText = applyStyle(displayText, detail.style);
            console.log(`   ${styledText}`);
        }
    }
    console.log(); // Empty line between items
}
const DETAIL_INDENT = "   ";
const DETAIL_MAX_LINES = 5;
function formatDetailText(text, indent = DETAIL_INDENT) {
    const columns = Math.max((process.stdout.columns || 80) - indent.length, 40);
    const wrapped = wrapAnsi(text, columns, { hard: true, trim: false });
    const lines = wrapped.split("\n");
    if (lines.length <= DETAIL_MAX_LINES) {
        return lines.join("\n" + indent);
    }
    return (lines.slice(0, DETAIL_MAX_LINES).join("\n" + indent) + "\n" + indent + "…");
}
function applyStyle(value, style) {
    switch (style) {
        case "dim":
            return chalk.dim(value);
        case "accent":
            return chalk.magenta(value);
        case "warning":
            return chalk.red(value);
        case "success":
            return chalk.green(value);
        case "normal":
        default:
            return chalk.blue(value);
    }
}
function convertGenericItemToFormattedItem(item) {
    const itemObj = item;
    return {
        title: itemObj.title || itemObj.name || itemObj.key || itemObj.id || "Item",
        id: itemObj.id,
        key: itemObj.key,
        description: itemObj.description,
        details: [],
    };
}
function formatItemsGeneric(items, startingNumber = 0) {
    // Convert generic items to FormattedItem and use formatSingleItem
    items.forEach((item, index) => {
        const formatted = convertGenericItemToFormattedItem(item);
        formatSingleItem(formatted, startingNumber + index);
    });
}
