// @zapier/zapier-sdk-cli/dist/src/utils/version-checker.js
import packageJsonLib from "package-json";
import chalk from "chalk";
import log from "./log";
import { getConfig } from "@zapier/zapier-sdk-cli-login";
import { getUpdateCommand } from "./package-manager-detector";
import semver from "semver";
const ONE_DAY_MS = 24 * 60 * 60 * 1000;
const CACHE_RESET_INTERVAL_MS = (() => {
    const { ZAPIER_SDK_UPDATE_CHECK_INTERVAL_MS = `${ONE_DAY_MS}` } = process.env;
    const interval = parseInt(ZAPIER_SDK_UPDATE_CHECK_INTERVAL_MS);
    if (isNaN(interval) || interval < 0) {
        return -1;
    }
    return interval;
})();
function getVersionCache() {
    try {
        const cache = getConfig().get("version_cache");
        const now = Date.now();
        // If no cache, missing timestamp, or it's been more than a day, reset the cache
        if (!cache ||
            !cache.last_reset_timestamp ||
            now - cache.last_reset_timestamp >= CACHE_RESET_INTERVAL_MS) {
            const newCache = {
                last_reset_timestamp: now,
                packages: {},
            };
            getConfig().set("version_cache", newCache);
            return newCache;
        }
        return cache;
    }
    catch (error) {
        log.debug(`Failed to read version cache: ${error}`);
        return {
            last_reset_timestamp: Date.now(),
            packages: {},
        };
    }
}
function setCachedPackageInfo(packageName, version, info) {
    try {
        const cache = getVersionCache();
        if (!cache.packages[packageName]) {
            cache.packages[packageName] = {};
        }
        cache.packages[packageName][version] = info;
        getConfig().set("version_cache", cache);
    }
    catch (error) {
        log.debug(`Failed to cache package info: ${error}`);
    }
}
function getCachedPackageInfo(packageName, version) {
    try {
        const cache = getVersionCache();
        return cache.packages[packageName]?.[version];
    }
    catch (error) {
        log.debug(`Failed to get cached package info: ${error}`);
        return undefined;
    }
}
async function fetchCachedPackageInfo(packageName, version) {
    const cacheKey = version || "latest";
    // Try cache first
    let cachedInfo = getCachedPackageInfo(packageName, cacheKey);
    if (cachedInfo) {
        return cachedInfo;
    }
    // Not in cache, fetch from npm
    const packageInfo = await packageJsonLib(packageName, {
        version,
        fullMetadata: true,
    });
    const info = {
        version: packageInfo.version,
        deprecated: packageInfo.deprecated,
        fetched_at: new Date().toISOString(),
    };
    // Cache for next time
    setCachedPackageInfo(packageName, cacheKey, info);
    return info;
}
export async function checkForUpdates({ packageName, currentVersion, }) {
    try {
        // Get latest version info (with caching)
        const latestPackageInfo = await fetchCachedPackageInfo(packageName);
        const latestVersion = latestPackageInfo.version;
        const hasUpdate = semver.gt(latestVersion, currentVersion);
        // Check deprecation status of the current version (with caching)
        let currentPackageInfo;
        try {
            currentPackageInfo = await fetchCachedPackageInfo(packageName, currentVersion);
        }
        catch (error) {
            // If we can't fetch the current version info, use the latest version info
            log.debug(`Failed to check deprecation for current version: ${error}`);
            currentPackageInfo = latestPackageInfo;
        }
        const isDeprecated = Boolean(currentPackageInfo.deprecated);
        const deprecationMessage = isDeprecated
            ? String(currentPackageInfo.deprecated)
            : undefined;
        return {
            hasUpdate,
            latestVersion,
            currentVersion,
            isDeprecated,
            deprecationMessage,
        };
    }
    catch (error) {
        // If we can't check for updates (network issues, etc.), fail silently
        log.debug(`Failed to check for updates: ${error}`);
        return {
            hasUpdate: false,
            currentVersion,
            isDeprecated: false,
        };
    }
}
export function displayUpdateNotification(versionInfo, packageName) {
    if (versionInfo.isDeprecated) {
        console.error();
        console.error(chalk.red.bold("⚠️  DEPRECATION WARNING") +
            chalk.red(` - ${packageName} v${versionInfo.currentVersion} is deprecated.`));
        if (versionInfo.deprecationMessage) {
            console.error(chalk.red(`   ${versionInfo.deprecationMessage}`));
        }
        console.error(chalk.red(`   Please update to the latest version.`));
        console.error();
    }
    // Use stderr so update notifications don't pollute stdout when piping JSON to jq
    if (versionInfo.hasUpdate) {
        console.error();
        console.error(chalk.yellow.bold("📦 Update available!") +
            chalk.yellow(` ${packageName} v${versionInfo.currentVersion} → v${versionInfo.latestVersion}`));
        console.error(chalk.yellow(`   Run ${chalk.bold(getUpdateCommand(packageName))} to update.`));
        console.error();
    }
}
export async function checkAndNotifyUpdates({ packageName, currentVersion, }) {
    if (CACHE_RESET_INTERVAL_MS < 0) {
        return;
    }
    const versionInfo = await checkForUpdates({ packageName, currentVersion });
    displayUpdateNotification(versionInfo, packageName);
}
