// @zapier/zapier-sdk-cli/dist/src/plugins/init/display.js
import chalk from "chalk";
import { buildNextSteps } from "./steps";
import { getPackageManagerCommands } from "./utils";
export function createConsoleDisplayHooks() {
    return {
        onItemComplete: (message) => console.log("   " + chalk.green("✓") + "  " + chalk.dim(message)),
        onWarn: (message) => console.warn(chalk.yellow("!") + "  " + message),
        onStepStart: ({ description, stepNumber, totalSteps, command, skipPrompts, }) => {
            const progressMessage = `${description}...`;
            const stepCounter = chalk.dim(`${stepNumber}/${totalSteps}`);
            if (skipPrompts) {
                console.log("\n" +
                    chalk.bold(`❯  ${progressMessage}`) +
                    "  " +
                    stepCounter +
                    "\n");
            }
            else {
                console.log(chalk.dim("→") + "  " + progressMessage + "  " + stepCounter);
            }
            if (command) {
                console.log("   " + chalk.cyan(`$ ${command}`));
            }
        },
        onStepSuccess: ({ stepNumber, totalSteps }) => console.log("\n" +
            chalk.green("✓") +
            "  " +
            chalk.dim(`Step ${stepNumber}/${totalSteps} complete`) +
            "\n"),
        onStepError: ({ description, command, err }) => {
            const detail = err instanceof Error && err.message
                ? `\n   ${chalk.dim(err.message)}`
                : "";
            const hint = command
                ? `\n   ${chalk.dim("run manually:")} ${chalk.cyan(`$ ${command}`)}`
                : "";
            console.error(`\n${chalk.red("✖")}  ${chalk.bold(description)}${chalk.dim(" failed")}${detail}${hint}`);
        },
    };
}
export function displaySummaryAndNextSteps({ projectName, steps, completedSetupStepIds, packageManager, }) {
    const formatStatus = (complete) => ({
        icon: complete ? chalk.green("✓") : chalk.yellow("!"),
        text: complete
            ? chalk.green("Setup complete")
            : chalk.yellow("Setup interrupted"),
    });
    const formatNextStep = (step, i) => "   " + chalk.dim(`${i + 1}.`) + "  " + chalk.bold(step.description);
    const formatCommand = (cmd) => "       " + chalk.cyan(`$ ${cmd}`);
    const formatCompletedStep = (step) => "   " + chalk.green("✓") + "  " + step.description;
    const { execCmd } = getPackageManagerCommands({ packageManager });
    const leftoverSteps = steps.filter((s) => !completedSetupStepIds.includes(s.id));
    const isComplete = leftoverSteps.length === 0;
    const status = formatStatus(isComplete);
    console.log("\n" + chalk.bold("❯  Summary") + "\n");
    console.log("   " + chalk.dim("Project") + "   " + chalk.bold(projectName));
    console.log("   " + chalk.dim("Status") + "    " + status.icon + "  " + status.text);
    const completedSteps = steps.filter((s) => completedSetupStepIds.includes(s.id));
    if (completedSteps.length > 0) {
        console.log();
        for (const step of completedSteps)
            console.log(formatCompletedStep(step));
    }
    const nextSteps = buildNextSteps({ projectName, leftoverSteps, execCmd });
    console.log("\n" + chalk.bold("❯  Next Steps") + "\n");
    nextSteps.forEach((step, i) => {
        console.log(formatNextStep(step, i));
        if (step.command)
            console.log(formatCommand(step.command));
        console.log();
    });
}
