// @zapier/zapier-sdk-cli/dist/src/plugins/init/index.js
import { createFunction } from "@zapier/zapier-sdk";
import { InitSchema } from "./schemas";
import { detectPackageManager } from "../../utils/package-manager-detector";
import { validateInitOptions, withInterruptCleanup } from "./utils";
import { getInitSteps, runStep } from "./steps";
import { createConsoleDisplayHooks, displaySummaryAndNextSteps, } from "./display";
import { ZapierCliExitError } from "../../utils/errors";
export { toProjectName } from "./utils";
export const initPlugin = () => {
    const init = createFunction(async function init(options) {
        const { projectName: rawName, skipPrompts = false } = options;
        const cwd = process.cwd();
        const { projectName, projectDir } = validateInitOptions({ rawName, cwd });
        const displayHooks = createConsoleDisplayHooks();
        const packageManagerInfo = detectPackageManager(cwd);
        if (packageManagerInfo.name === "unknown") {
            displayHooks.onWarn("Could not detect package manager, defaulting to npm.");
        }
        const packageManager = packageManagerInfo.name === "unknown" ? "npm" : packageManagerInfo.name;
        const steps = getInitSteps({
            projectDir,
            projectName,
            packageManager,
            displayHooks,
        });
        const completedSetupStepIds = [];
        for (let i = 0; i < steps.length; i++) {
            const step = steps[i];
            const succeeded = await withInterruptCleanup(step.cleanup, () => runStep({
                step,
                stepNumber: i + 1,
                totalSteps: steps.length,
                skipPrompts,
                displayHooks,
            }));
            if (!succeeded)
                break;
            completedSetupStepIds.push(step.id);
        }
        if (completedSetupStepIds.length === 0) {
            throw new ZapierCliExitError("Project setup failed — no steps completed.");
        }
        displaySummaryAndNextSteps({
            projectName,
            steps,
            completedSetupStepIds,
            packageManager,
        });
    }, InitSchema);
    return {
        init,
        context: {
            meta: {
                init: {
                    categories: ["utility"],
                    inputSchema: InitSchema,
                    supportsJsonOutput: false,
                },
            },
        },
    };
};
