// @zapier/zapier-sdk-cli/dist/src/plugins/init/steps.js
import { existsSync, mkdirSync, writeFileSync, readdirSync, copyFileSync, } from "fs";
import { join, dirname, relative } from "path";
import { promptYesNo, createExec, getPackageManagerCommands, getDirentParentPath, buildTemplateVariables, cleanupProject, createProjectDir, isTemplateFile, getDestRelPath, renderTemplate, } from "./utils";
import { TEMPLATES } from "./schemas";
import { TEMPLATES_DIR } from "../../paths";
const DEFAULT_TEMPLATE = "basic";
export function getTemplateDir({ template = DEFAULT_TEMPLATE, } = {}) {
    const dirPath = join(TEMPLATES_DIR, template);
    if (!existsSync(dirPath)) {
        throw new Error(`Template "${template}" not found at ${dirPath}. Available templates: ${TEMPLATES.join(", ")}`);
    }
    return dirPath;
}
export function scaffoldFiles({ projectDir, templatesDir, variables = {}, displayHooks, }) {
    const entries = readdirSync(templatesDir, {
        withFileTypes: true,
        recursive: true,
    });
    const files = entries.filter((e) => e.isFile());
    if (files.length === 0) {
        throw new Error(`Template directory "${templatesDir}" contains no files.`);
    }
    for (const entry of files) {
        const srcPath = join(getDirentParentPath(entry), entry.name);
        const relPath = relative(templatesDir, srcPath);
        const isTemplate = isTemplateFile(entry.name);
        const destRelPath = getDestRelPath(relPath, isTemplate);
        const destPath = join(projectDir, destRelPath);
        mkdirSync(dirname(destPath), { recursive: true });
        if (isTemplate) {
            writeFileSync(destPath, renderTemplate(srcPath, variables));
        }
        else {
            copyFileSync(srcPath, destPath);
        }
        displayHooks?.onItemComplete?.(destRelPath);
    }
}
export function scaffoldProject({ projectDir, projectName, packageManager, template, displayHooks, }) {
    const variables = buildTemplateVariables({ projectName, packageManager });
    const templatesDir = getTemplateDir({ template });
    createProjectDir({ projectDir, projectName });
    scaffoldFiles({ projectDir, templatesDir, variables, displayHooks });
}
export async function runStep({ step, stepNumber, totalSteps, skipPrompts, displayHooks, }) {
    if (step.askConfirmation) {
        const should = await promptYesNo({
            message: step.description,
            defaultValue: true,
            skipPrompts,
        });
        if (!should)
            return false;
    }
    displayHooks?.onStepStart({
        description: step.description,
        stepNumber,
        totalSteps,
        command: step.command,
        skipPrompts,
    });
    try {
        await step.run();
        displayHooks?.onStepSuccess({ stepNumber, totalSteps });
        return true;
    }
    catch (err) {
        step.cleanup?.();
        displayHooks?.onStepError({
            description: step.description,
            command: step.command,
            err,
        });
        return false;
    }
}
export function getInitSteps({ projectDir, projectName, packageManager, template, displayHooks, }) {
    if (template !== undefined &&
        !TEMPLATES.includes(template)) {
        throw new Error(`Unknown template "${template}". Available templates: ${TEMPLATES.join(", ")}`);
    }
    const { installCmd, execCmd, devCmd } = getPackageManagerCommands({
        packageManager,
    });
    const exec = createExec({ cwd: projectDir });
    return [
        {
            id: "scaffold",
            description: "Scaffold project files",
            cleanup: () => cleanupProject({ projectDir }),
            run: () => scaffoldProject({
                projectDir,
                projectName,
                packageManager,
                template,
                displayHooks,
            }),
        },
        {
            id: "install-deps",
            description: "Install dependencies",
            askConfirmation: true,
            command: installCmd,
            run: () => exec(installCmd),
        },
        {
            id: "login",
            description: "Log in to Zapier",
            askConfirmation: true,
            command: `${execCmd} zapier-sdk login`,
            run: () => exec(`${execCmd} zapier-sdk login`),
        },
        {
            id: "run",
            description: "Run your app",
            askConfirmation: true,
            command: devCmd,
            run: () => exec(devCmd),
        },
    ];
}
export function buildNextSteps({ projectName, leftoverSteps, execCmd, }) {
    return [
        {
            description: "Change into your project directory",
            command: `cd ${projectName}`,
        },
        ...leftoverSteps.map(({ description, command }) => ({
            description,
            command,
        })),
        {
            description: "Search for an app to integrate",
            command: `${execCmd} zapier-sdk list-apps --search "<app name>"`,
        },
        {
            description: "Add an app and generate TypeScript types",
            command: `${execCmd} zapier-sdk add <app-key>`,
        },
    ];
}
