// @zapier/zapier-sdk-cli/dist/src/plugins/init/utils.js
import { existsSync, mkdirSync, readFileSync, readdirSync, rmSync } from "fs";
import { execSync } from "child_process";
import { extname, join } from "path";
import chalk from "chalk";
import Handlebars from "handlebars";
import inquirer from "inquirer";
// Node 20 added `parentPath`; Node 18 used `path` for the same purpose.
export function getDirentParentPath(entry) {
    const e = entry;
    const parent = e.parentPath ?? e.path;
    if (!parent)
        throw new Error("readdirSync entry missing parentPath/path — unsupported Node version");
    return parent;
}
export function toProjectName(name) {
    // Essentially converts to kebab-case
    return (name
        .toLowerCase()
        // Replace anything that isn't a letter, digit, or hyphen with a hyphen.
        // Hyphens are kept as-is since they're valid in npm package names.
        .replace(/[^a-z0-9-]/g, "-")
        .replace(/-+/g, "-") // collapse consecutive hyphens produced by the step above
        .replace(/^-|-$/g, "")); // strip any leading/trailing hyphens
}
export function validateInitOptions({ rawName, cwd, }) {
    const projectName = toProjectName(rawName);
    if (!projectName) {
        throw new Error(`"${rawName}" results in an empty project name. Provide a name with at least one letter or number.`);
    }
    const projectDir = join(cwd, projectName);
    if (existsSync(projectDir)) {
        const contents = readdirSync(projectDir);
        if (contents.length > 0) {
            throw new Error(`Directory "${projectName}" already exists and is not empty. Choose a different name.`);
        }
    }
    return { projectName, projectDir };
}
// shell is required on Windows where npx/npm/pnpm/yarn are .cmd files
// that execSync cannot locate without a shell.
export function createExec({ cwd }) {
    return (cmd) => execSync(cmd, {
        cwd,
        stdio: "inherit",
        shell: process.platform === "win32" ? "cmd.exe" : "/bin/sh",
    });
}
export async function promptYesNo({ message, defaultValue, skipPrompts, }) {
    if (skipPrompts)
        return defaultValue;
    const { answer } = await inquirer.prompt([
        { type: "confirm", name: "answer", message, default: defaultValue },
    ]);
    return answer;
}
export function getPackageManagerCommands({ packageManager, }) {
    return {
        installCmd: `${packageManager} install`,
        devCmd: packageManager === "yarn" ? "yarn dev" : `${packageManager} run dev`,
        devScript: packageManager === "bun" ? "bun src/index.ts" : "tsx src/index.ts",
        execCmd: {
            npm: "npx",
            yarn: "yarn dlx",
            pnpm: "pnpm dlx",
            bun: "bunx",
        }[packageManager],
    };
}
export function isTemplateFile(name) {
    return extname(name) === ".hbs";
}
export function getDestRelPath(relPath, isTemplate) {
    return isTemplate ? relPath.slice(0, -extname(relPath).length) : relPath;
}
export function renderTemplate(srcPath, variables) {
    const source = readFileSync(srcPath, "utf-8");
    // Without this, Handlebars HTML-encodes output (& → &amp;), which corrupts JSON and TypeScript files.
    const template = Handlebars.compile(source, { noEscape: true });
    return template(variables);
}
export function buildTemplateVariables({ projectName, packageManager, }) {
    const { execCmd, devScript, devCmd } = getPackageManagerCommands({
        packageManager,
    });
    return {
        projectName,
        execCmd,
        devScript,
        devCmd,
        includeTsx: packageManager !== "bun",
    };
}
export function cleanupProject({ projectDir }) {
    console.log("\n" + chalk.yellow("!") + "  Cleaning up...");
    rmSync(projectDir, { recursive: true, force: true });
}
/**
 * Registers a cleanup callback that fires if the process receives SIGINT while
 * `fn` is running. The listener is always removed once `fn` settles, preventing
 * listener leaks across steps. The SIGINT is re-raised after cleanup so the
 * shell receives a proper signal-terminated exit code (130) rather than exit(1).
 */
export async function withInterruptCleanup(cleanup, fn) {
    if (!cleanup)
        return fn();
    const handler = () => {
        cleanup();
        process.removeListener("SIGINT", handler);
        process.kill(process.pid, "SIGINT");
    };
    process.on("SIGINT", handler);
    try {
        return await fn();
    }
    finally {
        process.removeListener("SIGINT", handler);
    }
}
export function createProjectDir({ projectDir, projectName, }) {
    try {
        mkdirSync(projectDir, { recursive: true });
    }
    catch (err) {
        if (err instanceof Error) {
            const code = err.code;
            if (code === "EACCES") {
                throw new Error(`Permission denied creating "${projectName}". Check directory permissions.`);
            }
            if (code === "ENOSPC") {
                throw new Error("No space left on device.");
            }
        }
        throw err;
    }
}
