// @zapier/zapier-sdk/dist/plugins/tables/getTable/index.js
import { createFunction } from "../../../utils/function-utils";
import { GetTableApiResponseSchema, GetTableOptionsSchema, GetTableOptionsInputSchema, TableItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver } from "../../../resolvers";
import { transformTableItem } from "../utils";
export const getTablePlugin = ({ context }) => {
    async function getTable(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const rawResponse = await api.get(`/tables/api/v1/tables/${tableId}`, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to access this table or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to access this table. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 404) {
                    return new Error(`Table not found: ${tableId}`);
                }
                return undefined;
            },
            authRequired: true,
        });
        const response = GetTableApiResponseSchema.parse(rawResponse);
        return {
            data: transformTableItem(response.data),
        };
    }
    const getTableDefinition = createFunction(getTable, GetTableOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, getTable.name));
    return {
        getTable: getTableDefinition,
        context: {
            meta: {
                getTable: {
                    categories: ["table"],
                    type: "item",
                    itemType: "Table",
                    inputSchema: GetTableOptionsSchema,
                    outputSchema: TableItemSchema,
                    resolvers: {
                        table: tableIdResolver,
                    },
                },
            },
        },
    };
};
