// @zapier/zapier-sdk/dist/plugins/tables/getTable/schemas.js
import { z } from "zod";
import { TablePropertySchema } from "../../../types/properties";
import { TableItemSchema, TableApiItemSchema, } from "../listTables/schemas";
// Re-export for consumers
export { TableItemSchema, TableApiItemSchema };
export const GetTableApiResponseSchema = z.object({
    data: TableApiItemSchema,
});
const GetTableDescription = "Get detailed information about a specific table";
// Schema for docs/registry (new param names only)
export const GetTableOptionsSchema = z
    .object({
    table: TablePropertySchema,
})
    .describe(GetTableDescription)
    .meta({ aliases: { tableId: "table" } });
// Deprecated schema (backward compat)
const GetTableOptionsSchemaDeprecated = z.object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table to retrieve"),
});
// Union for runtime validation
export const GetTableOptionsInputSchema = z
    .union([GetTableOptionsSchema, GetTableOptionsSchemaDeprecated])
    .describe(GetTableDescription);
