// @zapier/zapier-sdk-cli/dist/src/plugins/curl/index.js
import { createWriteStream } from "fs";
import { promises as fs } from "fs";
import { dirname } from "path";
import { CurlSchema } from "./schemas";
import { CurlExitError, parseHeaderLine, basicAuthHeader, deriveRemoteFilename, formatWriteOut, appendQueryParams, readAllStdin, resolveDataArgText, resolveDataArgBinary, buildFormData, } from "./utils";
export const curlPlugin = ({ sdk, }) => {
    async function curl(options) {
        const { url: rawUrl, request, header = [], data = [], dataRaw = [], dataAscii = [], dataBinary = [], dataUrlencode = [], json, form = [], formString = [], get: forceGet, head: forceHead, location, include, output, remoteName, verbose, silent, showError, fail, failWithBody, writeOut, maxTime, user, compressed, connection: connectionParam, connectionId, } = options;
        const connection = connectionParam ?? connectionId;
        const parsedUrl = new URL(rawUrl);
        const headers = {};
        for (const h of header) {
            const parsed = parseHeaderLine(h);
            if (parsed) {
                headers[parsed.key] = parsed.value;
            }
        }
        if (user) {
            headers["Authorization"] = basicAuthHeader(user);
        }
        if (compressed) {
            headers["Accept-Encoding"] = "gzip, deflate, br";
        }
        const rawTextDataArgs = [...data, ...dataRaw, ...dataAscii];
        const rawBinaryDataArgs = [...dataBinary];
        const rawUrlencodeArgs = [...dataUrlencode];
        const hasForm = form.length > 0 || formString.length > 0;
        const hasJson = json !== undefined;
        const hasAnyData = hasJson ||
            hasForm ||
            rawTextDataArgs.length > 0 ||
            rawBinaryDataArgs.length > 0 ||
            rawUrlencodeArgs.length > 0;
        let method = "GET";
        if (forceHead) {
            method = "HEAD";
        }
        else if (request) {
            method = request;
        }
        else if (forceGet) {
            method = "GET";
        }
        else if (hasAnyData) {
            method = "POST";
        }
        let body;
        let effectiveUrl = parsedUrl;
        if (hasJson) {
            if (!headers["Content-Type"]) {
                headers["Content-Type"] = "application/json";
            }
            if (!headers["Accept"]) {
                headers["Accept"] = "application/json";
            }
            body = json;
        }
        else if (hasForm) {
            body = await buildFormData(form, formString);
        }
        else {
            const resolvedTextParts = [];
            for (const raw of rawTextDataArgs) {
                resolvedTextParts.push(await resolveDataArgText(raw));
            }
            const resolvedUrlEncodeParts = [];
            for (const raw of rawUrlencodeArgs) {
                // @file => read file, URL-encode contents
                if (raw.startsWith("@")) {
                    const content = await resolveDataArgText(raw);
                    resolvedUrlEncodeParts.push(encodeURIComponent(content));
                    continue;
                }
                const atIdx = raw.indexOf("@");
                const eqIdx = raw.indexOf("=");
                if (eqIdx !== -1) {
                    // name=value => encode both
                    const key = raw.slice(0, eqIdx);
                    const value = raw.slice(eqIdx + 1);
                    resolvedUrlEncodeParts.push(`${encodeURIComponent(key)}=${encodeURIComponent(value)}`);
                }
                else if (atIdx !== -1) {
                    // name@file => read file, encode contents as value
                    const key = raw.slice(0, atIdx);
                    const filePath = raw.slice(atIdx + 1);
                    const buf = filePath === "-"
                        ? await readAllStdin()
                        : await fs.readFile(filePath);
                    resolvedUrlEncodeParts.push(`${encodeURIComponent(key)}=${encodeURIComponent(buf.toString("utf8"))}`);
                }
                else {
                    resolvedUrlEncodeParts.push(encodeURIComponent(raw));
                }
            }
            const resolvedBinaryParts = [];
            for (const raw of rawBinaryDataArgs) {
                resolvedBinaryParts.push(await resolveDataArgBinary(raw));
            }
            const allTextParts = [...resolvedTextParts, ...resolvedUrlEncodeParts];
            if (forceGet && allTextParts.length > 0) {
                effectiveUrl = appendQueryParams(parsedUrl, allTextParts);
            }
            else if (resolvedBinaryParts.length > 0) {
                body = Buffer.concat(resolvedBinaryParts);
            }
            else if (allTextParts.length > 0) {
                body = allTextParts.join("&");
                if (!headers["Content-Type"]) {
                    headers["Content-Type"] = "application/x-www-form-urlencoded";
                }
            }
        }
        // curl defaults to NOT following redirects
        const redirect = location ? "follow" : "manual";
        if (verbose && !silent) {
            process.stderr.write(`> ${method} ${effectiveUrl.toString()}\n`);
            for (const [k, v] of Object.entries(headers)) {
                process.stderr.write(`> ${k}: ${v}\n`);
            }
            process.stderr.write(">\n");
        }
        const signal = maxTime ? AbortSignal.timeout(maxTime * 1000) : undefined;
        const start = performance.now();
        const response = await sdk.fetch(effectiveUrl.toString(), {
            method,
            headers,
            body,
            redirect,
            signal,
            connection,
        });
        const timeTotalSeconds = (performance.now() - start) / 1000;
        if (verbose && !silent) {
            process.stderr.write(`< HTTP ${response.status} ${response.statusText}\n`);
            response.headers.forEach((value, key) => {
                process.stderr.write(`< ${key}: ${value}\n`);
            });
            process.stderr.write("<\n");
        }
        const isHttpError = response.status >= 400;
        const shouldFail = (fail || failWithBody) && isHttpError;
        const shouldOutputBody = !shouldFail || !!failWithBody;
        const headerText = include
            ? `HTTP ${response.status} ${response.statusText}\n${Array.from(response.headers.entries())
                .map(([k, v]) => `${k}: ${v}`)
                .join("\n")}\n\n`
            : "";
        let bodyBytes = 0;
        const buf = shouldOutputBody
            ? Buffer.from(await response.arrayBuffer())
            : Buffer.alloc(0);
        bodyBytes = buf.length;
        const outputFile = output && output !== "-"
            ? output
            : remoteName
                ? deriveRemoteFilename(parsedUrl)
                : undefined;
        if (outputFile) {
            const dir = dirname(outputFile);
            if (dir !== ".") {
                await fs.mkdir(dir, { recursive: true });
            }
            const ws = createWriteStream(outputFile);
            if (headerText) {
                ws.write(headerText);
            }
            if (buf.length) {
                ws.write(buf);
            }
            await new Promise((resolve, reject) => {
                ws.end(() => resolve());
                ws.on("error", reject);
            });
        }
        else {
            if (headerText) {
                process.stdout.write(headerText);
            }
            if (buf.length) {
                process.stdout.write(buf);
            }
        }
        if (writeOut) {
            const formatted = formatWriteOut({
                template: writeOut,
                urlEffective: response.url || effectiveUrl.toString(),
                httpCode: response.status,
                timeTotalSeconds,
                sizeDownloadBytes: bodyBytes,
                contentType: response.headers.get("content-type"),
            });
            process.stdout.write(formatted);
        }
        if (shouldFail) {
            if (!silent || showError) {
                process.stderr.write(`curl: (22) The requested URL returned error: ${response.status}\n`);
            }
            throw new CurlExitError("HTTP request failed", 22);
        }
        return undefined;
    }
    return {
        curl,
        context: {
            meta: {
                curl: {
                    description: "Make authenticated HTTP requests to any API through Zapier. " +
                        "Pass a connection ID to automatically inject the user's stored credentials " +
                        "(OAuth tokens, API keys, etc.) into the outgoing request. " +
                        "Use it in place of the native curl command with additional Zapier-specific options.",
                    categories: ["http"],
                    inputSchema: CurlSchema,
                    aliases: {
                        request: "X",
                        header: "H",
                        data: "d",
                        form: "F",
                        get: "G",
                        head: "I",
                        location: "L",
                        include: "i",
                        output: "o",
                        remoteName: "O",
                        verbose: "v",
                        silent: "s",
                        showError: "S",
                        writeOut: "w",
                        maxTime: "m",
                        user: "u",
                        fail: "f",
                    },
                },
            },
        },
    };
};
