// @zapier/zapier-sdk-cli/dist/src/plugins/curl/schemas.js
import { z } from "zod";
export const CurlSchema = z
    .object({
    url: z.string().describe("Request URL"),
    request: z
        .enum(["GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"])
        .optional()
        .describe("HTTP method (defaults to GET, or POST if data is provided)"),
    header: z
        .array(z.string())
        .optional()
        .describe("HTTP headers in 'Key: Value' format (repeatable)"),
    data: z
        .array(z.string())
        .optional()
        .describe("HTTP POST data (repeatable, joined with &)"),
    dataRaw: z
        .array(z.string())
        .optional()
        .describe("HTTP POST data without special interpretation (repeatable)"),
    dataAscii: z
        .array(z.string())
        .optional()
        .describe("HTTP POST ASCII data (repeatable)"),
    dataBinary: z
        .array(z.string())
        .optional()
        .describe("HTTP POST binary data (repeatable)"),
    dataUrlencode: z
        .array(z.string())
        .optional()
        .describe("HTTP POST data, URL-encoded (repeatable)"),
    json: z
        .string()
        .optional()
        .describe("Send JSON body (sets Content-Type and Accept headers)"),
    form: z
        .array(z.string())
        .optional()
        .describe("Multipart form data as 'name=value' (repeatable)"),
    formString: z
        .array(z.string())
        .optional()
        .describe("Multipart form string field (repeatable)"),
    get: z
        .boolean()
        .optional()
        .describe("Force GET method and append data to query string"),
    head: z.boolean().optional().describe("Fetch headers only (HEAD request)"),
    location: z.boolean().optional().describe("Follow redirects"),
    include: z
        .boolean()
        .optional()
        .describe("Include response headers in output"),
    output: z
        .string()
        .optional()
        .describe("Write output to file instead of stdout"),
    remoteName: z
        .boolean()
        .optional()
        .describe("Write output to file named like the remote file"),
    verbose: z
        .boolean()
        .optional()
        .describe("Verbose output (show request/response headers on stderr)"),
    silent: z.boolean().optional().describe("Silent mode (suppress errors)"),
    showError: z
        .boolean()
        .optional()
        .describe("Show errors even when in silent mode"),
    fail: z
        .boolean()
        .optional()
        .describe("Fail silently on HTTP errors (exit code 22)"),
    failWithBody: z
        .boolean()
        .optional()
        .describe("Fail on HTTP errors but still output the body"),
    writeOut: z
        .string()
        .optional()
        .describe("Output format string after completion (e.g., '%{http_code}')"),
    maxTime: z
        .number()
        .optional()
        .describe("Maximum time in seconds for the request"),
    user: z
        .string()
        .optional()
        .describe("Basic auth credentials as 'user:password'"),
    compressed: z
        .boolean()
        .optional()
        .describe("Request compressed response (sends Accept-Encoding header)"),
    connection: z
        .union([z.string(), z.number()])
        .optional()
        .describe("Zapier connection ID or alias for authentication"),
    /** @deprecated Use `connection` instead. */
    connectionId: z
        .union([z.string(), z.number()])
        .optional()
        .meta({ deprecated: true }),
})
    .describe("Make HTTP requests through Zapier Relay with curl-like options");
