// @zapier/zapier-sdk-cli/dist/src/plugins/curl/utils.js
import { createHash } from "crypto";
import { promises as fs } from "fs";
import { basename } from "path";
export class CurlExitError extends Error {
    constructor(message, exitCode) {
        super(message);
        this.exitCode = exitCode;
        this.name = "CurlExitError";
    }
}
export function parseHeaderLine(input) {
    const idx = input.indexOf(":");
    if (idx === -1) {
        return null;
    }
    const key = input.slice(0, idx).trim();
    const value = input.slice(idx + 1).trim();
    if (!key) {
        return null;
    }
    return { key, value };
}
export function basicAuthHeader(userpass) {
    const idx = userpass.indexOf(":");
    const user = idx === -1 ? userpass : userpass.slice(0, idx);
    const pass = idx === -1 ? "" : userpass.slice(idx + 1);
    const token = Buffer.from(`${user}:${pass}`, "utf8").toString("base64");
    return `Basic ${token}`;
}
export function deriveRemoteFilename(url) {
    const path = url.pathname;
    const candidate = path.endsWith("/") ? "index.html" : basename(path);
    return candidate || "index.html";
}
export function decodeWriteOutEscapes(input) {
    return input
        .replace(/\\n/g, "\n")
        .replace(/\\r/g, "\r")
        .replace(/\\t/g, "\t");
}
export function formatWriteOut(params) {
    const { template } = params;
    const replacements = {
        "%{http_code}": String(params.httpCode).padStart(3, "0"),
        "%{time_total}": params.timeTotalSeconds.toFixed(3),
        "%{size_download}": String(params.sizeDownloadBytes),
        "%{url_effective}": params.urlEffective,
        "%{content_type}": params.contentType ?? "",
    };
    let out = template;
    for (const [token, value] of Object.entries(replacements)) {
        out = out.split(token).join(value);
    }
    return decodeWriteOutEscapes(out);
}
export function appendQueryParams(url, dataParts) {
    const out = new URL(url.toString());
    for (const part of dataParts) {
        const segments = part.split("&");
        for (const seg of segments) {
            if (!seg)
                continue;
            const idx = seg.indexOf("=");
            if (idx === -1) {
                out.searchParams.append(seg, "");
            }
            else {
                out.searchParams.append(seg.slice(0, idx), seg.slice(idx + 1));
            }
        }
    }
    return out;
}
export async function readAllStdin() {
    const chunks = [];
    for await (const chunk of process.stdin) {
        chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
    }
    return Buffer.concat(chunks);
}
export async function resolveDataArgText(raw) {
    if (!raw.startsWith("@")) {
        return raw;
    }
    const source = raw.slice(1);
    if (source === "-") {
        const buf = await readAllStdin();
        return buf.toString("utf8");
    }
    const buf = await fs.readFile(source);
    return buf.toString("utf8");
}
export async function resolveDataArgBinary(raw) {
    if (!raw.startsWith("@")) {
        return Buffer.from(raw, "utf8");
    }
    const source = raw.slice(1);
    if (source === "-") {
        return await readAllStdin();
    }
    return await fs.readFile(source);
}
export async function buildFormData(formArgs, formStringArgs) {
    if (typeof FormData === "undefined") {
        throw new CurlExitError("FormData is not available in this runtime; cannot use --form.", 2);
    }
    const fd = new FormData();
    const addField = async (item, forceString) => {
        const idx = item.indexOf("=");
        if (idx === -1) {
            throw new CurlExitError(`Invalid form field: '${item}'. Expected 'name=value' or 'name=@file'.`, 2);
        }
        const name = item.slice(0, idx);
        const value = item.slice(idx + 1);
        if (!name) {
            throw new CurlExitError(`Invalid form field: '${item}'. Field name cannot be empty.`, 2);
        }
        if (!forceString && value.startsWith("@")) {
            const filePath = value.slice(1);
            const buf = filePath === "-" ? await readAllStdin() : await fs.readFile(filePath);
            if (typeof Blob === "undefined") {
                fd.append(name, buf.toString("utf8"));
                return;
            }
            const blob = new Blob([buf]);
            const filename = filePath === "-"
                ? `stdin-${createHash("sha1").update(buf).digest("hex").slice(0, 8)}`
                : basename(filePath);
            fd.append(name, blob, filename);
            return;
        }
        fd.append(name, value);
    };
    for (const item of formArgs) {
        await addField(item, false);
    }
    for (const item of formStringArgs) {
        await addField(item, true);
    }
    return fd;
}
