// @zapier/zapier-sdk/dist/plugins/listApps/index.js
import { createPaginatedFunction } from "../../utils/function-utils";
import { DEFAULT_PAGE_SIZE } from "../../constants";
import { ListAppsSchema, } from "./schemas";
import { AppItemSchema } from "../../schemas/App";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { stripPageSuffix } from "../../utils/string-utils";
export const listAppsPlugin = ({ context }) => {
    async function listAppsPage(options) {
        const { api, resolveAppKeys } = context;
        const appKeys = options.apps ?? options.appKeys ?? [];
        const appLocators = await resolveAppKeys({
            appKeys: [...appKeys],
        });
        const implementationNameToLocator = {};
        for (const locator of appLocators) {
            implementationNameToLocator[locator.implementationName] = [
                ...(implementationNameToLocator[locator.implementationName] ?? []),
                locator,
            ];
        }
        const duplicatedLookupAppKeys = Object.keys(implementationNameToLocator)
            .filter((key) => implementationNameToLocator[key].length > 1)
            .map((key) => implementationNameToLocator[key])
            .flat()
            .map((locator) => locator.lookupAppKey);
        if (duplicatedLookupAppKeys.length > 0) {
            throw new Error(`Duplicate lookup app keys found: ${duplicatedLookupAppKeys.join(", ")}`);
        }
        // Step 2: Early return if user provided appKeys but they resolved to nothing AND no search
        // This prevents the handler from returning all apps when the user explicitly filtered
        // If search is provided, let the handler augment with search results
        if (appKeys.length > 0 && appLocators.length === 0 && !options.search) {
            return {
                data: [],
                links: { next: null },
                meta: { count: 0, limit: 0, offset: 0 },
            };
        }
        const implementationIds = appLocators.map((locator) => {
            const version = locator.version || "latest";
            return `${locator.implementationName}@${version}`;
        });
        return await api.get("/api/v0/apps", {
            searchParams: {
                appKeys: implementationIds.join(","),
                ...(options.search && { search: options.search }),
                ...(options.pageSize !== undefined && {
                    pageSize: options.pageSize.toString(),
                }),
                ...(options.cursor && { offset: options.cursor }),
            },
        });
    }
    const methodName = stripPageSuffix(listAppsPage.name);
    const listAppsDefinition = createPaginatedFunction(listAppsPage, ListAppsSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName), methodName, DEFAULT_PAGE_SIZE);
    return {
        listApps: listAppsDefinition,
        context: {
            meta: {
                listApps: {
                    categories: ["app"],
                    type: "list",
                    itemType: "App",
                    inputSchema: ListAppsSchema,
                    outputSchema: AppItemSchema,
                },
            },
        },
    };
};
