// @zapier/zapier-sdk-cli/dist/src/plugins/bundleCode/index.js
import { BundleCodeSchema } from "./schemas";
import { createFunction } from "@zapier/zapier-sdk";
import { buildSync } from "esbuild";
import * as fs from "fs";
import * as path from "path";
export const bundleCodePlugin = () => {
    const bundleCodeWithSdk = createFunction(async function bundleCodeWithSdk(options) {
        return await bundleCode(options);
    }, BundleCodeSchema);
    return {
        bundleCode: bundleCodeWithSdk,
        context: {
            meta: {
                bundleCode: {
                    categories: ["utility", "deprecated"],
                    inputSchema: BundleCodeSchema,
                },
            },
        },
    };
};
class ZapierBundleError extends Error {
    constructor(message, details, originalError) {
        super(message);
        this.code = "ZAPIER_BUNDLE_ERROR";
        this.name = "ZapierBundleError";
        this.details = details;
        this.originalError = originalError;
    }
}
/**
 * Bundle TypeScript code into executable JavaScript (CLI version)
 */
async function bundleCode(options) {
    const { input, output, target = "es2020", cjs = false, minify = false, string: returnString = false, } = options;
    // Resolve input path
    const resolvedInput = path.resolve(process.cwd(), input);
    try {
        // Bundle with esbuild
        const result = buildSync({
            entryPoints: [resolvedInput],
            bundle: true,
            platform: "node",
            target: target,
            format: cjs ? "cjs" : "esm",
            minify: minify,
            write: false,
            external: [], // Bundle everything
            banner: {
                js: "#!/usr/bin/env node",
            },
        });
        if (result.errors.length > 0) {
            const errorMessages = result.errors.map((e) => e.text);
            throw new ZapierBundleError(`Bundle failed: ${errorMessages.join(", ")}`, errorMessages);
        }
        const bundledCode = result.outputFiles?.[0]?.text;
        if (!bundledCode) {
            throw new ZapierBundleError("No output generated");
        }
        let finalOutput = bundledCode;
        if (returnString) {
            // Output as quoted string for node -e using JSON.stringify
            finalOutput = JSON.stringify(bundledCode);
        }
        // Write to file if output path specified
        if (output) {
            fs.mkdirSync(path.dirname(output), { recursive: true });
            fs.writeFileSync(output, finalOutput, "utf8");
        }
        return finalOutput;
    }
    catch (error) {
        // If it's already a ZapierBundleError, re-throw as-is
        if (error instanceof ZapierBundleError) {
            throw error;
        }
        throw new ZapierBundleError(`Bundle failed: ${error instanceof Error ? error.message : "Unknown error"}`, undefined, error instanceof Error ? error : undefined);
    }
}
