// @zapier/zapier-sdk/dist/schemas/Action.js
import { z } from "zod";
import { withFormatter } from "../utils/schema-utils";
import { ActionSchema } from "../api/schemas";
// ============================================================================
// Action Item Schema (extends API schema with computed fields and formatting)
// ============================================================================
export const ActionItemSchema = withFormatter(ActionSchema.omit({ type: true, name: true, selected_api: true }).extend({
    app_key: z.string(), // App key without version (extracted from selected_api)
    app_version: z.string().optional(), // Version extracted from selected_api
    action_type: ActionSchema.shape.type, // Mapped from original 'type' field
    title: z.string(), // Mapped from original 'name' field
    type: z.literal("action"), // Fixed type identifier
}), {
    format: (item) => {
        const details = [];
        details.push({
            text: `Type: ${item.action_type}`,
            style: "accent",
        });
        if (item.app_key) {
            details.push({
                text: `App: ${item.app_key}`,
                style: "normal",
            });
        }
        if (item.description) {
            details.push({
                text: item.description,
                style: "dim",
            });
        }
        return {
            title: item.title || item.key,
            id: item.id,
            key: item.key,
            description: item.description,
            details,
        };
    },
});
