// @zapier/zapier-sdk/dist/schemas/App.js
import { withFormatter } from "../utils/schema-utils";
import { AppItemSchema as AppItemSchemaBase } from "@zapier/zapier-sdk-core/v0/schemas/apps";
import { toSnakeCase } from "../utils/string-utils";
// ============================================================================
// App Item Schema (wraps core schema with SDK-specific formatting)
// ============================================================================
export const AppItemSchema = withFormatter(AppItemSchemaBase, {
    format: (item) => {
        // Create additional keys if slug exists
        const additionalKeys = [];
        if (item.slug && item.slug !== item.key) {
            additionalKeys.push(item.slug);
            const snakeCaseSlug = toSnakeCase(item.slug);
            if (snakeCaseSlug !== item.slug && snakeCaseSlug !== item.key) {
                additionalKeys.push(snakeCaseSlug);
            }
        }
        return {
            title: item.title,
            key: item.key,
            keys: [item.key, ...additionalKeys],
            description: item.description,
            details: [],
        };
    },
});
