// @zapier/zapier-sdk/dist/schemas/ClientCredentials.js
import { withFormatter } from "../utils/schema-utils";
import { ClientCredentialsItemSchema as ClientCredentialsItemSchemaBase, ClientCredentialsCreatedItemSchema as ClientCredentialsCreatedItemSchemaBase, } from "@zapier/zapier-sdk-core/v0/schemas/client-credentials";
// ============================================================================
// Client Credentials Item Schema (extends API schema with formatting)
// ============================================================================
export const ClientCredentialsItemSchema = withFormatter(ClientCredentialsItemSchemaBase, {
    format: (item) => {
        const details = [];
        if (item.allowed_scopes && item.allowed_scopes.length > 0) {
            details.push({
                text: `Scopes: ${item.allowed_scopes.join(", ")}`,
                style: "accent",
            });
        }
        if (item.created_at) {
            details.push({
                text: `Created: ${new Date(item.created_at).toLocaleDateString()}`,
                style: "dim",
            });
        }
        return {
            title: item.name,
            id: item.client_id,
            details,
        };
    },
});
// ============================================================================
// Client Credentials Created Item Schema (includes client_secret)
// ============================================================================
export const ClientCredentialsCreatedItemSchema = withFormatter(ClientCredentialsCreatedItemSchemaBase, {
    format: (item) => {
        return {
            title: item.name,
            id: item.client_id,
            details: [
                {
                    text: `Client Secret: ${item.client_secret}`,
                    style: "accent",
                },
            ],
        };
    },
});
