// @zapier/zapier-sdk/dist/schemas/Connection.js
import { withFormatter } from "../utils/schema-utils";
import { ConnectionItemSchema as ConnectionItemSchemaBase } from "@zapier/zapier-sdk-core/v0/schemas/connections";
// ============================================================================
// Connection Item Schema (extends API schema with computed fields and formatting)
// ============================================================================
export const ConnectionItemSchema = withFormatter(ConnectionItemSchemaBase, {
    format: (item) => {
        const details = [];
        if (item.identifier) {
            details.push({
                text: `Identifier: ${item.identifier}`,
                style: "accent",
            });
        }
        details.push({
            text: `Account: ${item.account_id} | Private: ${item.is_private} | Shared: ${item.shared_with_all}`,
            style: "dim",
        });
        if (item.marked_stale_at) {
            details.push({
                text: `Warning: Marked stale: ${new Date(item.marked_stale_at).toLocaleDateString()}`,
                style: "warning",
            });
        }
        return {
            title: item.title || `Connection ${item.id}`,
            id: item.id?.toString(),
            details,
        };
    },
});
