// @zapier/zapier-sdk/dist/schemas/Field.js
import { z } from "zod";
import { withFormatter } from "../utils/schema-utils";
// ============================================================================
// Base Field Schema
// ============================================================================
export const BaseFieldItemSchema = z.object({
    type: z.string(), // "input_field", "info_field", or "fieldset"
    key: z.string(), // From need.key
});
// ============================================================================
// Input Field Item Schema (extends API Need schema with computed fields and formatting)
// ============================================================================
export const InputFieldItemSchema = withFormatter(BaseFieldItemSchema.extend({
    type: z.literal("input_field"),
    default_value: z.string(), // Mapped from 'default' with fallback to ""
    depends_on: z.array(z.string()), // Mapped from 'depends_on' with fallback to []
    description: z.string(), // Mapped from 'help_text' with fallback to ""
    invalidates_input_fields: z.boolean(), // Mapped from 'alters_custom_fields' with fallback to false
    is_required: z.boolean(), // Mapped from 'required' with fallback to false
    placeholder: z.string(), // Mapped from 'placeholder' with fallback to ""
    title: z.string(), // Mapped from 'label' with fallback to ""
    value_type: z.string(), // Computed from InputFieldType enum
    format: z.string().optional(), // Computed from InputFieldFormat enum (optional)
    items: z.object({ type: z.string() }).optional(), // Computed from ItemsType enum (optional, only for arrays)
}), {
    format: (item) => {
        const details = [];
        // Type and required status
        const typeInfo = [
            item.value_type && `Type: ${item.value_type}`,
            `Required: ${item.is_required ? "Yes" : "No"}`,
        ]
            .filter(Boolean)
            .join(" | ");
        if (typeInfo) {
            details.push({ text: typeInfo, style: "dim" });
        }
        // Description
        if (item.description) {
            details.push({
                text: item.description,
                style: "normal",
            });
        }
        // Default value
        if (item.default_value) {
            details.push({
                text: `Default: ${item.default_value}`,
                style: "accent",
            });
        }
        // Placeholder
        if (item.placeholder) {
            details.push({
                text: `Placeholder: ${item.placeholder}`,
                style: "dim",
            });
        }
        // Dependencies
        if (item.depends_on && item.depends_on.length > 0) {
            details.push({
                text: `Depends on: ${item.depends_on.join(", ")}`,
                style: "warning",
            });
        }
        // Format information
        if (item.format) {
            details.push({ text: `Format: ${item.format}`, style: "dim" });
        }
        // Items type for arrays
        if (item.items) {
            details.push({
                text: `Items type: ${item.items.type}`,
                style: "dim",
            });
        }
        return {
            title: item.title || item.key,
            key: item.key,
            description: item.description,
            details,
        };
    },
});
// ============================================================================
// Info Field Item Schema (for help text and copy fields)
// ============================================================================
export const InfoFieldItemSchema = withFormatter(BaseFieldItemSchema.extend({
    type: z.literal("info_field"),
    description: z.string(), // From need.help_text
    title: z.string().optional(), // Optional title
}), {
    format: (item) => ({
        title: item.title || "Info",
        key: item.key,
        description: item.description,
        details: [{ text: item.description, style: "normal" }],
    }),
});
export const FieldsetItemSchema = BaseFieldItemSchema.extend({
    type: z.literal("fieldset"),
    title: z.string(),
    fields: z.lazy(() => z.array(z.union([
        InputFieldItemSchema,
        InfoFieldItemSchema,
        FieldsetItemSchema,
    ]))),
});
// ============================================================================
// Root Field Item Schema (union of all possible field types)
// ============================================================================
export const RootFieldItemSchema = z.union([
    InputFieldItemSchema,
    InfoFieldItemSchema,
    FieldsetItemSchema,
]);
