// @zapier/zapier-sdk/dist/schemas/Run.js
import { z } from "zod";
import { withFormatter } from "../utils/schema-utils";
// Type guard to safely access object properties
function getStringProperty(obj, key) {
    if (typeof obj === "object" && obj !== null && key in obj) {
        const value = obj[key];
        return typeof value === "string" ? value : undefined;
    }
    return undefined;
}
// Formatter for action results - handles any data structure
function formatActionResult(item) {
    // Try to extract common fields for structured display
    const obj = typeof item === "object" && item !== null ? item : {};
    // Try common title fields
    const title = getStringProperty(obj, "title") ||
        getStringProperty(obj, "name") ||
        getStringProperty(obj, "label") ||
        getStringProperty(obj, "subject") ||
        "Action Result";
    return {
        title,
        id: getStringProperty(obj, "id"),
        key: getStringProperty(obj, "key"),
        description: getStringProperty(obj, "description"),
        data: item, // Let formatJsonOutput handle the JSON rendering
        details: [], // Not used when data is provided
    };
}
// Flexible schema for action results - can be any object
export const ActionResultItemSchema = withFormatter(z.unknown().describe("Action execution result"), { format: formatActionResult });
