// @zapier/zapier-sdk-cli/dist/src/utils/auth/login.js
import open from "open";
import crypto from "node:crypto";
import express from "express";
import pkceChallenge from "pkce-challenge";
import { AUTH_MODE_HEADER, LOGIN_PORTS, LOGIN_TIMEOUT_MS } from "../constants";
import { spinPromise } from "../spinner";
import log from "../log";
import api from "../api/client";
import getCallablePromise from "../getCallablePromise";
import inquirer from "inquirer";
import { updateLogin, logout, getPkceLoginConfig, getLoginStorageMode, getConfigPath, } from "@zapier/zapier-sdk-cli-login";
import { ZapierCliUserCancellationError } from "../errors";
const findAvailablePort = () => {
    return new Promise((resolve, reject) => {
        let portIndex = 0;
        const tryPort = (port) => {
            const server = express().listen(port, () => {
                server.close();
                resolve(port);
            });
            server.on("error", (err) => {
                if (err.code === "EADDRINUSE") {
                    if (portIndex < LOGIN_PORTS.length) {
                        // Try next predefined port
                        tryPort(LOGIN_PORTS[portIndex++]);
                    }
                    else {
                        // All configured ports are busy
                        reject(new Error(`All configured OAuth callback ports are busy: ${LOGIN_PORTS.join(", ")}. Please try again later or close applications using these ports.`));
                    }
                }
                else {
                    reject(err);
                }
            });
        };
        if (LOGIN_PORTS.length > 0) {
            tryPort(LOGIN_PORTS[portIndex++]);
        }
        else {
            reject(new Error("No OAuth callback ports configured"));
        }
    });
};
const generateRandomString = () => {
    const array = new Uint32Array(28);
    crypto.getRandomValues(array);
    return Array.from(array, (dec) => ("0" + dec.toString(16)).substring(-2)).join("");
};
// Ensure offline_access is included in scope for PKCE (needed for refresh tokens)
function ensureOfflineAccess(scope) {
    if (scope.includes("offline_access")) {
        return scope;
    }
    return `${scope} offline_access`;
}
const login = async ({ timeoutMs = LOGIN_TIMEOUT_MS, credentials, }) => {
    const { clientId, tokenUrl, authorizeUrl } = getPkceLoginConfig({
        credentials,
    });
    const scope = ensureOfflineAccess(credentials?.scope || "internal credentials");
    await logout();
    // Find an available port
    const availablePort = await findAvailablePort();
    const redirectUri = `http://localhost:${availablePort}/oauth`;
    log.info(`Using port ${availablePort} for OAuth callback`);
    const { promise: promisedCode, resolve: setCode, reject: rejectCode, } = getCallablePromise();
    const app = express();
    app.get("/oauth", (req, res) => {
        setCode(String(req.query.code));
        // Set headers to prevent keep-alive
        res.setHeader("Connection", "close");
        res.end("You can now close this tab and return to the CLI.");
    });
    const server = app.listen(availablePort);
    // Track connections to force close them if needed
    const connections = new Set();
    server.on("connection", (conn) => {
        connections.add(conn);
        conn.on("close", () => connections.delete(conn));
    });
    // Set up signal handlers for graceful shutdown
    const cleanup = () => {
        server.close();
        log.info("\n❌ Login cancelled by user");
        rejectCode(new ZapierCliUserCancellationError());
    };
    process.on("SIGINT", cleanup);
    process.on("SIGTERM", cleanup);
    const { code_verifier: codeVerifier, code_challenge: codeChallenge } = await pkceChallenge();
    const authUrl = `${authorizeUrl}?${new URLSearchParams({
        response_type: "code",
        client_id: clientId,
        redirect_uri: redirectUri,
        scope,
        state: generateRandomString(),
        code_challenge: codeChallenge,
        code_challenge_method: "S256",
    }).toString()}`;
    log.info("Opening your browser to log in.");
    log.info("If it doesn't open, visit:", authUrl);
    open(authUrl);
    // Track the timeout timer so we can cancel it after login succeeds
    let timeoutTimer;
    try {
        await spinPromise(Promise.race([
            promisedCode,
            new Promise((_resolve, reject) => {
                timeoutTimer = setTimeout(() => {
                    reject(new Error(`Login timed out after ${Math.round(timeoutMs / 1000)} seconds.`));
                }, timeoutMs);
            }),
        ]), "Waiting for you to login and authorize");
    }
    finally {
        // Clear the timeout timer to prevent it from keeping the process alive
        if (timeoutTimer) {
            clearTimeout(timeoutTimer);
        }
        // Remove signal handlers
        process.off("SIGINT", cleanup);
        process.off("SIGTERM", cleanup);
        // Close server with timeout and force-close connections if needed
        await new Promise((resolve) => {
            const timeout = setTimeout(() => {
                log.info("Server close timed out, forcing connection shutdown...");
                // Force close all connections
                connections.forEach((conn) => conn.destroy());
                resolve();
            }, 1000); // 1 second timeout
            server.close(() => {
                clearTimeout(timeout);
                resolve();
            });
        });
    }
    log.info("Exchanging authorization code for tokens...");
    const { data } = await api.post(tokenUrl, {
        grant_type: "authorization_code",
        code: await promisedCode,
        redirect_uri: redirectUri,
        client_id: clientId,
        code_verifier: codeVerifier,
    }, {
        headers: {
            [AUTH_MODE_HEADER]: "no",
            "Content-Type": "application/x-www-form-urlencoded",
        },
    });
    let targetStorage;
    if (getLoginStorageMode() === "config") {
        const { upgrade } = await inquirer.prompt([
            {
                type: "confirm",
                name: "upgrade",
                message: "Would you like to upgrade to system keychain storage? " +
                    "This is recommended to securely store your credentials. " +
                    "However, note that older SDK/CLI versions will NOT be able to read these credentials, " +
                    "so you will want to upgrade them to the latest version.",
                default: true,
            },
        ]);
        targetStorage = upgrade ? "keychain" : "config";
    }
    else {
        targetStorage = "keychain";
    }
    try {
        await updateLogin(data, { storage: targetStorage });
    }
    catch (err) {
        if (targetStorage === "keychain") {
            log.warn(`Could not store credentials in system keychain. Storing in plaintext at ${getConfigPath()}.`);
            await updateLogin(data, { storage: "config" });
        }
        else {
            throw err;
        }
    }
    log.info("Token exchange completed successfully");
    return data.access_token;
};
export default login;
