// @zapier/zapier-sdk/dist/plugins/tables/createTable/index.js
import { createFunction } from "../../../utils/function-utils";
import { CreateTableOptionsSchema, CreateTableApiResponseSchema, TableItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableNameResolver } from "../../../resolvers";
import { transformTableItem } from "../utils";
export const createTablePlugin = ({ context }) => {
    async function createTable(options) {
        const { api } = context;
        const rawResponse = await api.post("/tables/api/v1/tables", {
            name: options.name,
            description: options.description,
        }, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to create tables or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to create tables. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
        });
        const response = CreateTableApiResponseSchema.parse(rawResponse);
        return {
            data: transformTableItem(response.data),
        };
    }
    const createTableDefinition = createFunction(createTable, CreateTableOptionsSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, createTable.name));
    return {
        createTable: createTableDefinition,
        context: {
            meta: {
                createTable: {
                    categories: ["table"],
                    type: "create",
                    itemType: "Table",
                    inputSchema: CreateTableOptionsSchema,
                    outputSchema: TableItemSchema,
                    resolvers: {
                        name: tableNameResolver,
                    },
                },
            },
        },
    };
};
