// @zapier/zapier-sdk/dist/plugins/listConnections/index.js
import { ListConnectionsQuerySchema, } from "./schemas";
import { splitVersionedKey, resolveConnectionId, transformConnectionItem, } from "../../utils/domain-utils";
import { createPaginatedFunction } from "../../utils/function-utils";
import { DEFAULT_PAGE_SIZE } from "../../constants";
import { appKeyResolver } from "../../resolvers";
import { ConnectionItemSchema } from "../../schemas/Connection";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { stripPageSuffix } from "../../utils/string-utils";
import { ZapierAuthenticationError, ZapierValidationError, } from "../../types/errors";
import { setMethodMetadata } from "../../utils/telemetry-context";
export const listConnectionsPlugin = ({ context }) => {
    async function listConnectionsPage(options) {
        if (options.includeShared) {
            await context.checkCapability("canIncludeSharedConnections");
        }
        const { api, getVersionedImplementationId } = context;
        const searchParams = {};
        if (options.pageSize !== undefined) {
            searchParams.page_size = options.pageSize.toString();
        }
        // Resolve app/appKey to implementationId if provided
        const appKey = options.app ?? options.appKey;
        if (appKey) {
            const implementationId = await getVersionedImplementationId(appKey);
            if (implementationId) {
                setMethodMetadata({ selectedApi: implementationId });
                // Extract versionless app key to pass to handler
                const [versionlessSelectedApi] = splitVersionedKey(implementationId);
                searchParams.app_key = versionlessSelectedApi;
            }
            else {
                // If we can't resolve via manifest, pass the raw appKey and let the API handle it
                searchParams.app_key = appKey;
            }
        }
        // Resolve connection references (names from map, or direct IDs)
        // Which parameter the values came from determines the resolution behavior:
        // - connections: resolve names from map, fall back to direct IDs
        // - connectionIds/authenticationIds: use directly (legacy behavior)
        const connectionRefs = options.connections;
        const legacyConnectionIds = options.connectionIds ?? options.authenticationIds;
        if (connectionRefs && connectionRefs.length > 0) {
            const resolvedIds = await Promise.all(connectionRefs.map((ref) => resolveConnectionId({
                connection: ref,
                resolveConnection: context.resolveConnection,
            })));
            searchParams.connection_ids = resolvedIds
                .filter((id) => id != null)
                .join(",");
        }
        else if (legacyConnectionIds && legacyConnectionIds.length > 0) {
            searchParams.connection_ids = legacyConnectionIds.join(",");
        }
        if (options.search) {
            searchParams.search = options.search;
        }
        if (options.title) {
            searchParams.title = options.title;
        }
        const accountId = options.account ?? options.accountId;
        if (accountId) {
            searchParams.account_id = accountId;
        }
        if (options.owner && options.owner !== "me" && !options.includeShared) {
            throw new ZapierValidationError('The "owner" option requires "includeShared" to be true. Without includeShared, only your own connections are returned.');
        }
        const owner = options.includeShared ? options.owner : "me";
        if (owner) {
            searchParams.owner = owner;
        }
        // isExpired takes priority when explicitly set. Otherwise expired defaults to false,
        // meaning only non-expired connections are returned unless the caller opts in.
        if (options.isExpired !== undefined) {
            searchParams.is_expired = options.isExpired.toString();
        }
        else {
            searchParams.is_expired = (options.expired ?? false).toString();
        }
        if (options.cursor) {
            searchParams.offset = options.cursor;
        }
        const response = await api.get("/api/v0/connections", {
            searchParams,
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to access connections or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to list connections. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
        });
        return {
            ...response,
            data: response.data.map(transformConnectionItem),
        };
    }
    const methodName = stripPageSuffix(listConnectionsPage.name);
    const listConnectionsDefinition = createPaginatedFunction(listConnectionsPage, ListConnectionsQuerySchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName), methodName, DEFAULT_PAGE_SIZE);
    return {
        listConnections: listConnectionsDefinition,
        context: {
            meta: {
                listConnections: {
                    categories: ["connection"],
                    type: "list",
                    itemType: "Connection",
                    inputSchema: ListConnectionsQuerySchema,
                    outputSchema: ConnectionItemSchema,
                    resolvers: {
                        app: appKeyResolver,
                    },
                },
            },
        },
    };
};
