// @zapier/zapier-sdk/dist/plugins/listConnections/schemas.js
import { z } from "zod";
import { AppKeyPropertySchema, AppPropertySchema, ConnectionsPropertySchema, } from "../../types/properties";
import { ListConnectionsQuerySchema as ListConnectionsQueryBase } from "@zapier/zapier-sdk-core/v0/schemas/connections";
export const ListConnectionsQuerySchema = ListConnectionsQueryBase.omit({
    offset: true,
    // Omit snake_case fields, we'll add camelCase versions
    app_key: true,
    connection_ids: true,
    account_id: true,
    is_expired: true,
    page_size: true,
})
    .extend({
    // New name for appKey
    app: AppPropertySchema.optional().describe("App key of connections to list (e.g., 'SlackCLIAPI' or slug like 'github')"),
    /** @deprecated Use `app` instead. */
    appKey: AppKeyPropertySchema.optional()
        .describe("App key of connections to list (e.g., 'SlackCLIAPI' or slug like 'github')")
        .meta({ deprecated: true }),
    // New name for connectionIds
    connections: ConnectionsPropertySchema.optional().describe("List of connection IDs to filter by"),
    /** @deprecated Use `connections` instead. */
    connectionIds: z
        .array(z.string())
        .optional()
        .describe("List of connection IDs to filter by")
        .meta({ deprecated: true }),
    /** @deprecated Use `connections` instead. */
    authenticationIds: z
        .array(z.string())
        .optional()
        .describe("List of authentication IDs to filter by (deprecated, use connections)")
        .meta({ deprecated: true }),
    account: z.string().optional().describe("Account to filter by"),
    /** @deprecated Use `account` instead. */
    accountId: z
        .string()
        .optional()
        .describe("Account ID to filter by")
        .meta({ deprecated: true }),
    // Include shared connections (overrides default owner=me filter)
    includeShared: z
        .boolean()
        .optional()
        .describe("Include connections shared with you. By default, only your own connections are returned (owner=me). Set to true to also include shared connections."),
    /** @deprecated Use `expired` instead */
    isExpired: z
        .boolean()
        .optional()
        .describe("Filter by expired status")
        .meta({
        deprecated: true,
        deprecationMessage: "Use --expired instead to show only expired connections.",
    }),
    expired: z
        .boolean()
        .optional()
        .describe("Show only expired connections (default: only non-expired connections are returned)"),
    // Override pageSize to make optional
    pageSize: z
        .number()
        .min(1)
        .optional()
        .describe("Number of connections per page"),
    // SDK specific property for pagination/iterable helpers
    maxItems: z
        .number()
        .min(1)
        .optional()
        .describe("Maximum total items to return across all pages"),
    // SDK specific property for pagination/iterable helpers
    cursor: z.string().optional().describe("Cursor to start from"),
})
    .describe("List available connections with optional filtering");
