// @zapier/zapier-sdk/dist/plugins/findFirstConnection/index.js
import { FindFirstConnectionSchema, } from "./schemas";
import { createFunction } from "../../utils/function-utils";
import { ConnectionItemSchema } from "../../schemas/Connection";
import { ZapierResourceNotFoundError } from "../../types/errors";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
export const findFirstConnectionPlugin = ({ sdk, context }) => {
    async function findFirstConnection(options = {}) {
        // Use listConnections with maxItems 1 to get the first result
        const connectionsResponse = await sdk.listConnections({
            ...options,
            maxItems: 1,
        });
        if (connectionsResponse.data.length === 0) {
            throw new ZapierResourceNotFoundError("No connection found matching the specified criteria", { resourceType: "Connection" });
        }
        return {
            data: connectionsResponse.data[0],
        };
    }
    const findFirstConnectionDefinition = createFunction(findFirstConnection, FindFirstConnectionSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, findFirstConnection.name));
    return {
        findFirstConnection: findFirstConnectionDefinition,
        context: {
            meta: {
                findFirstConnection: {
                    categories: ["connection"],
                    type: "item",
                    itemType: "Connection",
                    inputSchema: FindFirstConnectionSchema,
                    outputSchema: ConnectionItemSchema,
                },
            },
        },
    };
};
