// @zapier/zapier-sdk/dist/plugins/listInputFieldChoices/index.js
import { ListInputFieldChoicesSchema, ListInputFieldChoicesInputSchema, InputFieldChoiceItemSchema, } from "./schemas";
import { ZapierApiError, ZapierConfigurationError } from "../../types/errors";
import { createPaginatedFunction } from "../../utils/function-utils";
import { DEFAULT_PAGE_SIZE } from "../../constants";
import { appKeyResolver, actionTypeResolver, actionKeyResolver, connectionIdResolver, inputFieldKeyResolver, inputsAllOptionalResolver, } from "../../resolvers";
import { fetchImplementationNeeds, fetchImplementationChoices, } from "../../services/implementations";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { resolveConnectionId } from "../../utils/domain-utils";
import { stripPageSuffix } from "../../utils/string-utils";
import { setMethodMetadata } from "../../utils/telemetry-context";
// Transform NeedChoices to InputFieldChoiceItem
function transformNeedChoicesToInputFieldChoiceItem(choice) {
    return {
        key: choice.key,
        label: choice.label,
        sample: choice.sample,
        value: choice.value,
    };
}
export const listInputFieldChoicesPlugin = ({ context, sdk }) => {
    async function listInputFieldChoicesPage(options) {
        const { api, getVersionedImplementationId } = context;
        const appKey = "app" in options ? options.app : options.appKey;
        const actionKey = "action" in options ? options.action : options.actionKey;
        const inputFieldKey = "inputField" in options ? options.inputField : options.inputFieldKey;
        const { actionType, connection, connectionId, authenticationId, inputs, page, cursor, } = options;
        const resolvedConnectionId = await resolveConnectionId({
            connection,
            connectionId,
            authenticationId,
            resolveConnection: context.resolveConnection,
        });
        // Get versioned implementation ID for the app
        const selectedApi = await getVersionedImplementationId(appKey);
        if (!selectedApi) {
            throw new ZapierConfigurationError("No current_implementation_id found for app", { configType: "current_implementation_id" });
        }
        setMethodMetadata({
            selectedApi,
            operationType: actionType,
            operationKey: actionKey,
        });
        // Get action details
        const { data: action } = await sdk.getAction({
            app: appKey,
            actionType,
            action: actionKey,
        });
        // First, check for static choices by fetching implementation needs directly
        const needsData = await fetchImplementationNeeds({
            api,
            selectedApi,
            action: action.key,
            actionType,
            connectionId: resolvedConnectionId ?? null,
            inputs,
        });
        // Find the specific Need by key - search through raw needs array
        const targetNeed = needsData.needs?.find((need) => need.key === inputFieldKey);
        // If the need has static choices, return them
        if (targetNeed?.choices && targetNeed.choices.length > 0) {
            return {
                data: targetNeed.choices.map(transformNeedChoicesToInputFieldChoiceItem),
                nextCursor: undefined,
            };
        }
        if (!action.id) {
            throw new ZapierApiError(`Action ${actionKey} does not have an ID - cannot retrieve input field choices`);
        }
        // Use cursor (from pagination) as page number if available, otherwise use explicit page or default to 0
        const requestPage = cursor ? parseInt(cursor, 10) : (page ?? 0);
        // Use service layer to fetch dynamic choices
        const choicesData = await fetchImplementationChoices({
            api,
            actionId: action.id,
            inputFieldId: inputFieldKey,
            connectionId: resolvedConnectionId ?? null,
            inputs,
            page: requestPage,
        });
        // Transform NeedChoices objects to InputFieldChoiceItem objects
        const choices = (choicesData.choices || []).map(transformNeedChoicesToInputFieldChoiceItem);
        // Handle pagination
        let nextCursor;
        if (choicesData.next_page != null) {
            nextCursor = choicesData.next_page.toString();
        }
        else if (choicesData.links?.next) {
            // Extract page from next URL for external actions
            try {
                const nextUrl = new URL(choicesData.links.next);
                const nextPage = nextUrl.searchParams.get("page");
                if (nextPage) {
                    nextCursor = nextPage;
                }
            }
            catch {
                // Handle malformed URLs gracefully by not setting nextCursor
                nextCursor = undefined;
            }
        }
        return {
            data: choices,
            nextCursor,
        };
    }
    const methodName = stripPageSuffix(listInputFieldChoicesPage.name);
    const listInputFieldChoicesDefinition = createPaginatedFunction(listInputFieldChoicesPage, ListInputFieldChoicesInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName), methodName, DEFAULT_PAGE_SIZE);
    return {
        listInputFieldChoices: listInputFieldChoicesDefinition,
        context: {
            meta: {
                listInputFieldChoices: {
                    categories: ["action"],
                    type: "list",
                    itemType: "InputFieldChoice",
                    inputSchema: ListInputFieldChoicesSchema,
                    outputSchema: InputFieldChoiceItemSchema,
                    resolvers: {
                        app: appKeyResolver,
                        actionType: actionTypeResolver,
                        action: actionKeyResolver,
                        connection: connectionIdResolver,
                        inputField: inputFieldKeyResolver,
                        inputs: inputsAllOptionalResolver,
                    },
                },
            },
        },
    };
};
