// @zapier/zapier-sdk/dist/plugins/tables/getTableRecord/index.js
import { createFunction } from "../../../utils/function-utils";
import { GetTableRecordOptionsSchema, GetTableRecordOptionsInputSchema, GetTableRecordApiResponseSchema, RecordItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver, tableRecordIdResolver } from "../../../resolvers";
import { tableRecordFormatter } from "../../../formatters";
import { createFieldKeyTranslator } from "../utils";
import { transformRecordItem } from "../utils";
export const getTableRecordPlugin = ({ context }) => {
    async function getTableRecord(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const recordId = "record" in options ? options.record : options.recordId;
        const rawResponse = await api.get(`/tables/api/v1/tables/${tableId}/records/${recordId}`, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to access this record or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to access this record. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 404) {
                    return new Error(`Record not found: ${recordId}`);
                }
                return undefined;
            },
            authRequired: true,
        });
        const response = GetTableRecordApiResponseSchema.parse(rawResponse);
        const translator = await createFieldKeyTranslator({
            api,
            tableId,
            keyMode: options.keyMode,
        });
        return {
            data: {
                ...transformRecordItem(response.data),
                data: translator.translateOutput(response.data.data),
            },
        };
    }
    const getTableRecordDefinition = createFunction(getTableRecord, GetTableRecordOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, getTableRecord.name));
    return {
        getTableRecord: getTableRecordDefinition,
        context: {
            meta: {
                getTableRecord: {
                    categories: ["table"],
                    type: "item",
                    itemType: "Record",
                    inputSchema: GetTableRecordOptionsSchema,
                    outputSchema: RecordItemSchema,
                    resolvers: {
                        table: tableIdResolver,
                        record: tableRecordIdResolver,
                    },
                    formatter: tableRecordFormatter,
                },
            },
        },
    };
};
