// @zapier/zapier-sdk/dist/plugins/tables/getTableRecord/schemas.js
import { z } from "zod";
import { withFormatter } from "../../../utils/schema-utils";
import { KeyModeSchema } from "../utils";
import { TablePropertySchema, RecordPropertySchema, } from "../../../types/properties";
export const RecordApiItemSchema = z.object({
    id: z.string().regex(/^[A-Z0-9]{26}$/, "Record ID must be a valid ULID"),
    data: z.record(z.string(), z.unknown()),
    created_at: z.string(),
    edited_at: z.string(),
    schema_revision_id: z.number(),
    errors: z.record(z.string(), z.unknown()).nullable().optional(),
    orig_data: z.record(z.string(), z.unknown()).nullable().optional(),
    is_source_record: z.boolean().nullable().optional(),
});
// SDK output schema (simplified)
const RecordItemSchemaBase = z.object({
    id: z.string(),
    data: z.record(z.string(), z.unknown()),
    created_at: z.string(),
    edited_at: z.string(),
});
export const RecordItemSchema = withFormatter(RecordItemSchemaBase, {
    format: (item) => ({
        title: `Record ${item.id}`,
        id: item.id,
        data: item.data,
        details: [],
    }),
});
export const GetTableRecordApiResponseSchema = z.object({
    data: RecordApiItemSchema,
});
const GetTableRecordDescription = "Get a single record from a table by ID";
// Base schema for shared (non-renamed) properties
const GetTableRecordOptionsBaseSchema = z.object({
    keyMode: KeyModeSchema,
});
// Schema for docs/registry (new param names only)
export const GetTableRecordOptionsSchema = z
    .object({
    table: TablePropertySchema,
    record: RecordPropertySchema,
})
    .merge(GetTableRecordOptionsBaseSchema)
    .describe(GetTableRecordDescription)
    .meta({ aliases: { tableId: "table", recordId: "record" } });
// Deprecated schema (backward compat)
const GetTableRecordOptionsSchemaDeprecated = z
    .object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
    recordId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Record ID must be a valid ULID")
        .describe("The unique identifier of the record"),
})
    .merge(GetTableRecordOptionsBaseSchema);
// Union for runtime validation
export const GetTableRecordOptionsInputSchema = z
    .union([GetTableRecordOptionsSchema, GetTableRecordOptionsSchemaDeprecated])
    .describe(GetTableRecordDescription);
