// @zapier/zapier-sdk-cli/dist/src/plugins/login/index.js
import { isCredentialsObject, buildApplicationLifecycleEvent, } from "@zapier/zapier-sdk";
import login from "../../utils/auth/login";
import { getLoggedInUser, } from "@zapier/zapier-sdk-cli-login";
import { LoginSchema } from "./schemas";
function toPkceCredentials(credentials) {
    if (credentials &&
        isCredentialsObject(credentials) &&
        !("clientSecret" in credentials)) {
        return {
            type: "pkce",
            clientId: credentials.clientId,
            baseUrl: credentials.baseUrl,
            scope: credentials.scope,
        };
    }
    return undefined;
}
export const loginPlugin = ({ context, }) => {
    const loginFn = async (options) => {
        const timeoutSeconds = options.timeout
            ? parseInt(options.timeout, 10)
            : 300;
        if (isNaN(timeoutSeconds) || timeoutSeconds <= 0) {
            throw new Error("Timeout must be a positive number");
        }
        const resolvedCredentials = await context.resolveCredentials();
        const pkceCredentials = toPkceCredentials(resolvedCredentials);
        await login({
            timeoutMs: timeoutSeconds * 1000,
            credentials: pkceCredentials,
        });
        const user = await getLoggedInUser();
        context.eventEmission.emit("platform.sdk.ApplicationLifecycleEvent", buildApplicationLifecycleEvent({ lifecycle_event_type: "login_success" }, { customuser_id: user.customUserId, account_id: user.accountId }));
        console.log(`✅ Successfully logged in as ${user.email}`);
    };
    return {
        login: loginFn,
        context: {
            meta: {
                login: {
                    categories: ["account"],
                    inputSchema: LoginSchema,
                    supportsJsonOutput: false,
                },
            },
        },
    };
};
