// @zapier/zapier-sdk/dist/plugins/getConnection/index.js
import { createFunction } from "../../utils/function-utils";
import { connectionIdGenericResolver } from "../../resolvers";
import { transformConnectionItem } from "../../utils/domain-utils";
import { ConnectionItemSchema } from "../../schemas/Connection";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { GetConnectionParamSchema } from "./schemas";
export const getConnectionPlugin = ({ context }) => {
    async function getConnection(options) {
        const { api } = context;
        const resolvedConnectionId = options.connection ?? options.connectionId ?? options.authenticationId;
        if (!resolvedConnectionId) {
            throw new Error("connection is required");
        }
        const response = await api.get(`/api/v0/connections/${encodeURIComponent(String(resolvedConnectionId))}`);
        return {
            ...response,
            data: transformConnectionItem(response.data),
        };
    }
    const getConnectionDefinition = createFunction(getConnection, GetConnectionParamSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, getConnection.name));
    return {
        getConnection: getConnectionDefinition,
        context: {
            meta: {
                getConnection: {
                    categories: ["connection"],
                    type: "item",
                    itemType: "Connection",
                    inputSchema: GetConnectionParamSchema,
                    outputSchema: ConnectionItemSchema,
                    resolvers: {
                        connection: connectionIdGenericResolver,
                    },
                },
            },
        },
    };
};
