// @zapier/zapier-sdk/dist/plugins/fetch/index.js
import { ZapierRelayError } from "../../types/errors";
import { FetchUrlSchema, FetchInitSchema } from "./schemas";
import { resolveConnectionId } from "../../utils/domain-utils";
import { isPlainObject } from "../../utils/type-guard-utils";
// fetch() doesn't use createFunction/createPaginatedFunction (its signature
// doesn't fit the single-options-object pattern), so we wire up telemetry
// context tracking manually.
import { isTelemetryNested, runWithTelemetryContext, } from "../../utils/telemetry-context";
/**
 * Transforms full URLs into Relay format: /relay/{domain}/{path}
 */
function transformUrlToRelayPath(url) {
    const targetUrl = new URL(url);
    return `/relay/${targetUrl.host}${targetUrl.pathname}${targetUrl.search}${targetUrl.hash}`;
}
/**
 * Normalizes various header formats into a plain Record<string, string>.
 */
function normalizeHeaders(optionsHeaders) {
    const headers = {};
    if (!optionsHeaders) {
        return headers;
    }
    const headerEntries = optionsHeaders instanceof Headers
        ? Array.from(optionsHeaders.entries())
        : Array.isArray(optionsHeaders)
            ? optionsHeaders
            : Object.entries(optionsHeaders);
    for (const [key, value] of headerEntries) {
        headers[key] = value;
    }
    return headers;
}
function inferContentType(body) {
    if (body instanceof URLSearchParams) {
        return "application/x-www-form-urlencoded; charset=utf-8";
    }
    if (typeof body === "string") {
        const trimmed = body.trimStart();
        if (!trimmed.startsWith("{") && !trimmed.startsWith("[")) {
            return undefined;
        }
        try {
            JSON.parse(trimmed);
            return "application/json; charset=utf-8";
        }
        catch {
            return undefined;
        }
    }
    if (isPlainObject(body) || Array.isArray(body)) {
        return "application/json; charset=utf-8";
    }
    return undefined;
}
/**
 * Fetch plugin — the primary way to make authenticated HTTP requests through Zapier's Relay service.
 * Mirrors the native fetch(url, init?) signature with additional Zapier-specific options.
 */
export const fetchPlugin = ({ context }) => {
    return {
        fetch: async function fetch(url, init) {
            return runWithTelemetryContext(async () => {
                const { api } = context;
                const startTime = Date.now();
                const isNested = isTelemetryNested();
                try {
                    const { connectionId, connection, authenticationId, callbackUrl, ...fetchInit } = init || {};
                    const resolvedConnectionId = await resolveConnectionId({
                        connectionId,
                        connection,
                        authenticationId,
                        resolveConnection: context.resolveConnection,
                        label: "fetch",
                    });
                    const relayPath = transformUrlToRelayPath(url);
                    const headers = normalizeHeaders(fetchInit.headers);
                    const hasContentType = Object.keys(headers).some((k) => k.toLowerCase() === "content-type");
                    if (fetchInit.body && !hasContentType) {
                        const inferred = inferContentType(fetchInit.body);
                        if (inferred) {
                            headers["Content-Type"] = inferred;
                        }
                    }
                    if (resolvedConnectionId) {
                        headers["X-Relay-Authentication-Id"] = String(resolvedConnectionId);
                    }
                    if (callbackUrl) {
                        headers["X-Relay-Callback-Url"] = callbackUrl;
                    }
                    const result = await api.fetch(relayPath, {
                        method: fetchInit.method ?? "GET",
                        body: fetchInit.body,
                        headers,
                        redirect: fetchInit.redirect,
                        signal: fetchInit.signal,
                        authRequired: true,
                    });
                    const relayError = result.headers.get("x-relay-error");
                    if (relayError) {
                        throw new ZapierRelayError(relayError, {
                            statusCode: result.status,
                        });
                    }
                    if (!isNested) {
                        context.eventEmission.emitMethodCalled({
                            method_name: "fetch",
                            execution_duration_ms: Date.now() - startTime,
                            success_flag: true,
                            error_message: null,
                            error_type: null,
                            argument_count: init ? 2 : 1,
                            is_paginated: false,
                        });
                    }
                    return result;
                }
                catch (error) {
                    if (!isNested) {
                        context.eventEmission.emitMethodCalled({
                            method_name: "fetch",
                            execution_duration_ms: Date.now() - startTime,
                            success_flag: false,
                            error_message: error instanceof Error ? error.message : String(error),
                            error_type: error instanceof Error ? error.constructor.name : "Unknown",
                            argument_count: init ? 2 : 1,
                            is_paginated: false,
                        });
                    }
                    throw error;
                }
            });
        },
        context: {
            meta: {
                fetch: {
                    description: "Make authenticated HTTP requests to any API through Zapier. " +
                        "Pass a connectionId to automatically inject the user's stored credentials " +
                        "(OAuth tokens, API keys, etc.) into the outgoing request. " +
                        "Mirrors the native fetch(url, init?) signature with additional Zapier-specific options.",
                    packages: ["sdk", "cli", "mcp"],
                    categories: ["http"],
                    returnType: "Response",
                    inputParameters: [
                        { name: "url", schema: FetchUrlSchema },
                        { name: "init", schema: FetchInitSchema },
                    ],
                },
            },
        },
    };
};
