// @zapier/zapier-sdk/dist/utils/function-utils.js
/**
 * Generic utility functions for creating paginated SDK functions
 */
import { ZapierError, ZapierUnknownError, ZapierAuthenticationError, } from "../types/errors";
import { paginate } from "./pagination-utils";
import { createValidator, validateOptions } from "./validation";
import { isTelemetryNested, runWithTelemetryContext, } from "./telemetry-context";
/**
 * Helper function to extract cursor from pagination data
 * Supports JSON:API style `links.next` URLs by extracting the `offset` parameter.
 */
export function extractCursor(data) {
    if (!data?.next) {
        return undefined;
    }
    try {
        const urlObj = new URL(data.next);
        const offset = urlObj.searchParams.get("offset");
        return offset || undefined;
    }
    catch {
        return undefined;
    }
}
/**
 * Helper function to convert unknown errors to SDK errors.
 * Recognizes errors by name (e.g., from cli-login package which can't
 * import SDK errors directly) and converts them to proper SDK errors.
 */
function normalizeError(error) {
    if (error instanceof ZapierError) {
        return error;
    }
    const message = error instanceof Error ? error.message : String(error);
    const cause = error instanceof Error ? error : undefined;
    // Check for known error names from packages that can't import SDK errors
    if (error instanceof Error) {
        if (error.name === "ZapierAuthenticationError") {
            return new ZapierAuthenticationError(message, { cause });
        }
    }
    return new ZapierUnknownError(message, { cause });
}
/**
 * Higher-order function that creates a function from a core function
 *
 * @param coreFn - Function that returns T directly or throws errors
 * @param schema - Optional Zod schema for validation
 * @param telemetry - Optional telemetry callbacks
 * @returns A function that normalizes and rethrows errors
 */
export function createFunction(coreFn, schema, telemetry) {
    const functionName = coreFn.name;
    // Create a named function using dynamic property access
    const namedFunctions = {
        [functionName]: async function (options) {
            return runWithTelemetryContext(async () => {
                const startTime = Date.now();
                const normalizedOptions = (options ?? {});
                const isNested = isTelemetryNested();
                try {
                    if (schema) {
                        const validatedOptions = validateOptions(schema, normalizedOptions);
                        const result = await coreFn({
                            ...normalizedOptions,
                            ...validatedOptions,
                        });
                        if (!isNested && telemetry?.onMethodCalled) {
                            const argumentCount = Object.keys(normalizedOptions).length;
                            telemetry.onMethodCalled({
                                methodName: functionName,
                                durationMs: Date.now() - startTime,
                                success: true,
                                argumentCount,
                                isPaginated: false,
                            });
                        }
                        return result;
                    }
                    const result = await coreFn(normalizedOptions);
                    if (!isNested && telemetry?.onMethodCalled) {
                        const argumentCount = Object.keys(normalizedOptions).length;
                        telemetry.onMethodCalled({
                            methodName: functionName,
                            durationMs: Date.now() - startTime,
                            success: true,
                            argumentCount,
                            isPaginated: false,
                        });
                    }
                    return result;
                }
                catch (error) {
                    const normalizedError = normalizeError(error);
                    if (!isNested && telemetry?.onMethodCalled) {
                        const argumentCount = Object.keys(normalizedOptions).length;
                        telemetry.onMethodCalled({
                            methodName: functionName,
                            durationMs: Date.now() - startTime,
                            success: false,
                            error: normalizedError,
                            argumentCount,
                            isPaginated: false,
                        });
                    }
                    throw normalizedError;
                }
            });
        },
    };
    return namedFunctions[functionName];
}
/**
 * Higher-order function that creates a page function that wraps results in {data}
 * Internal utility used by createPaginatedFunction
 */
function createPageFunction(coreFn) {
    const functionName = coreFn.name + "Page";
    // Create a named function using dynamic property access
    const namedFunctions = {
        [functionName]: async function (options) {
            try {
                const result = await coreFn(options);
                // If result already has a data property, ensure it's an array
                if (result && typeof result === "object" && "data" in result) {
                    const paginatedResult = result;
                    // Support both nextCursor and JSON:API links.next format
                    let nextCursor;
                    if (paginatedResult.nextCursor) {
                        nextCursor = paginatedResult.nextCursor;
                    }
                    else if (paginatedResult.links?.next) {
                        // Extract cursor from fully qualified URL (JSON:API spec format)
                        nextCursor = extractCursor({ next: paginatedResult.links.next });
                    }
                    return {
                        data: Array.isArray(paginatedResult.data)
                            ? paginatedResult.data
                            : [paginatedResult.data],
                        nextCursor,
                    };
                }
                // If result is an array, wrap in {data: result}
                if (Array.isArray(result)) {
                    return { data: result };
                }
                // If result is non-array, wrap in {data: [result]}
                return { data: [result] };
            }
            catch (error) {
                // Catch and normalize any errors, then rethrow
                throw normalizeError(error);
            }
        },
    };
    return namedFunctions[functionName];
}
export function createPaginatedFunction(coreFn, schema, telemetry, explicitFunctionName, defaultPageSize) {
    const pageFunction = createPageFunction(coreFn);
    const functionName = explicitFunctionName || coreFn.name;
    const validator = schema ? createValidator(schema) : null;
    // Create the main paginated function
    const namedFunctions = {
        [functionName]: function (options) {
            return runWithTelemetryContext(() => {
                const startTime = Date.now();
                // Convert undefined options to empty object
                const normalizedOptions = (options ?? {});
                const isNested = isTelemetryNested();
                // Validate options if schema provided (validated fields take precedence)
                const validatedOptions = {
                    ...normalizedOptions,
                    ...(validator ? validator(normalizedOptions) : normalizedOptions),
                };
                const pageSize = validatedOptions.pageSize ?? defaultPageSize;
                const optimizedOptions = {
                    ...validatedOptions,
                    pageSize,
                };
                const iterator = paginate(pageFunction, optimizedOptions);
                // Get a promise for the first iteration
                const firstPagePromise = iterator.next().then((result) => {
                    if (result.done) {
                        throw new Error("Paginate should always iterate at least once");
                    }
                    return result.value;
                });
                // Emit telemetry after first page completes
                if (!isNested && telemetry?.onMethodCalled) {
                    firstPagePromise
                        .then(() => {
                        const argumentCount = Object.keys(normalizedOptions).length;
                        telemetry.onMethodCalled({
                            methodName: functionName,
                            durationMs: Date.now() - startTime,
                            success: true,
                            argumentCount,
                            isPaginated: true,
                        });
                    })
                        .catch((error) => {
                        const argumentCount = Object.keys(normalizedOptions).length;
                        telemetry.onMethodCalled({
                            methodName: functionName,
                            durationMs: Date.now() - startTime,
                            success: false,
                            error: error instanceof Error ? error : new Error(String(error)),
                            argumentCount,
                            isPaginated: true,
                        });
                    });
                }
                // Add Symbol.asyncIterator to make the promise iterable
                return Object.assign(firstPagePromise, {
                    [Symbol.asyncIterator]: async function* () {
                        // Yield the first page (await the promise)
                        yield await firstPagePromise;
                        // Yield the rest of the pages
                        for await (const page of iterator) {
                            yield page;
                        }
                    },
                    items: function () {
                        return {
                            [Symbol.asyncIterator]: async function* () {
                                // Yield the first page items
                                const firstPage = await firstPagePromise;
                                for (const item of firstPage.data) {
                                    yield item;
                                }
                                // Yield items from the rest of the pages
                                for await (const page of iterator) {
                                    for (const item of page.data) {
                                        yield item;
                                    }
                                }
                            },
                        };
                    },
                });
            });
        },
    };
    return namedFunctions[functionName];
}
