// @zapier/zapier-sdk/dist/utils/pagination-utils.js
const CURSOR_VERSION = 1;
const CURSOR_SOURCE = {
    API: "api",
    SDK: "sdk",
};
function encodeBase64(str) {
    return btoa(Array.from(new TextEncoder().encode(str), (b) => String.fromCharCode(b)).join(""));
}
function decodeBase64(str) {
    return new TextDecoder().decode(Uint8Array.from(atob(str), (c) => c.charCodeAt(0)));
}
function encodeApiCursor(cursor) {
    const envelope = {
        v: CURSOR_VERSION,
        source: CURSOR_SOURCE.API,
        cursor,
    };
    return encodeBase64(JSON.stringify(envelope));
}
function encodeSdkCursor(offset, cursor) {
    const envelope = {
        v: CURSOR_VERSION,
        source: CURSOR_SOURCE.SDK,
        cursor,
        offset,
    };
    return encodeBase64(JSON.stringify(envelope));
}
export function decodeIncomingCursor(incoming) {
    if (!incoming) {
        return { offset: 0, cursor: undefined };
    }
    try {
        const decoded = decodeBase64(incoming);
        const envelope = JSON.parse(decoded);
        if (envelope.v !== CURSOR_VERSION) {
            return { offset: 0, cursor: incoming };
        }
        if (envelope.source === CURSOR_SOURCE.SDK) {
            return { offset: envelope.offset ?? 0, cursor: envelope.cursor };
        }
        if (envelope.source === CURSOR_SOURCE.API) {
            return { offset: 0, cursor: envelope.cursor };
        }
        return { offset: 0, cursor: incoming };
    }
    catch {
        return { offset: 0, cursor: incoming };
    }
}
export function createPrefixedCursor(prefix, cursor) {
    if (!cursor) {
        return `${prefix}::`;
    }
    return `${prefix}::${cursor}`;
}
export function splitPrefixedCursor(cursor, prefixes) {
    if (!cursor) {
        return [undefined, undefined];
    }
    const [prefix, ...rest] = cursor.split("::");
    if (prefixes && !prefixes.includes(prefix)) {
        return [undefined, cursor];
    }
    cursor = rest.join("::");
    if (!cursor) {
        return [prefix, undefined];
    }
    return [prefix, cursor];
}
/**
 * Internal primitive: paginates through a page function using raw (unencoded) cursors.
 * Encoding and decoding of cursors is the responsibility of the public wrappers.
 */
async function* paginateMaxItemsWithUnencodedCursor(pageFunction, pageOptions) {
    let cursor = pageOptions?.cursor;
    let totalItemsYielded = 0;
    const maxItems = pageOptions?.maxItems;
    const pageSize = pageOptions?.pageSize;
    do {
        const options = {
            ...(pageOptions || {}),
            cursor,
            pageSize: maxItems !== undefined && pageSize !== undefined
                ? Math.min(pageSize, maxItems)
                : pageSize,
        };
        const page = await pageFunction(options);
        if (maxItems !== undefined) {
            const remainingItems = maxItems - totalItemsYielded;
            if (page.data.length >= remainingItems) {
                yield {
                    ...page,
                    data: page.data.slice(0, remainingItems),
                    nextCursor: undefined,
                };
                break;
            }
        }
        yield page;
        totalItemsYielded += page.data.length;
        cursor = page.nextCursor;
    } while (cursor);
}
/**
 * Utility for paginating through API endpoints that return cursor-based pages.
 * Accepts and yields SDK-encoded cursor envelopes. Any incoming cursor is decoded
 * before being passed to the page function; all outgoing cursors are encoded.
 *
 * @param pageFunction - Function that fetches a single page with {data, nextCursor} structure
 * @param pageOptions - Options to pass to the page function (cursor will be managed automatically)
 * @returns Async iterator that yields pages with encoded cursors
 */
export async function* paginateMaxItems(pageFunction, pageOptions) {
    const { cursor } = decodeIncomingCursor(pageOptions?.cursor);
    const options = {
        ...(pageOptions || {}),
        cursor,
    };
    for await (const page of paginateMaxItemsWithUnencodedCursor(pageFunction, options)) {
        yield {
            ...page,
            nextCursor: page.nextCursor
                ? encodeApiCursor(page.nextCursor)
                : undefined,
        };
    }
}
export async function* paginateBuffered(pageFunction, pageOptions) {
    const pageSize = pageOptions?.pageSize;
    const { offset: cursorOffset, cursor: initialCursor } = decodeIncomingCursor(pageOptions?.cursor);
    const requestedMaxItems = pageOptions?.maxItems;
    const options = {
        ...(pageOptions || {}),
        cursor: initialCursor,
        // SDK cursors can carry an offset into a raw backend page. Since maxItems
        // is expected to be relative to the resumed position, we add that offset
        // so raw pagination still yields enough items after offset slicing.
        maxItems: requestedMaxItems !== undefined && cursorOffset > 0
            ? requestedMaxItems + cursorOffset
            : requestedMaxItems,
    };
    if (!pageSize) {
        for await (const page of paginateMaxItemsWithUnencodedCursor(pageFunction, options)) {
            yield {
                ...page,
                nextCursor: page.nextCursor
                    ? encodeApiCursor(page.nextCursor)
                    : undefined,
            };
        }
        return;
    }
    let bufferedPages = [];
    let isFirstPage = true;
    let rawCursor;
    for await (let page of paginateMaxItemsWithUnencodedCursor(pageFunction, options)) {
        const nextRawCursor = page.nextCursor;
        if (isFirstPage) {
            isFirstPage = false;
            if (cursorOffset) {
                page = {
                    ...page,
                    data: page.data.slice(cursorOffset),
                };
            }
        }
        const bufferedLength = bufferedPages.reduce((acc, p) => acc + p.data.length, 0);
        // If we don't have enough to fill a page, buffer this page.
        if (bufferedLength + page.data.length < pageSize) {
            bufferedPages.push(page);
            rawCursor = nextRawCursor;
            continue;
        }
        // Let's yield a page from our buffered pages.
        const bufferedItems = bufferedPages.map((p) => p.data).flat();
        const allItems = [...bufferedItems, ...page.data];
        const pageItems = allItems.slice(0, pageSize);
        const remainingItems = allItems.slice(pageItems.length);
        // No extra items to buffer, so we can just yield a normal page with a cursor to the next one.
        if (remainingItems.length === 0) {
            yield {
                ...page,
                data: pageItems,
                nextCursor: nextRawCursor ? encodeApiCursor(nextRawCursor) : undefined,
            };
            bufferedPages = [];
            rawCursor = nextRawCursor;
            continue;
        }
        // Yield our items with a cursor to offset into this page.
        yield {
            ...page,
            data: pageItems,
            nextCursor: encodeSdkCursor(page.data.length - remainingItems.length, rawCursor),
        };
        while (remainingItems.length > pageSize) {
            const chunkItems = remainingItems.splice(0, pageSize);
            yield {
                ...page,
                data: chunkItems,
                nextCursor: encodeSdkCursor(page.data.length - remainingItems.length, rawCursor),
            };
        }
        bufferedPages = [
            {
                ...page,
                data: remainingItems,
            },
        ];
        rawCursor = nextRawCursor;
    }
    if (bufferedPages.length > 0) {
        const lastBufferedPage = bufferedPages.slice(-1)[0];
        const bufferedItems = bufferedPages.map((p) => p.data).flat();
        yield {
            ...lastBufferedPage,
            data: bufferedItems,
        };
    }
}
export const paginate = paginateBuffered;
/**
 * Concatenate multiple paginated SDK results into a single paginated stream.
 * Each source is a function returning a dual Promise+AsyncIterable (as SDK
 * paginated methods return). Sources are drained in order.
 *
 * The optional `dedupe` key extractor filters items from source N against
 * all items seen in sources 0 through N-1.
 *
 * Uses paginateBuffered internally to normalize page sizes across source
 * boundaries — e.g. if the first source only has 2 items, they'll be
 * buffered with items from the next source into a full page.
 *
 * Returns the same dual Promise+AsyncIterable shape that resolvers expect.
 */
export function concatPaginated({ sources, dedupe, pageSize = 100, }) {
    if (sources.length === 0) {
        const empty = { data: [] };
        return Object.assign(Promise.resolve(empty), {
            [Symbol.asyncIterator]: async function* () {
                yield empty;
            },
        });
    }
    let sourceIndex = 0;
    let currentIterator = null;
    const seen = new Set();
    const pageFunction = async (_options) => {
        while (sourceIndex < sources.length) {
            if (!currentIterator) {
                const result = sources[sourceIndex]();
                currentIterator = result[Symbol.asyncIterator]();
            }
            const next = await currentIterator.next();
            if (next.done) {
                sourceIndex++;
                currentIterator = null;
                continue;
            }
            let items = next.value.data;
            if (dedupe) {
                if (sourceIndex > 0) {
                    items = items.filter((item) => !seen.has(dedupe(item)));
                }
                for (const item of items) {
                    seen.add(dedupe(item));
                }
            }
            const hasMoreInSource = next.value.nextCursor != null;
            const hasMoreSources = sourceIndex < sources.length - 1;
            // Synthetic sentinel — not a real API cursor. It just signals to
            // paginateBuffered that more data exists (from the current source or
            // the next one). This works because concatPaginated manages its own
            // iteration state internally. Currently only used by CLI resolvers;
            // if this is ever used in user-facing SDK methods, we should replace
            // it with a stateless composite cursor (e.g. sourceIndex + inner cursor).
            return {
                data: items,
                nextCursor: hasMoreInSource || hasMoreSources ? "__has_more__" : undefined,
            };
        }
        return { data: [] };
    };
    const iterator = paginateBuffered(pageFunction, { pageSize });
    const firstPagePromise = iterator.next().then((result) => {
        if (result.done) {
            return { data: [] };
        }
        return result.value;
    });
    return Object.assign(firstPagePromise, {
        [Symbol.asyncIterator]: async function* () {
            yield await firstPagePromise;
            for await (const page of { [Symbol.asyncIterator]: () => iterator }) {
                yield page;
            }
        },
    });
}
/**
 * Strip the PromiseLike from an async iterable, returning a plain
 * AsyncIterable. This prevents async functions from unwrapping the
 * iterable (since async only unwraps PromiseLike, not AsyncIterable).
 */
export function toIterable(source) {
    return { [Symbol.asyncIterator]: () => source[Symbol.asyncIterator]() };
}
