// @zapier/zapier-sdk/dist/utils/schema-utils.js
import { z } from "zod";
// Helper function to add format metadata to schemas
export function withFormatter(schema, formatMeta) {
    // Store format metadata on the schema definition
    Object.assign(schema._zod.def, {
        formatMeta: formatMeta,
    });
    return schema;
}
// Helper function to get output schema from an input schema
export function getOutputSchema(inputSchema) {
    return inputSchema._zod.def.outputSchema;
}
// Helper function to link input schemas to output schemas
export function withOutputSchema(inputSchema, outputSchema) {
    // Store output schema reference on the input schema
    Object.assign(inputSchema._zod.def, {
        outputSchema,
    });
    return inputSchema;
}
// Helper function to add resolver metadata to schemas
export function withResolver(schema, config) {
    // Store resolver metadata on the schema definition
    schema._zod.def.resolverMeta = config;
    return schema;
}
// ============================================================================
// Schema Description Utilities
// ============================================================================
export function getSchemaDescription(schema) {
    return schema.description;
}
export function getFieldDescriptions(schema) {
    const descriptions = {};
    const shape = schema.shape;
    for (const [key, fieldSchema] of Object.entries(shape)) {
        if (fieldSchema instanceof z.ZodType && fieldSchema.description) {
            descriptions[key] = fieldSchema.description;
        }
    }
    return descriptions;
}
// Helper function to mark a parameter as positional for CLI
export function withPositional(schema) {
    // Store positional metadata on the schema definition
    Object.assign(schema._zod.def, {
        positionalMeta: { positional: true },
    });
    return schema;
}
function schemaHasPositionalMeta(schema) {
    return "positionalMeta" in schema._zod.def;
}
// Helper function to check if a parameter should be positional
export function isPositional(schema) {
    // Check the current schema first
    if (schemaHasPositionalMeta(schema) &&
        schema._zod.def
            .positionalMeta?.positional) {
        return true;
    }
    // If this is a ZodOptional, check the inner type
    if (schema instanceof z.ZodOptional) {
        return isPositional(schema._zod.def.innerType);
    }
    // If this is a ZodDefault, check the inner type
    if (schema instanceof z.ZodDefault) {
        return isPositional(schema._zod.def.innerType);
    }
    return false;
}
