// @zapier/zapier-sdk/dist/utils/telemetry-utils.js
import { getMethodMetadata } from "./telemetry-context";
/**
 * Creates a standardized telemetry callback for SDK method invocations.
 *
 * This factory eliminates boilerplate by providing a consistent implementation
 * of the onMethodCalled callback used by createFunction and createPaginatedFunction.
 *
 * @param emitMethodCalled - The event emission function from EventEmissionContext
 * @param methodName - The method name to report in telemetry. Use `function.name` to derive
 *   this automatically from the function declaration, ensuring DRY principles and making
 *   refactoring easier. Function names are preserved through bundling.
 * @returns TelemetryCallbacks object with standardized onMethodCalled implementation
 *
 * @example
 * Basic usage with createFunction:
 * ```typescript
 * async function listApps(options: ListAppsOptions) { ... }
 *
 * const listAppsDefinition = createFunction(
 *   listApps,
 *   ListAppsSchema,
 *   createTelemetryCallback(context.eventEmission.emitMethodCalled, listApps.name),
 * );
 * ```
 *
 * @example
 * For paginated functions, use stripPageSuffix to remove the "Page" suffix:
 * ```typescript
 * import { stripPageSuffix } from "./string-utils";
 *
 * async function listAppsPage(options: ListAppsOptions & { pageSize: number }) { ... }
 *
 * const methodName = stripPageSuffix(listAppsPage.name);
 * const listAppsDefinition = createPaginatedFunction(
 *   listAppsPage,
 *   ListAppsSchema,
 *   createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName),
 *   methodName,
 * );
 * ```
 */
export function createTelemetryCallback(emitMethodCalled, methodName) {
    return {
        onMethodCalled: (data) => {
            const metadata = getMethodMetadata();
            emitMethodCalled({
                method_name: methodName,
                execution_duration_ms: data.durationMs,
                success_flag: data.success,
                error_message: data.error?.message ?? null,
                error_type: data.error?.constructor.name ?? null,
                argument_count: data.argumentCount,
                is_paginated: data.isPaginated,
                selected_api: metadata?.selectedApi ?? null,
                operation_type: metadata?.operationType ?? null,
                operation_key: metadata?.operationKey ?? null,
            });
        },
    };
}
