// @zapier/zapier-sdk/dist/api/auth.js
/**
 * Authentication Utilities
 *
 * This module provides utilities for handling authentication tokens,
 * including JWT detection and authorization header generation.
 */
export function isJwt(token) {
    // JWT tokens have exactly 3 parts separated by dots
    const parts = token.split(".");
    if (parts.length !== 3) {
        return false;
    }
    // Each part should be base64url encoded (no padding)
    // Basic validation - each part should be non-empty and contain valid base64url characters
    const base64UrlPattern = /^[A-Za-z0-9_-]+$/;
    return parts.every((part) => part.length > 0 && base64UrlPattern.test(part));
}
/**
 * Parse a JWT token into its three parts
 * Returns the parts array if valid, null otherwise
 */
function parseJwt(token) {
    if (!isJwt(token)) {
        return null;
    }
    return token.split(".");
}
export function getAuthorizationHeader(token) {
    // Check if token is a JWT (has 3 parts separated by dots)
    if (isJwt(token)) {
        return `JWT ${token}`;
    }
    // Default to Bearer for other token types
    return `Bearer ${token}`;
}
/**
 * Extract user IDs from JWT token
 * Decodes the JWT payload and extracts customuser_id and account_id
 * Handles nested JWTs for service tokens
 * Returns null values on any failure (silent-by-design)
 */
export function extractUserIdsFromJwt(token) {
    const parts = parseJwt(token);
    if (!parts) {
        return { customuser_id: null, account_id: null };
    }
    try {
        const payload = JSON.parse(Buffer.from(parts[1], "base64url").toString("utf-8"));
        let actualPayload = payload;
        if (payload.sub_type === "service" && payload.njwt) {
            const nestedParts = payload.njwt.split(".");
            if (nestedParts.length === 3) {
                actualPayload = JSON.parse(Buffer.from(nestedParts[1], "base64url").toString("utf-8"));
            }
        }
        const accountId = actualPayload["zap:acc"] != null
            ? parseInt(String(actualPayload["zap:acc"]), 10)
            : null;
        const customUserId = actualPayload.sub_type === "customuser" && actualPayload.sub != null
            ? parseInt(String(actualPayload.sub), 10)
            : null;
        return {
            customuser_id: customUserId !== null && !isNaN(customUserId) ? customUserId : null,
            account_id: accountId !== null && !isNaN(accountId) ? accountId : null,
        };
    }
    catch {
        return { customuser_id: null, account_id: null };
    }
}
