// @zapier/zapier-sdk/dist/api/client.js
/**
 * API Client Implementation
 *
 * This module contains the core API client implementation, including
 * HTTP method handlers, response processing, and client factory functions.
 */
import { getAuthorizationHeader } from "./auth";
import { createDebugLogger, createDebugFetch } from "./debug";
import { pollUntilComplete } from "./polling";
import { resolveAuthToken, invalidateCredentialsToken, isCliLoginAvailable, } from "../auth";
import { getZapierBaseUrl } from "../utils/url-utils";
import { sleep, calculateExponentialBackoffMs } from "../utils/retry-utils";
import { isPlainObject } from "../utils/type-guard-utils";
import { ZapierApiError, ZapierAuthenticationError, ZapierValidationError, ZapierNotFoundError, ZapierRateLimitError, } from "../types/errors";
import { ZAPIER_MAX_NETWORK_RETRIES, ZAPIER_MAX_NETWORK_RETRY_DELAY_MS, } from "../constants";
/**
 * Parse rate limit information from response headers
 */
function parseRateLimitHeaders(response) {
    const info = {};
    // Parse Retry-After header (can be seconds or HTTP date)
    const retryAfter = response.headers.get("retry-after");
    if (retryAfter) {
        const seconds = parseInt(retryAfter, 10);
        if (!isNaN(seconds)) {
            info.retryAfterMs = seconds * 1000;
        }
        else {
            // Try parsing as HTTP date
            const date = Date.parse(retryAfter);
            if (!isNaN(date)) {
                info.retryAfterMs = Math.max(0, date - Date.now());
            }
        }
    }
    // Parse X-RateLimit-Reset (Unix timestamp)
    const reset = response.headers.get("x-ratelimit-reset");
    if (reset) {
        const resetTimestamp = parseInt(reset, 10);
        if (!isNaN(resetTimestamp)) {
            // Convert to ms (timestamps are usually in seconds)
            info.resetMs = resetTimestamp * 1000;
            // If no Retry-After, calculate wait time from reset
            if (info.retryAfterMs === undefined) {
                info.retryAfterMs = Math.max(0, info.resetMs - Date.now());
            }
        }
    }
    // Parse X-RateLimit-Limit
    const limit = response.headers.get("x-ratelimit-limit");
    if (limit) {
        const limitNum = parseInt(limit, 10);
        if (!isNaN(limitNum)) {
            info.limit = limitNum;
        }
    }
    // Parse X-RateLimit-Remaining
    const remaining = response.headers.get("x-ratelimit-remaining");
    if (remaining) {
        const remainingNum = parseInt(remaining, 10);
        if (!isNaN(remainingNum)) {
            info.remaining = remainingNum;
        }
    }
    return info;
}
// Configuration for paths
const pathConfig = {
    // e.g. /relay -> https://sdkapi.zapier.com/api/v0/sdk/relay/...
    "/relay": {
        authHeader: "X-Relay-Authorization",
        pathPrefix: "/api/v0/sdk/relay",
    },
    // e.g. /zapier -> https://sdkapi.zapier.com/api/v0/sdk/zapier/...
    "/zapier": {
        authHeader: "Authorization",
        pathPrefix: "/api/v0/sdk/zapier",
    },
    // e.g. /tables -> https://sdkapi.zapier.com/api/v0/sdk/tables/...
    "/tables": {
        authHeader: "Authorization",
        pathPrefix: "/api/v0/sdk/tables",
    },
};
class ZapierApiClient {
    constructor(options) {
        this.options = options;
        this.fetch = async (path, init) => {
            if (!path.startsWith("/")) {
                throw new ZapierValidationError(`fetch expects a path starting with '/', got: ${path}`);
            }
            if (init?.body && (isPlainObject(init.body) || Array.isArray(init.body))) {
                init.body = JSON.stringify(init.body);
            }
            const { url, pathConfig } = this.buildUrl(path, init?.searchParams);
            const builtHeaders = await this.buildHeaders(init, pathConfig);
            const inputHeaders = new Headers(init?.headers ?? {});
            const mergedHeaders = new Headers();
            builtHeaders.forEach((value, key) => {
                mergedHeaders.set(key, value);
            });
            inputHeaders.forEach((value, key) => {
                mergedHeaders.set(key, value);
            });
            // Retry loop for rate limiting (429)
            let retries = 0;
            while (true) {
                const response = await this.options.fetch(url, {
                    ...init,
                    headers: mergedHeaders,
                });
                // Handle rate limiting (429) with retries
                if (response.status === 429) {
                    const rateLimitInfo = parseRateLimitHeaders(response);
                    // Calculate delay: prefer Retry-After/X-RateLimit-Reset, fall back to exponential backoff
                    const delayMs = rateLimitInfo.retryAfterMs ??
                        calculateExponentialBackoffMs(retries + 1);
                    // If delay exceeds maxNetworkRetryDelayMs or retries exhausted, throw immediately
                    if (delayMs > this.maxNetworkRetryDelayMs ||
                        retries >= this.maxNetworkRetries) {
                        throw new ZapierRateLimitError("Rate limited", {
                            statusCode: 429,
                            rateLimit: rateLimitInfo,
                            retries,
                        });
                    }
                    retries++;
                    // Emit retry event
                    this.emitEvent("api:rate_limit_retry", {
                        retry: retries,
                        maxNetworkRetries: this.maxNetworkRetries,
                        delayMs,
                        path,
                        method: init?.method ?? "GET",
                        rateLimit: rateLimitInfo,
                    });
                    // Wait before retrying
                    await sleep(delayMs);
                    continue;
                }
                return response;
            }
        };
        this.get = async (path, options = {}) => {
            return this.fetchJson("GET", path, undefined, options);
        };
        this.post = async (path, data, options = {}) => {
            return this.fetchJson("POST", path, data, options);
        };
        this.put = async (path, data, options = {}) => {
            return this.fetchJson("PUT", path, data, options);
        };
        this.patch = async (path, data, options = {}) => {
            return this.fetchJson("PATCH", path, data, options);
        };
        this.delete = async (path, data, options = {}) => {
            return this.fetchJson("DELETE", path, data, options);
        };
        this.poll = async (path, options = {}) => {
            return pollUntilComplete({
                fetchPoll: () => this.fetch(path, {
                    method: "GET",
                    searchParams: options.searchParams,
                    authRequired: options.authRequired,
                }),
                initialDelay: options.initialDelay,
                timeoutMs: options.timeoutMs,
                successStatus: options.successStatus,
                pendingStatus: options.pendingStatus,
                isPending: options.isPending,
                resultExtractor: options.resultExtractor,
            });
        };
        this.maxNetworkRetries =
            options.maxNetworkRetries ?? ZAPIER_MAX_NETWORK_RETRIES;
        this.maxNetworkRetryDelayMs =
            options.maxNetworkRetryDelayMs ?? ZAPIER_MAX_NETWORK_RETRY_DELAY_MS;
    }
    // Emit an event if onEvent handler is configured
    emitEvent(type, payload) {
        if (this.options.onEvent) {
            this.options.onEvent({ type, payload, timestamp: Date.now() });
        }
    }
    // Helper to parse response data
    async parseResult(response) {
        // Read as text first to avoid "Body has already been read" errors
        // if json() fails partway through parsing
        const text = await response.text();
        try {
            return { type: "json", data: JSON.parse(text) };
        }
        catch {
            return { type: "text", data: text };
        }
    }
    // Helper to get a token from the different places it could be gotten
    async getAuthToken(options) {
        return resolveAuthToken({
            credentials: this.options.credentials,
            token: this.options.token,
            onEvent: this.options.onEvent,
            fetch: this.options.fetch,
            baseUrl: this.options.baseUrl,
            requiredScopes: options?.requiredScopes,
            debug: this.options.debug,
        });
    }
    // Helper to handle responses
    async handleResponse(params) {
        const { response, customErrorHandler, wasMissingAuthToken, requiredScopes, } = params;
        const { data: responseData } = await this.parseResult(response);
        if (response.ok) {
            return responseData;
        }
        const errorInfo = {
            status: response.status,
            statusText: response.statusText,
            data: responseData,
        };
        // Check for custom error handling first
        if (customErrorHandler) {
            const customError = customErrorHandler(errorInfo);
            if (customError) {
                // javascript may not respect customErrorHandler return type, so we need to throw a proper error if it's not an Error
                if (customError instanceof Error) {
                    throw customError;
                }
                else {
                    throw new Error(`customErrorHandler returned a non-Error: ${JSON.stringify(customError)}`);
                }
            }
        }
        // Parse the error response
        const { message, errors } = this.parseErrorResponse(errorInfo);
        const errorOptions = {
            statusCode: response.status,
            errors,
        };
        // Use appropriate error type based on status code
        if (response.status === 404) {
            throw new ZapierNotFoundError(message, errorOptions);
        }
        if (response.status === 401 || response.status === 403) {
            // If we get a 401/403 error and no auth token was provided, give helpful message
            if (wasMissingAuthToken) {
                throw new ZapierAuthenticationError(`Authentication required (HTTP ${response.status}). Please provide credentials in options or set ZAPIER_CREDENTIALS environment variable.`, errorOptions);
            }
            // On 401, invalidate cached token so next request gets a fresh one
            if (response.status === 401) {
                await invalidateCredentialsToken({
                    credentials: this.options.credentials,
                    token: this.options.token,
                    requiredScopes,
                });
            }
            throw new ZapierAuthenticationError(message, errorOptions);
        }
        if (response.status === 400) {
            throw new ZapierValidationError(message, errorOptions);
        }
        // Generic API error for other status codes
        throw new ZapierApiError(message, errorOptions);
    }
    hasErrorArray(data) {
        return (typeof data === "object" &&
            data !== null &&
            "errors" in data &&
            Array.isArray(data.errors));
    }
    // Helper to check if data has API errors
    isApiErrorArray(dataArray) {
        const data = dataArray[0];
        return (typeof data === "object" &&
            data !== null &&
            "message" in data &&
            "code" in data &&
            "title" in data &&
            "detail" in data);
    }
    // Do our best to extract an error message from the response data
    extractErrorMessage(data) {
        if (typeof data === "string") {
            return data;
        }
        if (typeof data === "object" && data !== null) {
            if ("message" in data && typeof data.message === "string") {
                return data.message;
            }
            if ("error" in data) {
                if (typeof data.error === "string") {
                    return data.error;
                }
                if (typeof data.error === "object" && data.error !== null) {
                    if ("message" in data.error &&
                        typeof data.error.message === "string") {
                        return data.error.message;
                    }
                }
                try {
                    return JSON.stringify(data.error);
                }
                catch {
                    /* defend against circular objects, even though they shouldn't be coming back from the wire */
                }
            }
            if ("errors" in data && Array.isArray(data.errors)) {
                if (this.isApiErrorArray(data.errors)) {
                    return data.errors[0].detail || data.errors[0].title;
                }
                else if (data.errors.length > 0) {
                    // Handle simple string errors array
                    const firstError = data.errors[0];
                    if (typeof firstError === "string") {
                        return firstError;
                    }
                    // For non-string errors, stringify them
                    try {
                        return JSON.stringify(firstError);
                    }
                    catch {
                        return String(firstError);
                    }
                }
            }
        }
        return undefined;
    }
    // Helper to parse API error response
    parseErrorResponse(errorInfo) {
        // If we can't parse data, use status text
        const fallbackMessage = `HTTP ${errorInfo.status}: ${errorInfo.statusText}`;
        try {
            // Check if the response has the standard error format
            if (typeof errorInfo.data === "string") {
                return { message: `${fallbackMessage}: ${errorInfo.data}` };
            }
            const errorMessage = this.extractErrorMessage(errorInfo.data) || fallbackMessage;
            if (this.hasErrorArray(errorInfo.data)) {
                if (this.isApiErrorArray(errorInfo.data.errors)) {
                    return {
                        message: errorMessage,
                        errors: errorInfo.data.errors,
                    };
                }
                else {
                    return {
                        message: errorMessage,
                        errors: errorInfo.data.errors.map((e) => ({
                            status: errorInfo.status,
                            code: String(errorInfo.status),
                            title: errorInfo.statusText,
                            detail: JSON.stringify(e),
                        })),
                    };
                }
            }
            return { message: errorMessage };
        }
        catch {
            return { message: fallbackMessage };
        }
    }
    // Apply any special routing logic for configured paths.
    applyPathConfiguration(path) {
        // Find matching path configuration.
        const matchingPathKey = Object.keys(pathConfig).find((configPath) => path === configPath || path.startsWith(configPath + "/"));
        const config = matchingPathKey ? pathConfig[matchingPathKey] : undefined;
        // Apply pathPrefix if there's a matching config with pathPrefix.
        let finalPath = path;
        if (config &&
            "pathPrefix" in config &&
            config.pathPrefix &&
            matchingPathKey) {
            // Strip the matching path key, and use the pathPrefix instead.
            const pathWithoutPrefix = path.slice(matchingPathKey.length) || "/";
            finalPath = `${config.pathPrefix}${pathWithoutPrefix}`;
        }
        // Check if baseUrl is a Zapier-inferred base URL.
        const zapierBaseUrl = getZapierBaseUrl(this.options.baseUrl);
        // Let's remain compatible with a base URL that is set to a Zapier-inferred
        // domain, rather than requiring the base URL to go to our proxy. Later, the
        // proxy will be removed, so this should make that transition easier.
        if (zapierBaseUrl === this.options.baseUrl.replace(/\/$/, "")) {
            // If baseUrl is already the Zapier base URL, use sdkapi subdomain.
            const originalBaseUrl = new URL(this.options.baseUrl);
            const finalBaseUrl = `https://sdkapi.${originalBaseUrl.hostname}`;
            return {
                url: new URL(finalPath, finalBaseUrl),
                pathConfig: config,
            };
        }
        // For a base URL that isn't a Zapier-inferred domain (e.g., localhost),
        // preserve any path from the base URL and append the final path.
        const baseUrl = new URL(this.options.baseUrl);
        const basePath = baseUrl.pathname.replace(/\/$/, "");
        return {
            url: new URL(basePath + finalPath, baseUrl.origin),
            pathConfig: config,
        };
    }
    // Helper to build full URLs and return routing info
    buildUrl(path, searchParams) {
        const { url, pathConfig: config } = this.applyPathConfiguration(path);
        if (searchParams) {
            Object.entries(searchParams).forEach(([key, value]) => {
                url.searchParams.set(key, value);
            });
        }
        return { url: url.toString(), pathConfig: config };
    }
    // Helper to build headers
    async buildHeaders(options = {}, pathConfig) {
        const headers = new Headers(options.headers ?? {});
        // Even if auth is not required, we still want to add it in case it adds
        // useful context to the API. The session is a good example of this. Auth
        // is not required, but if we don't add auth, then we won't get the user's
        // session!
        const authToken = await this.getAuthToken({
            requiredScopes: options.requiredScopes,
        });
        if (authToken) {
            const authHeaderName = pathConfig && pathConfig.authHeader
                ? pathConfig.authHeader
                : "Authorization";
            headers.set(authHeaderName, getAuthorizationHeader(authToken));
        }
        // If we know auth is required, and we don't have a token, throw an error
        // before we even make a request.
        if (options.authRequired) {
            const authHeaderName = pathConfig?.authHeader ?? "Authorization";
            if (headers.get(authHeaderName) == null && authToken == null) {
                const message = isCliLoginAvailable() === false
                    ? "Authentication required but no credentials available. " +
                        "To use CLI login, install the CLI as a dev dependency " +
                        "(e.g. `npm install -D @zapier/zapier-sdk-cli`) and " +
                        "log in (e.g. `npx zapier-sdk login`). " +
                        "Alternatively, set the ZAPIER_CREDENTIALS environment variable " +
                        "or ZAPIER_CREDENTIALS_CLIENT_ID and " +
                        "ZAPIER_CREDENTIALS_CLIENT_SECRET environment variables."
                    : "Authentication required but no credentials available. " +
                        "Please log in (e.g. `npx zapier-sdk login`) or set the " +
                        "ZAPIER_CREDENTIALS environment variable or " +
                        "ZAPIER_CREDENTIALS_CLIENT_ID and " +
                        "ZAPIER_CREDENTIALS_CLIENT_SECRET environment variables.";
                throw new ZapierAuthenticationError(message);
            }
        }
        return headers;
    }
    // Helper to perform HTTP requests with JSON handling
    async fetchJson(method, path, data, options = {}) {
        const headers = { ...options.headers };
        // Add Content-Type for JSON requests with body data
        if (data && typeof data === "object") {
            headers["Content-Type"] = "application/json";
        }
        // Check if we have an auth token available
        const wasMissingAuthToken = options.authRequired &&
            (await this.getAuthToken({ requiredScopes: options.requiredScopes })) ==
                null;
        const response = await this.fetch(path, {
            ...options,
            method,
            body: data != null ? JSON.stringify(data) : undefined,
            headers,
        });
        const result = await this.handleResponse({
            response,
            customErrorHandler: options.customErrorHandler,
            wasMissingAuthToken,
            requiredScopes: options.requiredScopes,
        });
        // Allow empty responses for DELETE or 204 No Content
        if (typeof result === "string") {
            if (result === "" && (method === "DELETE" || response.status === 204)) {
                return undefined;
            }
            throw new ZapierValidationError(`Response could not be parsed as JSON: ${result}`);
        }
        return result;
    }
}
export const createZapierApi = (options) => {
    const { debug = false, fetch: originalFetch = globalThis.fetch } = options;
    const debugLog = createDebugLogger(debug);
    const debugFetch = createDebugFetch({ originalFetch, debugLog });
    return new ZapierApiClient({
        ...options,
        debug,
        fetch: debugFetch,
    });
};
