// @zapier/zapier-sdk/dist/api/debug.js
/**
 * Debug Logging Utilities
 *
 * This module provides debug logging capabilities for API requests and responses,
 * including request timing and detailed logging of HTTP interactions.
 */
// We don't want to import util since it might not work in a browser. So we'll lazily load it and cache the result so
// we can use it synchronously after that. The types above are just to make our linter happy.
let utilModule = null;
let utilPromise = null;
export function createDebugLogger(enabled) {
    if (!enabled) {
        return () => { }; // No-op function when debug is disabled
    }
    // Try to load util module on first use.
    if (!utilPromise) {
        utilPromise = import("util")
            .then((util) => {
            utilModule = util;
            return util;
        })
            .catch(() => {
            utilModule = null;
            return null;
        });
    }
    return (message, data) => {
        if (data === undefined || data === "") {
            console.log(`[Zapier SDK] ${message}`);
            return;
        }
        // Use cached module if available for sync logging. Yes, this means if the util module isn't yet loaded, we'll fall
        // back to a plain old console.log, which might mean we don't see proper formatting/nesting. But that's unlikely
        // since this logging happens during async fetching. Alternatively, we could switch to async logging, but this
        // seems fine since we have to have a fallback anyway.
        if (utilModule) {
            const formatted = utilModule.inspect(data, {
                colors: true,
                depth: null,
                breakLength: 80,
            });
            console.log(`[Zapier SDK] ${message}`, formatted);
        }
        else {
            // Fallback - browser will handle formatting automatically
            console.log(`[Zapier SDK] ${message}`, data);
        }
    };
}
export function censorHeaders(headers) {
    if (!headers)
        return headers;
    const headersObj = new Headers(headers);
    const authKeys = ["authorization", "x-api-key"];
    for (const [key, value] of headersObj.entries()) {
        if (authKeys.some((authKey) => key.toLowerCase() === authKey)) {
            // Find space after prefix (like "Bearer ")
            const spaceIndex = value.indexOf(" ");
            if (spaceIndex > 0 && spaceIndex < value.length - 1) {
                const prefix = value.substring(0, spaceIndex + 1); // Include the space
                const token = value.substring(spaceIndex + 1);
                if (token.length > 12) {
                    // Show first 4 and last 4 chars, truncate middle
                    const start = token.substring(0, 4);
                    const end = token.substring(token.length - 4);
                    headersObj.set(key, `${prefix}${start}...${end}`);
                }
                else {
                    // Short token, use first character + three dots
                    const firstChar = token.charAt(0);
                    headersObj.set(key, `${prefix}${firstChar}...`);
                }
            }
            else {
                // No prefix found, truncate the whole value
                if (value.length > 12) {
                    const start = value.substring(0, 4);
                    const end = value.substring(value.length - 4);
                    headersObj.set(key, `${start}...${end}`);
                }
                else {
                    // Short token, use first character + three dots
                    const firstChar = value.charAt(0);
                    headersObj.set(key, `${firstChar}...`);
                }
            }
        }
    }
    return Object.fromEntries(headersObj);
}
export function createDebugFetch(options) {
    const { originalFetch, debugLog } = options;
    return async (input, options) => {
        const startTime = Date.now();
        // Convert input to URL string for logging
        const url = typeof input === "string" ? input : input.toString();
        const method = options?.method || "GET";
        debugLog(`→ ${method} ${url}`, {
            headers: censorHeaders(options?.headers),
            body: options?.body && typeof options.body === "string"
                ? (() => {
                    try {
                        return JSON.parse(options.body);
                    }
                    catch {
                        return options.body;
                    }
                })()
                : options?.body,
        });
        try {
            const response = await originalFetch(input, options);
            const duration = Date.now() - startTime;
            debugLog(`← ${response.status} ${response.statusText} (${duration}ms)`, {
                url,
                method,
                status: response.status,
            });
            return response;
        }
        catch (error) {
            const duration = Date.now() - startTime;
            debugLog(`✖ Request failed (${duration}ms)`, {
                url,
                method,
                error: error instanceof Error ? error.message : error,
            });
            throw error;
        }
    };
}
