// @zapier/zapier-sdk/dist/api/schemas.js
import { z } from "zod";
/**
 * Base Zod schemas for API response models.
 * These schemas represent the raw data structures returned by Zapier API endpoints.
 * All TypeScript types in api/types.ts are inferred from these schemas.
 */
// ============================================================================
// Basic Building Block Schemas
// ============================================================================
export const NeedChoicesSchema = z.object({
    key: z.string().optional(),
    label: z.string().optional(),
    sample: z.string().optional(),
    value: z.string().optional(),
});
export const NeedSchema = z.object({
    key: z.string(),
    alters_custom_fields: z.boolean().nullable().optional(),
    capabilities: z.array(z.string()).optional(),
    choices: z.array(NeedChoicesSchema).optional(),
    computed: z.boolean().nullable().optional(),
    custom_field: z.boolean().optional(),
    default: z.string().optional(),
    depends_on: z.array(z.string()).optional(),
    format: z.literal("SELECT").optional(),
    from_search: z.boolean().optional(),
    from_write: z.boolean().optional(),
    help_text: z.string().optional(),
    help_text_html: z.string().optional(),
    input_format: z.array(z.string()).optional(),
    label: z.string().optional(),
    language: z.string().optional(),
    parent_key: z.string().optional(),
    placeholder: z.string().optional(),
    prefill: z.string().optional(),
    required: z.boolean().optional(),
    searchfill: z.string().optional(),
    send_in_json: z.boolean().optional(),
    regex: z.string().optional(),
    type: z
        .enum([
        "integer",
        "string",
        "text",
        "datetime",
        "boolean",
        "file",
        "decimal",
        "copy",
        "password",
        "dict",
        "code",
        "filter",
        "json",
    ])
        .optional(),
    list: z.boolean().optional(),
});
export const ActionLinksSchema = z.object({
    action_url: z.string().optional(),
});
export const ActionPermissionsSchema = z.object({
    can_use: z.boolean().optional(),
});
export const ActionSchema = z.object({
    id: z.string().optional(),
    type: z.enum([
        "filter",
        "read",
        "read_bulk",
        "run",
        "search",
        "search_and_write",
        "search_or_write",
        "write",
    ]),
    key: z.string(),
    name: z.string(),
    description: z.string(),
    is_important: z.boolean().optional(),
    is_hidden: z.boolean().optional(),
    selected_api: z.string().optional(),
});
export const ChoiceSchema = z.object({
    value: z.union([z.string(), z.number()]),
    label: z.string(),
});
export const FieldSchema = z.object({
    key: z.string(),
    label: z.string(),
    type: z.enum([
        "string",
        "number",
        "boolean",
        "datetime",
        "file",
        "object",
        "array",
    ]),
    required: z.boolean(),
    description: z.string().optional(),
    choices: z.array(ChoiceSchema).optional(),
});
export const ActionExecutionResultSchema = z.object({
    data: z.array(z.unknown()),
});
export const ActionFieldChoiceSchema = z.object({
    value: z.union([z.string(), z.number()]),
    label: z.string(),
});
export const ActionFieldSchema = z.object({
    key: z.string(),
    label: z.string().optional(),
    required: z.boolean(),
    type: z.string().optional(),
    helpText: z.string().optional(),
    helpTextHtml: z.string().optional(),
    choices: z.array(ActionFieldChoiceSchema).optional(),
    default: z.string().optional(),
    placeholder: z.string().optional(),
    computed: z.boolean().optional(),
    customField: z.boolean().optional(),
    dependsOn: z.array(z.string()).optional(),
    format: z.string().optional(),
    inputFormat: z.array(z.string()).optional(),
});
// ============================================================================
// Main API Entity Schemas
// ============================================================================
export const UserProfileSchema = z.object({
    id: z.number(),
    public_id: z.string().optional(),
    code: z.string(),
    user_id: z.number(),
    auto_provisioned: z.boolean(),
    first_name: z.string(),
    last_name: z.string(),
    username: z.string(),
    personas: z.string(),
    user_generated_personas: z.string(),
    last_login: z.string(),
    email: z.string(),
    email_hash: z.string(),
    email_confirmed: z.boolean(),
    timezone: z.string(),
    photo_url: z.string(),
    has_seen_notifications: z.record(z.string(), z.boolean().nullable()),
    signup: z.string(),
    since_signup: z.string(),
    has_activated: z.boolean(),
    enable_gz_creator: z.boolean(),
    should_see_nps_survey: z.boolean(),
    is_developer: z.boolean(),
    is_expert: z.boolean(),
    tos_agreement: z.boolean(),
    should_renew_tos: z.boolean(),
    is_gdpr_consented: z.boolean(),
    disable_ssl_check: z.boolean(),
    identity: z.number(),
    summary_schedule: z.string(),
    alert_triggers: z.string(),
    alert_actions: z.string(),
    is_staff: z.boolean(),
    is_zt_reviewer: z.boolean(),
    is_high_value: z.boolean(),
    is_temporary: z.boolean(),
    banner_message: z.string(),
    enable_totp_2fa: z.boolean(),
    viewed_help: z.record(z.string(), z.boolean()),
    show_editor_migration_mesaging: z.boolean(),
    switches: z.record(z.string(), z.unknown()),
    organizations: z.array(z.record(z.string(), z.unknown()).nullable()),
    primary_organization: z.record(z.string(), z.unknown()).nullable(),
    has_active_zaps: z.boolean(),
    has_google_sso: z.boolean(),
    auth_realm: z.string(),
    roles: z.array(z.object({
        account_id: z.number(),
        role: z.string(),
    })),
});
export const AppSchema = z.object({
    age_in_days: z.string().optional(),
    api_docs_url: z.string().nullable().optional(),
    app_profile_url: z.string(),
    banner: z.string().optional(),
    categories: z.array(z.string()).optional(), // Service category names
    canonical_id: z.string().optional(),
    current_implementation_id: z.string(),
    days_since_last_update: z.string().optional(),
    description: z.string(),
    external_url: z.string(),
    hashtag: z.string().optional(),
    id: z.number().optional(),
    image: z.string().optional(),
    images: z.string().optional(),
    integration_overview_html: z.string().nullable().optional(),
    internal_id: z.string(),
    invite_url: z.string().nullable().optional(),
    is_beta: z.string().optional(),
    is_built_in: z.string().optional(),
    is_featured: z.string().optional(),
    is_premium: z.boolean().optional(),
    is_public: z.string().optional(),
    is_upcoming: z.string().optional(),
    learn_more_url: z.string(),
    name: z.string(),
    popularity: z.number(),
    primary_color: z.string(),
    request_count: z.string().optional(),
    slug: z.string(),
    zap_usage_count: z.number().nullable().optional(),
});
export const ServiceSchema = z.object({
    id: z.number().optional(),
    canonical_id: z.string().optional(),
    current_implementation_id: z.string(),
    name: z.string(),
    slug: z.string(),
    app_url: z.string().optional(),
    learn_more_url: z.string().optional(),
    description: z.string(),
    primary_color: z.string(),
    popularity: z.number(),
    image: z.string().optional(),
    images: z.string().optional(),
});
export const ServicesResponseSchema = z.object({
    results: z.array(ServiceSchema),
    next: z.string().nullable().optional(),
    previous: z.string().nullable().optional(),
});
// ============================================================================
// Request/Response Schemas
// ============================================================================
export const NeedsRequestSchema = z.object({
    selected_api: z.string(),
    action: z.string(),
    type_of: z.string(),
    authentication_id: z.union([z.string(), z.number()]).optional(),
    params: z.record(z.string(), z.unknown()).optional(),
});
export const NeedsResponseSchema = z.object({
    success: z.boolean(),
    needs: z.array(NeedSchema).optional(),
    errors: z.array(z.string()).optional(),
    last_fetched_at: z.string().optional(),
    schema: z.record(z.string(), z.unknown()).optional(),
});
// ============================================================================
// Implementation Schema
// ============================================================================
export const ImplementationSchema = z.object({
    selected_api: z.string(),
    app_id: z.number().optional(),
    auth_type: z.string().optional(),
    auth_fields: z.string().optional(),
    actions: z.array(ActionSchema).optional(),
    is_deprecated: z.boolean().optional(),
    is_private_only: z.boolean().optional(),
    is_invite_only: z.boolean().optional(),
    is_beta: z.boolean().optional().default(false),
    is_premium: z.boolean().optional().default(false),
    is_hidden: z.string().optional(),
    name: z.string().optional(),
    slug: z.string().optional(),
    images: z.record(z.string(), z.string().nullable()).optional(),
    primary_color: z.string().optional(),
    secondary_color: z.string().optional(),
    current_implementation: z.string().optional(),
    other_implementations: z.string().optional(),
});
export const ImplementationsResponseSchema = z.object({
    count: z.number(),
    next: z.string().nullable().optional(),
    previous: z.string().nullable().optional(),
    results: z.array(ImplementationSchema),
});
// ============================================================================
// Need Choices Schemas (for listInputFieldChoices functionality)
// ============================================================================
export const NeedChoicesResponseMetaSchema = z.object({
    page: z.string().nullable().optional(),
});
export const NeedChoicesResponseLinksSchema = z.object({
    next: z.string().nullable().optional(),
    prev: z.string().nullable().optional(),
});
export const NeedChoicesRequestSchema = z.object({
    selected_api: z
        .string()
        .optional()
        .describe("Something like `SlackAPI` (for Python apps) or `SplitwiseCLIAPI@1.0.0` (for CLI apps). Non-public apps are fine as long as the authed user can access them."),
    authentication_id: z
        .union([z.string(), z.number()])
        .optional()
        .describe("If the app needs auth, provide an `authentication_id` that has the `selected_api` of the app you want to run. Can be any auth visible to the user (including shared)."),
    params: z
        .record(z.string(), z.unknown())
        .optional()
        .describe("Object that matches the input the node would normally get. Has all the same keys/types as the `needs` of the action."),
    page: z.number().optional().default(0),
    prefill: z
        .string()
        .optional()
        .describe("The prefill string to indicate what we're fetching choices for. Likely something like `spreadsheet.id.title`. Must be provided alongside `selected_api` if both `action_id` and `input_field_id` are not."),
    action_id: z
        .string()
        .optional()
        .describe("The id that will be used to lookup the Action for prefill lookup. If provided, `input_field_id` is required, else `prefill` must be provided."),
    input_field_id: z
        .string()
        .optional()
        .describe("The id (key) of the input field (Need) that dynamic choices are being retrieved for. If provided, `action_id` is required, else `prefill` must be provided."),
});
export const NeedChoicesResponseSchema = z.object({
    success: z.boolean(),
    choices: z.array(NeedChoicesSchema).optional(),
    next_page: z.number().optional(),
    errors: z.array(z.string()).optional(),
    meta: NeedChoicesResponseMetaSchema.optional(),
    links: NeedChoicesResponseLinksSchema.optional(),
});
