// @zapier/zapier-sdk/dist/plugins/getInputFieldsSchema/index.js
import { GetInputFieldsSchemaSchema, GetInputFieldsSchemaInputSchema, } from "./schemas";
import { ZapierConfigurationError } from "../../types/errors";
import { createFunction } from "../../utils/function-utils";
import { appKeyResolver, actionTypeResolver, actionKeyResolver, connectionIdResolver, inputsAllOptionalResolver, } from "../../resolvers";
import { fetchImplementationNeeds } from "../../services/implementations";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { resolveConnectionId } from "../../utils/domain-utils";
import { setMethodMetadata } from "../../utils/telemetry-context";
export const getInputFieldsSchemaPlugin = ({ sdk, context }) => {
    async function getInputFieldsSchema(options) {
        const { api, getVersionedImplementationId } = context;
        const appKey = "app" in options ? options.app : options.appKey;
        const actionKey = "action" in options ? options.action : options.actionKey;
        const { actionType, connection, connectionId, authenticationId, inputs } = options;
        const resolvedConnectionId = await resolveConnectionId({
            connection,
            connectionId,
            authenticationId,
            resolveConnection: context.resolveConnection,
        });
        const selectedApi = await getVersionedImplementationId(appKey);
        if (!selectedApi) {
            throw new ZapierConfigurationError("No current_implementation_id found for app", { configType: "current_implementation_id" });
        }
        setMethodMetadata({
            selectedApi,
            operationType: actionType,
            operationKey: actionKey,
        });
        const { data: action } = await sdk.getAction({
            app: appKey,
            actionType,
            action: actionKey,
        });
        const needsData = await fetchImplementationNeeds({
            api,
            selectedApi,
            action: action.key,
            actionType,
            connectionId: resolvedConnectionId ?? null,
            inputs,
        });
        return {
            data: needsData.schema || {},
        };
    }
    const getInputFieldsSchemaDefinition = createFunction(getInputFieldsSchema, GetInputFieldsSchemaInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, getInputFieldsSchema.name));
    return {
        getInputFieldsSchema: getInputFieldsSchemaDefinition,
        context: {
            meta: {
                getInputFieldsSchema: {
                    categories: ["action"],
                    type: "item",
                    itemType: "InputSchema",
                    inputSchema: GetInputFieldsSchemaSchema,
                    resolvers: {
                        app: appKeyResolver,
                        actionType: actionTypeResolver,
                        action: actionKeyResolver,
                        connection: connectionIdResolver,
                        inputs: inputsAllOptionalResolver,
                    },
                },
            },
        },
    };
};
