// @zapier/zapier-sdk/dist/plugins/tables/createTableRecords/index.js
import { createFunction } from "../../../utils/function-utils";
import { CreateTableRecordsOptionsSchema, CreateTableRecordsOptionsInputSchema, CreateTableRecordsApiResponseSchema, RecordItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver, tableRecordsResolver } from "../../../resolvers";
import { tableRecordFormatter } from "../../../formatters";
import { createFieldKeyTranslator, transformRecordItem, throwOnRecordErrors, throwOnResponseErrors, } from "../utils";
export const createTableRecordsPlugin = ({ context }) => {
    async function createTableRecords(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const translator = await createFieldKeyTranslator({
            api,
            tableId,
            keyMode: options.keyMode,
        });
        const rawResponse = await api.post(`/tables/api/v1/tables/${tableId}/records`, {
            new_records: options.records.map((record) => ({
                data: translator.translateInput(record.data),
            })),
        }, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to create records or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to create records. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 404) {
                    return new Error(`Table not found: ${tableId}`);
                }
                return undefined;
            },
            authRequired: true,
        });
        throwOnResponseErrors(rawResponse);
        const response = CreateTableRecordsApiResponseSchema.parse(rawResponse);
        for (const changeset of response.data) {
            throwOnRecordErrors(changeset.new);
        }
        return {
            data: response.data.map((changeset) => ({
                ...transformRecordItem(changeset.new),
                data: translator.translateOutput(changeset.new.data),
            })),
        };
    }
    const createTableRecordsDefinition = createFunction(createTableRecords, CreateTableRecordsOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, createTableRecords.name));
    return {
        createTableRecords: createTableRecordsDefinition,
        context: {
            meta: {
                createTableRecords: {
                    categories: ["table"],
                    type: "create",
                    itemType: "Record",
                    inputSchema: CreateTableRecordsOptionsSchema,
                    outputSchema: RecordItemSchema,
                    resolvers: {
                        table: tableIdResolver,
                        records: tableRecordsResolver,
                    },
                    formatter: tableRecordFormatter,
                },
            },
        },
    };
};
