// @zapier/zapier-sdk/dist/plugins/tables/createTableRecords/schemas.js
import { z } from "zod";
import { TablePropertySchema } from "../../../types/properties";
import { RecordApiItemSchema, RecordItemSchema, } from "../getTableRecord/schemas";
import { KeyModeSchema } from "../utils";
// Re-export for consumers
export { RecordItemSchema };
// The API returns changesets with old/new values
const RecordChangesetSchema = z.object({
    change_id: z.string(),
    old: RecordApiItemSchema.nullable(),
    new: RecordApiItemSchema,
    changed_field_ids: z.array(z.number()).nullable(),
    record_id: z.string(),
});
export const CreateTableRecordsApiResponseSchema = z.object({
    data: z.array(RecordChangesetSchema),
});
export const NewRecordSchema = z.object({
    data: z
        .record(z.string(), z.unknown())
        .describe("The field values for the record, keyed by field ID"),
});
const CreateTableRecordsDescription = "Create one or more records in a table";
const CreateTableRecordsBase = z.object({
    records: z
        .array(NewRecordSchema)
        .min(1)
        .max(100)
        .describe("Array of records to create (max 100)"),
    keyMode: KeyModeSchema,
});
// Schema for docs/registry (new param names only)
export const CreateTableRecordsOptionsSchema = z
    .object({
    table: TablePropertySchema,
})
    .merge(CreateTableRecordsBase)
    .describe(CreateTableRecordsDescription)
    .meta({ aliases: { tableId: "table" } });
// Deprecated schema (backward compat)
const CreateTableRecordsOptionsSchemaDeprecated = z
    .object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
})
    .merge(CreateTableRecordsBase);
// Union for runtime validation
export const CreateTableRecordsOptionsInputSchema = z
    .union([
    CreateTableRecordsOptionsSchema,
    CreateTableRecordsOptionsSchemaDeprecated,
])
    .describe(CreateTableRecordsDescription);
