// @zapier/zapier-sdk/dist/plugins/getAction/index.js
import { GetActionSchema, GetActionInputSchema, } from "./schemas";
import { ZapierResourceNotFoundError } from "../../types/errors";
import { createFunction } from "../../utils/function-utils";
import { appKeyResolver, actionTypeResolver, actionKeyResolver, } from "../../resolvers";
import { ActionItemSchema } from "../../schemas/Action";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { setMethodMetadata } from "../../utils/telemetry-context";
export const getActionPlugin = ({ sdk, context }) => {
    async function getAction(options) {
        const appKey = "app" in options ? options.app : options.appKey;
        const actionKey = "action" in options ? options.action : options.actionKey;
        const { actionType } = options;
        // selectedApi is intentionally omitted — getAction doesn't resolve a versioned
        // implementation ID; it delegates to sdk.listActions, which sets selectedApi in its own scope.
        setMethodMetadata({
            operationType: actionType,
            operationKey: actionKey,
        });
        // Use the listActions function from the SDK to search for the specific action across all pages
        for await (const action of sdk.listActions({ app: appKey }).items()) {
            if ((action.key === actionKey || action.id === actionKey) &&
                action.action_type === actionType) {
                return { data: action };
            }
        }
        // If we get here, action was not found
        throw new ZapierResourceNotFoundError(`Action not found: ${actionKey} with type ${actionType}`, { resourceType: "Action", resourceId: `${actionKey} (${actionType})` });
    }
    const getActionDefinition = createFunction(getAction, GetActionInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, getAction.name));
    return {
        getAction: getActionDefinition,
        context: {
            meta: {
                getAction: {
                    categories: ["action"],
                    type: "item",
                    itemType: "Action",
                    inputSchema: GetActionSchema,
                    outputSchema: ActionItemSchema,
                    resolvers: {
                        app: appKeyResolver,
                        actionType: actionTypeResolver,
                        action: actionKeyResolver,
                    },
                },
            },
        },
    };
};
