// @zapier/zapier-sdk/dist/plugins/tables/deleteTableRecords/index.js
import { createFunction } from "../../../utils/function-utils";
import { DeleteTableRecordsOptionsSchema, DeleteTableRecordsOptionsInputSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver, tableRecordIdsResolver } from "../../../resolvers";
export const deleteTableRecordsPlugin = ({ context }) => {
    async function deleteTableRecords(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const recordIds = "records" in options ? options.records : options.recordIds;
        await api.delete(`/tables/api/v1/tables/${tableId}/records`, { record_ids: recordIds }, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to delete records or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to delete records. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
        });
        return {
            success: true,
        };
    }
    const deleteTableRecordsDefinition = createFunction(deleteTableRecords, DeleteTableRecordsOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, deleteTableRecords.name));
    return {
        deleteTableRecords: deleteTableRecordsDefinition,
        context: {
            meta: {
                deleteTableRecords: {
                    categories: ["table"],
                    type: "delete",
                    itemType: "Record",
                    inputSchema: DeleteTableRecordsOptionsSchema,
                    resolvers: {
                        table: tableIdResolver,
                        records: tableRecordIdsResolver,
                    },
                    confirm: "delete",
                },
            },
        },
    };
};
