// @zapier/zapier-sdk/dist/plugins/tables/deleteTableRecords/schemas.js
import { z } from "zod";
import { TablePropertySchema, RecordsPropertySchema, } from "../../../types/properties";
const DeleteTableRecordsDescription = "Delete one or more records from a table";
// Schema for docs/registry (new param names only)
export const DeleteTableRecordsOptionsSchema = z
    .object({
    table: TablePropertySchema,
    records: RecordsPropertySchema,
})
    .describe(DeleteTableRecordsDescription)
    .meta({ aliases: { tableId: "table", recordIds: "records" } });
// Deprecated schema (backward compat)
const DeleteTableRecordsOptionsSchemaDeprecated = z.object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
    recordIds: z
        .array(z.string().regex(/^[A-Z0-9]{26}$/, "Record ID must be a valid ULID"))
        .min(1)
        .max(100)
        .describe("Array of record IDs to delete (max 100)"),
});
// Union for runtime validation
export const DeleteTableRecordsOptionsInputSchema = z
    .union([
    DeleteTableRecordsOptionsSchema,
    DeleteTableRecordsOptionsSchemaDeprecated,
])
    .describe(DeleteTableRecordsDescription);
