// @zapier/zapier-sdk/dist/types/sdk.js
/**
 * SDK-related types and interfaces
 */
import { z } from "zod";
import { CredentialsSchema } from "./credentials";
// SDK Configuration Types
export const BaseSdkOptionsSchema = z.object({
    credentials: CredentialsSchema.optional().describe("Authentication credentials. Can be a string (token or API key), a client credentials object ({ clientId, clientSecret }), a PKCE object ({ clientId }), or a function returning any of those."),
    debug: z.boolean().optional().describe("Enable debug logging."),
    baseUrl: z
        .string()
        .optional()
        .describe("Base URL for Zapier API endpoints.")
        .meta({ valueHint: "url" }),
    trackingBaseUrl: z
        .string()
        .optional()
        .describe("Base URL for Zapier tracking endpoints.")
        .meta({ valueHint: "url" }),
    /**
     * Maximum number of retries for rate-limited requests (429 responses).
     * Set to 0 to disable retries. Default is 3.
     */
    maxNetworkRetries: z
        .number()
        .optional()
        .describe("Max retries for rate-limited requests (default: 3).")
        .meta({ valueHint: "count" }),
    /**
     * Maximum delay in milliseconds to wait for a rate limit retry.
     * If the server requests a longer delay, the request fails immediately.
     * Default is 60000 (60 seconds).
     */
    maxNetworkRetryDelayMs: z
        .number()
        .optional()
        .describe("Max delay in ms to wait for retry (default: 60000).")
        .meta({ valueHint: "ms" }),
    // Internal
    manifestPath: z
        .string()
        .optional()
        .describe("Path to a .zapierrc manifest file for app version locking.")
        .meta({ internal: true }),
    manifest: z
        .custom()
        .optional()
        .describe("Manifest for app version locking.")
        .meta({ internal: true }),
    onEvent: z.custom().optional().meta({ internal: true }),
    fetch: z.custom().optional().meta({ internal: true }),
    eventEmission: z
        .custom()
        .optional()
        .meta({ internal: true }),
    canIncludeSharedConnections: z
        .boolean()
        .optional()
        .describe("Allow listing shared connections."),
    canIncludeSharedTables: z
        .boolean()
        .optional()
        .describe("Allow listing shared tables."),
    canDeleteTables: z.boolean().optional().describe("Allow deleting tables."),
    // Deprecated
    token: z.string().optional().meta({ deprecated: true }), // Use credentials instead
});
