// @zapier/zapier-sdk/dist/auth.js
/**
 * SDK Authentication Utilities
 *
 * This module provides SDK-level authentication utilities focused
 * on token acquisition. It uses the credentials system for resolution
 * and handles different credential types appropriately.
 *
 * CLI-specific functionality like login/logout is handled by the
 * @zapier/zapier-sdk-cli-login package.
 */
import { isClientCredentials, isPkceCredentials } from "./types/credentials";
import { resolveCredentials, getClientIdFromCredentials } from "./credentials";
import { ZAPIER_BASE_URL } from "./constants";
export { isClientCredentials, isPkceCredentials, isCredentialsObject, isCredentialsFunction, } from "./types/credentials";
/**
 * In-memory cache for tokens obtained via client credentials flow.
 * Keyed by clientId + sorted scopes.
 */
const tokenCache = new Map();
/**
 * In-flight token exchange promises to prevent duplicate exchanges.
 * When an exchange is in progress, subsequent calls await the same promise.
 * Keyed by clientId + sorted scopes.
 */
const pendingExchanges = new Map();
/**
 * Build a cache key from clientId and scopes.
 */
function buildCacheKey(clientId, scopes) {
    const sortedScopes = [...scopes].sort().join(",");
    return `${clientId}:${sortedScopes}`;
}
/**
 * Clear the token cache. Useful for testing or forcing re-authentication.
 */
export function clearTokenCache() {
    tokenCache.clear();
    pendingExchanges.clear();
    cachedCliLogin = undefined;
}
const TOKEN_EXPIRATION_BUFFER = 5 * 60 * 1000; // 5 minutes
/**
 * Get a cached token if it exists and is not expired.
 * Returns undefined if no valid cached token exists.
 */
function getCachedToken(clientId, scopes) {
    const cacheKey = buildCacheKey(clientId, scopes);
    const cached = tokenCache.get(cacheKey);
    if (!cached)
        return undefined;
    // Check if token is still valid (with expiration buffer)
    if (cached.expiresAt > Date.now() + TOKEN_EXPIRATION_BUFFER) {
        return cached.accessToken;
    }
    // Token expired, remove from cache
    tokenCache.delete(cacheKey);
    return undefined;
}
/**
 * Cache a token obtained from client credentials flow.
 */
function cacheToken(clientId, scopes, accessToken, expiresIn) {
    const cacheKey = buildCacheKey(clientId, scopes);
    tokenCache.set(cacheKey, {
        accessToken,
        expiresAt: Date.now() + expiresIn * 1000,
    });
}
/**
 * Invalidate a cached token for a specific clientId and scope combination.
 * Called when we get a 401 response.
 */
export function invalidateCachedToken(clientId, scopes) {
    const cacheKey = buildCacheKey(clientId, scopes);
    tokenCache.delete(cacheKey);
    pendingExchanges.delete(cacheKey);
}
/**
 * Get the token endpoint URL for client credentials exchange.
 */
function getTokenEndpointUrl(baseUrl) {
    const base = baseUrl || ZAPIER_BASE_URL;
    return `${base}/oauth/token/`;
}
/**
 * Merge and deduplicate scopes from credentials and required scopes.
 * If no credentials scope is specified, defaults to "external".
 * Returns a sorted array of unique scopes.
 */
function mergeScopes(credentialsScope, requiredScopes) {
    const scopeSet = new Set();
    // Add scopes from credentials (space-separated string)
    // If no credentials scope specified, default to "external"
    if (credentialsScope) {
        for (const s of credentialsScope.split(" ")) {
            if (s.trim()) {
                scopeSet.add(s.trim());
            }
        }
    }
    else {
        scopeSet.add("external");
    }
    // Add required scopes
    if (requiredScopes) {
        for (const s of requiredScopes) {
            scopeSet.add(s);
        }
    }
    return [...scopeSet].sort();
}
/**
 * Exchange client credentials for an access token.
 */
async function exchangeClientCredentials(options) {
    const { clientId, clientSecret, baseUrl, scope, requiredScopes, onEvent } = options;
    const fetchFn = options.fetch || globalThis.fetch;
    const tokenUrl = getTokenEndpointUrl(baseUrl);
    // Merge credentials scope with required scopes
    const mergedScopes = mergeScopes(scope, requiredScopes);
    const scopeString = mergedScopes.join(" ");
    onEvent?.({
        type: "auth_exchanging",
        payload: {
            message: "Exchanging client credentials for token...",
            operation: "client_credentials",
        },
        timestamp: Date.now(),
    });
    const response = await fetchFn(tokenUrl, {
        method: "POST",
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
        },
        body: new URLSearchParams({
            grant_type: "client_credentials",
            client_id: clientId,
            client_secret: clientSecret,
            scope: scopeString,
            audience: "zapier.com",
        }),
    });
    if (!response.ok) {
        const errorText = await response.text();
        onEvent?.({
            type: "auth_error",
            payload: {
                message: `Client credentials exchange failed: ${response.status}`,
                error: errorText,
                operation: "client_credentials",
            },
            timestamp: Date.now(),
        });
        throw new Error(`Client credentials exchange failed: ${response.status} ${response.statusText}`);
    }
    const data = await response.json();
    if (!data.access_token) {
        throw new Error("Client credentials response missing access_token");
    }
    // Cache the token with the scopes used
    const expiresIn = data.expires_in || 3600; // Default to 1 hour
    cacheToken(clientId, mergedScopes, data.access_token, expiresIn);
    onEvent?.({
        type: "auth_success",
        payload: {
            message: "Client credentials exchange successful",
            operation: "client_credentials",
        },
        timestamp: Date.now(),
    });
    return data.access_token;
}
let cachedCliLogin;
/**
 * Dynamically imports a CLI login package with caching.
 *
 * Tries `@zapier/zapier-sdk-cli/login` first (available when the CLI is a
 * direct dependency), then falls back to `@zapier/zapier-sdk-cli-login` (for
 * users who install the lightweight login package directly).
 *
 * Returns the module if found, or `undefined` if neither package is available.
 * The result is cached after the first attempt.
 */
async function getCliLogin() {
    if (cachedCliLogin !== undefined) {
        return cachedCliLogin || undefined;
    }
    try {
        const mod = await import("@zapier/zapier-sdk-cli/login");
        if (typeof mod.getToken === "function") {
            cachedCliLogin = mod;
            return cachedCliLogin;
        }
    }
    catch {
        // Not available, try fallback
    }
    try {
        const mod = await import("@zapier/zapier-sdk-cli-login");
        if (typeof mod.getToken === "function") {
            cachedCliLogin = mod;
            return cachedCliLogin;
        }
    }
    catch {
        // Not available
    }
    cachedCliLogin = false;
    return undefined;
}
/**
 * Inject an already-loaded CLI login module so the SDK skips its dynamic import.
 * This guarantees CLI and SDK share the same module instance in the same process.
 */
export function injectCliLogin(module) {
    cachedCliLogin = module;
}
/**
 * Returns whether a CLI login package is available.
 * `undefined` if no import has been attempted yet, `true` if found, `false` if not.
 */
export function isCliLoginAvailable() {
    if (cachedCliLogin === undefined)
        return undefined;
    return cachedCliLogin !== false;
}
/**
 * Attempts to get a token from the CLI login package.
 *
 * Returns undefined if no valid token is found or no login package is available.
 * Throws if there's an actual error (like token refresh failure).
 */
export async function getTokenFromCliLogin(options) {
    const cliLogin = await getCliLogin();
    if (!cliLogin)
        return undefined;
    return await cliLogin.getToken(options);
}
/**
 * Resolves an auth token from wherever it can be found.
 *
 * Resolution order:
 * 1. Explicit credentials option (or deprecated token option)
 * 2. Environment variables
 * 3. CLI login package (if available)
 *
 * For different credential types:
 * - String: Used directly as the token
 * - Client credentials: Exchanged for an access token (with caching)
 * - PKCE: Delegates to CLI login (throws if not available)
 *
 * Returns undefined if no valid token is found.
 */
export async function resolveAuthToken(options = {}) {
    // Resolve credentials from options and env vars
    const credentials = await resolveCredentials({
        credentials: options.credentials,
        token: options.token,
        baseUrl: options.baseUrl,
    });
    // If we got credentials, process them based on type
    if (credentials !== undefined) {
        return resolveAuthTokenFromCredentials(credentials, options);
    }
    // No credentials from options or env, try CLI login for stored token
    return getTokenFromCliLogin({
        onEvent: options.onEvent,
        fetch: options.fetch,
        debug: options.debug,
    });
}
/**
 * Resolve an auth token from resolved credentials.
 */
async function resolveAuthTokenFromCredentials(credentials, options) {
    // String credentials are used directly as tokens
    if (typeof credentials === "string") {
        return credentials;
    }
    // Client credentials: exchange for token
    if (isClientCredentials(credentials)) {
        const { clientId } = credentials;
        // Compute merged scopes for cache lookup
        const mergedScopes = mergeScopes(credentials.scope, options.requiredScopes);
        const cacheKey = buildCacheKey(clientId, mergedScopes);
        // Check cache first
        const cached = getCachedToken(clientId, mergedScopes);
        if (cached) {
            return cached;
        }
        // Check if there's already an exchange in progress for this clientId + scopes
        const pending = pendingExchanges.get(cacheKey);
        if (pending) {
            return pending;
        }
        // Start new exchange and cache the promise to prevent duplicate exchanges
        const exchangePromise = exchangeClientCredentials({
            clientId: credentials.clientId,
            clientSecret: credentials.clientSecret,
            baseUrl: credentials.baseUrl || options.baseUrl,
            scope: credentials.scope,
            requiredScopes: options.requiredScopes,
            fetch: options.fetch,
            onEvent: options.onEvent,
        }).finally(() => {
            // Remove from pending when done (success or failure)
            pendingExchanges.delete(cacheKey);
        });
        pendingExchanges.set(cacheKey, exchangePromise);
        return exchangePromise;
    }
    // PKCE credentials: delegate to CLI login
    if (isPkceCredentials(credentials)) {
        // Try to get stored token from CLI login
        const storedToken = await getTokenFromCliLogin({
            onEvent: options.onEvent,
            fetch: options.fetch,
            credentials,
            debug: options.debug,
        });
        if (storedToken) {
            return storedToken;
        }
        // No stored token - user needs to run login
        throw new Error("PKCE credentials require interactive login. " +
            "Please run the 'login' command with the CLI first, or use client_credentials flow.");
    }
    // Should not reach here, but handle gracefully
    throw new Error("Unknown credentials type");
}
/**
 * Invalidate the cached token for the given credentials, if applicable.
 * This is called when we receive a 401 response, indicating the token
 * is no longer valid. Only affects client_credentials flow tokens.
 */
export async function invalidateCredentialsToken(options) {
    const resolved = await resolveCredentials(options);
    if (!resolved)
        return;
    const clientId = getClientIdFromCredentials(resolved);
    if (clientId && isClientCredentials(resolved)) {
        const scopes = mergeScopes(resolved.scope, options.requiredScopes);
        invalidateCachedToken(clientId, scopes);
    }
}
