// @zapier/zapier-sdk/dist/constants.js
/**
 * SDK Constants
 *
 * This module contains shared constants used throughout the SDK.
 */
/**
 * Base URL for Zapier API endpoints
 */
export const ZAPIER_BASE_URL = globalThis.process?.env?.ZAPIER_BASE_URL || "https://zapier.com";
/**
 * Maximum number of items that can be requested per page across all paginated functions
 */
export const MAX_PAGE_LIMIT = 10000;
/**
 * Default number of items per page for paginated functions
 */
export const DEFAULT_PAGE_SIZE = 100;
/**
 * Default timeout for action execution (in milliseconds)
 */
export const DEFAULT_ACTION_TIMEOUT_MS = 180000;
function parseIntEnvVar(name) {
    const value = globalThis.process?.env?.[name];
    if (value === undefined)
        return undefined;
    const parsed = parseInt(value, 10);
    if (isNaN(parsed)) {
        console.warn(`[zapier-sdk] Invalid value for ${name}: "${value}" (expected integer)`);
        return undefined;
    }
    return parsed;
}
/**
 * Network retry configuration from environment variables
 */
export const ZAPIER_MAX_NETWORK_RETRIES = parseIntEnvVar("ZAPIER_MAX_NETWORK_RETRIES") ?? 3;
export const ZAPIER_MAX_NETWORK_RETRY_DELAY_MS = parseIntEnvVar("ZAPIER_MAX_NETWORK_RETRY_DELAY_MS") ?? 60000;
