// @zapier/zapier-sdk/dist/credentials.js
/**
 * Credentials Resolution
 *
 * This module provides the core logic for resolving credentials from various sources:
 * - Explicit credentials option
 * - Deprecated token option
 * - Environment variables (new and deprecated)
 * - CLI login stored tokens
 */
import { isCredentialsFunction, isCredentialsObject, } from "./types/credentials";
import { logDeprecation } from "./utils/logging";
import { getZapierBaseUrl } from "./utils/url-utils";
/**
 * Derive the auth base URL from the SDK base URL.
 * Returns the Zapier root domain (e.g., https://zapier.com from https://api.zapier.com).
 */
function deriveAuthBaseUrl(sdkBaseUrl) {
    if (!sdkBaseUrl)
        return undefined;
    return getZapierBaseUrl(sdkBaseUrl);
}
/**
 * Normalize a credentials object by ensuring it has a type and resolved baseUrl.
 * Env vars provide defaults, then explicit credentials override them.
 */
function normalizeCredentialsObject(obj, sdkBaseUrl) {
    // Build defaults from env vars for optional fields, then override with explicit credentials.
    // clientId is required in CredentialsObject so it always comes from obj.
    const merged = {
        clientSecret: globalThis.process?.env?.ZAPIER_CREDENTIALS_CLIENT_SECRET,
        baseUrl: globalThis.process?.env?.ZAPIER_CREDENTIALS_BASE_URL,
        scope: globalThis.process?.env?.ZAPIER_CREDENTIALS_SCOPE,
        ...obj,
    };
    // Derive baseUrl from SDK if still not set
    const resolvedBaseUrl = merged.baseUrl || deriveAuthBaseUrl(sdkBaseUrl);
    // Explicitly construct the correct typed object to satisfy TypeScript
    if (obj.type === "client_credentials" ||
        ("clientSecret" in merged && merged.clientSecret)) {
        return {
            type: "client_credentials",
            clientId: obj.clientId,
            clientSecret: merged.clientSecret,
            baseUrl: resolvedBaseUrl,
            scope: merged.scope,
        };
    }
    return {
        type: "pkce",
        clientId: obj.clientId,
        baseUrl: resolvedBaseUrl,
        scope: merged.scope,
    };
}
/**
 * Resolve credentials from environment variables.
 *
 * Precedence:
 * 1. ZAPIER_CREDENTIALS (string token)
 * 2. ZAPIER_CREDENTIALS_CLIENT_ID + ZAPIER_CREDENTIALS_CLIENT_SECRET (client credentials)
 * 3. ZAPIER_CREDENTIALS_CLIENT_ID alone (PKCE)
 * 4. Deprecated ZAPIER_TOKEN, ZAPIER_AUTH_* vars (with warnings)
 *
 * @param sdkBaseUrl - SDK base URL used to derive auth base URL if not specified
 */
export function resolveCredentialsFromEnv(sdkBaseUrl) {
    // 1. Check ZAPIER_CREDENTIALS (string token only)
    if (globalThis.process?.env?.ZAPIER_CREDENTIALS) {
        return globalThis.process?.env?.ZAPIER_CREDENTIALS;
    }
    // 2. Check ZAPIER_CREDENTIALS_* individual vars
    if (globalThis.process?.env?.ZAPIER_CREDENTIALS_CLIENT_ID) {
        const resolvedBaseUrl = globalThis.process?.env?.ZAPIER_CREDENTIALS_BASE_URL ||
            deriveAuthBaseUrl(sdkBaseUrl);
        // Infer type from presence of clientSecret
        if (globalThis.process?.env?.ZAPIER_CREDENTIALS_CLIENT_SECRET) {
            return {
                type: "client_credentials",
                clientId: globalThis.process?.env?.ZAPIER_CREDENTIALS_CLIENT_ID,
                clientSecret: globalThis.process?.env?.ZAPIER_CREDENTIALS_CLIENT_SECRET,
                baseUrl: resolvedBaseUrl,
                scope: globalThis.process?.env?.ZAPIER_CREDENTIALS_SCOPE,
            };
        }
        else {
            return {
                type: "pkce",
                clientId: globalThis.process?.env?.ZAPIER_CREDENTIALS_CLIENT_ID,
                baseUrl: resolvedBaseUrl,
                scope: globalThis.process?.env?.ZAPIER_CREDENTIALS_SCOPE,
            };
        }
    }
    // 3. Check deprecated env vars (with warnings)
    if (globalThis.process?.env?.ZAPIER_TOKEN) {
        logDeprecation("ZAPIER_TOKEN is deprecated. Use ZAPIER_CREDENTIALS instead.");
        return globalThis.process?.env?.ZAPIER_TOKEN;
    }
    if (globalThis.process?.env?.ZAPIER_AUTH_CLIENT_ID) {
        logDeprecation("ZAPIER_AUTH_CLIENT_ID is deprecated. Use ZAPIER_CREDENTIALS_CLIENT_ID instead.");
        if (globalThis.process?.env?.ZAPIER_AUTH_BASE_URL) {
            logDeprecation("ZAPIER_AUTH_BASE_URL is deprecated. Use ZAPIER_CREDENTIALS_BASE_URL instead.");
        }
        const resolvedBaseUrl = globalThis.process?.env?.ZAPIER_AUTH_BASE_URL ||
            deriveAuthBaseUrl(sdkBaseUrl);
        return {
            type: "pkce",
            clientId: globalThis.process?.env?.ZAPIER_AUTH_CLIENT_ID,
            baseUrl: resolvedBaseUrl,
        };
    }
    return undefined;
}
/**
 * Resolve credentials from all possible sources.
 *
 * Precedence:
 * 1. Explicit credentials option
 * 2. Deprecated token option (with warning)
 * 3. Environment variables
 * 4. CLI login stored token (handled separately in auth.ts)
 *
 * If credentials is a function, it is called and must return
 * a string or credentials object (not another function).
 *
 * The baseUrl option is used to derive the auth base URL if not
 * specified in credentials.
 */
export async function resolveCredentials(options = {}) {
    const { baseUrl } = options;
    // 1. Check explicit credentials option
    if (options.credentials !== undefined) {
        // If it's a function, call it
        if (isCredentialsFunction(options.credentials)) {
            const resolved = await options.credentials();
            // Validate that the function didn't return another function
            if (typeof resolved === "function") {
                throw new Error("Credentials function returned another function. " +
                    "Credentials functions must return a string or credentials object.");
            }
            // Normalize object if needed
            if (isCredentialsObject(resolved)) {
                return normalizeCredentialsObject(resolved, baseUrl);
            }
            return resolved;
        }
        // If it's an object, normalize it
        if (isCredentialsObject(options.credentials)) {
            return normalizeCredentialsObject(options.credentials, baseUrl);
        }
        // It's a string token
        return options.credentials;
    }
    // 2. Check deprecated token option (with warning)
    if (options.token !== undefined) {
        logDeprecation("The `token` option is deprecated. Use `credentials` instead.");
        return options.token;
    }
    // 3. Check environment variables
    const envCredentials = resolveCredentialsFromEnv(baseUrl);
    if (envCredentials !== undefined) {
        return envCredentials;
    }
    // 4. CLI login is handled separately in auth.ts
    return undefined;
}
/**
 * Extract the base URL from credentials for use in auth flows.
 */
export function getBaseUrlFromCredentials(credentials) {
    if (credentials && isCredentialsObject(credentials)) {
        return credentials.baseUrl;
    }
    return undefined;
}
/**
 * Extract the client ID from credentials for use in auth flows.
 */
export function getClientIdFromCredentials(credentials) {
    if (credentials && isCredentialsObject(credentials)) {
        return credentials.clientId;
    }
    return undefined;
}
