// @zapier/zapier-sdk/dist/define.cjs
'use strict';

// src/define/define.ts
function toJsonSchema(schema) {
  return schema.toJSONSchema();
}
function buildDescribe(meta) {
  return () => ({
    kind: meta.kind,
    name: meta.name,
    description: meta.description,
    inputSchema: meta.inputSchema ? toJsonSchema(meta.inputSchema) : void 0,
    connection: meta.connection,
    connections: meta.connections
  });
}
function defineTool(config) {
  const execute = () => {
    throw new Error(
      `defineTool("${config.name}") is not callable yet. Runtime resolution is coming in Phase 2.`
    );
  };
  Object.defineProperty(execute, "name", {
    value: config.name,
    writable: false
  });
  return Object.assign(execute, {
    kind: "tool",
    description: config.description,
    connection: config.connection,
    connections: config.connections,
    inputSchema: config.inputSchema,
    inputDependencies: config.inputDependencies,
    run: config.run,
    options: config.options,
    schema: config.schema,
    digest: config.digest,
    describe: buildDescribe({
      kind: "tool",
      name: config.name,
      description: config.description,
      inputSchema: config.inputSchema,
      connection: config.connection,
      connections: config.connections
    })
  });
}
function defineDurable(config) {
  const execute = () => {
    throw new Error(
      `defineDurable("${config.name}") is not callable yet. Runtime resolution is coming in Phase 2.`
    );
  };
  Object.defineProperty(execute, "name", {
    value: config.name,
    writable: false
  });
  return Object.assign(execute, {
    kind: "durable",
    description: config.description,
    connection: config.connection,
    connections: config.connections,
    inputSchema: config.inputSchema,
    inputDependencies: config.inputDependencies,
    run: config.run,
    describe: buildDescribe({
      kind: "durable",
      name: config.name,
      description: config.description,
      inputSchema: config.inputSchema,
      connection: config.connection,
      connections: config.connections
    })
  });
}
function defineComponent(config) {
  const component = () => {
    throw new Error(
      `defineComponent("${config.name}") is not callable yet. Runtime resolution is coming in Phase 2.`
    );
  };
  Object.defineProperty(component, "name", {
    value: config.name,
    writable: false
  });
  return Object.assign(component, {
    kind: "component",
    description: config.description,
    connection: "connection" in config ? config.connection : void 0,
    connections: "connections" in config ? config.connections : void 0,
    describe: buildDescribe({
      kind: "component",
      name: config.name,
      description: config.description,
      connection: "connection" in config ? config.connection : void 0,
      connections: "connections" in config ? config.connections : void 0
    })
  });
}

// src/define/helpers.ts
function optionsFrom(tool, binds) {
  return { kind: "options", tool, binds };
}
function schemaFrom(tool, binds) {
  return { kind: "schema", tool, binds };
}

exports.defineComponent = defineComponent;
exports.defineDurable = defineDurable;
exports.defineTool = defineTool;
exports.optionsFrom = optionsFrom;
exports.schemaFrom = schemaFrom;
