// @zapier/zapier-sdk/dist/index.js
// Export everything from types and plugins
export * from "./types/domain";
export * from "./types/functions";
export * from "./types/properties";
export * from "./types/errors";
export * from "./plugins/apps";
export * from "./plugins/fetch";
export * from "./plugins/listApps";
export * from "./plugins/listActions";
export * from "./plugins/listInputFields";
export * from "./plugins/listConnections";
export * from "./plugins/listClientCredentials";
export * from "./plugins/createClientCredentials";
export * from "./plugins/deleteClientCredentials";
export * from "./plugins/getApp";
export * from "./plugins/getAction";
export * from "./plugins/getConnection";
export * from "./plugins/findFirstConnection";
export * from "./plugins/findUniqueConnection";
export * from "./plugins/runAction";
export * from "./plugins/request";
export * from "./plugins/manifest";
export * from "./plugins/getProfile";
export * from "./plugins/api";
// Export schema utilities for CLI
export { isPositional } from "./utils/schema-utils";
export { createFunction } from "./utils/function-utils";
// Export telemetry utilities for CLI
export { runWithTelemetryContext } from "./utils/telemetry-context";
// Export string utilities
export { toSnakeCase, toTitleCase } from "./utils/string-utils";
// Export batch utilities
export { batch } from "./utils/batch-utils";
// Export resolver utilities for CLI
export * from "./resolvers";
// Export auth utilities for CLI use
export * from "./auth";
// Export credentials types and utilities
export * from "./credentials";
export * from "./types/credentials";
// Export connections types and utilities
export * from "./types/connections";
export * from "./plugins/connections";
// Export capabilities
export { buildCapabilityMessage } from "./plugins/capabilities";
// Export logging utilities
export { logDeprecation, resetDeprecationWarnings } from "./utils/logging";
// Export constants
export * from "./constants";
// All functions now available through the plugin system via createZapierSdk()
// Export plugin schemas for external use
export { RelayRequestSchema, RelayFetchSchema, } from "./plugins/request/schemas";
// Export table plugins
export * from "./plugins/tables/listTables";
export * from "./plugins/tables/getTable";
export * from "./plugins/tables/createTable";
export * from "./plugins/tables/deleteTable";
export * from "./plugins/tables/listTableFields";
export * from "./plugins/tables/createTableFields";
export * from "./plugins/tables/deleteTableFields";
export * from "./plugins/tables/getTableRecord";
export * from "./plugins/tables/listTableRecords";
export * from "./plugins/tables/createTableRecords";
export * from "./plugins/tables/deleteTableRecords";
export * from "./plugins/tables/updateTableRecords";
// Export the main combined SDK and new flexible SDK creator
export { createZapierSdk, createZapierSdkWithoutRegistry, createSdk, } from "./sdk";
export { BaseSdkOptionsSchema } from "./types/sdk";
// Export registry plugin for manual use
export { registryPlugin } from "./plugins/registry";
export { generateEventId, getCurrentTimestamp, getReleaseId, getOsInfo, getPlatformVersions, isCi, getCiPlatform, getMemoryUsage, getCpuTime, buildApplicationLifecycleEvent, buildErrorEventWithContext, buildErrorEvent, createBaseEvent, buildMethodCalledEvent, } from "./plugins/eventEmission";
