// @zapier/zapier-sdk/dist/sdk.js
// Import plugin functions
import { appsPlugin } from "./plugins/apps/index";
import { fetchPlugin } from "./plugins/fetch/index";
import { registryPlugin } from "./plugins/registry/index";
import { apiPlugin } from "./plugins/api/index";
import { getProfilePlugin } from "./plugins/getProfile";
import { listAppsPlugin } from "./plugins/listApps";
import { getAppPlugin } from "./plugins/getApp";
import { listActionsPlugin } from "./plugins/listActions";
import { getActionPlugin } from "./plugins/getAction";
import { runActionPlugin } from "./plugins/runAction";
import { listConnectionsPlugin } from "./plugins/listConnections";
import { getConnectionPlugin } from "./plugins/getConnection";
import { findFirstConnectionPlugin } from "./plugins/findFirstConnection";
import { findUniqueConnectionPlugin } from "./plugins/findUniqueConnection";
// Deprecated authentication plugins (thin wrappers for backward compatibility)
import { listAuthenticationsPlugin, getAuthenticationPlugin, findFirstAuthenticationPlugin, findUniqueAuthenticationPlugin, } from "./plugins/deprecated/authentications";
import { listClientCredentialsPlugin } from "./plugins/listClientCredentials";
import { createClientCredentialsPlugin } from "./plugins/createClientCredentials";
import { deleteClientCredentialsPlugin } from "./plugins/deleteClientCredentials";
import { listInputFieldsPlugin } from "./plugins/listInputFields";
import { getInputFieldsSchemaPlugin } from "./plugins/getInputFieldsSchema";
import { listInputFieldChoicesPlugin } from "./plugins/listInputFieldChoices";
import { listTablesPlugin } from "./plugins/tables/listTables";
import { getTablePlugin } from "./plugins/tables/getTable";
import { deleteTablePlugin } from "./plugins/tables/deleteTable";
import { createTablePlugin } from "./plugins/tables/createTable";
import { listTableFieldsPlugin } from "./plugins/tables/listTableFields";
import { createTableFieldsPlugin } from "./plugins/tables/createTableFields";
import { deleteTableFieldsPlugin } from "./plugins/tables/deleteTableFields";
import { getTableRecordPlugin } from "./plugins/tables/getTableRecord";
import { listTableRecordsPlugin } from "./plugins/tables/listTableRecords";
import { createTableRecordsPlugin } from "./plugins/tables/createTableRecords";
import { deleteTableRecordsPlugin } from "./plugins/tables/deleteTableRecords";
import { updateTableRecordsPlugin } from "./plugins/tables/updateTableRecords";
import { requestPlugin } from "./plugins/request";
import { manifestPlugin } from "./plugins/manifest";
import { capabilitiesPlugin } from "./plugins/capabilities";
import { connectionsPlugin } from "./plugins/connections";
import { eventEmissionPlugin } from "./plugins/eventEmission";
// Create SDK that supports flat plugins - returns an SDK, not a builder
export function createSdk(options = {}, initialSdk = {}, initialContext = { meta: {} }) {
    return {
        ...initialSdk,
        getContext: () => initialContext,
        addPlugin(plugin, addPluginOptions = {}) {
            // Create SDK with getContext method for plugins to use
            const currentSdkWithContext = {
                ...initialSdk,
                getContext: () => initialContext,
            };
            // Apply the plugin with options merged in
            const pluginResult = plugin({
                sdk: currentSdkWithContext,
                context: {
                    ...initialContext,
                    // Add the options that createSdk was called with to context
                    options,
                },
                ...addPluginOptions,
            });
            // Extract context from plugin result if present
            const { context: pluginContext, ...sdkProperties } = pluginResult;
            // Merge SDK properties directly (flat structure)
            const newSdk = { ...initialSdk, ...sdkProperties };
            // Merge context additions if plugin returned any
            // Ensure meta is always present, starting with existing or empty object
            let newContext = {
                ...initialContext,
                ...addPluginOptions,
                meta: initialContext.meta || {},
            };
            if (pluginContext) {
                // Extract meta from plugin context for special merging
                const { meta: pluginMeta, ...otherPluginContext } = pluginContext;
                newContext = {
                    ...newContext,
                    ...otherPluginContext,
                };
                // Merge meta separately to avoid overwriting
                if (pluginMeta) {
                    newContext = {
                        ...newContext,
                        meta: {
                            ...newContext.meta, // Existing meta (now guaranteed to exist)
                            ...pluginMeta, // New meta from plugin
                        },
                    };
                }
            }
            // Recursively create new SDK with updated properties
            return createSdk(options, newSdk, newContext);
        },
    };
}
export function createZapierSdkWithoutRegistry(options = {}) {
    return (createSdk(options)
        // Event emission (must be first to be available to other plugins)
        .addPlugin(eventEmissionPlugin)
        // Provides the API client in context
        .addPlugin(apiPlugin)
        // Manifest plugin (provides version locking context) - must come after apiPlugin
        .addPlugin(manifestPlugin)
        // Capabilities gates (must come after manifestPlugin for config access)
        .addPlugin(capabilitiesPlugin)
        // Named connections resolution
        .addPlugin(connectionsPlugin)
        // Apps/actions/fields
        .addPlugin(listAppsPlugin)
        .addPlugin(getAppPlugin)
        .addPlugin(listActionsPlugin)
        .addPlugin(getActionPlugin)
        .addPlugin(listInputFieldsPlugin)
        .addPlugin(getInputFieldsSchemaPlugin)
        .addPlugin(listInputFieldChoicesPlugin)
        // Run action
        .addPlugin(runActionPlugin)
        // Connections
        .addPlugin(listConnectionsPlugin)
        .addPlugin(getConnectionPlugin)
        .addPlugin(findFirstConnectionPlugin)
        .addPlugin(findUniqueConnectionPlugin)
        // Deprecated authentication plugins (thin wrappers for backward compatibility)
        .addPlugin(listAuthenticationsPlugin)
        .addPlugin(getAuthenticationPlugin)
        .addPlugin(findFirstAuthenticationPlugin)
        .addPlugin(findUniqueAuthenticationPlugin)
        // Client Credentials
        .addPlugin(listClientCredentialsPlugin)
        .addPlugin(createClientCredentialsPlugin)
        .addPlugin(deleteClientCredentialsPlugin)
        // Relay fetching — fetch is the primary HTTP plugin; request is a deprecated wrapper
        .addPlugin(fetchPlugin)
        .addPlugin(requestPlugin)
        // Tables
        .addPlugin(listTablesPlugin)
        .addPlugin(getTablePlugin)
        .addPlugin(deleteTablePlugin)
        .addPlugin(createTablePlugin)
        .addPlugin(listTableFieldsPlugin)
        .addPlugin(createTableFieldsPlugin)
        .addPlugin(deleteTableFieldsPlugin)
        .addPlugin(getTableRecordPlugin)
        .addPlugin(listTableRecordsPlugin)
        .addPlugin(createTableRecordsPlugin)
        .addPlugin(deleteTableRecordsPlugin)
        .addPlugin(updateTableRecordsPlugin)
        // Magic apps plugin
        .addPlugin(appsPlugin)
        // Profile
        .addPlugin(getProfilePlugin));
}
export function createZapierSdk(options = {}) {
    return (createZapierSdkWithoutRegistry(options)
        // Register plugins for CLI/MCP metadata
        .addPlugin(registryPlugin));
}
