// @zapier/zapier-sdk-core/dist/v0/common/responses.js
// src/v0/common/responses.ts
var rateLimitHeaders = {
  "X-RateLimit-Limit": {
    schema: {
      type: "integer"
    },
    description: "The number of requests remaining in the current rate limit window."
  },
  "X-RateLimit-Remaining": {
    schema: {
      type: "integer"
    },
    description: "The number of requests remaining in the current rate limit window."
  },
  "X-RateLimit-Reset": {
    schema: {
      type: "string"
    },
    description: "The timestamp when the rate limit will be reset."
  },
  "Retry-After": {
    schema: {
      type: "string"
    },
    description: "The number of seconds to wait before retrying."
  }
};
var retryAfterHeader = {
  "Retry-After": {
    schema: {
      type: "string"
    },
    description: "The number of seconds to wait before retrying."
  }
};
var commonErrorResponses = {
  /**
   * 400 Bad Request - Invalid request parameters or body
   */
  400: {
    description: "Bad Request",
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 401 Unauthorized - Missing or invalid authentication
   */
  401: {
    description: "Unauthorized",
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 429 Too Many Requests - Rate limit exceeded
   * Includes rate limit headers
   */
  429: {
    description: "Too Many Requests",
    headers: rateLimitHeaders,
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 500 Server Error - Internal server error
   */
  500: {
    description: "Server Error",
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 503 Service Unavailable - Service temporarily unavailable
   * Includes Retry-After header
   */
  503: {
    description: "Service Unavailable",
    headers: retryAfterHeader,
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  }
};
export {
  commonErrorResponses,
  rateLimitHeaders,
  retryAfterHeader
};
