// @zapier/zapier-sdk-core/dist/v0/common/responses.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/v0/common/responses.ts
var responses_exports = {};
__export(responses_exports, {
  commonErrorResponses: () => commonErrorResponses,
  rateLimitHeaders: () => rateLimitHeaders,
  retryAfterHeader: () => retryAfterHeader
});
module.exports = __toCommonJS(responses_exports);
var rateLimitHeaders = {
  "X-RateLimit-Limit": {
    schema: {
      type: "integer"
    },
    description: "The number of requests remaining in the current rate limit window."
  },
  "X-RateLimit-Remaining": {
    schema: {
      type: "integer"
    },
    description: "The number of requests remaining in the current rate limit window."
  },
  "X-RateLimit-Reset": {
    schema: {
      type: "string"
    },
    description: "The timestamp when the rate limit will be reset."
  },
  "Retry-After": {
    schema: {
      type: "string"
    },
    description: "The number of seconds to wait before retrying."
  }
};
var retryAfterHeader = {
  "Retry-After": {
    schema: {
      type: "string"
    },
    description: "The number of seconds to wait before retrying."
  }
};
var commonErrorResponses = {
  /**
   * 400 Bad Request - Invalid request parameters or body
   */
  400: {
    description: "Bad Request",
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 401 Unauthorized - Missing or invalid authentication
   */
  401: {
    description: "Unauthorized",
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 429 Too Many Requests - Rate limit exceeded
   * Includes rate limit headers
   */
  429: {
    description: "Too Many Requests",
    headers: rateLimitHeaders,
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 500 Server Error - Internal server error
   */
  500: {
    description: "Server Error",
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  },
  /**
   * 503 Service Unavailable - Service temporarily unavailable
   * Includes Retry-After header
   */
  503: {
    description: "Service Unavailable",
    headers: retryAfterHeader,
    content: {
      "application/json": {
        schema: {
          $ref: "#/components/schemas/ErrorsResponse"
        }
      }
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  commonErrorResponses,
  rateLimitHeaders,
  retryAfterHeader
});
