// @zapier/zapier-sdk/dist/plugins/tables/listTableRecords/index.js
import { createPaginatedFunction } from "../../../utils/function-utils";
import { ListTableRecordsApiResponseSchema, ListTableRecordsOptionsSchema, ListTableRecordsOptionsInputSchema, RecordItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { stripPageSuffix } from "../../../utils/string-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver, tableFiltersResolver, tableSortResolver, } from "../../../resolvers";
import { tableRecordFormatter } from "../../../formatters";
import { createFieldKeyTranslator, transformRecordItem, throwOnResponseErrors, } from "../utils";
import { DEFAULT_PAGE_SIZE } from "../../../constants";
function extractNextCursor(meta) {
    if (!meta?.pagination?.has_more || !meta.pagination.end_cursor) {
        return undefined;
    }
    return meta.pagination.end_cursor;
}
export const listTableRecordsPlugin = ({ context }) => {
    async function listTableRecordsPage(options) {
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        const translator = await createFieldKeyTranslator({
            api,
            tableId,
            keyMode: options.keyMode,
        });
        const body = {};
        if (options.pageSize !== undefined) {
            body.limit = options.pageSize;
        }
        if (options.filters) {
            body.filters = options.filters.map((f) => ({
                key: translator.translateFieldKey(f.fieldKey),
                operator: f.operator,
                value: f.value,
            }));
        }
        if (options.sort) {
            body.orders = [
                {
                    key: translator.translateFieldKey(options.sort.fieldKey),
                    direction: options.sort.direction,
                },
            ];
        }
        if (options.cursor) {
            body.cursor_query = {
                query_type: "window",
                start_cursor: options.cursor,
            };
        }
        const rawResponse = await api.post(`/tables/api/v1/tables/${tableId}/records/query`, body, {
            searchParams: { allow_nested_queries: "true" },
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to access this table or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to access this table. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 404) {
                    return new Error(`Table not found: ${tableId}`);
                }
                return undefined;
            },
            authRequired: true,
        });
        throwOnResponseErrors(rawResponse);
        const response = ListTableRecordsApiResponseSchema.parse(rawResponse);
        return {
            data: response.data.map((item) => ({
                ...transformRecordItem(item),
                data: translator.translateOutput(item.data),
            })),
            nextCursor: extractNextCursor(response.meta),
        };
    }
    const methodName = stripPageSuffix(listTableRecordsPage.name);
    const listTableRecordsDefinition = createPaginatedFunction(listTableRecordsPage, ListTableRecordsOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName), methodName, DEFAULT_PAGE_SIZE);
    return {
        listTableRecords: listTableRecordsDefinition,
        context: {
            meta: {
                listTableRecords: {
                    categories: ["table"],
                    type: "list",
                    itemType: "Record",
                    inputSchema: ListTableRecordsOptionsSchema,
                    outputSchema: RecordItemSchema,
                    resolvers: {
                        table: tableIdResolver,
                        filters: tableFiltersResolver,
                        sort: tableSortResolver,
                    },
                    formatter: tableRecordFormatter,
                },
            },
        },
    };
};
