// @zapier/zapier-sdk/dist/plugins/tables/listTableRecords/schemas.js
import { z } from "zod";
import { TablePropertySchema } from "../../../types/properties";
import { RecordApiItemSchema, RecordItemSchema, } from "../getTableRecord/schemas";
import { KeyModeSchema } from "../utils";
// Re-export for consumers
export { RecordItemSchema };
export const ListTableRecordsApiResponseSchema = z.object({
    data: z.array(RecordApiItemSchema),
    meta: z
        .object({
        pagination: z
            .object({
            start_cursor: z.string().nullable().optional(),
            end_cursor: z.string().nullable().optional(),
            has_more: z.boolean().optional(),
            has_less: z.boolean().optional(),
        })
            .optional(),
    })
        .optional(),
});
const FilterOperatorSchema = z.enum([
    "exact",
    "different",
    "contains",
    "icontains",
    "gte",
    "gt",
    "lt",
    "lte",
    "range",
    "in",
    "isnull",
    "startswith",
    "search",
    "is_within",
]);
const FilterConditionSchema = z.object({
    fieldKey: z.string().describe("The field key to filter on (e.g. f1, f2)"),
    operator: FilterOperatorSchema.describe("The comparison operator"),
    value: z.unknown().optional().describe("The value to compare against"),
});
const SortDirectionSchema = z.enum(["asc", "desc"]);
export const SortConditionSchema = z.object({
    fieldKey: z.string().describe("The field key to sort by"),
    direction: SortDirectionSchema.optional()
        .default("asc")
        .describe("Sort direction"),
});
const ListTableRecordsDescription = "List records in a table with optional filtering and sorting";
const ListTableRecordsBase = z.object({
    filters: z
        .array(FilterConditionSchema)
        .optional()
        .describe("Filter conditions for the query"),
    sort: SortConditionSchema.optional().describe("Sort records by a field"),
    pageSize: z
        .number()
        .min(1)
        .max(1000)
        .optional()
        .describe("Number of records per page (max 1000)"),
    maxItems: z
        .number()
        .min(1)
        .optional()
        .describe("Maximum total items to return across all pages"),
    cursor: z.string().optional().describe("Cursor to start from"),
    keyMode: KeyModeSchema,
});
// Schema for docs/registry (new param names only)
export const ListTableRecordsOptionsSchema = z
    .object({
    table: TablePropertySchema,
})
    .merge(ListTableRecordsBase)
    .describe(ListTableRecordsDescription)
    .meta({ aliases: { tableId: "table" } });
// Deprecated schema (backward compat)
const ListTableRecordsOptionsSchemaDeprecated = z
    .object({
    tableId: z
        .string()
        .regex(/^[A-Z0-9]{26}$/, "Table ID must be a valid ULID")
        .describe("The unique identifier of the table"),
})
    .merge(ListTableRecordsBase);
// Union for runtime validation
export const ListTableRecordsOptionsInputSchema = z
    .union([
    ListTableRecordsOptionsSchema,
    ListTableRecordsOptionsSchemaDeprecated,
])
    .describe(ListTableRecordsDescription);
