// @zapier/zapier-sdk/dist/plugins/tables/listTables/index.js
import { createPaginatedFunction } from "../../../utils/function-utils";
import { ListTablesApiResponseSchema, ListTablesOptionsSchema, TableItemSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { stripPageSuffix } from "../../../utils/string-utils";
import { ZapierAuthenticationError, ZapierValidationError, } from "../../../types/errors";
import { transformTableItem } from "../utils";
import { DEFAULT_PAGE_SIZE } from "../../../constants";
function extractNextCursor(links) {
    if (!links?.next) {
        return undefined;
    }
    try {
        const url = new URL(links.next);
        const offset = url.searchParams.get("offset");
        return offset || undefined;
    }
    catch {
        return undefined;
    }
}
export const listTablesPlugin = ({ context }) => {
    async function listTablesPage(options) {
        if (options.includeShared) {
            await context.checkCapability("canIncludeSharedTables");
        }
        const { api } = context;
        const searchParams = {};
        if (options.pageSize !== undefined) {
            searchParams.limit = options.pageSize.toString();
        }
        const tableIds = "tables" in options && options.tables ? options.tables : options.tableIds;
        if (tableIds && tableIds.length > 0) {
            searchParams.ids = tableIds.join(",");
        }
        if (options?.kind) {
            searchParams.kind = options.kind;
        }
        if (options?.search) {
            searchParams.q = options.search;
        }
        if (options?.owner && options.owner !== "me" && !options.includeShared) {
            throw new ZapierValidationError('The "owner" option requires "includeShared" to be true. Without includeShared, only your own tables are returned.');
        }
        const owner = options.includeShared ? options.owner : "me";
        if (owner) {
            if (owner === "me") {
                const profile = await api.get("/zapier/api/v4/profile/", {
                    authRequired: true,
                });
                searchParams.owner_customuser_id = String(profile.id);
            }
            else {
                searchParams.owner_customuser_id = owner;
            }
        }
        if (options.cursor) {
            searchParams.offset = options.cursor;
        }
        const rawResponse = await api.get("/tables/api/v1/tables", {
            searchParams,
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to list tables or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to list tables. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
        });
        const response = ListTablesApiResponseSchema.parse(rawResponse);
        return {
            data: response.data.map(transformTableItem),
            nextCursor: extractNextCursor(response.links),
        };
    }
    const methodName = stripPageSuffix(listTablesPage.name);
    const listTablesDefinition = createPaginatedFunction(listTablesPage, ListTablesOptionsSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, methodName), methodName, DEFAULT_PAGE_SIZE);
    return {
        listTables: listTablesDefinition,
        context: {
            meta: {
                listTables: {
                    categories: ["table"],
                    type: "list",
                    itemType: "Table",
                    inputSchema: ListTablesOptionsSchema,
                    outputSchema: TableItemSchema,
                },
            },
        },
    };
};
