// @zapier/zapier-sdk/dist/plugins/tables/listTables/schemas.js
import { z } from "zod";
import { TablesPropertySchema } from "../../../types/properties";
export const TableApiItemSchema = z.object({
    id: z.string(),
    name: z.string(),
    description: z.string().nullable().optional(),
    created_at: z.string(),
    edited_at: z.string(),
    kind: z.enum(["table", "virtual_table"]),
    owner_account_id: z.number(),
    owner_account_public_id: z.string().optional(),
    owner_user_id: z.number().nullable().optional(),
    owner_zapier_customuser_id: z.number(),
    owner_zapier_customuser_public_id: z.string().optional(),
    parent_table_id: z.string().nullable().optional(),
});
export const ListTablesApiResponseSchema = z.object({
    data: z.array(TableApiItemSchema),
    links: z
        .object({
        next: z.string().nullable().optional(),
    })
        .optional(),
});
export const TableItemSchema = z.object({
    id: z.string(),
    name: z.string(),
    description: z.string().optional(),
    created_at: z.string(),
    edited_at: z.string(),
    kind: z.enum(["table", "virtual_table"]),
    account_id: z.string(),
    profile_id: z.string(),
    parent_table_id: z.string().optional(),
});
export const ListTablesOptionsSchema = z
    .object({
    tables: TablesPropertySchema.optional(),
    tableIds: z
        .array(z.string())
        .optional()
        .describe("Filter by specific table IDs")
        .meta({ deprecated: true }),
    kind: z
        .enum(["table", "virtual_table", "both"])
        .optional()
        .describe("Filter by table type"),
    search: z
        .string()
        .optional()
        .describe("Search term to filter tables by name"),
    owner: z
        .string()
        .optional()
        .describe('Filter by table owner. Use "me" for the current user, or a numeric user ID. Requires includeShared to be true.'),
    includeShared: z
        .boolean()
        .optional()
        .describe("Include tables shared with you. Without this, only your own tables are returned."),
    pageSize: z
        .number()
        .min(1)
        .optional()
        .describe("Number of tables per page"),
    maxItems: z
        .number()
        .min(1)
        .optional()
        .describe("Maximum total items to return across all pages"),
    cursor: z.string().optional().describe("Cursor to start from"),
})
    .describe("List tables available to the authenticated user");
