// @zapier/zapier-sdk/dist/plugins/tables/deleteTable/index.js
import { createFunction } from "../../../utils/function-utils";
import { DeleteTableOptionsSchema, DeleteTableOptionsInputSchema, } from "./schemas";
import { createTelemetryCallback } from "../../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../../types/errors";
import { tableIdResolver } from "../../../resolvers";
export const deleteTablePlugin = ({ context }) => {
    async function deleteTable(options) {
        await context.checkCapability("canDeleteTables");
        const { api } = context;
        const tableId = "table" in options ? options.table : options.tableId;
        await api.delete(`/tables/api/v1/tables/${tableId}`, undefined, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to delete tables or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to delete tables. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
        });
        return {
            success: true,
        };
    }
    const deleteTableDefinition = createFunction(deleteTable, DeleteTableOptionsInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, deleteTable.name));
    return {
        deleteTable: deleteTableDefinition,
        context: {
            meta: {
                deleteTable: {
                    categories: ["table"],
                    type: "delete",
                    itemType: "Table",
                    inputSchema: DeleteTableOptionsSchema,
                    resolvers: {
                        table: tableIdResolver,
                    },
                    confirm: "delete",
                },
            },
        },
    };
};
