// @zapier/zapier-sdk-cli/dist/src/plugins/feedback/index.js
import { FeedbackSchema } from "./schemas";
import { createFunction } from "@zapier/zapier-sdk";
import { getLoggedInUser } from "@zapier/zapier-sdk-cli-login";
const feedbackResolver = {
    type: "static",
    inputType: "text",
    placeholder: "Enter your feedback",
};
const DEFAULT_FEEDBACK_WEBHOOK_URL = "https://hooks.zapier.com/hooks/catch/20279515/uc98k9m/";
const MAX_RETRIES = 2;
const RETRY_DELAY_MS = 1000;
function getFeedbackWebhookUrl() {
    return (process.env.ZAPIER_FEEDBACK_WEBHOOK_URL || DEFAULT_FEEDBACK_WEBHOOK_URL);
}
// This is a feedback-specific retry wrapper. A generic fetchWithRetry util
// would be the better long-term solution but should be a separate MR so it's
// built with broader use cases in mind.
async function postWithRetry({ body, attemptsLeft, }) {
    const response = await fetch(getFeedbackWebhookUrl(), {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body,
    });
    if (!response.ok && attemptsLeft > 1) {
        await new Promise((r) => setTimeout(r, RETRY_DELAY_MS));
        return postWithRetry({ body, attemptsLeft: attemptsLeft - 1 });
    }
    return response;
}
export const feedbackPlugin = ({ context, }) => {
    const debug = context.options.debug;
    const feedbackWithSdk = createFunction(async function feedback(options) {
        const user = await getLoggedInUser();
        const body = JSON.stringify({
            email: user.email,
            customuser_id: user.customUserId,
            feedback: options.feedback,
        });
        const response = await postWithRetry({ body, attemptsLeft: MAX_RETRIES });
        if (debug) {
            const text = await response.text();
            console.error("[debug] Webhook response:", text);
        }
        return "Thank you for your feedback!";
    }, FeedbackSchema);
    return {
        feedback: feedbackWithSdk,
        context: {
            meta: {
                feedback: {
                    categories: ["utility"],
                    inputSchema: FeedbackSchema,
                    resolvers: {
                        feedback: feedbackResolver,
                    },
                },
            },
        },
    };
};
