// @zapier/zapier-sdk/dist/plugins/deleteClientCredentials/index.js
import { DeleteClientCredentialsSchema, } from "./schemas";
import { createFunction } from "../../utils/function-utils";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
import { ZapierAuthenticationError } from "../../types/errors";
import { clientIdResolver } from "../../resolvers";
export const deleteClientCredentialsPlugin = ({ context }) => {
    async function deleteClientCredentials(options) {
        const { api } = context;
        await api.delete(`/api/v0/client-credentials/${options.clientId}`, undefined, {
            customErrorHandler: ({ status }) => {
                if (status === 401) {
                    return new ZapierAuthenticationError(`Authentication failed. Your token may not have permission to delete client credentials or may be expired. (HTTP ${status})`, { statusCode: status });
                }
                if (status === 403) {
                    return new ZapierAuthenticationError(`Access forbidden. Your token may not have the required scopes to delete client credentials. (HTTP ${status})`, { statusCode: status });
                }
                return undefined;
            },
            authRequired: true,
            requiredScopes: ["credentials"],
        });
        return {
            success: true,
        };
    }
    const deleteClientCredentialsDefinition = createFunction(deleteClientCredentials, DeleteClientCredentialsSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, deleteClientCredentials.name));
    return {
        deleteClientCredentials: deleteClientCredentialsDefinition,
        context: {
            meta: {
                deleteClientCredentials: {
                    categories: ["client-credentials"],
                    type: "delete",
                    itemType: "ClientCredentials",
                    inputSchema: DeleteClientCredentialsSchema,
                    resolvers: {
                        clientId: clientIdResolver,
                    },
                },
            },
        },
    };
};
