// @zapier/zapier-sdk-cli/dist/src/plugins/add/index.js
import { createFunction } from "@zapier/zapier-sdk";
import { AddSchema } from "./schemas";
import { detectTypesOutputDirectory } from "../../utils/directory-detection";
import { resolve } from "path";
export const addPlugin = ({ sdk }) => {
    const add = createFunction(async function add(options) {
        const { apps: appKeys, connections: connectionIds, configPath, typesOutput = await detectTypesOutputDirectory(), } = options;
        const resolvedTypesOutput = resolve(typesOutput);
        console.log(`📦 Adding ${appKeys.length} app(s)...`);
        // Track app names for better console output
        const appSlugAndKeyMap = new Map();
        // Define progress handler for manifest building
        const handleManifestProgress = (event) => {
            switch (event.type) {
                case "apps_lookup_start":
                    console.log(`📦 Looking up ${event.count} app(s)...`);
                    break;
                case "app_found":
                    // Store the app display name for later use
                    const displayName = event.app.slug
                        ? `${event.app.slug} (${event.app.key})`
                        : event.app.key;
                    appSlugAndKeyMap.set(event.app.key, displayName);
                    break;
                case "apps_lookup_complete":
                    if (event.count === 0) {
                        console.warn("⚠️ No apps found");
                    }
                    break;
                case "app_processing_start":
                    const appName = event.slug
                        ? `${event.slug} (${event.app})`
                        : event.app;
                    console.log(`📦 Adding ${appName}...`);
                    break;
                case "manifest_updated":
                    const appDisplay = appSlugAndKeyMap.get(event.app) || event.app;
                    console.log(`📝 Locked ${appDisplay} to ${event.app}@${event.version} using key '${event.manifestKey}'`);
                    break;
                case "app_processing_error":
                    const errorApp = appSlugAndKeyMap.get(event.app) || event.app;
                    console.warn(`⚠️ ${event.error} for ${errorApp}`);
                    break;
            }
        };
        // Define progress handler for type generation
        const handleTypesProgress = (event) => {
            switch (event.type) {
                case "connections_lookup_start":
                    console.log(`🔐 Looking up ${event.count} connection(s)...`);
                    break;
                case "connections_lookup_complete":
                    console.log(`🔐 Found ${event.count} connection(s)`);
                    break;
                case "connection_matched":
                    const appWithConnection = appSlugAndKeyMap.get(event.app) || event.app;
                    console.log(`🔐 Using connection ${event.connectionId} (${event.connectionTitle}) for ${appWithConnection}`);
                    break;
                case "connection_not_matched":
                    const appWithoutConnection = appSlugAndKeyMap.get(event.app) || event.app;
                    console.warn(`⚠️ No matching connection found for ${appWithoutConnection}`);
                    break;
                case "file_written":
                    console.log(`🔧 Generated types for ${event.manifestKey} at ${event.filePath}`);
                    break;
                case "app_processing_error":
                    const errorApp = appSlugAndKeyMap.get(event.app) || event.app;
                    console.warn(`⚠️ ${event.error} for ${errorApp}`);
                    break;
            }
        };
        const manifestResult = await sdk.buildManifest({
            apps: appKeys,
            skipWrite: false,
            configPath,
            onProgress: handleManifestProgress,
        });
        const typesResult = await sdk.generateAppTypes({
            apps: appKeys,
            connections: connectionIds,
            skipWrite: false,
            typesOutputDirectory: resolvedTypesOutput,
            onProgress: handleTypesProgress,
        });
        const results = manifestResult.manifest?.apps || {};
        // Final summary
        const successfulApps = Object.keys(results).filter((manifestKey) => typesResult.writtenFiles?.[manifestKey]);
        if (successfulApps.length > 0) {
            console.log(`✅ Added ${successfulApps.length} app(s) to manifest`);
        }
    }, AddSchema);
    return {
        add,
        context: {
            meta: {
                add: {
                    categories: ["utility"],
                    inputSchema: AddSchema,
                },
            },
        },
    };
};
