// @zapier/zapier-sdk-mcp/dist/index.mjs
// src/index.ts
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema
} from "@modelcontextprotocol/sdk/types.js";
import {
  createZapierSdk
} from "@zapier/zapier-sdk";

// src/utils/schema-converter.ts
function getZodDef(schema) {
  const schemaAny = schema;
  return schemaAny._zod?.def;
}
function getDescription(schema) {
  if (!schema) return void 0;
  return schema.description;
}
function convertZodToMcpSchema(zodSchema) {
  const def = getZodDef(zodSchema);
  if (!def) {
    return {
      type: "string",
      description: "Unsupported schema type"
    };
  }
  return convertZodType(def, zodSchema);
}
function convertZodType(def, schema) {
  const description = getDescription(schema);
  if (def.type === "object" && schema) {
    const schemaWithShape = schema;
    const properties = {};
    const required = [];
    if (schemaWithShape.shape) {
      for (const [key, value] of Object.entries(schemaWithShape.shape)) {
        const fieldDef = getZodDef(value);
        if (fieldDef) {
          properties[key] = convertZodType(fieldDef, value);
          if (fieldDef.type !== "optional" && fieldDef.type !== "nullable") {
            required.push(key);
          }
        }
      }
    }
    const result = {
      type: "object",
      properties
    };
    if (description) result.description = description;
    if (required.length > 0) result.required = required;
    return result;
  }
  if (def.type === "string") {
    const result = { type: "string" };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "number") {
    const result = { type: "number" };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "boolean") {
    const result = { type: "boolean" };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "array" && def.element) {
    const elementDef = getZodDef(def.element);
    const result = {
      type: "array",
      items: elementDef ? convertZodType(elementDef, def.element) : { type: "string" }
    };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "optional" && def.innerType) {
    const innerDef = getZodDef(def.innerType);
    if (innerDef) {
      return convertZodType(innerDef, def.innerType);
    }
  }
  if (def.type === "nullable" && def.innerType) {
    const innerDef = getZodDef(def.innerType);
    if (innerDef) {
      const innerSchema = convertZodType(innerDef, def.innerType);
      if (innerSchema.type !== void 0) {
        return {
          ...innerSchema,
          type: Array.isArray(innerSchema.type) ? [...innerSchema.type, "null"] : [innerSchema.type, "null"]
        };
      }
      return {
        anyOf: [innerSchema, { type: "null" }]
      };
    }
  }
  if (def.type === "enum" && def.entries) {
    const result = {
      type: "string",
      enum: Object.keys(def.entries)
    };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "union" && def.options) {
    const result = {
      oneOf: def.options.map((option) => {
        const optionDef = getZodDef(option);
        return optionDef ? convertZodType(optionDef, option) : null;
      }).filter((s) => s !== null)
    };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "intersection" && def.left && def.right) {
    const leftDef = getZodDef(def.left);
    const rightDef = getZodDef(def.right);
    const result = {
      allOf: [
        leftDef ? convertZodType(leftDef, def.left) : { type: "string" },
        rightDef ? convertZodType(rightDef, def.right) : { type: "string" }
      ]
    };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "literal" && def.values && def.values.length > 0) {
    const value = def.values[0];
    const result = {
      type: typeof value,
      enum: [value]
    };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "record") {
    let additionalProperties = true;
    if (def.valueType) {
      const valueDef = getZodDef(def.valueType);
      if (valueDef) {
        additionalProperties = convertZodType(valueDef, def.valueType);
      }
    }
    const result = {
      type: "object",
      additionalProperties
    };
    if (description) result.description = description;
    return result;
  }
  if (def.type === "any") {
    const result = {};
    if (description) result.description = description;
    return result;
  }
  if (def.type === "unknown") {
    const result = {};
    if (description) result.description = description;
    return result;
  }
  if (def.type === "date") {
    return {
      type: "string",
      format: "date-time",
      description: description || "ISO 8601 date-time string"
    };
  }
  if (def.type === "bigint") {
    return {
      type: "string",
      description: description || "BigInt as string"
    };
  }
  if (def.type === "undefined") {
    return {
      type: "null",
      description: description || "Undefined value"
    };
  }
  if (def.type === "null") {
    return {
      type: "null",
      description: description || "Null value"
    };
  }
  if (def.type === "void") {
    return {
      type: "null",
      description: description || "Void (null) value"
    };
  }
  if (def.type === "nan") {
    return {
      type: "number",
      description: description || "Not-a-Number value"
    };
  }
  if (def.type === "default" && def.innerType) {
    const innerDef = getZodDef(def.innerType);
    if (innerDef) {
      return convertZodType(innerDef, def.innerType);
    }
  }
  if (def.type === "catch" && def.innerType) {
    const innerDef = getZodDef(def.innerType);
    if (innerDef) {
      return convertZodType(innerDef, def.innerType);
    }
  }
  console.warn(
    `Unhandled Zod type: ${def.type}. Please consider adding support for this type.`
  );
  return {
    type: "string",
    description: description || `Unsupported type: ${def.type}`
  };
}

// src/index.ts
function convertInputParametersToMcpSchema(inputParameters) {
  const properties = {};
  const required = [];
  for (const param of inputParameters) {
    let schema = param.schema;
    let isOptional = false;
    const optionalDef = schema._zod?.def;
    if (optionalDef?.type === "optional" && optionalDef.innerType) {
      isOptional = true;
      schema = optionalDef.innerType;
    }
    const objectDef = schema._zod?.def;
    if (objectDef?.type === "object") {
      const schemaWithShape = schema;
      if (schemaWithShape.shape) {
        for (const [key, fieldSchema] of Object.entries(
          schemaWithShape.shape
        )) {
          properties[key] = convertZodToMcpSchema(fieldSchema);
          const fieldDef = fieldSchema._zod?.def;
          if (!isOptional && fieldDef?.type !== "optional" && fieldDef?.type !== "nullable") {
            required.push(key);
          }
        }
      }
    } else {
      properties[param.name] = convertZodToMcpSchema(param.schema);
      if (!isOptional) {
        required.push(param.name);
      }
    }
  }
  const result = {
    type: "object",
    properties
  };
  if (required.length > 0) {
    result.required = required;
  }
  return result;
}
function reconstructMcpPositionalArgs(inputParameters, flatArgs) {
  const args = [];
  for (const param of inputParameters) {
    let schema = param.schema;
    const optionalDef = schema._zod?.def;
    if (optionalDef?.type === "optional" && optionalDef.innerType) {
      schema = optionalDef.innerType;
    }
    const objectDef = schema._zod?.def;
    if (objectDef?.type === "object") {
      const schemaWithShape = schema;
      if (schemaWithShape.shape) {
        const obj = {};
        let hasValues = false;
        for (const key of Object.keys(schemaWithShape.shape)) {
          if (key in flatArgs && flatArgs[key] !== void 0) {
            obj[key] = flatArgs[key];
            hasValues = true;
          }
        }
        args.push(hasValues ? obj : void 0);
      } else {
        args.push(void 0);
      }
    } else {
      args.push(flatArgs[param.name]);
    }
  }
  return args;
}
function createZapierMcpServer({ debug = false } = {}) {
  const zapierSdk = createZapierSdk({
    debug,
    eventEmission: { callContext: "mcp" }
  });
  const server = new Server(
    {
      name: "zapier-sdk",
      version: "1.0.0"
    },
    {
      capabilities: {
        tools: {}
      }
    }
  );
  server.setRequestHandler(ListToolsRequestSchema, async () => {
    const tools = zapierSdk.getRegistry({ package: "mcp" }).functions.map((functionInfo) => {
      const toolName = functionInfo.name.replace(/([A-Z])/g, "-$1").toLowerCase();
      const description = functionInfo.description || functionInfo.inputSchema?.description || `Execute ${functionInfo.name}`;
      if (functionInfo.inputSchema) {
        return {
          name: toolName,
          description,
          inputSchema: convertZodToMcpSchema(functionInfo.inputSchema)
        };
      }
      return {
        name: toolName,
        description,
        inputSchema: convertInputParametersToMcpSchema(
          functionInfo.inputParameters
        )
      };
    });
    return { tools };
  });
  server.setRequestHandler(CallToolRequestSchema, async (request) => {
    const { name, arguments: args } = request.params;
    const functionName = name.replace(
      /-([a-z])/g,
      (_, letter) => letter.toUpperCase()
    );
    const functionInfo = zapierSdk.getRegistry({ package: "mcp" }).functions.find(
      (info) => info.name === functionName
    );
    if (!functionInfo) {
      const availableTools = zapierSdk.getRegistry({ package: "mcp" }).functions.map(
        (info) => info.name.replace(/([A-Z])/g, "-$1").toLowerCase()
      ).join(", ");
      throw new Error(
        `Unknown tool: ${name}. Available tools: ${availableTools}`
      );
    }
    try {
      let result;
      if (functionInfo.inputSchema) {
        const validatedArgs = functionInfo.inputSchema.parse(args || {});
        const sdkFn = zapierSdk[functionInfo.name];
        result = await sdkFn(validatedArgs);
      } else if (functionInfo.inputParameters) {
        const flatArgs = args || {};
        const positionalArgs = reconstructMcpPositionalArgs(
          functionInfo.inputParameters,
          flatArgs
        );
        const sdkFn = zapierSdk[functionInfo.name];
        result = await sdkFn(...positionalArgs);
      } else {
        const sdkFn = zapierSdk[functionInfo.name];
        result = await sdkFn();
      }
      return {
        content: [
          {
            type: "text",
            text: JSON.stringify(result, null, 2)
          }
        ]
      };
    } catch (error) {
      if (debug) {
        console.error(
          `Error executing function '${functionInfo.name}' (tool: ${name}):`,
          error
        );
      }
      let errorMessage;
      let isValidationError = false;
      if (error instanceof Error) {
        if (error.name === "ZodError") {
          isValidationError = true;
          errorMessage = `Validation error for ${functionInfo.name}: ${error.message}`;
        } else {
          errorMessage = `Error in ${functionInfo.name}: ${error.message}`;
        }
      } else {
        errorMessage = `Unexpected error in ${functionInfo.name}: ${String(error)}`;
      }
      const errorSchema = functionInfo.inputSchema ? convertZodToMcpSchema(functionInfo.inputSchema) : functionInfo.inputParameters ? convertInputParametersToMcpSchema(functionInfo.inputParameters) : {};
      return {
        content: [
          {
            type: "text",
            text: isValidationError ? `${errorMessage}

Expected input format:
${JSON.stringify(errorSchema, null, 2)}` : errorMessage
          }
        ],
        isError: true
      };
    }
  });
  return server;
}
async function startMcpServer(options = {}) {
  const server = createZapierMcpServer(options);
  const transport = new StdioServerTransport();
  return new Promise((resolve) => {
    let isShuttingDown = false;
    const cleanup = async () => {
      if (isShuttingDown) return;
      isShuttingDown = true;
      try {
        await server.close();
      } finally {
        resolve();
      }
    };
    server.onclose = () => {
      cleanup();
    };
    process.once("SIGINT", cleanup);
    process.once("SIGTERM", cleanup);
    server.connect(transport).catch((error) => {
      console.error("Failed to start MCP server:", error);
      process.exit(1);
    });
  });
}
export {
  createZapierMcpServer,
  startMcpServer
};
