// @zapier/zapier-sdk/dist/plugins/getApp/index.js
import { createFunction } from "../../utils/function-utils";
import { GetAppSchema, GetAppInputSchema } from "./schemas";
import { AppItemSchema } from "../../schemas/App";
import { ZapierAppNotFoundError } from "../../types/errors";
import { appKeyResolver } from "../../resolvers";
import { createTelemetryCallback } from "../../utils/telemetry-utils";
// GetApp plugin depends on listApps SDK function
export const getAppPlugin = ({ sdk, context }) => {
    async function getApp(options) {
        const appKey = "app" in options ? options.app : options.appKey;
        const appsIterable = sdk
            .listApps({
            appKeys: [appKey],
        })
            .items();
        for await (const app of appsIterable) {
            return {
                data: app,
            };
        }
        throw new ZapierAppNotFoundError("App not found", {
            appKey,
        });
    }
    const getAppDefinition = createFunction(getApp, GetAppInputSchema, createTelemetryCallback(context.eventEmission.emitMethodCalled, getApp.name));
    // Return flat structure - getApp goes directly to SDK
    return {
        getApp: getAppDefinition,
        context: {
            meta: {
                getApp: {
                    categories: ["app"],
                    type: "item",
                    itemType: "App",
                    inputSchema: GetAppSchema,
                    outputSchema: AppItemSchema,
                    resolvers: {
                        app: appKeyResolver,
                    },
                },
            },
        },
    };
};
