// @zapier/zapier-sdk-cli/dist/src/plugins/generateAppTypes/index.js
import { createFunction, ZapierValidationError, ZapierUnknownError, } from "@zapier/zapier-sdk";
import { GenerateAppTypesSchema, } from "./schemas";
import { AstTypeGenerator } from "../../generators/ast-generator";
import { getManifestKey } from "../../utils/manifest-helpers";
import { mkdir, writeFile } from "fs/promises";
import { join, resolve } from "path";
import { detectTypesOutputDirectory } from "../../utils/directory-detection";
export const generateAppTypesPlugin = ({ sdk }) => {
    const generateAppTypes = createFunction(async function generateAppTypes(options) {
        const { apps: appKeys, connections: connectionIds, skipWrite = false, typesOutputDirectory = await detectTypesOutputDirectory(), onProgress, } = options;
        const resolvedTypesOutput = resolve(typesOutputDirectory);
        const result = {
            typeDefinitions: {},
        };
        // Get apps using listApps (which respects existing manifest for version locking)
        onProgress?.({ type: "apps_lookup_start", count: appKeys.length });
        const appsIterable = sdk.listApps({ apps: appKeys }).items();
        const apps = [];
        for await (const app of appsIterable) {
            apps.push(app);
            onProgress?.({ type: "app_found", app });
        }
        onProgress?.({ type: "apps_lookup_complete", count: apps.length });
        if (apps.length === 0) {
            return result;
        }
        // Fetch connections if provided
        const connections = [];
        if (connectionIds && connectionIds.length > 0) {
            onProgress?.({
                type: "connections_lookup_start",
                count: connectionIds.length,
            });
            const connectionsIterable = sdk
                .listConnections({ connections: connectionIds })
                .items();
            for await (const connection of connectionsIterable) {
                connections.push(connection);
            }
            onProgress?.({
                type: "connections_lookup_complete",
                count: connections.length,
            });
        }
        // Ensure output directory exists if we're writing files
        if (!skipWrite && resolvedTypesOutput) {
            await mkdir(resolvedTypesOutput, { recursive: true });
        }
        // Initialize writtenFiles if we're writing to disk
        if (!skipWrite) {
            result.writtenFiles = {};
        }
        // Process each app
        for (const app of apps) {
            onProgress?.({
                type: "app_processing_start",
                app: app.key,
                slug: app.slug,
            });
            try {
                if (!app.version) {
                    const errorMessage = `Invalid implementation ID format: ${app.implementation_id}. Expected format: <implementationName>@<version>`;
                    onProgress?.({
                        type: "app_processing_error",
                        app: app.key,
                        error: errorMessage,
                    });
                    throw new ZapierValidationError(errorMessage, {
                        details: {
                            appKey: app.key,
                            implementationId: app.implementation_id,
                        },
                    });
                }
                // Find matching connection for this app if connections were provided
                let connectionId;
                if (connections.length > 0) {
                    const matchingConnection = connections.find((conn) => {
                        return conn.app_key === app.key;
                    });
                    if (matchingConnection) {
                        connectionId = matchingConnection.id;
                        onProgress?.({
                            type: "connection_matched",
                            app: app.key,
                            connectionId: matchingConnection.id,
                            connectionTitle: matchingConnection.title || "",
                        });
                    }
                    else {
                        onProgress?.({
                            type: "connection_not_matched",
                            app: app.key,
                        });
                    }
                }
                const manifestKey = getManifestKey(app);
                // Generate type definitions
                const generator = new AstTypeGenerator();
                const typeDefinitionString = await generator.generateTypes({
                    app,
                    connectionId,
                    sdk,
                });
                result.typeDefinitions[manifestKey] = typeDefinitionString;
                onProgress?.({
                    type: "type_generated",
                    manifestKey,
                    sizeBytes: typeDefinitionString.length,
                });
                // Write to file if skipWrite is false
                if (!skipWrite && resolvedTypesOutput && result.writtenFiles) {
                    const filePath = join(resolvedTypesOutput, `${manifestKey}.d.ts`);
                    await writeFile(filePath, typeDefinitionString, "utf8");
                    result.writtenFiles[manifestKey] = filePath;
                    onProgress?.({
                        type: "file_written",
                        manifestKey,
                        filePath,
                    });
                }
                onProgress?.({
                    type: "app_processing_complete",
                    app: app.key,
                });
            }
            catch (error) {
                const errorMessage = `Failed to process app ${app.key}: ${error instanceof Error ? error.message : String(error)}`;
                onProgress?.({
                    type: "app_processing_error",
                    app: app.key,
                    error: errorMessage,
                });
                if (error instanceof ZapierValidationError) {
                    throw error; // Preserve the specific error type
                }
                else {
                    throw new ZapierUnknownError(errorMessage, {
                        cause: error, // Works for both Error and non-Error
                    });
                }
            }
        }
        return result;
    }, GenerateAppTypesSchema);
    return {
        generateAppTypes,
        context: {
            meta: {
                generateAppTypes: {
                    categories: ["utility"],
                    inputSchema: GenerateAppTypesSchema,
                },
            },
        },
    };
};
