// @zapier/zapier-sdk-cli/dist/src/plugins/buildManifest/index.js
import { createFunction, ZapierValidationError, ZapierUnknownError, } from "@zapier/zapier-sdk";
import { BuildManifestSchema, } from "./schemas";
import { createManifestEntry } from "../../utils/manifest-helpers";
export const buildManifestPlugin = ({ sdk, context }) => {
    const buildManifest = createFunction(async function buildManifest(options) {
        const { apps: appKeys, skipWrite = false, configPath, onProgress, } = options;
        onProgress?.({ type: "apps_lookup_start", count: appKeys.length });
        const appsIterable = sdk.listApps({ apps: appKeys }).items();
        const apps = [];
        for await (const app of appsIterable) {
            apps.push(app);
            onProgress?.({ type: "app_found", app });
        }
        onProgress?.({ type: "apps_lookup_complete", count: apps.length });
        if (apps.length === 0) {
            return {};
        }
        let updatedManifest;
        // Process each app
        for (const app of apps) {
            onProgress?.({
                type: "app_processing_start",
                app: app.key,
                slug: app.slug,
            });
            try {
                const manifestEntry = createManifestEntry(app);
                onProgress?.({
                    type: "manifest_entry_built",
                    app: app.key,
                    manifestKey: manifestEntry.implementationName,
                    version: manifestEntry.version || "",
                });
                const { key: updatedManifestKey, manifest } = await context.updateManifestEntry({
                    appKey: app.key,
                    entry: manifestEntry,
                    configPath,
                    skipWrite,
                    manifest: updatedManifest,
                });
                updatedManifest = manifest;
                onProgress?.({
                    type: "manifest_updated",
                    app: app.key,
                    manifestKey: updatedManifestKey,
                    version: manifestEntry.version || "",
                });
                onProgress?.({ type: "app_processing_complete", app: app.key });
            }
            catch (error) {
                const errorMessage = `Failed to process app ${app.key}: ${error instanceof Error ? error.message : String(error)}`;
                onProgress?.({
                    type: "app_processing_error",
                    app: app.key,
                    error: errorMessage,
                });
                if (error instanceof ZapierValidationError) {
                    throw error; // Preserve the specific error type
                }
                else {
                    throw new ZapierUnknownError(errorMessage, {
                        cause: error, // Works for both Error and non-Error
                    });
                }
            }
        }
        return { manifest: updatedManifest };
    }, BuildManifestSchema);
    return {
        buildManifest,
        context: {
            meta: {
                buildManifest: {
                    categories: ["utility"],
                    inputSchema: BuildManifestSchema,
                },
            },
        },
    };
};
