// @zapier/zapier-sdk/dist/plugins/apps/schemas.js
import { z } from "zod";
import { ConnectionIdPropertySchema, ConnectionPropertySchema, AuthenticationIdPropertySchema, ActionTimeoutMsPropertySchema, } from "../../types/properties";
export const ActionExecutionInputSchema = z
    .object({
    inputs: z.record(z.string(), z.unknown()).optional(),
    /** @deprecated Use `connection` instead. */
    connectionId: ConnectionIdPropertySchema.optional().meta({
        deprecated: true,
    }),
    connection: ConnectionPropertySchema.optional(),
    /** @deprecated Use connection instead */
    authenticationId: AuthenticationIdPropertySchema.optional().meta({
        deprecated: true,
    }),
    timeoutMs: ActionTimeoutMsPropertySchema,
})
    .describe("Execute an action with the given inputs for the bound app, as an alternative to runAction")
    .meta({
    aliases: { connectionId: "connection", authenticationId: "connection" },
});
export const AppFactoryInputSchema = z
    .object({
    /** @deprecated Use `connection` instead. */
    connectionId: ConnectionIdPropertySchema.optional().meta({
        deprecated: true,
    }),
    connection: ConnectionPropertySchema.optional(),
    /** @deprecated Use connection instead */
    authenticationId: AuthenticationIdPropertySchema.optional().meta({
        deprecated: true,
    }),
})
    .describe("Bind a connection alias or numeric connectionId to an app")
    .meta({
    aliases: { connectionId: "connection", authenticationId: "connection" },
});
// Note: AppsPluginSdkExtension removed - now using AppsPluginProvides in index.ts
