// @zapier/zapier-sdk/dist/plugins/registry/index.js
// Registry plugin requires no context but collects SDK metadata
export const registryPlugin = ({ sdk, context }) => {
    const metaKeys = Object.keys(context.meta || {});
    const categoryDefinitions = {
        account: {
            title: "Account",
        },
        app: {
            title: "App",
            titlePlural: "Apps",
        },
        connection: {
            title: "Connection",
        },
        action: {
            title: "Action",
        },
        "client-credentials": {
            title: "Client Credentials",
            titlePlural: "Client Credentials",
        },
        table: {
            title: "Table",
        },
        http: {
            title: "HTTP Request",
        },
        utility: {
            title: "Utility",
            titlePlural: "Utilities",
        },
        other: {
            title: "Other",
        },
    };
    const functions = metaKeys
        .filter((key) => {
        const property = sdk[key];
        if (typeof property === "function") {
            return true;
        }
        const [rootKey] = key.split(".");
        const rootProperty = sdk[rootKey];
        if (typeof rootProperty === "object" && rootProperty !== null) {
            return true;
        }
        return false;
    })
        .map((key) => {
        const meta = context.meta[key];
        return {
            name: key,
            description: meta.description,
            type: meta.type,
            itemType: meta.itemType,
            returnType: meta.returnType,
            inputSchema: meta.inputSchema,
            inputParameters: meta.inputParameters,
            outputSchema: meta.outputSchema,
            categories: meta.categories || [],
            resolvers: meta.resolvers,
            formatter: meta.formatter,
            packages: meta.packages,
            // Auto-infer confirm: "delete" from type: "delete" (all deletes should confirm)
            confirm: meta.confirm ?? (meta.type === "delete" ? "delete" : undefined),
            deprecation: meta.deprecation,
            aliases: meta.aliases,
            supportsJsonOutput: meta.supportsJsonOutput ?? true,
        };
    })
        .sort((a, b) => a.name.localeCompare(b.name));
    const knownCategories = Object.keys(categoryDefinitions);
    // Cache the entire registry result for each package
    const registryCache = new Map();
    function getRegistry(options) {
        const packageFilter = options?.package;
        const cacheKey = packageFilter || "__all__";
        // Return cached result if available
        if (registryCache.has(cacheKey)) {
            return registryCache.get(cacheKey);
        }
        // Filter functions by package if specified
        const filteredFunctions = packageFilter
            ? functions.filter((f) => 
            // Include if packages is undefined (belongs to all packages) or includes the specified package
            !f.packages || f.packages.includes(packageFilter))
            : functions;
        // Build categories with filtered functions
        const filteredCategories = knownCategories
            .sort((a, b) => {
            // Keep "other" category last
            if (a === "other")
                return 1;
            if (b === "other")
                return -1;
            // Alphabetize by title, not key
            const titleA = categoryDefinitions[a].title;
            const titleB = categoryDefinitions[b].title;
            return titleA.localeCompare(titleB);
        })
            .map((categoryKey) => {
            // Find the functions that are in this category and match the package filter
            const categoryFunctions = filteredFunctions
                .filter((f) => f.categories.includes(categoryKey) ||
                // If the category is "other" and the function is not in any other category, include it
                (categoryKey === "other" &&
                    !f.categories.some((c) => knownCategories.includes(c))))
                .map((f) => f.name)
                .sort(); // Alphabetize functions within each category
            const definition = categoryDefinitions[categoryKey];
            const title = definition.title;
            return {
                key: categoryKey,
                title,
                titlePlural: definition.titlePlural ?? `${title}s`,
                functions: categoryFunctions,
            };
        })
            .filter((category) => category.functions.length > 0); // Only include categories with functions
        const result = {
            functions: filteredFunctions,
            categories: filteredCategories,
        };
        // Cache the result
        registryCache.set(cacheKey, result);
        return result;
    }
    return {
        getRegistry,
    };
};
